{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "9caf0b1a-c3b5-4352-8928-4858a306ef4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "from Bio import SeqIO, SeqFeature, SeqRecord\n",
    "from collections import defaultdict\n",
    "import pickle\n",
    "import glob\n",
    "import os\n",
    "import pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "033552b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CP124100.1': {'ID': 'CP124100.1', 'name': 'chromosome', 'length': 910422, 'strain': 'NE_5261'}, 'CP124098.1': {'ID': 'CP124098.1', 'name': 'lp17', 'length': 17222, 'strain': 'NE_5261'}, 'CP124097.1': {'ID': 'CP124097.1', 'name': 'cp26', 'length': 26454, 'strain': 'NE_5261'}, 'CP124099.1': {'ID': 'CP124099.1', 'name': 'lp54', 'length': 53730, 'strain': 'NE_5261'}, 'CP019844.1': {'ID': 'CP019844.1', 'name': 'chromosome', 'length': 909921, 'strain': 'PAli'}, 'CP019851.1': {'ID': 'CP019851.1', 'name': 'lp17', 'length': 16820, 'strain': 'PAli'}, 'CP019845.1': {'ID': 'CP019845.1', 'name': 'cp26', 'length': 26498, 'strain': 'PAli'}, 'CP019846.1': {'ID': 'CP019846.1', 'name': 'cp32-1', 'length': 30726, 'strain': 'PAli'}, 'CP019847.1': {'ID': 'CP019847.1', 'name': 'cp32-3', 'length': 30418, 'strain': 'PAli'}, 'CP019848.1': {'ID': 'CP019848.1', 'name': 'cp32-4', 'length': 30299, 'strain': 'PAli'}, 'CP019849.1': {'ID': 'CP019849.1', 'name': 'cp32-5', 'length': 30674, 'strain': 'PAli'}, 'CP019850.1': {'ID': 'CP019850.1', 'name': 'cp32-9', 'length': 30580, 'strain': 'PAli'}, 'CP019852.1': {'ID': 'CP019852.1', 'name': 'lp36', 'length': 36714, 'strain': 'PAli'}, 'CP019853.1': {'ID': 'CP019853.1', 'name': 'lp38', 'length': 38841, 'strain': 'PAli'}, 'CP019854.1': {'ID': 'CP019854.1', 'name': 'lp54', 'length': 53650, 'strain': 'PAli'}, 'CP019855.1': {'ID': 'CP019855.1', 'name': 'lp56', 'length': 52971, 'strain': 'PAli'}, 'CP002312.1': {'ID': 'CP002312.1', 'name': 'chromosome', 'length': 922801, 'strain': 'JD1'}, 'CP002316.1': {'ID': 'CP002316.1', 'name': 'cp26', 'length': 26525, 'strain': 'JD1'}, 'CP002310.1': {'ID': 'CP002310.1', 'name': 'cp32-1+5', 'length': 60739, 'strain': 'JD1'}, 'CP002323.1': {'ID': 'CP002323.1', 'name': 'cp32-10', 'length': 30314, 'strain': 'JD1'}, 'CP002309.1': {'ID': 'CP002309.1', 'name': 'cp32-11', 'length': 29902, 'strain': 'JD1'}, 'CP002321.1': {'ID': 'CP002321.1', 'name': 'cp32-12', 'length': 30019, 'strain': 'JD1'}, 'CP002320.1': {'ID': 'CP002320.1', 'name': 'cp32-3', 'length': 29707, 'strain': 'JD1'}, 'CP002307.1': {'ID': 'CP002307.1', 'name': 'cp32-6', 'length': 30804, 'strain': 'JD1'}, 'CP002311.1': {'ID': 'CP002311.1', 'name': 'cp32-8', 'length': 31085, 'strain': 'JD1'}, 'CP002322.1': {'ID': 'CP002322.1', 'name': 'cp32-9', 'length': 30697, 'strain': 'JD1'}, 'CP002313.1': {'ID': 'CP002313.1', 'name': 'lp17', 'length': 17522, 'strain': 'JD1'}, 'CP002308.1': {'ID': 'CP002308.1', 'name': 'lp25', 'length': 22827, 'strain': 'JD1'}, 'CP002306.1': {'ID': 'CP002306.1', 'name': 'lp28-1', 'length': 24354, 'strain': 'JD1'}, 'CP002314.1': {'ID': 'CP002314.1', 'name': 'lp28-3', 'length': 28797, 'strain': 'JD1'}, 'CP002318.1': {'ID': 'CP002318.1', 'name': 'lp28-4', 'length': 30096, 'strain': 'JD1'}, 'CP002317.1': {'ID': 'CP002317.1', 'name': 'lp28-5', 'length': 24614, 'strain': 'JD1'}, 'CP002324.1': {'ID': 'CP002324.1', 'name': 'lp28-6', 'length': 26981, 'strain': 'JD1'}, 'CP002319.1': {'ID': 'CP002319.1', 'name': 'lp28-7', 'length': 29993, 'strain': 'JD1'}, 'CP002315.1': {'ID': 'CP002315.1', 'name': 'lp36', 'length': 22808, 'strain': 'JD1'}, 'CP002325.1': {'ID': 'CP002325.1', 'name': 'lp38', 'length': 27590, 'strain': 'JD1'}, 'CP001652.1': {'ID': 'CP001652.1', 'name': 'lp54', 'length': 52916, 'strain': 'JD1'}, 'CP094597.1': {'ID': 'CP094597.1', 'name': 'chromosome', 'length': 909655, 'strain': 'B31-5A4'}, 'CP094606.1': {'ID': 'CP094606.1', 'name': 'cp9', 'length': 9387, 'strain': 'B31-5A4'}, 'CP094607.1': {'ID': 'CP094607.1', 'name': 'lp17', 'length': 16850, 'strain': 'B31-5A4'}, 'CP094608.1': {'ID': 'CP094608.1', 'name': 'lp21', 'length': 18211, 'strain': 'B31-5A4'}, 'CP094617.1': {'ID': 'CP094617.1', 'name': 'lp25', 'length': 24404, 'strain': 'B31-5A4'}, 'CP094598.1': {'ID': 'CP094598.1', 'name': 'cp26', 'length': 26499, 'strain': 'B31-5A4'}, 'CP094609.1': {'ID': 'CP094609.1', 'name': 'lp28-1', 'length': 28447, 'strain': 'B31-5A4'}, 'CP094610.1': {'ID': 'CP094610.1', 'name': 'lp28-2', 'length': 29936, 'strain': 'B31-5A4'}, 'CP094611.1': {'ID': 'CP094611.1', 'name': 'lp28-3', 'length': 28332, 'strain': 'B31-5A4'}, 'CP094612.1': {'ID': 'CP094612.1', 'name': 'lp28-4', 'length': 27109, 'strain': 'B31-5A4'}, 'CP094599.1': {'ID': 'CP094599.1', 'name': 'cp32-1', 'length': 30749, 'strain': 'B31-5A4'}, 'CP094600.1': {'ID': 'CP094600.1', 'name': 'cp32-3', 'length': 30224, 'strain': 'B31-5A4'}, 'CP094601.1': {'ID': 'CP094601.1', 'name': 'cp32-4', 'length': 30300, 'strain': 'B31-5A4'}, 'CP094602.1': {'ID': 'CP094602.1', 'name': 'cp32-6', 'length': 29839, 'strain': 'B31-5A4'}, 'CP094603.1': {'ID': 'CP094603.1', 'name': 'cp32-7', 'length': 30801, 'strain': 'B31-5A4'}, 'CP094604.1': {'ID': 'CP094604.1', 'name': 'cp32-8', 'length': 30886, 'strain': 'B31-5A4'}, 'CP094605.1': {'ID': 'CP094605.1', 'name': 'cp32-9', 'length': 30652, 'strain': 'B31-5A4'}, 'CP094613.1': {'ID': 'CP094613.1', 'name': 'lp36', 'length': 36662, 'strain': 'B31-5A4'}, 'CP094614.1': {'ID': 'CP094614.1', 'name': 'lp38', 'length': 38617, 'strain': 'B31-5A4'}, 'CP094615.1': {'ID': 'CP094615.1', 'name': 'lp54', 'length': 53445, 'strain': 'B31-5A4'}, 'CP094616.1': {'ID': 'CP094616.1', 'name': 'lp56', 'length': 52696, 'strain': 'B31-5A4'}, 'CP031412.1': {'ID': 'CP031412.1', 'name': 'chromosome', 'length': 908512, 'strain': 'MM1'}, 'CP031399.1': {'ID': 'CP031399.1', 'name': 'cp9', 'length': 11401, 'strain': 'MM1'}, 'CP031400.1': {'ID': 'CP031400.1', 'name': 'lp17', 'length': 14437, 'strain': 'MM1'}, 'CP031401.1': {'ID': 'CP031401.1', 'name': 'lp25', 'length': 24154, 'strain': 'MM1'}, 'CP031398.1': {'ID': 'CP031398.1', 'name': 'cp26', 'length': 26507, 'strain': 'MM1'}, 'CP031402.1': {'ID': 'CP031402.1', 'name': 'lp28-3', 'length': 26768, 'strain': 'MM1'}, 'CP031403.1': {'ID': 'CP031403.1', 'name': 'lp28-4', 'length': 27467, 'strain': 'MM1'}, 'CP031409.1': {'ID': 'CP031409.1', 'name': 'lp28-8', 'length': 28515, 'strain': 'MM1'}, 'CP031408.1': {'ID': 'CP031408.1', 'name': 'cp32-1', 'length': 30883, 'strain': 'MM1'}, 'CP031410.1': {'ID': 'CP031410.1', 'name': 'cp32-4', 'length': 13328, 'strain': 'MM1'}, 'CP031411.1': {'ID': 'CP031411.1', 'name': 'cp32-5', 'length': 21898, 'strain': 'MM1'}, 'CP031405.1': {'ID': 'CP031405.1', 'name': 'cp32-6', 'length': 19997, 'strain': 'MM1'}, 'CP031407.1': {'ID': 'CP031407.1', 'name': 'cp32-7', 'length': 14018, 'strain': 'MM1'}, 'CP031406.1': {'ID': 'CP031406.1', 'name': 'cp32-9', 'length': 30606, 'strain': 'MM1'}, 'CP031404.1': {'ID': 'CP031404.1', 'name': 'lp36', 'length': 27951, 'strain': 'MM1'}, 'CP031397.1': {'ID': 'CP031397.1', 'name': 'lp54', 'length': 53798, 'strain': 'MM1'}, 'CP124092.1': {'ID': 'CP124092.1', 'name': 'chromosome', 'length': 910655, 'strain': 'Am293'}, 'CP124090.1': {'ID': 'CP124090.1', 'name': 'lp17', 'length': 16928, 'strain': 'Am293'}, 'CP124089.1': {'ID': 'CP124089.1', 'name': 'cp26', 'length': 26498, 'strain': 'Am293'}, 'CP124091.1': {'ID': 'CP124091.1', 'name': 'lp54', 'length': 53672, 'strain': 'Am293'}, 'CP019916.1': {'ID': 'CP019916.1', 'name': 'chromosome', 'length': 910728, 'strain': 'PAbe'}, 'CP019922.1': {'ID': 'CP019922.1', 'name': 'lp17', 'length': 16800, 'strain': 'PAbe'}, 'CP019917.1': {'ID': 'CP019917.1', 'name': 'cp26', 'length': 26498, 'strain': 'PAbe'}, 'CP019923.1': {'ID': 'CP019923.1', 'name': 'lp28-1', 'length': 28157, 'strain': 'PAbe'}, 'CP019919.1': {'ID': 'CP019919.1', 'name': 'cp32-2', 'length': 31245, 'strain': 'PAbe'}, 'CP019920.1': {'ID': 'CP019920.1', 'name': 'cp32-3', 'length': 25789, 'strain': 'PAbe'}, 'CP019918.1': {'ID': 'CP019918.1', 'name': 'cp32-5-1', 'length': 56632, 'strain': 'PAbe'}, 'CP019921.1': {'ID': 'CP019921.1', 'name': 'cp32-9-4', 'length': 62238, 'strain': 'PAbe'}, 'CP019924.1': {'ID': 'CP019924.1', 'name': 'lp36', 'length': 36846, 'strain': 'PAbe'}, 'CP019925.1': {'ID': 'CP019925.1', 'name': 'lp54', 'length': 53650, 'strain': 'PAbe'}, 'CP019926.1': {'ID': 'CP019926.1', 'name': 'lp56', 'length': 52971, 'strain': 'PAbe'}, 'pBSV2_2': {'ID': 'pBSV2_2', 'name': 'pBSV2', 'length': 5843, 'strain': 'Synthetic_Vector'}, 'CP001446.1': {'ID': 'CP001446.1', 'name': 'cp26', 'length': 26506, 'strain': 'WI91-23'}, 'CP001445.1': {'ID': 'CP001445.1', 'name': 'cp32-1', 'length': 30026, 'strain': 'WI91-23'}, 'CP001449.1': {'ID': 'CP001449.1', 'name': 'cp32-10', 'length': 29756, 'strain': 'WI91-23'}, 'CP001451.1': {'ID': 'CP001451.1', 'name': 'cp32-11', 'length': 30772, 'strain': 'WI91-23'}, 'CP001444.1': {'ID': 'CP001444.1', 'name': 'cp32-4', 'length': 13540, 'strain': 'WI91-23'}, 'CP001453.1': {'ID': 'CP001453.1', 'name': 'cp32-5', 'length': 18338, 'strain': 'WI91-23'}, 'CP001450.1': {'ID': 'CP001450.1', 'name': 'cp32-6', 'length': 30017, 'strain': 'WI91-23'}, 'CP001457.1': {'ID': 'CP001457.1', 'name': 'cp32-7', 'length': 30052, 'strain': 'WI91-23'}, 'CP001460.1': {'ID': 'CP001460.1', 'name': 'cp9-3', 'length': 8706, 'strain': 'WI91-23'}, 'CP001461.1': {'ID': 'CP001461.1', 'name': 'lp17', 'length': 15386, 'strain': 'WI91-23'}, 'CP001459.1': {'ID': 'CP001459.1', 'name': 'lp25', 'length': 24661, 'strain': 'WI91-23'}, 'CP001456.1': {'ID': 'CP001456.1', 'name': 'lp28-1', 'length': 26634, 'strain': 'WI91-23'}, 'CP001452.1': {'ID': 'CP001452.1', 'name': 'lp28-3', 'length': 27543, 'strain': 'WI91-23'}, 'CP001454.1': {'ID': 'CP001454.1', 'name': 'lp28-4', 'length': 28237, 'strain': 'WI91-23'}, 'CP001455.1': {'ID': 'CP001455.1', 'name': 'lp28-6', 'length': 29782, 'strain': 'WI91-23'}, 'CP001448.1': {'ID': 'CP001448.1', 'name': 'lp36', 'length': 30134, 'strain': 'WI91-23'}, 'CP001458.1': {'ID': 'CP001458.1', 'name': 'lp38', 'length': 38893, 'strain': 'WI91-23'}, 'CP001462.1': {'ID': 'CP001462.1', 'name': 'lp5', 'length': 5252, 'strain': 'WI91-23'}, 'CP001447.1': {'ID': 'CP001447.1', 'name': 'lp54', 'length': 53783, 'strain': 'WI91-23'}, 'CP001443.1': {'ID': 'CP001443.1', 'name': 'lp56', 'length': 27581, 'strain': 'WI91-23'}, 'CP074054.1': {'ID': 'CP074054.1', 'name': 'chromosome', 'length': 903660, 'strain': 'B500'}, 'CP074063.1': {'ID': 'CP074063.1', 'name': 'lp17', 'length': 15547, 'strain': 'B500'}, 'CP074072.1': {'ID': 'CP074072.1', 'name': 'lp21-cp9', 'length': 24387, 'strain': 'B500'}, 'CP074059.1': {'ID': 'CP074059.1', 'name': 'lp25', 'length': 24206, 'strain': 'B500'}, 'CP074055.1': {'ID': 'CP074055.1', 'name': 'cp26', 'length': 26512, 'strain': 'B500'}, 'CP074060.1': {'ID': 'CP074060.1', 'name': 'lp28-3', 'length': 27653, 'strain': 'B500'}, 'CP074061.1': {'ID': 'CP074061.1', 'name': 'lp28-4', 'length': 28165, 'strain': 'B500'}, 'CP074062.1': {'ID': 'CP074062.1', 'name': 'lp28-6', 'length': 29288, 'strain': 'B500'}, 'CP074071.1': {'ID': 'CP074071.1', 'name': 'cp32-11', 'length': 30176, 'strain': 'B500'}, 'CP074065.1': {'ID': 'CP074065.1', 'name': 'cp32-3', 'length': 30189, 'strain': 'B500'}, 'CP074066.1': {'ID': 'CP074066.1', 'name': 'cp32-4', 'length': 30617, 'strain': 'B500'}, 'CP074067.1': {'ID': 'CP074067.1', 'name': 'cp32-5', 'length': 18341, 'strain': 'B500'}, 'CP074068.1': {'ID': 'CP074068.1', 'name': 'cp32-6', 'length': 30008, 'strain': 'B500'}, 'CP074069.1': {'ID': 'CP074069.1', 'name': 'cp32-7', 'length': 30054, 'strain': 'B500'}, 'CP074070.1': {'ID': 'CP074070.1', 'name': 'cp32-8', 'length': 29890, 'strain': 'B500'}, 'CP074057.1': {'ID': 'CP074057.1', 'name': 'lp36', 'length': 30208, 'strain': 'B500'}, 'CP074058.1': {'ID': 'CP074058.1', 'name': 'lp38', 'length': 38930, 'strain': 'B500'}, 'CP074056.1': {'ID': 'CP074056.1', 'name': 'lp54', 'length': 53758, 'strain': 'B500'}, 'CP074064.1': {'ID': 'CP074064.1', 'name': 'lp56', 'length': 27715, 'strain': 'B500'}, 'CP001205.1': {'ID': 'CP001205.1', 'name': 'chromosome', 'length': 906707, 'strain': 'ZS7'}, 'CP001212.1': {'ID': 'CP001212.1', 'name': 'cp26', 'length': 26514, 'strain': 'ZS7'}, 'CP001206.1': {'ID': 'CP001206.1', 'name': 'cp32-1', 'length': 30330, 'strain': 'ZS7'}, 'CP001200.1': {'ID': 'CP001200.1', 'name': 'cp32-12', 'length': 29806, 'strain': 'ZS7'}, 'CP001202.1': {'ID': 'CP001202.1', 'name': 'cp32-3+10', 'length': 48168, 'strain': 'ZS7'}, 'CP001208.1': {'ID': 'CP001208.1', 'name': 'cp32-4', 'length': 30964, 'strain': 'ZS7'}, 'CP001211.1': {'ID': 'CP001211.1', 'name': 'cp32-9', 'length': 30467, 'strain': 'ZS7'}, 'CP001204.1': {'ID': 'CP001204.1', 'name': 'lp17', 'length': 17266, 'strain': 'ZS7'}, 'CP001198.1': {'ID': 'CP001198.1', 'name': 'lp25', 'length': 24326, 'strain': 'ZS7'}, 'CP001210.1': {'ID': 'CP001210.1', 'name': 'lp28-1', 'length': 23422, 'strain': 'ZS7'}, 'CP001209.1': {'ID': 'CP001209.1', 'name': 'lp28-2', 'length': 29758, 'strain': 'ZS7'}, 'CP001203.1': {'ID': 'CP001203.1', 'name': 'lp28-3', 'length': 28414, 'strain': 'ZS7'}, 'CP001207.1': {'ID': 'CP001207.1', 'name': 'lp28-4', 'length': 28885, 'strain': 'ZS7'}, 'CP001201.1': {'ID': 'CP001201.1', 'name': 'lp36', 'length': 36852, 'strain': 'ZS7'}, 'CP001199.1': {'ID': 'CP001199.1', 'name': 'lp54', 'length': 53615, 'strain': 'ZS7'}, 'CP124084.1': {'ID': 'CP124084.1', 'name': 'chromosome', 'length': 910563, 'strain': 'Am388'}, 'CP124082.1': {'ID': 'CP124082.1', 'name': 'lp17', 'length': 16904, 'strain': 'Am388'}, 'CP124081.1': {'ID': 'CP124081.1', 'name': 'cp26', 'length': 26498, 'strain': 'Am388'}, 'CP124083.1': {'ID': 'CP124083.1', 'name': 'lp54', 'length': 53667, 'strain': 'Am388'}, 'CP001568.1': {'ID': 'CP001568.1', 'name': 'cp26', 'length': 26523, 'strain': 'Bol26'}, 'CP001565.1': {'ID': 'CP001565.1', 'name': 'cp32-12', 'length': 31433, 'strain': 'Bol26'}, 'CP001572.1': {'ID': 'CP001572.1', 'name': 'cp32-5', 'length': 30186, 'strain': 'Bol26'}, 'CP001570.1': {'ID': 'CP001570.1', 'name': 'cp32-9', 'length': 30597, 'strain': 'Bol26'}, 'CP001564.1': {'ID': 'CP001564.1', 'name': 'lp17', 'length': 17287, 'strain': 'Bol26'}, 'CP001573.1': {'ID': 'CP001573.1', 'name': 'lp28-3', 'length': 17941, 'strain': 'Bol26'}, 'CP001569.1': {'ID': 'CP001569.1', 'name': 'lp28-4', 'length': 28812, 'strain': 'Bol26'}, 'CP001566.1': {'ID': 'CP001566.1', 'name': 'lp28-9', 'length': 28200, 'strain': 'Bol26'}, 'CP001567.1': {'ID': 'CP001567.1', 'name': 'lp36', 'length': 36936, 'strain': 'Bol26'}, 'CP001571.1': {'ID': 'CP001571.1', 'name': 'lp54', 'length': 53655, 'strain': 'Bol26'}, 'CP094579.1': {'ID': 'CP094579.1', 'name': 'chromosome', 'length': 920773, 'strain': 'B379'}, 'CP094589.1': {'ID': 'CP094589.1', 'name': 'lp17', 'length': 17965, 'strain': 'B379'}, 'CP094580.1': {'ID': 'CP094580.1', 'name': 'cp26', 'length': 26515, 'strain': 'B379'}, 'CP094590.1': {'ID': 'CP094590.1', 'name': 'lp28-3', 'length': 28768, 'strain': 'B379'}, 'CP094591.1': {'ID': 'CP094591.1', 'name': 'lp28-4', 'length': 29187, 'strain': 'B379'}, 'CP094592.1': {'ID': 'CP094592.1', 'name': 'lp28-5', 'length': 27673, 'strain': 'B379'}, 'CP094593.1': {'ID': 'CP094593.1', 'name': 'lp28-6', 'length': 26653, 'strain': 'B379'}, 'CP094581.1': {'ID': 'CP094581.1', 'name': 'cp32-11', 'length': 30287, 'strain': 'B379'}, 'CP094582.1': {'ID': 'CP094582.1', 'name': 'cp32-12', 'length': 30796, 'strain': 'B379'}, 'CP094583.1': {'ID': 'CP094583.1', 'name': 'cp32-2', 'length': 30302, 'strain': 'B379'}, 'CP094584.1': {'ID': 'CP094584.1', 'name': 'cp32-3', 'length': 30263, 'strain': 'B379'}, 'CP094585.1': {'ID': 'CP094585.1', 'name': 'cp32-5', 'length': 30637, 'strain': 'B379'}, 'CP094586.1': {'ID': 'CP094586.1', 'name': 'cp32-6', 'length': 30642, 'strain': 'B379'}, 'CP094587.1': {'ID': 'CP094587.1', 'name': 'cp32-7', 'length': 21166, 'strain': 'B379'}, 'CP094588.1': {'ID': 'CP094588.1', 'name': 'cp32-9', 'length': 21173, 'strain': 'B379'}, 'CP094594.1': {'ID': 'CP094594.1', 'name': 'lp36', 'length': 31386, 'strain': 'B379'}, 'CP094595.1': {'ID': 'CP094595.1', 'name': 'lp38', 'length': 29988, 'strain': 'B379'}, 'CP094596.1': {'ID': 'CP094596.1', 'name': 'lp54', 'length': 53532, 'strain': 'B379'}, 'CP005925.1': {'ID': 'CP005925.1', 'name': 'chromosome', 'length': 910736, 'strain': 'CA382'}, 'CP124108.1': {'ID': 'CP124108.1', 'name': 'chromosome', 'length': 902811, 'strain': '80a'}, 'CP124106.1': {'ID': 'CP124106.1', 'name': 'lp17', 'length': 16096, 'strain': '80a'}, 'CP124105.1': {'ID': 'CP124105.1', 'name': 'cp26', 'length': 26506, 'strain': '80a'}, 'CP124107.1': {'ID': 'CP124107.1', 'name': 'lp54', 'length': 54709, 'strain': '80a'}, 'CP124096.1': {'ID': 'CP124096.1', 'name': 'chromosome', 'length': 910422, 'strain': 'NE_5267'}, 'CP124094.1': {'ID': 'CP124094.1', 'name': 'lp17', 'length': 17221, 'strain': 'NE_5267'}, 'CP124093.1': {'ID': 'CP124093.1', 'name': 'cp26', 'length': 26454, 'strain': 'NE_5267'}, 'CP124095.1': {'ID': 'CP124095.1', 'name': 'lp54', 'length': 53730, 'strain': 'NE_5267'}, 'CP124080.1': {'ID': 'CP124080.1', 'name': 'chromosome', 'length': 907764, 'strain': 'Sh-2-82'}, 'CP124078.1': {'ID': 'CP124078.1', 'name': 'lp17', 'length': 17944, 'strain': 'Sh-2-82'}, 'CP124077.1': {'ID': 'CP124077.1', 'name': 'cp26', 'length': 26514, 'strain': 'Sh-2-82'}, 'CP124079.1': {'ID': 'CP124079.1', 'name': 'lp54', 'length': 53843, 'strain': 'Sh-2-82'}, 'CP094561.1': {'ID': 'CP094561.1', 'name': 'chromosome', 'length': 903249, 'strain': 'B408'}, 'CP094569.1': {'ID': 'CP094569.1', 'name': 'lp17', 'length': 15444, 'strain': 'B408'}, 'CP094570.1': {'ID': 'CP094570.1', 'name': 'lp25', 'length': 21301, 'strain': 'B408'}, 'CP094562.1': {'ID': 'CP094562.1', 'name': 'cp26', 'length': 26507, 'strain': 'B408'}, 'CP094571.1': {'ID': 'CP094571.1', 'name': 'lp28-1', 'length': 22481, 'strain': 'B408'}, 'CP094572.1': {'ID': 'CP094572.1', 'name': 'lp28-3', 'length': 27525, 'strain': 'B408'}, 'CP094573.1': {'ID': 'CP094573.1', 'name': 'lp28-4', 'length': 27922, 'strain': 'B408'}, 'CP094574.1': {'ID': 'CP094574.1', 'name': 'lp28-6', 'length': 29294, 'strain': 'B408'}, 'CP094563.1': {'ID': 'CP094563.1', 'name': 'cp32-1', 'length': 30177, 'strain': 'B408'}, 'CP094564.1': {'ID': 'CP094564.1', 'name': 'cp32-3', 'length': 30193, 'strain': 'B408'}, 'CP094565.1': {'ID': 'CP094565.1', 'name': 'cp32-5', 'length': 18341, 'strain': 'B408'}, 'CP094566.1': {'ID': 'CP094566.1', 'name': 'cp32-6', 'length': 30019, 'strain': 'B408'}, 'CP094567.1': {'ID': 'CP094567.1', 'name': 'cp32-7', 'length': 30055, 'strain': 'B408'}, 'CP094568.1': {'ID': 'CP094568.1', 'name': 'cp32-8', 'length': 29893, 'strain': 'B408'}, 'CP094575.1': {'ID': 'CP094575.1', 'name': 'lp36', 'length': 30028, 'strain': 'B408'}, 'CP094576.1': {'ID': 'CP094576.1', 'name': 'lp38', 'length': 38839, 'strain': 'B408'}, 'CP094577.1': {'ID': 'CP094577.1', 'name': 'lp54', 'length': 53561, 'strain': 'B408'}, 'CP094578.1': {'ID': 'CP094578.1', 'name': 'lp56', 'length': 27231, 'strain': 'B408'}, 'CP017201.2': {'ID': 'CP017201.2', 'name': 'chromosome', 'length': 903654, 'strain': 'B331'}, 'CP017202.2': {'ID': 'CP017202.2', 'name': 'cp26', 'length': 26506, 'strain': 'B331'}, 'CP017203.1': {'ID': 'CP017203.1', 'name': 'cp32-8', 'length': 29892, 'strain': 'B331'}, 'CP017204.1': {'ID': 'CP017204.1', 'name': 'cp32-3', 'length': 30192, 'strain': 'B331'}, 'CP017205.1': {'ID': 'CP017205.1', 'name': 'cp32-11', 'length': 30176, 'strain': 'B331'}, 'CP017206.1': {'ID': 'CP017206.1', 'name': 'cp32-4', 'length': 30617, 'strain': 'B331'}, 'CP017207.1': {'ID': 'CP017207.1', 'name': 'cp32-5', 'length': 18340, 'strain': 'B331'}, 'CP017208.1': {'ID': 'CP017208.1', 'name': 'cp32-6', 'length': 30017, 'strain': 'B331'}, 'CP017209.1': {'ID': 'CP017209.1', 'name': 'cp32-7', 'length': 30054, 'strain': 'B331'}, 'CP017210.1': {'ID': 'CP017210.1', 'name': 'cp9', 'length': 8714, 'strain': 'B331'}, 'CP017211.1': {'ID': 'CP017211.1', 'name': 'lp17', 'length': 15594, 'strain': 'B331'}, 'CP017212.1': {'ID': 'CP017212.1', 'name': 'lp25', 'length': 24723, 'strain': 'B331'}, 'CP017213.1': {'ID': 'CP017213.1', 'name': 'lp28-3', 'length': 27643, 'strain': 'B331'}, 'CP017214.1': {'ID': 'CP017214.1', 'name': 'lp28-11', 'length': 23104, 'strain': 'B331'}, 'CP017215.2': {'ID': 'CP017215.2', 'name': 'lp28-6', 'length': 29288, 'strain': 'B331'}, 'CP017216.1': {'ID': 'CP017216.1', 'name': 'lp36', 'length': 30210, 'strain': 'B331'}, 'CP017217.1': {'ID': 'CP017217.1', 'name': 'lp38', 'length': 38925, 'strain': 'B331'}, 'CP017218.1': {'ID': 'CP017218.1', 'name': 'lp54', 'length': 53759, 'strain': 'B331'}, 'CP017219.1': {'ID': 'CP017219.1', 'name': 'lp56', 'length': 27681, 'strain': 'B331'}, 'CP001271.1': {'ID': 'CP001271.1', 'name': 'cp26', 'length': 26522, 'strain': '156a'}, 'CP001261.1': {'ID': 'CP001261.1', 'name': 'cp32-1', 'length': 14883, 'strain': '156a'}, 'CP001270.1': {'ID': 'CP001270.1', 'name': 'cp32-11', 'length': 30439, 'strain': '156a'}, 'CP001266.1': {'ID': 'CP001266.1', 'name': 'cp32-12', 'length': 30652, 'strain': '156a'}, 'CP001264.1': {'ID': 'CP001264.1', 'name': 'cp32-3', 'length': 30280, 'strain': '156a'}, 'CP001263.1': {'ID': 'CP001263.1', 'name': 'cp32-4', 'length': 30300, 'strain': '156a'}, 'CP001259.1': {'ID': 'CP001259.1', 'name': 'cp32-6', 'length': 30293, 'strain': '156a'}, 'CP001268.1': {'ID': 'CP001268.1', 'name': 'cp32-7', 'length': 21329, 'strain': '156a'}, 'CP001269.1': {'ID': 'CP001269.1', 'name': 'cp32-8', 'length': 30946, 'strain': '156a'}, 'CP001258.1': {'ID': 'CP001258.1', 'name': 'lp17', 'length': 17849, 'strain': '156a'}, 'CP001265.1': {'ID': 'CP001265.1', 'name': 'lp25', 'length': 24910, 'strain': '156a'}, 'CP001273.1': {'ID': 'CP001273.1', 'name': 'lp28-1', 'length': 24159, 'strain': '156a'}, 'CP001256.1': {'ID': 'CP001256.1', 'name': 'lp28-3', 'length': 28770, 'strain': '156a'}, 'CP001272.1': {'ID': 'CP001272.1', 'name': 'lp28-4', 'length': 29501, 'strain': '156a'}, 'CP001274.1': {'ID': 'CP001274.1', 'name': 'lp28-5', 'length': 27342, 'strain': '156a'}, 'CP001262.1': {'ID': 'CP001262.1', 'name': 'lp28-6', 'length': 26705, 'strain': '156a'}, 'CP001260.1': {'ID': 'CP001260.1', 'name': 'lp36', 'length': 23638, 'strain': '156a'}, 'CP001267.1': {'ID': 'CP001267.1', 'name': 'lp38', 'length': 30302, 'strain': '156a'}, 'CP001257.1': {'ID': 'CP001257.1', 'name': 'lp54', 'length': 53885, 'strain': '156a'}, 'CP001493.1': {'ID': 'CP001493.1', 'name': 'cp26', 'length': 26507, 'strain': '94a'}, 'CP001498.1': {'ID': 'CP001498.1', 'name': 'cp32-10', 'length': 30162, 'strain': '94a'}, 'CP001491.1': {'ID': 'CP001491.1', 'name': 'cp32-11', 'length': 30648, 'strain': '94a'}, 'CP001496.1': {'ID': 'CP001496.1', 'name': 'cp32-5', 'length': 29459, 'strain': '94a'}, 'CP001499.1': {'ID': 'CP001499.1', 'name': 'cp32-6', 'length': 30816, 'strain': '94a'}, 'CP001492.1': {'ID': 'CP001492.1', 'name': 'cp32-7', 'length': 30433, 'strain': '94a'}, 'CP001501.1': {'ID': 'CP001501.1', 'name': 'cp9', 'length': 8696, 'strain': '94a'}, 'CP001497.1': {'ID': 'CP001497.1', 'name': 'lp17', 'length': 18110, 'strain': '94a'}, 'CP001502.1': {'ID': 'CP001502.1', 'name': 'lp28-3', 'length': 26181, 'strain': '94a'}, 'CP001490.1': {'ID': 'CP001490.1', 'name': 'lp28-4', 'length': 28038, 'strain': '94a'}, 'CP001494.1': {'ID': 'CP001494.1', 'name': 'lp36', 'length': 28234, 'strain': '94a'}, 'CP001500.1': {'ID': 'CP001500.1', 'name': 'lp54', 'length': 53756, 'strain': '94a'}, 'CP001495.1': {'ID': 'CP001495.1', 'name': 'lp56', 'length': 21082, 'strain': '94a'}, 'CP124088.1': {'ID': 'CP124088.1', 'name': 'chromosome', 'length': 910229, 'strain': 'Am315'}, 'CP124086.1': {'ID': 'CP124086.1', 'name': 'lp17', 'length': 16517, 'strain': 'Am315'}, 'CP124085.1': {'ID': 'CP124085.1', 'name': 'cp26', 'length': 26525, 'strain': 'Am315'}, 'CP124087.1': {'ID': 'CP124087.1', 'name': 'lp54', 'length': 53232, 'strain': 'Am315'}, 'CP019767.1': {'ID': 'CP019767.1', 'name': 'chromosome', 'length': 909921, 'strain': 'B31_NRZ'}, 'CP019761.1': {'ID': 'CP019761.1', 'name': 'lp17', 'length': 16821, 'strain': 'B31_NRZ'}, 'CP019755.1': {'ID': 'CP019755.1', 'name': 'cp26', 'length': 26498, 'strain': 'B31_NRZ'}, 'CP019762.1': {'ID': 'CP019762.1', 'name': 'lp28-1', 'length': 21891, 'strain': 'B31_NRZ'}, 'CP019756.1': {'ID': 'CP019756.1', 'name': 'cp32-5+1', 'length': 61418, 'strain': 'B31_NRZ'}, 'CP019757.1': {'ID': 'CP019757.1', 'name': 'cp32-2', 'length': 31245, 'strain': 'B31_NRZ'}, 'CP019758.1': {'ID': 'CP019758.1', 'name': 'cp32-3', 'length': 30222, 'strain': 'B31_NRZ'}, 'CP019759.1': {'ID': 'CP019759.1', 'name': 'cp32-4', 'length': 30300, 'strain': 'B31_NRZ'}, 'CP019760.1': {'ID': 'CP019760.1', 'name': 'cp32-9', 'length': 30651, 'strain': 'B31_NRZ'}, 'CP019763.1': {'ID': 'CP019763.1', 'name': 'lp36', 'length': 36780, 'strain': 'B31_NRZ'}, 'CP019764.1': {'ID': 'CP019764.1', 'name': 'lp38', 'length': 38834, 'strain': 'B31_NRZ'}, 'CP019765.1': {'ID': 'CP019765.1', 'name': 'lp54', 'length': 53650, 'strain': 'B31_NRZ'}, 'CP019766.1': {'ID': 'CP019766.1', 'name': 'lp56', 'length': 52951, 'strain': 'B31_NRZ'}, 'CP124104.1': {'ID': 'CP124104.1', 'name': 'chromosome', 'length': 910663, 'strain': 'NE_5248'}, 'CP124102.1': {'ID': 'CP124102.1', 'name': 'lp17', 'length': 17282, 'strain': 'NE_5248'}, 'CP124101.1': {'ID': 'CP124101.1', 'name': 'cp26', 'length': 26504, 'strain': 'NE_5248'}, 'CP124103.1': {'ID': 'CP124103.1', 'name': 'lp54', 'length': 53675, 'strain': 'NE_5248'}, 'CP002228.1': {'ID': 'CP002228.1', 'name': 'chromosome', 'length': 902191, 'strain': 'N40'}, 'CP002239.1': {'ID': 'CP002239.1', 'name': 'cp26', 'length': 26498, 'strain': 'N40'}, 'CP002240.1': {'ID': 'CP002240.1', 'name': 'cp32-10', 'length': 30154, 'strain': 'N40'}, 'CP002238.1': {'ID': 'CP002238.1', 'name': 'cp32-12', 'length': 27767, 'strain': 'N40'}, 'CP002235.1': {'ID': 'CP002235.1', 'name': 'cp32-4', 'length': 15915, 'strain': 'N40'}, 'CP002227.1': {'ID': 'CP002227.1', 'name': 'cp32-5', 'length': 29615, 'strain': 'N40'}, 'CP002229.1': {'ID': 'CP002229.1', 'name': 'cp32-7', 'length': 17483, 'strain': 'N40'}, 'CP002242.1': {'ID': 'CP002242.1', 'name': 'cp32-9', 'length': 30658, 'strain': 'N40'}, 'CP002231.1': {'ID': 'CP002231.1', 'name': 'cp9', 'length': 8722, 'strain': 'N40'}, 'CP002241.1': {'ID': 'CP002241.1', 'name': 'lp17', 'length': 20672, 'strain': 'N40'}, 'CP002234.1': {'ID': 'CP002234.1', 'name': 'lp25', 'length': 23394, 'strain': 'N40'}, 'CP002232.1': {'ID': 'CP002232.1', 'name': 'lp28-2', 'length': 29521, 'strain': 'N40'}, 'CP002236.1': {'ID': 'CP002236.1', 'name': 'lp28-4', 'length': 29424, 'strain': 'N40'}, 'CP002233.1': {'ID': 'CP002233.1', 'name': 'lp28-5', 'length': 26642, 'strain': 'N40'}, 'CP002230.1': {'ID': 'CP002230.1', 'name': 'lp36', 'length': 29194, 'strain': 'N40'}, 'CP002237.1': {'ID': 'CP002237.1', 'name': 'lp38', 'length': 37903, 'strain': 'N40'}, 'CP001651.1': {'ID': 'CP001651.1', 'name': 'lp54', 'length': 53786, 'strain': 'N40'}, 'AE000783.1': {'ID': 'AE000783.1', 'name': 'chromosome', 'length': 910724, 'strain': 'B31'}, 'AE001583.1': {'ID': 'AE001583.1', 'name': 'lp5', 'length': 5228, 'strain': 'B31'}, 'AE000791.1': {'ID': 'AE000791.1', 'name': 'cp9', 'length': 9386, 'strain': 'B31'}, 'AE000793.2': {'ID': 'AE000793.2', 'name': 'lp17', 'length': 16821, 'strain': 'B31'}, 'AE001582.2': {'ID': 'AE001582.2', 'name': 'lp21', 'length': 18777, 'strain': 'B31'}, 'AE000785.1': {'ID': 'AE000785.1', 'name': 'lp25', 'length': 24177, 'strain': 'B31'}, 'AE000792.1': {'ID': 'AE000792.1', 'name': 'cp26', 'length': 26498, 'strain': 'B31'}, 'AE000794.2': {'ID': 'AE000794.2', 'name': 'lp28-1', 'length': 28155, 'strain': 'B31'}, 'AE000786.1': {'ID': 'AE000786.1', 'name': 'lp28-2', 'length': 29766, 'strain': 'B31'}, 'AE000784.1': {'ID': 'AE000784.1', 'name': 'lp28-3', 'length': 28601, 'strain': 'B31'}, 'AE000789.1': {'ID': 'AE000789.1', 'name': 'lp28-4', 'length': 27323, 'strain': 'B31'}, 'AE001575.1': {'ID': 'AE001575.1', 'name': 'cp32-1', 'length': 30750, 'strain': 'B31'}, 'AE001576.1': {'ID': 'AE001576.1', 'name': 'cp32-3', 'length': 30223, 'strain': 'B31'}, 'AE001577.1': {'ID': 'AE001577.1', 'name': 'cp32-4', 'length': 30299, 'strain': 'B31'}, 'AE001578.1': {'ID': 'AE001578.1', 'name': 'cp32-6', 'length': 29838, 'strain': 'B31'}, 'AE001579.1': {'ID': 'AE001579.1', 'name': 'cp32-7', 'length': 30800, 'strain': 'B31'}, 'AE001580.1': {'ID': 'AE001580.1', 'name': 'cp32-8', 'length': 30885, 'strain': 'B31'}, 'AE001581.1': {'ID': 'AE001581.1', 'name': 'cp32-9', 'length': 30651, 'strain': 'B31'}, 'AE000788.1': {'ID': 'AE000788.1', 'name': 'lp36', 'length': 36849, 'strain': 'B31'}, 'AE000787.1': {'ID': 'AE000787.1', 'name': 'lp38', 'length': 38829, 'strain': 'B31'}, 'AE000790.2': {'ID': 'AE000790.2', 'name': 'lp54', 'length': 53657, 'strain': 'B31'}, 'AE001584.1': {'ID': 'AE001584.1', 'name': 'lp56', 'length': 52971, 'strain': 'B31'}}\n"
     ]
    }
   ],
   "source": [
    "with open('../../plasmid_id/blast/blast_parsing_dict.pickle', 'rb') as infile:\n",
    "    id_parsing_dict = pickle.load(infile)\n",
    "print(id_parsing_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "7043900a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>contig</th>\n",
       "      <th>best_hit</th>\n",
       "      <th>completeness</th>\n",
       "      <th>new_contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>URI47</td>\n",
       "      <td>NODE_1_length_463767_cov_77.611941</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>incomplete</td>\n",
       "      <td>contig_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>URI47</td>\n",
       "      <td>NODE_2_length_221463_cov_78.896272</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>incomplete</td>\n",
       "      <td>contig_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>URI47</td>\n",
       "      <td>NODE_3_length_213568_cov_76.158203</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>incomplete</td>\n",
       "      <td>contig_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>URI47</td>\n",
       "      <td>NODE_4_length_53880_cov_109.176347</td>\n",
       "      <td>lp54</td>\n",
       "      <td>presumed</td>\n",
       "      <td>contig_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>URI47</td>\n",
       "      <td>NODE_5_length_39870_cov_168.736612</td>\n",
       "      <td>lp36</td>\n",
       "      <td>presumed</td>\n",
       "      <td>contig_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16824</th>\n",
       "      <td>URI93H</td>\n",
       "      <td>contig000025</td>\n",
       "      <td>NaN</td>\n",
       "      <td>incomplete</td>\n",
       "      <td>contig_25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16825</th>\n",
       "      <td>URI93H</td>\n",
       "      <td>contig000026</td>\n",
       "      <td>NaN</td>\n",
       "      <td>incomplete</td>\n",
       "      <td>contig_26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16826</th>\n",
       "      <td>URI93H</td>\n",
       "      <td>contig000027</td>\n",
       "      <td>NaN</td>\n",
       "      <td>incomplete</td>\n",
       "      <td>contig_27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16827</th>\n",
       "      <td>URI93H</td>\n",
       "      <td>contig000028</td>\n",
       "      <td>NaN</td>\n",
       "      <td>incomplete</td>\n",
       "      <td>contig_28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16828</th>\n",
       "      <td>URI93H</td>\n",
       "      <td>contig000029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>incomplete</td>\n",
       "      <td>contig_29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16829 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         name                              contig    best_hit completeness  \\\n",
       "0       URI47  NODE_1_length_463767_cov_77.611941  chromosome   incomplete   \n",
       "1       URI47  NODE_2_length_221463_cov_78.896272  chromosome   incomplete   \n",
       "2       URI47  NODE_3_length_213568_cov_76.158203  chromosome   incomplete   \n",
       "3       URI47  NODE_4_length_53880_cov_109.176347        lp54     presumed   \n",
       "4       URI47  NODE_5_length_39870_cov_168.736612        lp36     presumed   \n",
       "...       ...                                 ...         ...          ...   \n",
       "16824  URI93H                        contig000025         NaN   incomplete   \n",
       "16825  URI93H                        contig000026         NaN   incomplete   \n",
       "16826  URI93H                        contig000027         NaN   incomplete   \n",
       "16827  URI93H                        contig000028         NaN   incomplete   \n",
       "16828  URI93H                        contig000029         NaN   incomplete   \n",
       "\n",
       "      new_contig  \n",
       "0       contig_1  \n",
       "1       contig_2  \n",
       "2       contig_3  \n",
       "3       contig_4  \n",
       "4       contig_5  \n",
       "...          ...  \n",
       "16824  contig_25  \n",
       "16825  contig_26  \n",
       "16826  contig_27  \n",
       "16827  contig_28  \n",
       "16828  contig_29  \n",
       "\n",
       "[16829 rows x 5 columns]"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "plasmid_calls_csv = '/Users/mf019/bioinformatics/longread_pangenome/plasmid_id/output/plasmid_calls_V6/best_matches_v6_1000kb.tsv'\n",
    "contig2plasmid_df = pandas.read_csv(plasmid_calls_csv, delimiter='\\t')\n",
    "contig2plasmid_df = contig2plasmid_df[['name', 'contig', 'best_hit', 'completeness']]\n",
    "contig2plasmid_df['new_contig'] = contig2plasmid_df.groupby('name').cumcount() + 1\n",
    "contig2plasmid_df['new_contig'] = contig2plasmid_df['new_contig'].apply(lambda x: f'contig_{x}')\n",
    "contig2plasmid_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "6e4cb628",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "contig2plasmid_dict = defaultdict(lambda: defaultdict(dict))\n",
    "for _, row in contig2plasmid_df.iterrows():\n",
    "    contig2plasmid_dict[row['name']][row['new_contig']] = {\n",
    "        'contig' : row ['contig'],\n",
    "        'best_hit': row['best_hit'],\n",
    "        'completeness': row['completeness']\n",
    "    }\n",
    "    contig2plasmid_dict[row['name']][row['contig']] = {\n",
    "        'contig' : row ['new_contig'],\n",
    "        'best_hit': row['best_hit'],\n",
    "        'completeness': row['completeness']\n",
    "    }\n",
    "contig2plasmid_dict = {name: dict(contigs) for name, contigs in contig2plasmid_dict.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "988075e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         name                              contig    best_hit completeness  \\\n",
      "0       URI47  NODE_1_length_463767_cov_77.611941  chromosome   incomplete   \n",
      "1       URI47  NODE_2_length_221463_cov_78.896272  chromosome   incomplete   \n",
      "2       URI47  NODE_3_length_213568_cov_76.158203  chromosome   incomplete   \n",
      "3       URI47  NODE_4_length_53880_cov_109.176347        lp54     presumed   \n",
      "4       URI47  NODE_5_length_39870_cov_168.736612        lp36     presumed   \n",
      "...       ...                                 ...         ...          ...   \n",
      "16824  URI93H                        contig000025         NaN   incomplete   \n",
      "16825  URI93H                        contig000026         NaN   incomplete   \n",
      "16826  URI93H                        contig000027         NaN   incomplete   \n",
      "16827  URI93H                        contig000028         NaN   incomplete   \n",
      "16828  URI93H                        contig000029         NaN   incomplete   \n",
      "\n",
      "      new_contig  \n",
      "0       contig_1  \n",
      "1       contig_2  \n",
      "2       contig_3  \n",
      "3       contig_4  \n",
      "4       contig_5  \n",
      "...          ...  \n",
      "16824  contig_25  \n",
      "16825  contig_26  \n",
      "16826  contig_27  \n",
      "16827  contig_28  \n",
      "16828  contig_29  \n",
      "\n",
      "[16829 rows x 5 columns]\n"
     ]
    }
   ],
   "source": [
    "print(contig2plasmid_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "f8e4fa90-546a-497a-aa3f-db3a768059fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lets set up our parsing function. This is just the barebones function. I'll expand it with additional helpers\n",
    "# once I reach that point :)\n",
    "# primarily to include the plasmid name for the replicon.\n",
    "# and to\n",
    "def parse_genbank(file_path, assembly_id):\n",
    "    set_of_keys = set()\n",
    "    ref_flag = True if assembly_id == \"REF\" else False\n",
    "    assemblies = SeqIO.parse(file_path, \"genbank\")\n",
    "    genes = []\n",
    "    for record in assemblies:\n",
    "        if ref_flag == False and assembly_id not in contig2plasmid_dict.keys():\n",
    "            break\n",
    "        for feature in record.features:\n",
    "            if feature.type in [\"gene\", \"CDS\", \"mRNA\", \"tRNA\", \"rRNA\", \"signal_peptide\"]:\n",
    "                if ref_flag:\n",
    "                    assembly_id = id_parsing_dict[record.id]['strain']\n",
    "                    replicon_name = id_parsing_dict[record.id]['name']\n",
    "                else:\n",
    "                    replicon_name = contig2plasmid_dict[assembly_id][record.id]['best_hit']\n",
    "\n",
    "                locus_tag = feature.qualifiers.get(\"locus_tag\", ['unknown'])[0]\n",
    "                db_xrefs = feature.qualifiers.get(\"db_xref\", [])\n",
    "                db_xref_dict = {xref.split(\":\")[0]: xref.split(\":\")[1] for xref in db_xrefs}\n",
    "\n",
    "                gene_info = {\n",
    "                    \"feature_type\": feature.type,\n",
    "                    \"gene\": feature.qualifiers.get(\"gene\", [\"unknown\"])[0],\n",
    "                    \"locus_tag\": locus_tag,\n",
    "                    \"note\": feature.qualifiers.get(\"note\", [\"unknown\"])[0],\n",
    "                    \"protein_id\": feature.qualifiers.get(\"protein_id\", [\"unknown\"])[0],\n",
    "                    \"product\": feature.qualifiers.get(\"product\", [\"unknown\"])[0],\n",
    "                    \"sequence\": str(feature.extract(record.seq)),\n",
    "                    \"replicon\": record.id,\n",
    "                    \"replicon_name\": replicon_name,\n",
    "                    \"start\": feature.location.start,\n",
    "                    \"end\": feature.location.end,\n",
    "                    \"strand\": feature.location.strand,\n",
    "                    \"assembly\": assembly_id,\n",
    "                    \"translation\": str(feature.qualifiers.get('translation',[''])),\n",
    "                    \"inference\": feature.qualifiers.get(\"inference\",[\"\"])[0],\n",
    "                    \"transl_table\": feature.qualifiers.get(\"transl_table\",[\"\"])[0],\n",
    "                    **db_xref_dict  # Unpack db_xref_dict to include db_xrefs as columns\n",
    "                }\n",
    "                genes.append(gene_info)\n",
    "                set_of_keys.update(gene_info.keys())\n",
    "\n",
    "    # fill out empty dicts with None!\n",
    "    for gene in genes:\n",
    "        for key in set_of_keys:\n",
    "            if key not in gene:\n",
    "                gene[key] = None\n",
    "\n",
    "    return genes\n",
    "\n",
    "def create_and_populate_database(genes, db_name):\n",
    "    df = pandas.DataFrame(genes)\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    df.to_sql('annotations', conn, if_exists='replace', index=False)\n",
    "    conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "7dd0f159-90d9-4bae-bbd9-9819e958eae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#def create_table(db_name, table_name):\n",
    "#    conn = sqlite3.connect(db_name)\n",
    "#    cursor = conn.cursor()\n",
    "#    cursor.execute(f'''CREATE TABLE IF NOT EXISTS {table_name} (\n",
    "#                      id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "#                      feature_type TEXT,\n",
    "#                      gene TEXT,\n",
    "#                      locus_tag TEXT,\n",
    "#                      note TEXT,\n",
    "#                      protein_id TEXT,\n",
    "#                      product TEXT,\n",
    "#                      sequence TEXT,\n",
    "#                      replicon TEXT,\n",
    "#                      replicon_name TEXT,\n",
    "#                      start INTEGER,\n",
    "#                      end INTEGER,\n",
    "#                      strand INTEGER,\n",
    "#                      db_xrefs TEXT,\n",
    "#                      assembly TEXT,\n",
    "#                      translation TEXT,\n",
    "#                      inference TEXT,\n",
    "#                      transl_table INT)''')\n",
    "#    conn.commit()\n",
    "#    conn.close()\n",
    "#\n",
    "#def populate_table(db_name, table_name, genes):\n",
    "#    conn = sqlite3.connect(db_name)\n",
    "#    cursor = conn.cursor()\n",
    "#    for gene in genes:\n",
    "#        cursor.execute(f'''INSERT INTO {table_name}\n",
    "#                          (feature_type, gene, locus_tag, note, protein_id, product, sequence, replicon, replicon_name, start, end, strand, db_xrefs, assembly, translation, inference, transl_table)\n",
    "#                          VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)''',\n",
    "#                       (gene[\"feature_type\"], gene[\"gene\"], gene[\"locus_tag\"], gene[\"note\"],\n",
    "#                        gene[\"protein_id\"], gene[\"product\"], gene[\"sequence\"], gene[\"replicon\"],\n",
    "#                        gene[\"replicon_name\"], gene[\"start\"], gene[\"end\"], gene[\"strand\"], gene[\"db_xrefs\"],\n",
    "#                        gene[\"assembly\"], gene[\"translation\"], gene[\"inference\"], gene[\"transl_table\"]))\n",
    "#    conn.commit()\n",
    "#    conn.close()\n",
    "\n",
    "def get_tables(db_name):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    tables = pandas.read_sql_query(\"SELECT name FROM sqlite_master WHERE type='table';\", conn)\n",
    "    conn.close()\n",
    "    return tables\n",
    "\n",
    "def get_columns(db_name, table):\n",
    "    # Describe the structure of the 'genes' table\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    columns = pandas.read_sql_query(f\"PRAGMA table_info({table});\", conn)\n",
    "    conn.close()\n",
    "    return columns\n",
    "\n",
    "def table_head(db_name, table):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    table_head = pandas.read_sql_query(f\"SELECT * FROM {table} LIMIT 10;\", conn)\n",
    "    conn.close()\n",
    "    return table_head\n",
    "\n",
    "def get_all_of_gene(db_name, table, gene):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    all_genes = pandas.read_sql_query(f\"SELECT * FROM {table} WHERE feature_type = 'CDS' AND gene='{gene}'\", conn)\n",
    "    conn.close()\n",
    "    return all_genes\n",
    "\n",
    "def get_all_of_product(db_name, table, product):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    all_products = pandas.read_sql_query(f\"SELECT * FROM {table} WHERE feature_type = 'CDS' AND product='{product}'\", conn)\n",
    "    conn.close()\n",
    "    return all_products\n",
    "\n",
    "def dump_table_to_df(db_name, table):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    dataframe = pandas.read_sql_query(f\"SELECT * FROM {table};\", conn)\n",
    "    conn.close()\n",
    "    return dataframe\n",
    "\n",
    "def get_gene_from_baktaID(db_or_conn, table, bakta_id):\n",
    "    # is arg1 a string or a connection?\n",
    "    if isinstance(db_or_conn, str):\n",
    "        # it is string, thus db path. Create a new connection\n",
    "        conn = sqlite3.connect(db_or_conn)\n",
    "        close_conn = True\n",
    "    else:\n",
    "        # Ain't a string so must be connection. Use the existing connection\n",
    "        conn = db_or_conn\n",
    "        close_conn = False\n",
    "    all_genes = pandas.read_sql_query(f\"SELECT * FROM {table} WHERE locus_tag='{bakta_id}'\", conn)\n",
    "    # Pack it in pack it out!\n",
    "    if close_conn:\n",
    "        conn.close()\n",
    "    return all_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "f7417072-48bc-420e-838e-194651d49877",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "105\n",
      "317\n"
     ]
    }
   ],
   "source": [
    "list_of_assemblies = glob.glob('/Users/mf019/bioinformatics/longread_pangenome/assemblies/paired_assemblies/paired_only/*/annotation/*.gbff')\n",
    "print(len(list_of_assemblies))\n",
    "bakta_reference_genbank = '/Users/mf019/bioinformatics/longread_pangenome/ref/all_plasmids_bbss/Bbss.gbff'\n",
    "list_of_ncbi_ref_genbanks = glob.glob('/Users/mf019/bioinformatics/longread_pangenome/plasmid_id/plasmid_seqs/raw_gbs/*.gb')\n",
    "print(len(list_of_ncbi_ref_genbanks))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "1627a4a0-4316-4e1a-90b0-8006b6551c48",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ESI26\n",
      "URI112\n",
      "UCT110\n",
      "URI86\n",
      "UCT29\n",
      "URI44\n",
      "URI87\n",
      "URI91\n",
      "UCT32\n",
      "UWI263\n",
      "URI48\n",
      "URI33\n",
      "URI118\n",
      "URI102\n",
      "URI34\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception ignored in: <function tqdm.__del__ at 0x320110b80>\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/homebrew/lib/python3.11/site-packages/tqdm/std.py\", line 1147, in __del__\n",
      "    def __del__(self):\n",
      "\n",
      "KeyboardInterrupt: \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UNY172\n",
      "UWI248\n",
      "UCT35\n",
      "UNY149\n",
      "URI103\n",
      "UCT96\n",
      "URI42\n",
      "URI39\n",
      "UNY169\n",
      "URI120\n",
      "URI41\n",
      "UNY193\n",
      "URI36\n",
      "URI101\n",
      "URI117\n",
      "UNY203\n",
      "UWI247\n",
      "URI40\n",
      "URI56\n",
      "UCT31\n",
      "URI88\n",
      "UNY208\n",
      "URI47\n",
      "UWI283\n",
      "UCT50\n",
      "URI93\n",
      "UCT113\n",
      "URI46\n",
      "UCT92\n",
      "UCT109\n",
      "URI107\n",
      "URI111\n",
      "URI89\n",
      "UCT30\n",
      "URI101H\n",
      "ESI26H\n",
      "URI118H\n",
      "URI89H\n",
      "URI33H\n",
      "UNY1032P\n",
      "URI103H\n",
      "URI107H\n",
      "URI56H\n",
      "UNY203P\n",
      "URI93H\n",
      "URI102H\n",
      "URI91H\n",
      "UNY1090P\n",
      "URI88H\n",
      "URI48H\n",
      "UNY1128P\n",
      "URI34H\n",
      "URI120H\n",
      "UWI248P\n",
      "URI36H\n",
      "UNY1083P\n",
      "UCT30H\n",
      "URI42H\n",
      "UNY193P\n",
      "UCT29H\n",
      "UCT32H\n",
      "URI40H\n",
      "URI111H\n",
      "UCT113H\n",
      "URI44H\n",
      "UNY1085P\n",
      "UNY208P\n",
      "URI86H\n",
      "URI117H\n",
      "URI46H\n",
      "UCT92H\n",
      "UNY1038P\n",
      "UNY149P\n",
      "URI39H\n",
      "URI41H\n",
      "UCT31H\n",
      "UWI263P\n",
      "URI112H\n",
      "UWI247P\n",
      "UCT50H\n",
      "UCT109H\n",
      "UNY990P\n",
      "UWI283P\n",
      "UCT35H\n",
      "URI47H\n",
      "UCT110H\n",
      "UNY172P\n",
      "URI87H\n",
      "UNY169P\n",
      "UCT96H\n",
      "list: 268857\n",
      "actual: 268857\n"
     ]
    }
   ],
   "source": [
    "all_assembly_genes = []\n",
    "total_feats = 0\n",
    "for assembly in list_of_assemblies:\n",
    "    current_genes = []\n",
    "    assembly_id = os.path.basename(assembly).split('.')[0]\n",
    "    print(assembly_id)\n",
    "    current_genes = parse_genbank(assembly, assembly_id)\n",
    "    num_feats = len(current_genes)\n",
    "    #print(assembly_id)\n",
    "    #print(num_feats)\n",
    "    total_feats += num_feats\n",
    "    all_assembly_genes.append(current_genes)\n",
    "all_assembly_genes_flat = [feature for features in all_assembly_genes for feature in features]\n",
    "all_feats = len(all_assembly_genes_flat)\n",
    "print(f'list: {all_feats}')\n",
    "print(f'actual: {total_feats}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49115278",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "bakta_ref_genes = parse_genbank(bakta_reference_genbank, \"REF\")\n",
    "print(len(bakta_ref_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9099c9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "list: 58093\n",
      "actual: 58093\n"
     ]
    }
   ],
   "source": [
    "all_ref_ncbi_genes = []\n",
    "total_feats = 0\n",
    "for plasmid in list_of_ncbi_ref_genbanks:\n",
    "    current_genes = []\n",
    "    current_genes = parse_genbank(plasmid, \"REF\")\n",
    "    num_feats = len(current_genes)\n",
    "    total_feats += num_feats\n",
    "    all_ref_ncbi_genes.append(current_genes)\n",
    "all_ref_ncbi_genes_flat = [feature for features in all_ref_ncbi_genes for feature in features]\n",
    "all_feats = len(all_ref_ncbi_genes_flat)\n",
    "print(f'list: {all_feats}')\n",
    "print(f'actual: {total_feats}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "715fb19b-5e02-426e-bb3b-12232ed72cc2",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'dbxrefs'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[125], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m db \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnewgenes.db\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      2\u001b[0m create_table(db, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124massemblies\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m----> 3\u001b[0m \u001b[43mpopulate_table\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdb\u001b[49m\u001b[43m,\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43massemblies\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mall_assembly_genes_flat\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      5\u001b[0m create_table(db, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreference\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      6\u001b[0m populate_table(db,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreference\u001b[39m\u001b[38;5;124m\"\u001b[39m, bakta_ref_genes)\n",
      "Cell \u001b[0;32mIn[33], line 35\u001b[0m, in \u001b[0;36mpopulate_table\u001b[0;34m(db_name, table_name, genes)\u001b[0m\n\u001b[1;32m     28\u001b[0m cursor \u001b[38;5;241m=\u001b[39m conn\u001b[38;5;241m.\u001b[39mcursor()\n\u001b[1;32m     29\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m gene \u001b[38;5;129;01min\u001b[39;00m genes:\n\u001b[1;32m     30\u001b[0m     cursor\u001b[38;5;241m.\u001b[39mexecute(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'''\u001b[39m\u001b[38;5;124mINSERT INTO \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtable_name\u001b[38;5;132;01m}\u001b[39;00m\n\u001b[1;32m     31\u001b[0m \u001b[38;5;124m                      (feature_type, gene, locus_tag, note, protein_id, product, sequence, replicon, replicon_name, start, end, strand, dbxrefs, assembly, translation, inference, transl_table)\u001b[39m\n\u001b[1;32m     32\u001b[0m \u001b[38;5;124m                      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)\u001b[39m\u001b[38;5;124m'''\u001b[39m,\n\u001b[1;32m     33\u001b[0m                    (gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfeature_type\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgene\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlocus_tag\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnote\u001b[39m\u001b[38;5;124m\"\u001b[39m],\n\u001b[1;32m     34\u001b[0m                     gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprotein_id\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mproduct\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msequence\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreplicon\u001b[39m\u001b[38;5;124m\"\u001b[39m],\n\u001b[0;32m---> 35\u001b[0m                     gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreplicon_name\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstart\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mend\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstrand\u001b[39m\u001b[38;5;124m\"\u001b[39m], \u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdbxrefs\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m,\n\u001b[1;32m     36\u001b[0m                     gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124massembly\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtranslation\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minference\u001b[39m\u001b[38;5;124m\"\u001b[39m], gene[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtransl_table\u001b[39m\u001b[38;5;124m\"\u001b[39m]))\n\u001b[1;32m     37\u001b[0m conn\u001b[38;5;241m.\u001b[39mcommit()\n\u001b[1;32m     38\u001b[0m conn\u001b[38;5;241m.\u001b[39mclose()\n",
      "\u001b[0;31mKeyError\u001b[0m: 'dbxrefs'"
     ]
    }
   ],
   "source": [
    "db = \"newgenes.db\"\n",
    "create_table(db, \"assemblies\")\n",
    "populate_table(db,\"assemblies\", all_assembly_genes_flat)\n",
    "\n",
    "create_table(db, \"reference\")\n",
    "populate_table(db,\"reference\", bakta_ref_genes)\n",
    "\n",
    "create_table(db, \"ncbi_gb\")\n",
    "populate_table(db, \"ncbi_gb\", all_ref_ncbi_genes_flat)\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0b5582f",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_db = \"newannotations.db\"\n",
    "create_and_populate_database(all_assembly_genes_flat, new_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb86ff36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          name\n",
      "0  annotations\n"
     ]
    }
   ],
   "source": [
    "# List all tables\n",
    "print(get_tables(new_db))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c14b55ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    cid           name     type  notnull dflt_value  pk\n",
      "0     0   feature_type     TEXT        0       None   0\n",
      "1     1           gene     TEXT        0       None   0\n",
      "2     2      locus_tag     TEXT        0       None   0\n",
      "3     3           note     TEXT        0       None   0\n",
      "4     4     protein_id     TEXT        0       None   0\n",
      "5     5        product     TEXT        0       None   0\n",
      "6     6       sequence     TEXT        0       None   0\n",
      "7     7       replicon     TEXT        0       None   0\n",
      "8     8  replicon_name     TEXT        0       None   0\n",
      "9     9          start  INTEGER        0       None   0\n",
      "10   10            end  INTEGER        0       None   0\n",
      "11   11         strand  INTEGER        0       None   0\n",
      "12   12       assembly     TEXT        0       None   0\n",
      "13   13    translation     TEXT        0       None   0\n",
      "14   14      inference     TEXT        0       None   0\n",
      "15   15   transl_table     TEXT        0       None   0\n",
      "16   16     BlastRules     TEXT        0       None   0\n",
      "17   17             SO     TEXT        0       None   0\n",
      "18   18         UniRef     TEXT        0       None   0\n",
      "19   19             EC     TEXT        0       None   0\n",
      "20   20            COG     TEXT        0       None   0\n",
      "21   21           RFAM     TEXT        0       None   0\n",
      "22   22           KEGG     TEXT        0       None   0\n",
      "23   23             GO     TEXT        0       None   0\n",
      "24   24         RefSeq     TEXT        0       None   0\n",
      "25   25           PFAM     TEXT        0       None   0\n",
      "26   26        UniParc     TEXT        0       None   0\n",
      "  feature_type     gene     locus_tag     note              protein_id  \\\n",
      "0         gene  unknown  CKDGNL_00005  unknown                 unknown   \n",
      "1          CDS  unknown  CKDGNL_00005  unknown  gnl|Bakta|CKDGNL_00005   \n",
      "2         gene  unknown  CKDGNL_00010  unknown                 unknown   \n",
      "3          CDS  unknown  CKDGNL_00010  unknown  gnl|Bakta|CKDGNL_00010   \n",
      "4         gene  unknown  CKDGNL_00015  unknown                 unknown   \n",
      "5          CDS  unknown  CKDGNL_00015  unknown  gnl|Bakta|CKDGNL_00015   \n",
      "6         gene     yfcC  CKDGNL_00020  unknown                 unknown   \n",
      "7          CDS     yfcC  CKDGNL_00020  unknown  gnl|Bakta|CKDGNL_00020   \n",
      "8         gene     argF  CKDGNL_00025  unknown                 unknown   \n",
      "9          CDS     argF  CKDGNL_00025  unknown  gnl|Bakta|CKDGNL_00025   \n",
      "\n",
      "                                             product  \\\n",
      "0                                            unknown   \n",
      "1                  DUF5655 domain-containing protein   \n",
      "2                                            unknown   \n",
      "3  Type I restriction enzyme R N terminus family ...   \n",
      "4                                            unknown   \n",
      "5                             P12 family lipoprotein   \n",
      "6                                            unknown   \n",
      "7  putative membrane transporter YfcC, affects gl...   \n",
      "8                                            unknown   \n",
      "9                     ornithine carbamoyltransferase   \n",
      "\n",
      "                                            sequence  replicon replicon_name  \\\n",
      "0  ATGAAAGCAATATATAAAGAATTAGAAGATTTTATTTTTGTTTTAA...  contig_1    chromosome   \n",
      "1  ATGAAAGCAATATATAAAGAATTAGAAGATTTTATTTTTGTTTTAA...  contig_1    chromosome   \n",
      "2  ATGCCAAATGAATACAAAGATATAAATTTTGAAAATTCAATAAAAG...  contig_1    chromosome   \n",
      "3  ATGCCAAATGAATACAAAGATATAAATTTTGAAAATTCAATAAAAG...  contig_1    chromosome   \n",
      "4  ATGCTAATAAGTTTATTATCATGTAATACAAGTGACCCCAATGAAT...  contig_1    chromosome   \n",
      "5  ATGCTAATAAGTTTATTATCATGTAATACAAGTGACCCCAATGAAT...  contig_1    chromosome   \n",
      "6  ATGATCAAAATGCCAAGTAGTTTTACAATAATATTTTCTTTAATTG...  contig_1    chromosome   \n",
      "7  ATGATCAAAATGCCAAGTAGTTTTACAATAATATTTTCTTTAATTG...  contig_1    chromosome   \n",
      "8  ATGTATAATTTACGAAATAGGAGCTTTTTAAATCTTTTAGATTTTA...  contig_1    chromosome   \n",
      "9  ATGTATAATTTACGAAATAGGAGCTTTTTAAATCTTTTAGATTTTA...  contig_1    chromosome   \n",
      "\n",
      "   start  ...       SO                  UniRef       EC   COG  RFAM  KEGG  \\\n",
      "0   1061  ...     None                    None     None  None  None  None   \n",
      "1   1061  ...  0001217  UniRef50_UPI0022771FDB     None  None  None  None   \n",
      "2   1510  ...     None                    None     None  None  None  None   \n",
      "3   1510  ...  0001217         UniRef50_Q0SLH2     None  None  None  None   \n",
      "4   5775  ...     None                    None     None  None  None  None   \n",
      "5   5775  ...  0001217         UniRef90_Q6W6N9     None  None  None  None   \n",
      "6   7384  ...     None                    None     None  None  None  None   \n",
      "7   7384  ...  0001217     UniRef90_A0A1L6VUG6     None     R  None  None   \n",
      "8   8860  ...     None                    None     None  None  None  None   \n",
      "9   8860  ...  0001217         UniRef90_O51782  2.1.3.3     E  None  None   \n",
      "\n",
      "        GO          RefSeq  PFAM        UniParc  \n",
      "0     None            None  None           None  \n",
      "1     None            None  None           None  \n",
      "2     None            None  None           None  \n",
      "3     None            None  None           None  \n",
      "4     None            None  None           None  \n",
      "5     None            None  None           None  \n",
      "6     None            None  None           None  \n",
      "7     None  WP_002657298.1  None  UPI00016B2F52  \n",
      "8     None            None  None           None  \n",
      "9  0042450  WP_002657297.1  None  UPI00016B2DF0  \n",
      "\n",
      "[10 rows x 27 columns]\n"
     ]
    }
   ],
   "source": [
    "# Describe the structure of the 'genes' table\n",
    "print(get_columns(new_db, \"annotations\"))\n",
    "# Query some data from the 'genes' table\n",
    "print(table_head(new_db, \"annotations\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e7905de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    cid           name     type  notnull dflt_value  pk\n",
      "0     0             id  INTEGER        0       None   1\n",
      "1     1   feature_type     TEXT        0       None   0\n",
      "2     2           gene     TEXT        0       None   0\n",
      "3     3      locus_tag     TEXT        0       None   0\n",
      "4     4           note     TEXT        0       None   0\n",
      "5     5     protein_id     TEXT        0       None   0\n",
      "6     6        product     TEXT        0       None   0\n",
      "7     7       sequence     TEXT        0       None   0\n",
      "8     8       replicon     TEXT        0       None   0\n",
      "9     9  replicon_name     TEXT        0       None   0\n",
      "10   10          start  INTEGER        0       None   0\n",
      "11   11            end  INTEGER        0       None   0\n",
      "12   12         strand  INTEGER        0       None   0\n",
      "13   13        dbxrefs     TEXT        0       None   0\n",
      "14   14       assembly     TEXT        0       None   0\n",
      "15   15    translation     TEXT        0       None   0\n",
      "16   16      inference     TEXT        0       None   0\n",
      "17   17   transl_table      INT        0       None   0\n",
      "   id feature_type     gene     locus_tag     note              protein_id  \\\n",
      "0   1         gene  unknown  FMBKCJ_00005  unknown                 unknown   \n",
      "1   2          CDS  unknown  FMBKCJ_00005  unknown  gnl|Bakta|FMBKCJ_00005   \n",
      "2   3         gene  unknown  FMBKCJ_00010  unknown                 unknown   \n",
      "3   4          CDS  unknown  FMBKCJ_00010  unknown  gnl|Bakta|FMBKCJ_00010   \n",
      "4   5         gene  unknown  FMBKCJ_00015  unknown                 unknown   \n",
      "5   6          CDS  unknown  FMBKCJ_00015  unknown  gnl|Bakta|FMBKCJ_00015   \n",
      "6   7         gene     manB  FMBKCJ_00020  unknown                 unknown   \n",
      "7   8          CDS     manB  FMBKCJ_00020  unknown  gnl|Bakta|FMBKCJ_00020   \n",
      "8   9         gene     trpS  FMBKCJ_00025  unknown                 unknown   \n",
      "9  10          CDS     trpS  FMBKCJ_00025  unknown  gnl|Bakta|FMBKCJ_00025   \n",
      "\n",
      "                                             product  \\\n",
      "0                                            unknown   \n",
      "1                    Uncharacterized protein BB_0001   \n",
      "2                                            unknown   \n",
      "3  Glycoside hydrolase family 3 N-terminal domain...   \n",
      "4                                            unknown   \n",
      "5       UTP--glucose-1-phosphate uridylyltransferase   \n",
      "6                                            unknown   \n",
      "7                                 Phosphomannomutase   \n",
      "8                                            unknown   \n",
      "9                            tryptophan--tRNA ligase   \n",
      "\n",
      "                                            sequence    replicon  \\\n",
      "0  ATGAAGTATAGTGCTATTTTATTAATATGTAGCGTTAATTTATTTT...  CP124100.1   \n",
      "1  ATGAAGTATAGTGCTATTTTATTAATATGTAGCGTTAATTTATTTT...  CP124100.1   \n",
      "2  ATGGACTTTTTAAAAACCTTTTCTTTTTTGTTTTTTAGCTTTTTTT...  CP124100.1   \n",
      "3  ATGGACTTTTTAAAAACCTTTTCTTTTTTGTTTTTTAGCTTTTTTT...  CP124100.1   \n",
      "4  ATGAATGTTAAAGTTGATAAAATTTTTTCTGAAATGATACTTGAAA...  CP124100.1   \n",
      "5  ATGAATGTTAAAGTTGATAAAATTTTTTCTGAAATGATACTTGAAA...  CP124100.1   \n",
      "6  ATGCTTAAACAATATTCACTTAACATGAAAAATTTTAAAAAAGCTT...  CP124100.1   \n",
      "7  ATGCTTAAACAATATTCACTTAACATGAAAAATTTTAAAAAAGCTT...  CP124100.1   \n",
      "8  TTGAAAAGAAAGGTTATGCTTACGGGAGATAGGCCTACTGGCGCTC...  CP124100.1   \n",
      "9  TTGAAAAGAAAGGTTATGCTTACGGGAGATAGGCCTACTGGCGCTC...  CP124100.1   \n",
      "\n",
      "  replicon_name  start   end  strand  \\\n",
      "0    chromosome    112   676       1   \n",
      "1    chromosome    112   676       1   \n",
      "2    chromosome    766  1795      -1   \n",
      "3    chromosome    766  1795      -1   \n",
      "4    chromosome   1782  3249      -1   \n",
      "5    chromosome   1782  3249      -1   \n",
      "6    chromosome   3394  5188       1   \n",
      "7    chromosome   3394  5188       1   \n",
      "8    chromosome   5250  6306      -1   \n",
      "9    chromosome   5250  6306      -1   \n",
      "\n",
      "                                             dbxrefs assembly  \\\n",
      "0                                      No crossrefs?  NE_5261   \n",
      "1  RefSeq:WP_020948668.1;SO:0001217;UniParc:UPI00...  NE_5261   \n",
      "2                                      No crossrefs?  NE_5261   \n",
      "3  RefSeq:WP_002658389.1;SO:0001217;UniParc:UPI00...  NE_5261   \n",
      "4                                      No crossrefs?  NE_5261   \n",
      "5  RefSeq:WP_002658386.1;SO:0001217;UniParc:UPI00...  NE_5261   \n",
      "6                                      No crossrefs?  NE_5261   \n",
      "7  COG:COG1109;COG:G;RefSeq:WP_002658384.1;SO:000...  NE_5261   \n",
      "8                                      No crossrefs?  NE_5261   \n",
      "9  COG:COG0180;COG:J;EC:6.1.1.2;GO:0004830;GO:000...  NE_5261   \n",
      "\n",
      "                                         translation  \\\n",
      "0                                               ['']   \n",
      "1  ['MKYSAILLICSVNLFCFQNKLTTSRWEFPKEDLIKKKIKIGIIY...   \n",
      "2                                               ['']   \n",
      "3  ['MDFLKTFSFLFFSFFCLNLIAIESLPEIDYEYFNKDKSDLVDLI...   \n",
      "4                                               ['']   \n",
      "5  ['MNVKVDKIFSEMILEKLNSGEICPNNFESIKYFPCDSHENIFNI...   \n",
      "6                                               ['']   \n",
      "7  ['MLKQYSLNMKNFKKAFDEMIFSPSGFRKIFAKSKNEDSTENEIN...   \n",
      "8                                               ['']   \n",
      "9  ['MKRKVMLTGDRPTGALHLGHYVGSVVNRLKFQEEYETYFIIADL...   \n",
      "\n",
      "                           inference transl_table  \n",
      "0                                                  \n",
      "1  ab initio prediction:Prodigal:2.6           11  \n",
      "2                                                  \n",
      "3  ab initio prediction:Prodigal:2.6           11  \n",
      "4                                                  \n",
      "5  ab initio prediction:Prodigal:2.6           11  \n",
      "6                                                  \n",
      "7  ab initio prediction:Prodigal:2.6           11  \n",
      "8                                                  \n",
      "9  ab initio prediction:Prodigal:2.6           11  \n"
     ]
    }
   ],
   "source": [
    "# Describe the structure of the 'reference' table\n",
    "print(get_columns(db, \"reference\"))\n",
    "# Query some data from the 'reference' table\n",
    "print(table_head(db, \"reference\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9636f344",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    cid           name     type  notnull dflt_value  pk\n",
      "0     0             id  INTEGER        0       None   1\n",
      "1     1   feature_type     TEXT        0       None   0\n",
      "2     2           gene     TEXT        0       None   0\n",
      "3     3      locus_tag     TEXT        0       None   0\n",
      "4     4           note     TEXT        0       None   0\n",
      "5     5     protein_id     TEXT        0       None   0\n",
      "6     6        product     TEXT        0       None   0\n",
      "7     7       sequence     TEXT        0       None   0\n",
      "8     8       replicon     TEXT        0       None   0\n",
      "9     9  replicon_name     TEXT        0       None   0\n",
      "10   10          start  INTEGER        0       None   0\n",
      "11   11            end  INTEGER        0       None   0\n",
      "12   12         strand  INTEGER        0       None   0\n",
      "13   13        dbxrefs     TEXT        0       None   0\n",
      "14   14       assembly     TEXT        0       None   0\n",
      "15   15    translation     TEXT        0       None   0\n",
      "16   16      inference     TEXT        0       None   0\n",
      "17   17   transl_table      INT        0       None   0\n",
      "   id feature_type     gene locus_tag  \\\n",
      "0   1         gene  unknown    BB_U01   \n",
      "1   2          CDS  unknown    BB_U01   \n",
      "2   3         gene  unknown    BB_U02   \n",
      "3   4          CDS  unknown    BB_U02   \n",
      "4   5         gene  unknown    BB_U04   \n",
      "5   6          CDS  unknown    BB_U04   \n",
      "6   7         gene  unknown    BB_U05   \n",
      "7   8          CDS  unknown    BB_U05   \n",
      "8   9         gene  unknown    BB_U06   \n",
      "9  10          CDS  unknown    BB_U06   \n",
      "\n",
      "                                                note  protein_id  \\\n",
      "0                                            unknown     unknown   \n",
      "1                                            unknown  AAF07699.2   \n",
      "2                                            unknown     unknown   \n",
      "3                                            unknown  AAF07700.1   \n",
      "4                                            unknown     unknown   \n",
      "5  identified by match to protein family HMM PF02414  AAF07693.1   \n",
      "6                                            unknown     unknown   \n",
      "7  identified by match to protein family HMM PF01656  AAF07694.1   \n",
      "8                                            unknown     unknown   \n",
      "9  identified by match to protein family HMM PF01672  AAF07695.2   \n",
      "\n",
      "                              product  \\\n",
      "0                             unknown   \n",
      "1      conserved hypothetical protein   \n",
      "2                             unknown   \n",
      "3      conserved hypothetical protein   \n",
      "4                             unknown   \n",
      "5                               BBC01   \n",
      "6                             unknown   \n",
      "7                       PF-32 protein   \n",
      "8                             unknown   \n",
      "9  putative plasmid partition protein   \n",
      "\n",
      "                                            sequence    replicon  \\\n",
      "0  ATGTATAAGTCTGTAAAAGAACAACAAGAAAAAGGAATAGATCATA...  AE001582.2   \n",
      "1  ATGTATAAGTCTGTAAAAGAACAACAAGAAAAAGGAATAGATCATA...  AE001582.2   \n",
      "2  ATGGATGGAGTAATTAACGATACATTGGTCGCAAGAATGAAAAAGC...  AE001582.2   \n",
      "3  ATGGATGGAGTAATTAACGATACATTGGTCGCAAGAATGAAAAAGC...  AE001582.2   \n",
      "4  TTGCTTATTTCATTCGTTAAGGAGAGTTTGCCAATGCATAAAATCA...  AE001582.2   \n",
      "5  TTGCTTATTTCATTCGTTAAGGAGAGTTTGCCAATGCATAAAATCA...  AE001582.2   \n",
      "6  ATGAATAAAAAACCAAATATTTTAACAATAGCAAGTTTGAAAGGCG...  AE001582.2   \n",
      "7  ATGAATAAAAAACCAAATATTTTAACAATAGCAAGTTTGAAAGGCG...  AE001582.2   \n",
      "8  ATGAGCAAAAAAAATAAAAAAAAATTATTTCTTAATGATAGGATTG...  AE001582.2   \n",
      "9  ATGAGCAAAAAAAATAAAAAAAAATTATTTCTTAATGATAGGATTG...  AE001582.2   \n",
      "\n",
      "  replicon_name  start    end  strand        dbxrefs assembly  \\\n",
      "0          lp21    252    618       1  No crossrefs?      B31   \n",
      "1          lp21    252    618       1  No crossrefs?      B31   \n",
      "2          lp21    745   1114       1  No crossrefs?      B31   \n",
      "3          lp21    745   1114       1  No crossrefs?      B31   \n",
      "4          lp21   1485   2628       1  No crossrefs?      B31   \n",
      "5          lp21   1485   2628       1  No crossrefs?      B31   \n",
      "6          lp21   2867   3656       1  No crossrefs?      B31   \n",
      "7          lp21   2867   3656       1  No crossrefs?      B31   \n",
      "8          lp21  14647  15235       1  No crossrefs?      B31   \n",
      "9          lp21  14647  15235       1  No crossrefs?      B31   \n",
      "\n",
      "                                         translation inference transl_table  \n",
      "0                                               ['']                         \n",
      "1  ['MYKSVKEQQEKGIDHTCRILILTETIFEINLILENYSQKTLLKK...                     11  \n",
      "2                                               ['']                         \n",
      "3  ['MDGVINDTLVARMKKQIKFNKNKLIILVKTLDHMNKELLYSANK...                     11  \n",
      "4                                               ['']                         \n",
      "5  ['MLISFVKESLPMHKIKTTNNNPRNCYNKVQYKLIVLISTICYLN...                     11  \n",
      "6                                               ['']                         \n",
      "7  ['MNKKPNILTIASLKGGVGKSTLTILFSYLLKDLGKKILIVDMDP...                     11  \n",
      "8                                               ['']                         \n",
      "9  ['MSKKNKKKLFLNDRIEHNIIDNLSNLDTNLDKNLMVYKDLKEQL...                     11  \n"
     ]
    }
   ],
   "source": [
    "# Describe the structure of the 'reference' table\n",
    "print(get_columns(db, \"ncbi_gb\"))\n",
    "# Query some data from the 'reference' table\n",
    "print(table_head(db, \"ncbi_gb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81afc322",
   "metadata": {},
   "outputs": [],
   "source": [
    "specific_genes = ['ospC', 'cspZ', 'cspA']\n",
    "#all_cspA_df = get_all_of_gene(db, \"assemblies\", \"cspA\") # IT IS NOT ANNOTATED AS CSPA SO THANKS REFSEQ!\n",
    "all_cspa_df = get_all_of_product(db, \"assemblies\", \"complement regulator-acquiring protein\")\n",
    "all_cspZ_df = get_all_of_gene(db, \"assemblies\", \"cspZ\")\n",
    "all_ospC_df = get_all_of_gene(db, \"assemblies\", \"ospC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcb5fa13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [id, feature_type, gene, locus_tag, note, protein_id, product, sequence, replicon, replicon_name, start, end, strand, dbxrefs, assembly, translation, inference, transl_table]\n",
      "Index: []\n",
      "         id feature_type  gene     locus_tag     note              protein_id  \\\n",
      "0     12699          CDS  cspZ  DNBOAP_05820  unknown  gnl|Bakta|DNBOAP_05820   \n",
      "1     17993          CDS  cspZ  NCOCKI_05785  unknown  gnl|Bakta|NCOCKI_05785   \n",
      "2     20630          CDS  cspZ  DMHAMM_05695  unknown  gnl|Bakta|DMHAMM_05695   \n",
      "3     25598          CDS  cspZ  EMEHFM_04770  unknown  gnl|Bakta|EMEHFM_04770   \n",
      "4     36062          CDS  cspZ  BOHBDJ_05645  unknown  gnl|Bakta|BOHBDJ_05645   \n",
      "..      ...          ...   ...           ...      ...                     ...   \n",
      "124  829355          CDS  cspZ  FMOGGM_06175  unknown  gnl|Bakta|FMOGGM_06175   \n",
      "125  834861          CDS  cspZ  PCFIDF_05865  unknown  gnl|Bakta|PCFIDF_05865   \n",
      "126  840673          CDS  cspZ  AIHKLI_07085  unknown  gnl|Bakta|AIHKLI_07085   \n",
      "127  859966          CDS  cspZ  JLNNBJ_06910  unknown  gnl|Bakta|JLNNBJ_06910   \n",
      "128  865864          CDS  cspZ  HILFBG_07420  unknown  gnl|Bakta|HILFBG_07420   \n",
      "\n",
      "                           product  \\\n",
      "0    factor H-binding protein CspZ   \n",
      "1    factor H-binding protein CspZ   \n",
      "2    factor H-binding protein CspZ   \n",
      "3    factor H-binding protein CspZ   \n",
      "4    factor H-binding protein CspZ   \n",
      "..                             ...   \n",
      "124  factor H-binding protein CspZ   \n",
      "125  factor H-binding protein CspZ   \n",
      "126  factor H-binding protein CspZ   \n",
      "127  factor H-binding protein CspZ   \n",
      "128  factor H-binding protein CspZ   \n",
      "\n",
      "                                              sequence    replicon  \\\n",
      "0    ATGAAAAAAAGTTTTTTATCAATATACATGTTAATTTCAATAAGTT...   contig_18   \n",
      "1    ATGAAAAAAAGTTTTTTATCAATATACATGTTAATTTCAATAAGTT...   contig_18   \n",
      "2    ATGAAAAAAAGTTTTTTATCAATATACATGTTAATTTCAATAAGTT...   contig_17   \n",
      "3    ATGAAAAAAAGTTTTTTATCAATATACATGTTAATTTCAATAAGTT...  contig_118   \n",
      "4    ATGAAAAAAAGTTTTTTATCAATATACATGTTAATTTCAATAAGTT...   contig_17   \n",
      "..                                                 ...         ...   \n",
      "124  ATGAAAAAAAGTTTTTTATCAATATACATGTTAATTTCAATAAGTT...   contig_14   \n",
      "125  ATGAAAAAAAGTTTTTTATCAATATACATGTTAATTTCAATAAGTT...   contig_12   \n",
      "126  ATGAAAAAAAGTTTTTTATCAATATACATGTTAATTTCAATAAGTT...   contig_15   \n",
      "127  ATGAAAAAAAGTTTTTTATCAATATACATGTTAATTTCAATAAGTT...   contig_15   \n",
      "128  ATGAAAAAAAGTTTTTTATCAATATACATGTTAATTTCAATAAGTT...   contig_22   \n",
      "\n",
      "    replicon_name  start    end  strand  \\\n",
      "0     placeholder   7120   7831       1   \n",
      "1     placeholder    703   1414      -1   \n",
      "2     placeholder   1157   1868      -1   \n",
      "3     placeholder   2271   2982      -1   \n",
      "4     placeholder   7120   7831       1   \n",
      "..            ...    ...    ...     ...   \n",
      "124   placeholder   2265   2976      -1   \n",
      "125   placeholder  25614  26334       1   \n",
      "126   placeholder    626   1337      -1   \n",
      "127   placeholder    626   1337      -1   \n",
      "128   placeholder   3123   3834       1   \n",
      "\n",
      "                                               dbxrefs assembly  \\\n",
      "0    BlastRules:WP_010890313;RefSeq:WP_010890313.1;...    UCT29   \n",
      "1    BlastRules:WP_010890313;RefSeq:WP_010890313.1;...    URI87   \n",
      "2    BlastRules:WP_010890313;RefSeq:WP_010890313.1;...    URI91   \n",
      "3    BlastRules:WP_010890313;RefSeq:WP_014540433.1;...   UWI263   \n",
      "4    BlastRules:WP_010890313;RefSeq:WP_010890313.1;...   URI102   \n",
      "..                                                 ...      ...   \n",
      "124  BlastRules:WP_010890313;RefSeq:WP_014540433.1;...  UWI263P   \n",
      "125  BlastRules:WP_010890313;RefSeq:WP_012672257.1;...  UWI247P   \n",
      "126  BlastRules:WP_010890313;RefSeq:WP_010890313.1;...  UCT109H   \n",
      "127  BlastRules:WP_010890313;RefSeq:WP_010890313.1;...   URI87H   \n",
      "128  BlastRules:WP_010890313;RefSeq:WP_010890313.1;...   UCT96H   \n",
      "\n",
      "                                           translation  \\\n",
      "0    ['MKKSFLSIYMLISISLLSCDVSRLNQRNINELKIFVEKAKYYSI...   \n",
      "1    ['MKKSFLSIYMLISISLLSCDVSRLNQRNINELKIFVEKAKYYSI...   \n",
      "2    ['MKKSFLSIYMLISISLLSCDVSRLNQRNINELKIFVEKAKYYSI...   \n",
      "3    ['MKKSFLSIYMLISISLLSCDVSRLNQRNINELKIFVEKAKYYSI...   \n",
      "4    ['MKKSFLSIYMLISISLLSCDVSRLNQRNINELKIFVEKAKYYSI...   \n",
      "..                                                 ...   \n",
      "124  ['MKKSFLSIYMLISISLLSCDVSRLNQRNINELKIFVEKAKYYSI...   \n",
      "125  ['MKKSFLSIYMLISISLLSCDVSRLNQRNIDELKIFVEKAKYYSI...   \n",
      "126  ['MKKSFLSIYMLISISLLSCDVSRLNQRNINELKIFVEKAKYYSI...   \n",
      "127  ['MKKSFLSIYMLISISLLSCDVSRLNQRNINELKIFVEKAKYYSI...   \n",
      "128  ['MKKSFLSIYMLISISLLSCDVSRLNQRNINELKIFVEKAKYYSI...   \n",
      "\n",
      "                             inference  transl_table  \n",
      "0    ab initio prediction:Prodigal:2.6            11  \n",
      "1    ab initio prediction:Prodigal:2.6            11  \n",
      "2    ab initio prediction:Prodigal:2.6            11  \n",
      "3    ab initio prediction:Prodigal:2.6            11  \n",
      "4    ab initio prediction:Prodigal:2.6            11  \n",
      "..                                 ...           ...  \n",
      "124  ab initio prediction:Prodigal:2.6            11  \n",
      "125  ab initio prediction:Prodigal:2.6            11  \n",
      "126  ab initio prediction:Prodigal:2.6            11  \n",
      "127  ab initio prediction:Prodigal:2.6            11  \n",
      "128  ab initio prediction:Prodigal:2.6            11  \n",
      "\n",
      "[129 rows x 18 columns]\n",
      "         id feature_type  gene     locus_tag     note              protein_id  \\\n",
      "0      4512          CDS  ospC  JIDBHP_04835  unknown  gnl|Bakta|JIDBHP_04835   \n",
      "1      7139          CDS  ospC  APIDPD_04985  unknown  gnl|Bakta|APIDPD_04985   \n",
      "2      9764          CDS  ospC  GEBMJB_04900  unknown  gnl|Bakta|GEBMJB_04900   \n",
      "3     12451          CDS  ospC  DNBOAP_05200  unknown  gnl|Bakta|DNBOAP_05200   \n",
      "4     17749          CDS  ospC  NCOCKI_05175  unknown  gnl|Bakta|NCOCKI_05175   \n",
      "..      ...          ...   ...           ...      ...                     ...   \n",
      "205  840735          CDS  ospC  AIHKLI_07240  unknown  gnl|Bakta|AIHKLI_07240   \n",
      "206  851467          CDS  ospC  KGIAJM_06830  unknown  gnl|Bakta|KGIAJM_06830   \n",
      "207  854100          CDS  ospC  LNEFEP_05835  unknown  gnl|Bakta|LNEFEP_05835   \n",
      "208  860068          CDS  ospC  JLNNBJ_07165  unknown  gnl|Bakta|JLNNBJ_07165   \n",
      "209  865654          CDS  ospC  HILFBG_06895  unknown  gnl|Bakta|HILFBG_06895   \n",
      "\n",
      "                        product  \\\n",
      "0                          OspC   \n",
      "1                          OspC   \n",
      "2                          OspC   \n",
      "3    outer surface protein OspC   \n",
      "4    outer surface protein OspC   \n",
      "..                          ...   \n",
      "205  outer surface protein OspC   \n",
      "206                        OspC   \n",
      "207                        OspC   \n",
      "208  outer surface protein OspC   \n",
      "209     Outer surface protein C   \n",
      "\n",
      "                                              sequence   replicon  \\\n",
      "0    ATGAAAAAGAATACATTAAGTGCAATATTAATGACTTTATTTTTAT...   contig_6   \n",
      "1    ATGAAAAAGAATACATTAAGTGCAATATTAATGACTTTATTTTTAT...   contig_7   \n",
      "2    ATGAAAAAGAATACATTAAGTGCAATATTAATGACTTTATTTTTAT...   contig_7   \n",
      "3    ATGAAAAAGAATACATTAAGTGCAATATTAATGACTTTATTTTTAT...   contig_9   \n",
      "4    ATGAAAAAGAATACATTAAGTGCAATATTAATGACTTTATTTTTAT...   contig_9   \n",
      "..                                                 ...        ...   \n",
      "205  ATGAAAAAGAATACATTAAGTGCAATATTAATGACTTTATTTTTAT...  contig_16   \n",
      "206  ATGAAAAAGAATACATTAAGTGCAATATTAATGACTTTATTTTTAT...  contig_13   \n",
      "207  ATGAAAAAGAATACATTAAGTGCAATATTAATGACTTTATTTTTAT...  contig_10   \n",
      "208  ATGAAAAAGAATACATTAAGTGCAATATTAATGACTTTATTTTTAT...  contig_17   \n",
      "209  ATGAAAAAGAATACATTAAGTGCAATATTAATGACTTTATTTTTAT...  contig_15   \n",
      "\n",
      "    replicon_name  start    end  strand  \\\n",
      "0     placeholder  25883  26519      -1   \n",
      "1     placeholder     72    708       1   \n",
      "2     placeholder    942   1578       1   \n",
      "3     placeholder  16993  17626      -1   \n",
      "4     placeholder  22610  23243       1   \n",
      "..            ...    ...    ...     ...   \n",
      "205   placeholder  15557  16190       1   \n",
      "206   placeholder  16064  16700      -1   \n",
      "207   placeholder   7577   8213       1   \n",
      "208   placeholder   6341   6974       1   \n",
      "209   placeholder  13436  14072       1   \n",
      "\n",
      "                                               dbxrefs assembly  \\\n",
      "0    RefSeq:WP_015056312.1;SO:0001217;UniParc:UPI00...   URI112   \n",
      "1    RefSeq:WP_015056312.1;SO:0001217;UniParc:UPI00...   UCT110   \n",
      "2    RefSeq:WP_015056312.1;SO:0001217;UniParc:UPI00...    URI86   \n",
      "3    BlastRules:WP_010890595;RefSeq:WP_010890595.1;...    UCT29   \n",
      "4    BlastRules:WP_010890595;RefSeq:WP_010890595.1;...    URI87   \n",
      "..                                                 ...      ...   \n",
      "205  BlastRules:WP_010890595;RefSeq:WP_010890595.1;...  UCT109H   \n",
      "206  RefSeq:WP_015056312.1;SO:0001217;UniParc:UPI00...   URI47H   \n",
      "207  RefSeq:WP_015056312.1;SO:0001217;UniParc:UPI00...  UCT110H   \n",
      "208  BlastRules:WP_010890595;RefSeq:WP_010890595.1;...   URI87H   \n",
      "209  GO:0009279;GO:0009986;SO:0001217;UniRef:UniRef...   UCT96H   \n",
      "\n",
      "                                           translation  \\\n",
      "0    ['MKKNTLSAILMTLFLFISCNNSGKDGNTSANSADESVKGPNLTE...   \n",
      "1    ['MKKNTLSAILMTLFLFISCNNSGKDGNTSANSADESVKGPNLTE...   \n",
      "2    ['MKKNTLSAILMTLFLFISCNNSGKDGNTSANSADESVKGPNLTE...   \n",
      "3    ['MKKNTLSAILMTLFLFISCNNSGKDGNTSANSADESVKGPNLTE...   \n",
      "4    ['MKKNTLSAILMTLFLFISCNNSGKDGNTSANSADESVKGPNLTE...   \n",
      "..                                                 ...   \n",
      "205  ['MKKNTLSAILMTLFLFISCNNSGKDGNTSANSADESVKGPNLTE...   \n",
      "206  ['MKKNTLSAILMTLFLFISCNNSGKDGNTSANSADESVKGPNLTE...   \n",
      "207  ['MKKNTLSAILMTLFLFISCNNSGKDGNTSANSADESVKGPNLTE...   \n",
      "208  ['MKKNTLSAILMTLFLFISCNNSGKDGNTSANSADESVKGPNLTE...   \n",
      "209  ['MKKNTLSAILMTLFLFISCNNSGKDGNTSANSADESVKGPNLTE...   \n",
      "\n",
      "                             inference  transl_table  \n",
      "0    ab initio prediction:Prodigal:2.6            11  \n",
      "1    ab initio prediction:Prodigal:2.6            11  \n",
      "2    ab initio prediction:Prodigal:2.6            11  \n",
      "3    ab initio prediction:Prodigal:2.6            11  \n",
      "4    ab initio prediction:Prodigal:2.6            11  \n",
      "..                                 ...           ...  \n",
      "205  ab initio prediction:Prodigal:2.6            11  \n",
      "206  ab initio prediction:Prodigal:2.6            11  \n",
      "207  ab initio prediction:Prodigal:2.6            11  \n",
      "208  ab initio prediction:Prodigal:2.6            11  \n",
      "209  ab initio prediction:Prodigal:2.6            11  \n",
      "\n",
      "[210 rows x 18 columns]\n"
     ]
    }
   ],
   "source": [
    "print(all_cspA_df)\n",
    "print(all_cspZ_df)\n",
    "print(all_ospC_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "387434cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        id feature_type  gene    locus_tag     note  protein_id  \\\n",
      "0     2938          CDS  cspA   BbuN40_A68  unknown  ACS94810.1   \n",
      "1     7979          CDS  cspA  Bbu156a_A60  unknown  ACL33808.1   \n",
      "2    34597          CDS  cspA       BB_A68  unknown  AAC66286.1   \n",
      "3    36640          CDS  cspA   BbuZS7_A59  unknown  ACK74237.1   \n",
      "4    61031          CDS  cspA   BbuN40_A68  unknown  ACS94810.1   \n",
      "5    66072          CDS  cspA  Bbu156a_A60  unknown  ACL33808.1   \n",
      "6    92690          CDS  cspA       BB_A68  unknown  AAC66286.1   \n",
      "7    94733          CDS  cspA   BbuZS7_A59  unknown  ACK74237.1   \n",
      "8   119124          CDS  cspA   BbuN40_A68  unknown  ACS94810.1   \n",
      "9   124165          CDS  cspA  Bbu156a_A60  unknown  ACL33808.1   \n",
      "10  150783          CDS  cspA       BB_A68  unknown  AAC66286.1   \n",
      "11  152826          CDS  cspA   BbuZS7_A59  unknown  ACK74237.1   \n",
      "\n",
      "                                              product  \\\n",
      "0   complement regulator acquiring protein 1; puta...   \n",
      "1   complement regulator-acquiring surface protein...   \n",
      "2   complement regulator-acquiring surface protein...   \n",
      "3   complement regulator-acquiring surface protein...   \n",
      "4   complement regulator acquiring protein 1; puta...   \n",
      "5   complement regulator-acquiring surface protein...   \n",
      "6   complement regulator-acquiring surface protein...   \n",
      "7   complement regulator-acquiring surface protein...   \n",
      "8   complement regulator acquiring protein 1; puta...   \n",
      "9   complement regulator-acquiring surface protein...   \n",
      "10  complement regulator-acquiring surface protein...   \n",
      "11  complement regulator-acquiring surface protein...   \n",
      "\n",
      "                                             sequence    replicon  \\\n",
      "0   TTGAAAAAAGAAAAACTAAATAGAATCAAGATTAATACTATTGCAA...  CP001651.1   \n",
      "1   TTGAAAAAAGCCAAACTAAATATAATCAAGATTAATATTATTGCAA...  CP001257.1   \n",
      "2   TTGAAAAAAGCCAAACTAAATATAATCAAGATTAATATTATTGCAA...  AE000790.2   \n",
      "3   TTGACAAAAGCCAAACTAAATATAATCAAGCTTAATATTATTGCAA...  CP001199.1   \n",
      "4   TTGAAAAAAGAAAAACTAAATAGAATCAAGATTAATACTATTGCAA...  CP001651.1   \n",
      "5   TTGAAAAAAGCCAAACTAAATATAATCAAGATTAATATTATTGCAA...  CP001257.1   \n",
      "6   TTGAAAAAAGCCAAACTAAATATAATCAAGATTAATATTATTGCAA...  AE000790.2   \n",
      "7   TTGACAAAAGCCAAACTAAATATAATCAAGCTTAATATTATTGCAA...  CP001199.1   \n",
      "8   TTGAAAAAAGAAAAACTAAATAGAATCAAGATTAATACTATTGCAA...  CP001651.1   \n",
      "9   TTGAAAAAAGCCAAACTAAATATAATCAAGATTAATATTATTGCAA...  CP001257.1   \n",
      "10  TTGAAAAAAGCCAAACTAAATATAATCAAGATTAATATTATTGCAA...  AE000790.2   \n",
      "11  TTGACAAAAGCCAAACTAAATATAATCAAGCTTAATATTATTGCAA...  CP001199.1   \n",
      "\n",
      "   replicon_name  start    end  strand        dbxrefs assembly  \\\n",
      "0           lp54  47549  48305      -1  No crossrefs?      N40   \n",
      "1           lp54  47650  48406      -1  No crossrefs?     156a   \n",
      "2           lp54  46472  47228      -1  No crossrefs?      B31   \n",
      "3           lp54  46449  47205      -1  No crossrefs?      ZS7   \n",
      "4           lp54  47549  48305      -1  No crossrefs?      N40   \n",
      "5           lp54  47650  48406      -1  No crossrefs?     156a   \n",
      "6           lp54  46472  47228      -1  No crossrefs?      B31   \n",
      "7           lp54  46449  47205      -1  No crossrefs?      ZS7   \n",
      "8           lp54  47549  48305      -1  No crossrefs?      N40   \n",
      "9           lp54  47650  48406      -1  No crossrefs?     156a   \n",
      "10          lp54  46472  47228      -1  No crossrefs?      B31   \n",
      "11          lp54  46449  47205      -1  No crossrefs?      ZS7   \n",
      "\n",
      "                                          translation inference  transl_table  \n",
      "0   ['MKKEKLNRIKINTIAMILTLICTSCAPFSKIDPKANANTKPKKI...                      11  \n",
      "1   ['MKKAKLNIIKINIIAMILTLICTSCAPFSKIDPKANANTKPKKI...                      11  \n",
      "2   ['MKKAKLNIIKINIIAMILTLICTSCAPFSKIDPKANANTKPKKI...                      11  \n",
      "3   ['MTKAKLNIIKLNIIAMILTLICTSCAPFSKIDPKANANTKPKKI...                      11  \n",
      "4   ['MKKEKLNRIKINTIAMILTLICTSCAPFSKIDPKANANTKPKKI...                      11  \n",
      "5   ['MKKAKLNIIKINIIAMILTLICTSCAPFSKIDPKANANTKPKKI...                      11  \n",
      "6   ['MKKAKLNIIKINIIAMILTLICTSCAPFSKIDPKANANTKPKKI...                      11  \n",
      "7   ['MTKAKLNIIKLNIIAMILTLICTSCAPFSKIDPKANANTKPKKI...                      11  \n",
      "8   ['MKKEKLNRIKINTIAMILTLICTSCAPFSKIDPKANANTKPKKI...                      11  \n",
      "9   ['MKKAKLNIIKINIIAMILTLICTSCAPFSKIDPKANANTKPKKI...                      11  \n",
      "10  ['MKKAKLNIIKINIIAMILTLICTSCAPFSKIDPKANANTKPKKI...                      11  \n",
      "11  ['MTKAKLNIIKLNIIAMILTLICTSCAPFSKIDPKANANTKPKKI...                      11  \n"
     ]
    }
   ],
   "source": [
    "conn = sqlite3.connect(db)\n",
    "all_genes = pandas.read_sql_query(f\"SELECT * FROM ncbi_gb WHERE feature_type = 'CDS' AND UPPER(gene) LIKE 'cspA'\", conn)\n",
    "conn.close()\n",
    "\n",
    "print(all_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d872f73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       id feature_type     gene     locus_tag     note  \\\n",
      "0    1716         gene  unknown  CKDGNL_04300  unknown   \n",
      "1    1717          CDS  unknown  CKDGNL_04300  unknown   \n",
      "2  290342         gene  unknown  CKDGNL_04300  unknown   \n",
      "3  290343          CDS  unknown  CKDGNL_04300  unknown   \n",
      "4  578968         gene  unknown  CKDGNL_04300  unknown   \n",
      "5  578969          CDS  unknown  CKDGNL_04300  unknown   \n",
      "\n",
      "               protein_id                                 product  \\\n",
      "0                 unknown                                 unknown   \n",
      "1  gnl|Bakta|CKDGNL_04300  complement regulator-acquiring protein   \n",
      "2                 unknown                                 unknown   \n",
      "3  gnl|Bakta|CKDGNL_04300  complement regulator-acquiring protein   \n",
      "4                 unknown                                 unknown   \n",
      "5  gnl|Bakta|CKDGNL_04300  complement regulator-acquiring protein   \n",
      "\n",
      "                                            sequence  replicon replicon_name  \\\n",
      "0  TTGAAAAATAATAAATTAATTGCAATATTTTTATTGCATGTATTAA...  contig_4   placeholder   \n",
      "1  TTGAAAAATAATAAATTAATTGCAATATTTTTATTGCATGTATTAA...  contig_4   placeholder   \n",
      "2  TTGAAAAATAATAAATTAATTGCAATATTTTTATTGCATGTATTAA...  contig_4   placeholder   \n",
      "3  TTGAAAAATAATAAATTAATTGCAATATTTTTATTGCATGTATTAA...  contig_4   placeholder   \n",
      "4  TTGAAAAATAATAAATTAATTGCAATATTTTTATTGCATGTATTAA...  contig_4   placeholder   \n",
      "5  TTGAAAAATAATAAATTAATTGCAATATTTTTATTGCATGTATTAA...  contig_4   placeholder   \n",
      "\n",
      "   start    end  strand                                            dbxrefs  \\\n",
      "0  10250  11159       1                                      No crossrefs?   \n",
      "1  10250  11159       1  RefSeq:WP_012614997.1;SO:0001217;UniParc:UPI00...   \n",
      "2  10250  11159       1                                      No crossrefs?   \n",
      "3  10250  11159       1  RefSeq:WP_012614997.1;SO:0001217;UniParc:UPI00...   \n",
      "4  10250  11159       1                                      No crossrefs?   \n",
      "5  10250  11159       1  RefSeq:WP_012614997.1;SO:0001217;UniParc:UPI00...   \n",
      "\n",
      "  assembly                                        translation  \\\n",
      "0    ESI26                                               ['']   \n",
      "1    ESI26  ['MKNNKLIAIFLLHVLTVLILISCSLEVKDSNESKKHKKEKRKGK...   \n",
      "2    ESI26                                               ['']   \n",
      "3    ESI26  ['MKNNKLIAIFLLHVLTVLILISCSLEVKDSNESKKHKKEKRKGK...   \n",
      "4    ESI26                                               ['']   \n",
      "5    ESI26  ['MKNNKLIAIFLLHVLTVLILISCSLEVKDSNESKKHKKEKRKGK...   \n",
      "\n",
      "                           inference transl_table  \n",
      "0                                                  \n",
      "1  ab initio prediction:Prodigal:2.6           11  \n",
      "2                                                  \n",
      "3  ab initio prediction:Prodigal:2.6           11  \n",
      "4                                                  \n",
      "5  ab initio prediction:Prodigal:2.6           11  \n"
     ]
    }
   ],
   "source": [
    "print(get_gene_from_baktaID(db, 'assemblies', 'CKDGNL_04300'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "ea6cdf8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SQL should never be written more than once and here is my function.\n",
    "def search_terms_in_db(connection, table, term, columns):\n",
    "    # Construct the SQL query\n",
    "    query = f\"\"\"\n",
    "    SELECT *\n",
    "    FROM {table}\n",
    "    WHERE {' OR '.join([f\"{col} = ?\" for col in columns])}\n",
    "    \"\"\"\n",
    "    # try to find the term in each column.\n",
    "    try:\n",
    "        result_df = pandas.read_sql_query(query, connection, params=[term] * len(columns))\n",
    "        #print(f\"Query for term '{term}' succeeded.\")\n",
    "        return result_df # return our dataframe!\n",
    "    except Exception as e:\n",
    "        print(f\"Query for term '{term}' failed with error: {e}\")\n",
    "        return pandas.DataFrame() # Empty on error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "41e80144",
   "metadata": {},
   "outputs": [],
   "source": [
    "pangenome_df  = pandas.read_csv('/Users/mf019/bioinformatics/longread_pangenome/longread_analysis/v4/longread_paired_v4/gene_presence_absence.Rtab', delimiter='\\t')\n",
    "clustered_proteins_file = '/Users/mf019/bioinformatics/longread_pangenome/longread_analysis/v4/longread_paired_v4/clustered_proteins'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "c07e4fca",
   "metadata": {},
   "outputs": [],
   "source": [
    "pangene_groups = pangenome_df['Gene'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "5ea33d72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "488bb871b35e4defbbe4949afc76f9f2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Searching for those groups!:   0%|          | 0/3921 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['group_1108', 'group_1198', 'group_1199', 'group_1206', 'group_1208', 'group_1217', 'group_1228', 'group_1253', 'group_1299', 'group_1382', 'group_1387', 'group_1390', 'group_1391', 'group_1392', 'group_1406', 'group_1407', 'group_1408', 'group_1415', 'group_1418', 'group_1434', 'group_1438', 'group_1439', 'group_1471', 'group_1472', 'group_1482', 'group_2521', 'group_2532', 'group_2541', 'group_2544', 'group_2555', 'group_2556', 'group_2557', 'group_2558', 'group_2596', 'group_2598', 'group_2602', 'group_2667', 'group_2668', 'group_2693', 'group_2701', 'group_2711', 'group_2773', 'group_2778', 'group_2782', 'group_2784', 'group_2810', 'group_2811', 'group_2854', 'group_2894', 'group_2909', 'group_2911', 'group_2935', 'group_2952', 'group_2968', 'group_2990', 'group_2992', 'group_3022', 'group_3075', 'group_3099', 'group_3110', 'group_3116', 'group_3169', 'group_3172', 'group_3179', 'group_3181', 'group_3222', 'group_3224', 'group_3253', 'group_3263', 'group_3265', 'group_3281', 'group_3289', 'group_3319', 'group_3340', 'group_3349', 'group_3383', 'group_3386', 'group_3387', 'group_3395', 'group_3398', 'group_3403', 'group_3408', 'group_3413', 'group_3416', 'group_3447', 'group_3451', 'group_3453', 'group_3454', 'group_3479', 'group_3481', 'group_3484', 'group_3491', 'group_3492', 'group_3494', 'group_3536', 'group_3552', 'group_3558', 'group_3559', 'group_3562', 'group_3568', 'group_3569', 'group_3574', 'group_3576', 'group_3581', 'group_3584', 'group_3598', 'group_3599', 'group_3627', 'group_3628', 'group_3654', 'group_3655', 'group_3658', 'group_3660', 'group_3676', 'group_3679', 'group_3686', 'group_3708', 'group_3725', 'group_3732', 'group_3753', 'group_3806', 'group_3834', 'group_3839', 'group_3841', 'group_3878', 'group_3888', 'group_3891', 'group_3897', 'group_3907', 'group_3911', 'group_981', 'group_1001', 'group_1111', 'group_1249', 'group_1250', 'group_1274', 'group_1298', 'group_1465', 'group_1497', 'group_1499', 'group_1500', 'group_1606', 'group_1793', 'group_1794', 'group_1795', 'group_1796', 'group_1798', 'group_1800', 'group_1801', 'group_1824', 'group_1845', 'group_1846', 'group_1847', 'group_1851', 'group_1854', 'group_1856', 'group_1857', 'group_1858', 'group_1860', 'group_1861', 'group_1866', 'group_1869', 'group_1870', 'group_1871', 'group_1873', 'group_1875', 'group_1876', 'group_1883', 'group_1888', 'group_3525', 'group_1002', 'group_1109', 'group_1143', 'group_1144', 'group_1380', 'group_1381', 'group_1457', 'group_1543', 'group_1739', 'group_1742', 'group_1979', 'group_1980', 'group_3291', 'group_3566', 'group_862', 'group_1191', 'group_1402', 'group_1293', 'group_1294', 'group_1384', 'group_1532', 'group_1950', 'group_1956', 'group_1243', 'group_1441', 'group_1529', 'group_1459', 'group_1460', 'group_1534', 'group_1535', 'group_1977', 'group_1483', 'group_1547', 'group_1115', 'group_1133', 'group_1295', 'group_1972', 'group_776', 'group_1286', 'group_1520', 'group_1555', 'group_1933', 'group_1999', 'group_1817', 'group_1445', 'group_266', 'group_1102', 'group_624', 'group_1300', 'group_895', 'group_1479', 'group_1827', 'group_3163', 'group_1101', 'group_1467', 'group_261', 'group_774', 'group_1233', 'group_1242', 'group_13', 'group_1493', 'group_1494', 'group_1738', 'group_1799', 'group_1970', 'group_2960', 'group_797', 'group_1251', 'group_1308', 'group_3290', 'group_3828', 'group_604', 'group_657', 'group_943', 'group_1039', 'group_1279', 'group_1468', 'group_1502', 'group_1830', 'group_2588', 'group_3761', 'group_699', 'group_1505', 'group_1548', 'group_679', 'group_756', 'group_1063', 'group_1489', 'group_1680', 'group_2885', 'group_3674', 'group_573', 'group_598', 'group_666', 'group_1333', 'group_1983', 'group_2053', 'group_229', 'group_3272', 'group_3313', 'group_350', 'group_3861', 'group_431', 'group_843', 'group_933', 'group_1009', 'group_1121', 'group_1122', 'group_1150', 'group_1266', 'group_1267', 'group_1292', 'group_1475', 'group_1490', 'group_1531', 'group_1544', 'group_165', 'group_1829', 'group_1835', 'group_1836', 'group_1838', 'group_185', 'group_1947', 'group_1949', 'group_2088', 'group_2580', 'group_3126', 'group_364', 'group_599', 'group_631', 'group_1012', 'group_1013', 'group_1036', 'group_1047', 'group_1120', 'group_1280', 'group_1284', 'group_1287', 'group_1322', 'group_1346', 'group_1352', 'group_1484', 'group_1485', 'group_1486', 'group_1509', 'group_1510', 'group_1512', 'group_1514', 'group_1546', 'group_1559', 'group_1828', 'group_1905', 'group_1908', 'group_1909', 'group_1910', 'group_1911', 'group_1914', 'Late embryogenesis abundant protein LEA-2 subgroup domain-containing protein', 'group_1937', 'group_1938', 'group_2000', 'group_2007', 'group_2008', 'group_2010', 'group_2011', 'group_2012', 'group_2013', 'group_2014', 'group_2015', 'group_2016', 'group_2019', 'group_2042', 'group_2043', 'group_2044', 'group_2045', 'group_2046', 'group_2047', 'group_2048', 'group_2049', 'group_2050', 'group_2074', 'group_2372', 'group_2636', 'group_3037', 'group_3119', 'group_316', 'group_3208', 'group_344', 'group_3531', 'group_3881', 'group_478', 'group_721', 'group_745', 'group_781', 'group_813', 'group_868', 'group_1040', 'group_1155', 'group_1231', 'group_1232', 'group_1270', 'group_1281', 'group_1282', 'group_1323', 'group_1466', 'group_1488', 'group_1516', 'group_1564', 'group_1566', 'group_1570', 'group_1575', 'group_1832', 'group_2094', 'group_257', 'group_268', 'group_2890', 'group_3532', 'group_3692', 'group_3829', 'group_410', 'group_460', 'group_796', 'group_811', 'group_828', 'group_865', 'group_867', 'group_925', 'group_942', 'group_958', 'group_1006', 'group_1116', 'group_1167', 'group_1239', 'group_1289', 'group_1454', 'group_1455', 'group_1498', 'group_1528', 'group_1617', 'group_1735', 'group_1839', 'group_1906', 'group_1907', 'group_1948', 'group_1984', 'group_2005', 'group_2026', 'group_2039', 'group_2040', 'group_2041', 'group_2052', 'group_2650', 'group_2865', 'group_2896', 'group_3007', 'group_3244', 'group_358', 'group_3613', 'group_3756', 'group_3772', 'group_385', 'group_3874', 'group_3883', 'group_545', 'group_801', 'group_836', 'group_848', 'group_863', 'group_994', 'group_104', 'group_105', 'group_1100', 'group_130', 'group_1444', 'group_1458', 'group_1481', 'group_1504', 'group_1563', 'group_1599', 'group_1609', 'group_162', 'group_1814', 'group_1816', 'group_1991', 'group_1995', 'group_1996', 'group_2031', 'group_2032', 'group_2051', 'group_2059', 'group_2633', 'group_2643', 'group_2657', 'group_2680', 'group_2733', 'group_2887', 'group_2988', 'group_3056', 'group_3121', 'group_3160', 'group_3165', 'group_3210', 'group_323', 'group_3822', 'group_627', 'group_651', 'group_726', 'group_794', 'group_861', 'group_871', 'group_952', 'group_988', 'group_1007', 'group_1037', 'group_1105', 'group_1244', 'group_1321', 'group_1347', 'group_1518', 'group_1588', 'group_1594', 'group_1686', 'group_1807', 'group_1833', 'group_1932', 'group_1934', 'group_1965', 'group_1968', 'group_1993', 'group_2003', 'group_2004', 'group_201', 'group_2065', 'group_2068', 'group_231', 'group_2607', 'group_2622', 'group_2677', 'group_2694', 'group_2750', 'group_2852', 'group_29', 'group_2943', 'group_2946', 'group_3005', 'group_3086', 'group_3135', 'group_3147', 'group_3149', 'group_3366', 'group_346', 'group_3810', 'group_467', 'group_706', 'group_723', 'group_75', 'group_771', 'group_773', 'group_902', 'group_939', 'group_1026', 'group_1043', 'group_107', 'group_1241', 'group_1533', 'group_1549', 'group_1562', 'group_1590', 'group_18', 'group_1826', 'group_1874', 'group_1890', 'group_1891', 'group_1892', 'group_1893', 'group_1896', 'group_1897', 'group_1898', 'group_1899', 'group_1900', 'group_1901', 'group_1903', 'group_1920', 'group_1928', 'group_1929', 'group_1930', 'group_1931', 'group_1940', 'group_1989', 'group_223', 'group_2574', 'group_2585', 'group_2641', 'group_2719', 'group_2724', 'group_2760', 'group_3053', 'group_3059', 'group_3352', 'group_3501', 'group_3593', 'group_363', 'group_3668', 'group_374', 'group_380', 'group_3803', 'group_3860', 'group_526', 'group_58', 'group_680', 'group_685', 'group_695', 'group_777', 'group_778', 'group_945', 'group_96', 'group_1131', 'group_1136', 'group_1156', 'group_132', 'group_1511', 'group_1552', 'group_1561', 'group_1745', 'group_1840', 'group_1973', 'group_1987', 'group_200', 'group_2022', 'group_2025', 'group_2035', 'group_2054', 'group_2060', 'group_2093', 'group_2095', 'group_2101', 'group_24', 'group_260', 'group_2646', 'group_2842', 'group_2903', 'group_298', 'group_3043', 'group_3047', 'group_3140', 'group_3153', 'group_3195', 'group_332', 'group_3432', 'group_3547', 'group_3597', 'group_3611', 'group_3671', 'group_3739', 'group_3757', 'group_3815', 'group_3824', 'group_441', 'group_556', 'group_618', 'group_644', 'group_674', 'group_727', 'group_834', 'group_845', 'group_853', 'group_866', 'group_911', 'group_918', 'group_1016', 'group_1106', 'group_1264', 'group_1335', 'group_1513', 'group_1545', 'group_1811', 'group_1894', 'group_1917', 'group_1939', 'group_2063', 'group_2632', 'group_2639', 'group_2695', 'group_2860', 'group_2897', 'group_3085', 'group_3087', 'group_3189', 'group_3196', 'group_3309', 'group_3322', 'group_3701', 'group_3705', 'group_372', 'group_3804', 'group_3868', 'group_398', 'group_502', 'group_617', 'group_632', 'group_649', 'group_79', 'group_790', 'group_800', 'group_803', 'group_86', 'group_967', 'group_10', 'group_1042', 'group_1085', 'group_1173', 'group_1259', 'group_1263', 'group_1309', 'group_151', 'group_1523', 'group_1581', 'group_1597', 'group_1687', 'group_176', 'group_1821', 'group_191', 'group_1927', 'group_1975', 'group_1986', 'group_1998', 'group_2104', 'group_251', 'group_256', 'group_2638', 'group_2652', 'group_277', 'group_2919', 'group_2924', 'group_2961', 'group_2984', 'group_3038', 'group_31', 'group_3237', 'group_3242', 'group_3275', 'group_3353', 'group_349', 'group_3510', 'group_3742', 'group_3800', 'group_3808', 'group_395', 'group_418', 'group_439', 'group_484', 'group_520', 'group_615', 'group_764', 'group_766', 'group_807', 'group_832', 'group_956', 'group_1015', 'group_1053', 'group_1104', 'group_1326', 'group_1585', 'group_1586', 'group_1646', 'group_1743', 'group_179', 'group_1946', 'group_2110', 'group_2616', 'group_2645', 'group_2798', 'group_2821', 'group_2833', 'group_2867', 'group_2889', 'group_289', 'group_2965', 'group_3064', 'group_3079', 'group_3095', 'group_3157', 'group_3184', 'group_321', 'group_3248', 'group_325', 'group_3332', 'group_336', 'group_3460', 'group_3774', 'group_3871', 'group_401', 'group_420', 'group_459', 'group_540', 'group_553', 'group_605', 'group_622', 'group_659', 'group_747', 'group_821', 'group_830', 'group_846', 'group_887', 'group_905', 'group_1052', 'group_1060', 'group_1112', 'group_1129', 'group_1139', 'group_1146', 'group_1341', 'group_1603', 'group_1607', 'group_161', 'group_2572', 'group_2608', 'group_2609', 'group_2614', 'group_264', 'group_2921', 'group_2938', 'group_3009', 'group_3088', 'group_310', 'group_3125', 'group_3192', 'group_3199', 'group_320', 'group_3364', 'group_3370', 'group_345', 'group_3594', 'group_3595', 'group_3869', 'group_3882', 'group_399', 'group_443', 'group_448', 'group_473', 'group_588', 'group_619', 'group_694', 'group_804', 'group_898', 'group_94', 'group_1044', 'group_1058', 'group_1086', 'group_1097', 'group_1123', 'group_114', 'group_1290', 'group_1328', 'group_1334', 'group_144', 'group_1556', 'group_1596', 'group_163', 'group_182', 'group_1913', 'group_1941', 'group_1978', 'group_2002', 'group_2103', 'group_230', 'group_2648', 'group_2697', 'group_2723', 'group_276', 'group_2764', 'group_2766', 'group_280', 'group_2823', 'group_3008', 'group_3036', 'group_3084', 'group_3124', 'group_3156', 'group_3202', 'group_3271', 'group_3351', 'group_3367', 'group_3371', 'group_3515', 'group_352', 'group_3550', 'group_3672', 'group_3691', 'group_3694', 'group_3859', 'group_3864', 'group_405', 'group_446', 'group_468', 'group_507', 'group_512', 'group_538', 'group_539', 'group_578', 'group_65', 'group_675', 'group_715', 'group_995', 'group_1022', 'group_1246', 'group_1247', 'group_127', 'group_143', 'group_1446', 'group_1462', 'group_1463', 'group_1568', 'group_1572', 'group_1587', 'group_1601', 'group_1619', 'group_1620', 'group_1951', 'group_198', 'group_234', 'group_244', 'group_253', 'group_2566', 'group_2682', 'group_2740', 'group_2789', 'group_285', 'group_2851', 'group_2899', 'group_2918', 'group_293', 'group_2937', 'group_2982', 'group_3033', 'group_3120', 'group_3310', 'group_3350', 'group_347', 'group_3509', 'group_3544', 'group_3545', 'group_3641', 'group_3745', 'group_377', 'group_3786', 'group_3796', 'group_406', 'group_436', 'group_451', 'group_452', 'group_455', 'group_457', 'group_547', 'group_572', 'group_613', 'group_635', 'group_661', 'group_709', 'group_752', 'group_82', 'group_910', 'group_953', 'group_954', 'group_971', 'group_1031', 'group_1056', 'group_1124', 'group_1128', 'group_1166', 'group_1177', 'group_1182', 'group_1443', 'group_150', 'group_1577', 'group_16', 'group_1623', 'group_1624', 'group_1625', 'group_1626', 'group_1628', 'group_1632', 'group_1684', 'group_186', 'group_1969', 'group_2139', 'group_2142', 'group_2143', 'group_2144', 'group_2145', 'group_2149', 'group_2150', 'group_2151', 'group_2154', 'group_2155', 'group_2204', 'group_221', 'group_2611', 'group_2644', 'group_2658', 'group_273', 'group_2790', 'group_2815', 'group_283', 'group_2843', 'group_2872', 'group_2906', 'group_2973', 'group_3010', 'group_3039', 'group_3058', 'group_306', 'group_3061', 'group_3094', 'group_3123', 'group_3325', 'group_3357', 'group_3380', 'group_3381', 'group_348', 'group_3512', 'group_3662', 'group_376', 'group_3781', 'group_3788', 'group_382', 'group_39', 'group_416', 'group_464', 'group_476', 'group_485', 'group_491', 'group_492', 'group_495', 'group_508', 'group_510', 'group_514', 'group_606', 'group_621', 'group_650', 'group_655', 'group_669', 'group_718', 'group_724', 'group_837', 'group_841', 'group_860', 'group_896', 'group_921', 'group_980', 'group_990', 'group_1018', 'group_1107', 'group_115', 'group_1157', 'group_1364', 'group_1449', 'group_1612', 'group_1650', 'group_1654', 'group_184', 'group_190', 'group_1945', 'group_1959', 'group_197', 'group_2086', 'group_2087', 'group_2091', 'group_2100', 'group_2118', 'group_2148', 'group_219', 'group_220', 'group_254', 'group_2568', 'group_2656', 'group_2683', 'group_2721', 'group_2732', 'group_2741', 'group_2759', 'group_2762', 'group_2768', 'group_279', 'group_2814', 'group_2828', 'group_284', 'group_2847', 'group_2902', 'group_2904', 'group_2922', 'group_3003', 'group_3041', 'group_3045', 'group_3069', 'group_3078', 'group_3133', 'group_3185', 'group_3186', 'group_319', 'group_3190', 'group_3200', 'group_3232', 'group_3233', 'group_3247', 'group_326', 'group_3270', 'group_3323', 'group_3365', 'group_3379', 'group_3426', 'group_3465', 'group_3608', 'group_3631', 'group_3636', 'group_3640', 'group_3661', 'group_3664', 'group_3665', 'group_3693', 'group_3698', 'group_37', 'group_3737', 'group_3743', 'group_3758', 'group_3759', 'group_3783', 'group_3797', 'group_3820', 'group_3870', 'group_3872', 'group_3884', 'group_403', 'group_421', 'group_45', 'group_46', 'group_47', 'group_496', 'group_497', 'group_506', 'group_522', 'group_524', 'group_585', 'group_587', 'group_60', 'group_668', 'group_681', 'group_824', 'group_840', 'group_903', 'group_970', 'group_973', 'group_978', 'group_98', 'group_1033', 'group_108', 'group_1091', 'group_1094', 'group_1130', 'group_1175', 'group_124', 'group_1450', 'group_1456', 'group_15', 'group_1522', 'group_1551', 'group_1574', 'group_1631', 'group_167', 'group_1747', 'group_1922', 'group_194', 'group_2069', 'group_2189', 'group_2192', 'group_2201', 'group_2202', 'group_2203', 'group_2240', 'group_232', 'group_2582', 'group_2621', 'group_2634', 'group_2647', 'group_2649', 'group_2676', 'group_2698', 'group_2718', 'group_2738', 'group_2742', 'group_2788', 'group_2792', 'group_2813', 'group_2817', 'group_286', 'group_2873', 'group_2876', 'group_290', 'group_2941', 'group_2944', 'group_2971', 'group_3052', 'group_3067', 'group_3128', 'group_3129', 'group_3137', 'group_3209', 'group_3230', 'group_3308', 'group_3356', 'group_3363', 'group_3430', 'group_3498', 'group_3530', 'group_3632', 'group_3633', 'group_3637', 'group_3688', 'group_3755', 'group_3773', 'group_3776', 'group_3819', 'group_386', 'group_462', 'group_471', 'group_50', 'group_51', 'group_533', 'group_535', 'group_542', 'group_554', 'group_564', 'group_593', 'group_597', 'group_61', 'group_614', 'group_653', 'group_654', 'group_664', 'group_667', 'group_683', 'group_684', 'group_707', 'group_708', 'group_713', 'group_748', 'group_788', 'group_810', 'group_84', 'group_876', 'group_97', 'group_975', 'group_1023', 'group_1034', 'group_1072', 'group_1093', 'group_112', 'group_1148', 'group_1152', 'group_1178', 'group_125', 'group_1340', 'group_1343', 'group_135', 'group_1363', 'group_140', 'group_1565', 'group_1608', 'group_1611', 'group_1618', 'group_1622', 'group_1629', 'group_169', 'group_1813', 'group_1944', 'group_2073', 'group_2077', 'group_2078', 'group_2079', 'group_2080', 'group_2085', 'group_209', 'group_2124', 'group_2130', 'group_2138', 'group_2140', 'group_2141', 'group_2152', 'group_2153', 'group_2320', 'group_2321', 'group_2322', 'group_233', 'group_235', 'group_2356', 'group_245', 'group_2573', 'group_263', 'group_2635', 'group_2654', 'group_2678', 'group_2686', 'group_27', 'group_2744', 'group_2767', 'group_2786', 'group_2787', 'group_2791', 'group_2794', 'group_2818', 'group_2820', 'group_2831', 'group_2859', 'group_2861', 'group_2863', 'group_2869', 'group_2900', 'group_2901', 'group_2929', 'group_2969', 'group_3034', 'group_3044', 'group_3048', 'group_3062', 'group_3068', 'group_3076', 'group_3127', 'group_3148', 'group_3204', 'group_3229', 'group_33', 'group_3324', 'group_335', 'group_3374', 'group_3376', 'group_3427', 'group_3429', 'group_3434', 'group_3463', 'group_3496', 'group_3497', 'group_3504', 'group_3505', 'group_351', 'group_354', 'group_356', 'group_3592', 'group_3596', 'group_3609', 'group_3768', 'group_3779', 'group_384', 'group_413', 'group_435', 'group_438', 'group_445', 'group_449', 'group_458', 'group_548', 'group_568', 'group_569', 'group_586', 'group_590', 'group_648', 'group_652', 'group_665', 'group_671', 'group_731', 'group_732', 'group_742', 'group_757', 'group_77', 'group_775', 'group_808', 'group_816', 'group_85', 'group_881', 'group_886', 'group_899', 'group_917', 'group_923', 'group_932', 'group_948', 'group_101', 'group_1014', 'group_102', 'group_1030', 'group_1051', 'group_1054', 'group_1067', 'group_1073', 'group_1092', 'group_1154', 'group_1179', 'group_1190', 'group_120', 'group_126', 'group_1291', 'group_1310', 'group_1338', 'group_1348', 'group_1351', 'group_137', 'group_145', 'group_1524', 'group_1527', 'group_1536', 'group_1540', 'group_1582', 'group_1600', 'group_1614', 'group_1621', 'group_1633', 'group_1637', 'group_1640', 'group_1744', 'group_1916', 'group_1943', 'group_1966', 'group_199', 'group_2024', 'group_2071', 'group_2084', 'group_2096', 'group_2106', 'group_2119', 'group_2121', 'group_2128', 'group_2136', 'group_2137', 'group_2181', 'group_2190', 'group_2191', 'group_2199', 'group_2200', 'group_2220', 'group_2224', 'group_2230', 'group_2232', 'group_2233', 'group_225', 'group_2349', 'group_2350', 'group_2351', 'group_2352', 'group_2354', 'group_2355', 'group_2366', 'group_241', 'group_2571', 'group_2584', 'group_2610', 'group_2617', 'group_2637', 'group_267', 'group_2696', 'group_2739', 'group_2745', 'group_2746', 'group_2765', 'group_2796', 'group_2812', 'group_2822', 'group_2841', 'group_2846', 'group_2848', 'group_2862', 'group_2875', 'group_2884', 'group_2905', 'group_2923', 'group_2927', 'group_2959', 'group_296', 'group_2970', 'group_2972', 'group_2980', 'group_3011', 'group_3055', 'group_3071', 'group_3077', 'group_308', 'group_3089', 'group_313', 'group_3132', 'group_3138', 'group_3139', 'group_3159', 'group_32', 'group_3201', 'group_322', 'group_3238', 'group_3240', 'group_3245', 'group_3274', 'group_329', 'group_3306', 'group_3307', 'group_3312', 'group_3329', 'group_3359', 'group_3373', 'group_343', 'group_3464', 'group_3514', 'group_3534', 'group_3543', 'group_365', 'group_3667', 'group_367', 'group_3675', 'group_3697', 'group_3699', 'group_3744', 'group_3789', 'group_3818', 'group_3821', 'group_3825', 'group_3875', 'group_3885', 'group_393', 'group_434', 'group_437', 'group_440', 'group_442', 'group_461', 'group_463', 'group_472', 'group_503', 'group_516', 'group_549', 'group_561', 'group_575', 'group_610', 'group_620', 'group_634', 'group_638', 'group_658', 'group_66', 'group_67', 'group_686', 'group_688', 'group_692', 'group_696', 'group_705', 'group_716', 'group_717', 'group_720', 'group_734', 'group_735', 'group_736', 'group_746', 'group_750', 'group_751', 'group_753', 'group_762', 'group_78', 'group_806', 'group_81', 'group_829', 'group_838', 'group_854', 'group_877', 'group_88', 'group_880', 'group_885', 'group_891', 'group_950', 'group_951', 'group_963', 'group_983', 'group_989', 'group_996', 'group_103', 'group_1055', 'group_1057', 'group_106', 'group_1064', 'group_1069', 'group_1078', 'group_1079', 'group_11', 'group_1159', 'group_118', 'group_122', 'group_1265', 'group_131', 'group_1447', 'group_146', 'group_1480', 'group_1521', 'group_1569', 'group_1634', 'group_1635', 'group_1636', 'group_164', 'group_1644', 'group_1647', 'group_1648', 'group_1649', 'group_1667', 'group_1669', 'group_1673', 'group_174', 'group_1805', 'group_1819', 'group_1822', 'group_188', 'group_19', 'group_192', 'group_1953', 'group_196', 'group_1971', 'group_2028', 'group_2034', 'group_204', 'group_205', 'group_206', 'group_2097', 'group_2108', 'group_212', 'group_2125', 'group_2127', 'group_2185', 'group_2212', 'group_2213', 'group_2214', 'group_2215', 'group_2216', 'group_2217', 'group_2218', 'group_2223', 'group_2226', 'group_2227', 'group_2228', 'group_2243', 'group_2248', 'group_2249', 'group_2262', 'group_2264', 'group_2307', 'group_2315', 'group_2317', 'group_2318', 'group_238', 'group_239', 'group_240', 'group_2434', 'group_247', 'group_250', 'group_252', 'group_255', 'group_2569', 'group_2570', 'group_2586', 'group_259', 'group_2623', 'group_2640', 'group_2642', 'group_265', 'group_2726', 'group_2747', 'group_2769', 'group_2770', 'group_28', 'group_2819', 'group_2835', 'group_2864', 'group_2868', 'group_2886', 'group_2898', 'group_291', 'group_2962', 'group_2964', 'group_2977', 'group_299', 'group_2997', 'group_2998', 'group_3001', 'group_3002', 'group_3013', 'group_302', 'group_3049', 'group_3057', 'group_307', 'group_3090', 'group_3093', 'group_3098', 'group_3122', 'group_3143', 'group_3164', 'group_318', 'group_3188', 'group_3193', 'group_3203', 'group_3241', 'group_3273', 'group_331', 'group_3311', 'group_3326', 'group_3328', 'group_3331', 'group_3337', 'group_3354', 'group_3375', 'group_3378', 'group_34', 'group_342', 'group_3456', 'group_3459', 'group_3506', 'group_3517', 'group_3551', 'group_36', 'group_362', 'group_3663', 'group_3666', 'group_3669', 'group_3703', 'group_3740', 'group_3741', 'group_3760', 'group_3780', 'group_3782', 'group_3811', 'group_3823', 'group_3865', 'group_390', 'group_392', 'group_408', 'group_423', 'group_429', 'group_469', 'group_474', 'group_494', 'group_509', 'group_513', 'group_515', 'group_541', 'group_560', 'group_562', 'group_565', 'group_567', 'group_611', 'group_623', 'group_637', 'group_663', 'group_673', 'group_702', 'group_722', 'group_749', 'group_763', 'group_809', 'group_814', 'group_820', 'group_831', 'group_847', 'group_870', 'group_878', 'group_89', 'group_90', 'group_924', 'group_929', 'group_941', 'group_944', 'group_959', 'group_99', 'group_100', 'group_1011', 'group_1017', 'group_1028', 'group_1049', 'group_1059', 'group_1061', 'group_1062', 'group_1065', 'group_1070', 'group_1080', 'group_1087', 'group_109', 'group_1090', 'group_111', 'group_1126', 'group_1138', 'group_1145', 'group_1149', 'group_1153', 'group_1168', 'group_1169', 'group_1185', 'group_1236', 'group_1269', 'group_1304', 'group_1313', 'group_1330', 'group_1331', 'group_1332', 'group_1339', 'group_1354', 'group_1358', 'group_1359', 'group_136', 'group_1360', 'group_1362', 'group_139', 'group_14', 'group_149', 'group_1553', 'group_1554', 'group_1567', 'group_157', 'group_1571', 'group_1573', 'group_1576', 'group_160', 'group_1602', 'group_1613', 'group_1639', 'group_1645', 'group_1653', 'group_1659', 'group_1660', 'group_1662', 'group_1663', 'group_1665', 'group_1666', 'group_1668', 'group_1670', 'group_1671', 'group_1672', 'group_1674', 'group_1681', 'group_177', 'group_180', 'group_1809', 'group_1810', 'group_1820', 'group_183', 'group_189', 'group_1921', 'group_193', 'group_1974', 'group_1976', 'group_202', 'group_2036', 'group_2058', 'group_207', 'group_2090', 'group_210', 'group_2105', 'group_2107', 'group_211', 'group_2116', 'group_2120', 'group_2123', 'group_2131', 'group_2182', 'group_2183', 'group_2184', 'group_2187', 'group_2188', 'group_2196', 'group_2206', 'group_2207', 'group_2208', 'group_2209', 'group_2210', 'group_2221', 'group_2222', 'group_2225', 'group_2235', 'group_2236', 'group_2237', 'group_2241', 'group_2245', 'group_2250', 'group_226', 'group_2263', 'group_2265', 'group_2266', 'group_2268', 'group_2269', 'group_227', 'group_2270', 'group_2277', 'group_2278', 'group_2279', 'group_2280', 'group_2281', 'group_2282', 'group_2283', 'group_2284', 'group_2285', 'group_2286', 'group_2287', 'group_2288', 'group_2289', 'group_2290', 'group_2291', 'group_2292', 'group_2293', 'group_2294', 'group_2295', 'group_2296', 'group_2298', 'group_2299', 'group_2300', 'group_2309', 'group_2312', 'group_2314', 'group_2319', 'group_2324', 'group_2325', 'group_2326', 'group_2327', 'group_2328', 'group_2329', 'group_2330', 'group_2331', 'group_2332', 'group_2333', 'group_2334', 'group_2335', 'group_2337', 'group_2338', 'group_2339', 'group_2340', 'group_2341', 'group_2342', 'group_2343', 'group_2348', 'group_2357', 'group_2368', 'group_2369', 'group_237', 'group_2370', 'group_2371', 'group_2373', 'group_2374', 'group_2375', 'group_2376', 'group_2377', 'group_2378', 'group_2379', 'group_2380', 'group_2381', 'group_2382', 'group_2383', 'group_2384', 'group_2386', 'group_2387', 'group_2388', 'group_2389', 'group_2390', 'group_2391', 'group_2392', 'group_2393', 'group_2394', 'group_2398', 'group_2399', 'group_2400', 'group_2404', 'group_2405', 'group_2406', 'group_2411', 'group_2412', 'group_2413', 'group_2414', 'group_2416', 'group_2418', 'group_2419', 'group_242', 'group_2420', 'group_2421', 'group_243', 'group_2460', 'group_2461', 'group_2463', 'group_249', 'group_25', 'group_2567', 'group_2583', 'group_2613', 'group_2618', 'group_2619', 'group_262', 'group_2620', 'group_2651', 'group_2653', 'group_2659', 'group_2681', 'group_2684', 'group_270', 'group_271', 'group_2725', 'group_2728', 'group_2729', 'group_2730', 'group_2731', 'group_2734', 'group_2743', 'group_2748', 'group_275', 'group_2751', 'group_2761', 'group_2785', 'group_2824', 'group_2830', 'group_2850', 'group_2866', 'group_287', 'group_2926', 'group_2957', 'group_2963', 'group_2975', 'group_2978', 'group_2979', 'group_2981', 'group_2983', 'group_2987', 'group_2995', 'group_2996', 'group_3', 'group_3006', 'group_301', 'group_3014', 'group_304', 'group_3040', 'group_3042', 'group_3050', 'group_3054', 'group_3060', 'group_3065', 'group_3080', 'group_3083', 'group_3097', 'group_311', 'group_3130', 'group_3131', 'group_3134', 'group_3136', 'group_314', 'group_3154', 'group_3158', 'group_3161', 'group_317', 'group_3187', 'group_3194', 'group_3205', 'group_3231', 'group_3249', 'group_3250', 'group_328', 'group_330', 'group_3327', 'group_333', 'group_3330', 'group_334', 'group_3360', 'group_3361', 'group_3362', 'group_3372', 'group_3377', 'group_338', 'group_3436', 'group_3500', 'group_353', 'group_3546', 'group_3548', 'group_3564', 'group_357', 'group_359', 'group_3638', 'group_3673', 'group_3700', 'group_3702', 'group_3704', 'group_371', 'group_3738', 'group_375', 'group_3762', 'group_3767', 'group_3769', 'group_3771', 'group_3777', 'group_3785', 'group_3787', 'group_3798', 'group_381', 'group_3812', 'group_3816', 'group_3866', 'group_3873', 'group_394', 'group_400', 'group_402', 'group_41', 'group_419', 'group_42', 'group_424', 'group_426', 'group_427', 'group_432', 'group_433', 'group_44', 'group_450', 'group_456', 'group_465', 'group_466', 'group_470', 'group_475', 'group_480', 'group_486', 'group_489', 'group_490', 'group_499', 'group_500', 'group_501', 'group_517', 'group_523', 'group_536', 'group_537', 'group_543', 'group_550', 'group_552', 'group_563', 'group_570', 'group_589', 'group_592', 'group_595', 'group_6', 'group_600', 'group_601', 'group_607', 'group_609', 'group_612', 'group_616', 'group_625', 'group_630', 'group_672', 'group_677', 'group_689', 'group_728', 'group_739', 'group_740', 'group_743', 'group_754', 'group_759', 'group_760', 'group_767', 'group_798', 'group_815', 'group_819', 'group_83', 'group_842', 'group_844', 'group_851', 'group_859', 'group_864', 'group_872', 'group_874', 'group_882', 'group_884', 'group_894', 'group_901', 'group_906', 'group_907', 'group_909', 'group_91', 'group_913', 'group_914', 'group_919', 'group_92', 'group_920', 'group_930', 'group_955', 'group_957', 'group_965', 'group_966', 'group_968', 'group_977', 'group_991', 'group_993', 'group_1000', 'group_1005', 'group_1019', 'group_1021', 'group_1024', 'group_1025', 'group_1027', 'group_1029', 'group_1032', 'group_1035', 'group_1038', 'group_1041', 'group_1045', 'group_1046', 'group_1066', 'group_1068', 'group_1071', 'group_1074', 'group_1075', 'group_1076', 'group_1077', 'group_1088', 'group_1089', 'group_1095', 'group_1096', 'group_1098', 'group_1099', 'group_110', 'group_1113', 'group_1114', 'group_1117', 'group_1125', 'group_113', 'group_1132', 'group_1135', 'group_1137', 'group_1140', 'group_1141', 'group_1147', 'group_1151', 'group_1158', 'group_1160', 'group_1161', 'group_1162', 'group_1165', 'group_117', 'group_1170', 'group_1171', 'group_1172', 'group_1176', 'group_1180', 'group_1181', 'group_1184', 'group_1186', 'group_1187', 'group_1188', 'group_1189', 'group_1192', 'group_12', 'group_1205', 'group_121', 'group_1218', 'group_1220', 'group_1226', 'group_1230', 'group_1234', 'group_1235', 'group_1260', 'group_1261', 'group_1271', 'group_1272', 'group_128', 'group_1283', 'group_129', 'group_1302', 'group_1303', 'group_1305', 'group_1306', 'group_1312', 'group_1314', 'group_1316', 'group_1317', 'group_1318', 'group_1319', 'group_1327', 'group_133', 'group_1336', 'group_1337', 'group_134', 'group_1342', 'group_1349', 'group_1361', 'group_1365', 'group_1366', 'group_1367', 'group_1368', 'group_141', 'group_142', 'group_1448', 'group_1452', 'group_147', 'group_1474', 'group_148', 'group_1491', 'group_1492', 'group_1496', 'group_1508', 'group_1515', 'group_152', 'group_1525', 'group_1526', 'group_153', 'group_154', 'group_155', 'group_156', 'group_1560', 'group_1578', 'group_1579', 'group_158', 'group_1580', 'group_1583', 'group_1584', 'group_159', 'group_1592', 'group_1595', 'group_1598', 'group_1604', 'group_1610', 'group_1630', 'group_1641', 'group_1642', 'group_1643', 'group_1655', 'group_1656', 'group_1657', 'group_1658', 'group_166', 'group_1661', 'group_1664', 'group_1677', 'group_168', 'group_1689', 'group_1690', 'group_1693', 'group_1694', 'group_1695', 'group_1696', 'group_1697', 'group_1698', 'group_1699', 'group_17', 'group_170', 'group_1700', 'group_1701', 'group_1702', 'group_1703', 'group_1704', 'group_1705', 'group_1706', 'group_1707', 'group_1708', 'group_1709', 'group_171', 'group_1710', 'group_1711', 'group_1712', 'group_1713', 'group_1714', 'group_1715', 'group_1716', 'group_1717', 'group_1718', 'group_1719', 'group_172', 'group_1720', 'group_1722', 'group_1723', 'group_1724', 'group_1725', 'group_1727', 'group_1728', 'group_1729', 'group_173', 'group_1730', 'group_1731', 'group_1734', 'group_1736', 'group_1748', 'group_1749', 'group_175', 'group_1750', 'group_1751', 'group_1752', 'group_1753', 'group_1754', 'group_1755', 'group_1756', 'group_1757', 'group_1758', 'group_1759', 'group_1760', 'group_1761', 'group_1762', 'group_1763', 'group_1764', 'group_1765', 'group_1766', 'group_1767', 'group_1768', 'group_1769', 'group_1770', 'group_1771', 'group_1773', 'group_1774', 'group_1775', 'group_1776', 'group_1777', 'group_1779', 'group_178', 'group_1780', 'group_1781', 'group_1783', 'group_1784', 'group_1785', 'group_1786', 'group_1787', 'group_1788', 'group_1790', 'group_1791', 'group_1802', 'group_1803', 'group_1804', 'group_1806', 'group_1808', 'group_181', 'group_1812', 'group_1823', 'group_187', 'group_1915', 'group_1919', 'group_1923', 'group_1924', 'group_1925', 'group_195', 'group_1958', 'group_1960', 'group_1962', 'group_1988', 'group_2', 'group_2018', 'group_2020', 'group_2021', 'group_2023', 'group_2027', 'group_203', 'group_2033', 'group_2037', 'group_2055', 'group_2056', 'group_2089', 'group_2098', 'group_2099', 'group_21', 'group_2102', 'group_2109', 'group_2111', 'group_2113', 'group_2115', 'group_2122', 'group_2129', 'group_213', 'group_2132', 'group_2134', 'group_2158', 'group_2159', 'group_216', 'group_2160', 'group_2161', 'group_2162', 'group_2163', 'group_2164', 'group_2165', 'group_2166', 'group_2167', 'group_2168', 'group_2169', 'group_217', 'group_2170', 'group_2171', 'group_2172', 'group_2173', 'group_2174', 'group_2175', 'group_2176', 'group_2177', 'group_2178', 'group_2179', 'group_2180', 'group_2186', 'group_2193', 'group_2194', 'group_2195', 'group_2198', 'group_22', 'group_2205', 'group_2211', 'group_2219', 'group_2229', 'group_224', 'group_2242', 'group_2244', 'group_2246', 'group_2247', 'group_2252', 'group_2253', 'group_2254', 'group_2255', 'group_2256', 'group_2257', 'group_2258', 'group_2259', 'group_2260', 'group_2261', 'group_2267', 'group_2271', 'group_2272', 'group_2273', 'group_2274', 'group_2275', 'group_2276', 'group_228', 'group_23', 'group_2301', 'group_2302', 'group_2303', 'group_2304', 'group_2305', 'group_2308', 'group_2311', 'group_2316', 'group_2323', 'group_2346', 'group_2347', 'group_2358', 'group_2359', 'group_2360', 'group_2361', 'group_2362', 'group_2363', 'group_2364', 'group_2365', 'group_2367', 'group_2395', 'group_2396', 'group_2397', 'group_2401', 'group_2402', 'group_2403', 'group_2408', 'group_2409', 'group_2417', 'group_2422', 'group_2423', 'group_2424', 'group_2425', 'group_2426', 'group_2427', 'group_2428', 'group_2429', 'group_2430', 'group_2431', 'group_2432', 'group_2433', 'group_2435', 'group_2436', 'group_2437', 'group_2438', 'group_2439', 'group_2440', 'group_2441', 'group_2442', 'group_2443', 'group_2444', 'group_2445', 'group_2446', 'group_2447', 'group_2448', 'group_2449', 'group_2450', 'group_2452', 'group_2453', 'group_2454', 'group_2455', 'group_2456', 'group_2457', 'group_2458', 'group_2459', 'group_2464', 'group_2465', 'group_2466', 'group_2467', 'group_2468', 'group_2469', 'group_2470', 'group_2472', 'group_2473', 'group_2475', 'group_2476', 'group_2477', 'group_2478', 'group_2479', 'group_2480', 'group_2481', 'group_2483', 'group_2484', 'group_2485', 'group_2486', 'group_2487', 'group_2488', 'group_2575', 'group_258', 'group_2581', 'group_2587', 'group_26', 'group_2612', 'group_2615', 'group_2624', 'group_2655', 'group_2660', 'group_2679', 'group_2685', 'group_269', 'group_2699', 'group_2727', 'group_2735', 'group_2749', 'group_2752', 'group_2763', 'group_2771', 'group_2772', 'group_278', 'group_2793', 'group_2795', 'group_2797', 'group_281', 'group_2816', 'group_282', 'group_2825', 'group_2826', 'group_2827', 'group_2829', 'group_2832', 'group_2834', 'group_2845', 'group_2849', 'group_2870', 'group_2871', 'group_2874', 'group_288', 'group_2888', 'group_2907', 'group_2908', 'group_292', 'group_2920', 'group_2925', 'group_2928', 'group_2930', 'group_2939', 'group_294', 'group_2940', 'group_2942', 'group_2945', 'group_2947', 'group_295', 'group_2958', 'group_297', 'group_2974', 'group_2976', 'group_2985', 'group_2986', 'group_2989', 'group_300', 'group_3000', 'group_3004', 'group_3012', 'group_3015', 'group_3017', 'group_303', 'group_3035', 'group_3046', 'group_3063', 'group_3066', 'group_3070', 'group_3072', 'group_3081', 'group_3082', 'group_309', 'group_3091', 'group_3092', 'group_312', 'group_3141', 'group_3142', 'group_3144', 'group_3145', 'group_3150', 'group_3151', 'group_3152', 'group_3155', 'group_3162', 'group_3191', 'group_3197', 'group_3198', 'group_3206', 'group_3207', 'group_3234', 'group_3235', 'group_3236', 'group_3239', 'group_324', 'group_3243', 'group_3246', 'group_3251', 'group_327', 'group_3277', 'group_3292', 'group_3293', 'group_3314', 'group_3315', 'group_3333', 'group_3334', 'group_3335', 'group_3336', 'group_3355', 'group_3358', 'group_3368', 'group_3369', 'group_337', 'group_339', 'group_340', 'group_341', 'group_3428', 'group_3431', 'group_3433', 'group_3435', 'group_3458', 'group_3461', 'group_3462', 'group_3466', 'group_3467', 'group_3468', 'group_3499', 'group_3502', 'group_3503', 'group_3507', 'group_3511', 'group_3513', 'group_3516', 'group_3524', 'group_3533', 'group_3549', 'group_355', 'group_3565', 'group_360', 'group_361', 'group_3610', 'group_3612', 'group_3614', 'group_3634', 'group_3635', 'group_3639', 'group_366', 'group_3689', 'group_369', 'group_3690', 'group_370', 'group_3706', 'group_3707', 'group_373', 'group_3746', 'group_3763', 'group_3770', 'group_3775', 'group_378', 'group_3784', 'group_3790', 'group_3799', 'group_38', 'group_3801', 'group_3802', 'group_3809', 'group_3813', 'group_3814', 'group_3817', 'group_3826', 'group_383', 'group_3830', 'group_3831', 'group_3867', 'group_388', 'group_389', 'group_397', 'group_40', 'group_404', 'group_407', 'group_409', 'group_411', 'group_412', 'group_414', 'group_415', 'group_417', 'group_422', 'group_425', 'group_43', 'group_430', 'group_444', 'group_447', 'group_453', 'group_454', 'group_477', 'group_479', 'group_48', 'group_481', 'group_482', 'group_483', 'group_487', 'group_488', 'group_49', 'group_493', 'group_498', 'group_5', 'group_504', 'group_505', 'group_511', 'group_518', 'group_519', 'group_52', 'group_521', 'group_525', 'group_527', 'group_528', 'group_529', 'group_53', 'group_530', 'group_531', 'group_532', 'group_534', 'group_54', 'group_544', 'group_546', 'group_55', 'group_551', 'group_555', 'group_557', 'group_558', 'group_559', 'group_566', 'group_57', 'group_571', 'group_574', 'group_576', 'group_577', 'group_579', 'group_581', 'group_582', 'group_583', 'group_59', 'group_591', 'group_594', 'group_596', 'group_602', 'group_603', 'group_608', 'group_62', 'group_626', 'group_628', 'group_629', 'group_63', 'group_633', 'group_636', 'group_64', 'group_640', 'group_641', 'group_643', 'group_645', 'group_656', 'group_660', 'group_662', 'group_670', 'group_68', 'group_682', 'group_687', 'group_69', 'group_690', 'group_691', 'group_693', 'group_7', 'group_70', 'group_703', 'group_704', 'group_71', 'group_710', 'group_712', 'group_714', 'group_719', 'group_72', 'group_725', 'group_729', 'group_73', 'group_730', 'group_733', 'group_737', 'group_738', 'group_74', 'group_758', 'group_761', 'group_768', 'group_769', 'group_772', 'group_779', 'group_780', 'group_782', 'group_783', 'group_784', 'group_785', 'group_789', 'group_792', 'group_793', 'group_795', 'group_799', 'group_8', 'group_80', 'group_805', 'group_812', 'group_817', 'group_822', 'group_823', 'group_825', 'group_826', 'group_827', 'group_833', 'group_835', 'group_839', 'group_849', 'group_850', 'group_852', 'group_856', 'group_869', 'group_87', 'group_875', 'group_879', 'group_883', 'group_888', 'group_889', 'group_890', 'group_9', 'group_900', 'group_904', 'group_908', 'group_926', 'group_928', 'group_93', 'group_931', 'group_936', 'group_937', 'group_940', 'group_946', 'group_947', 'group_949', 'group_95', 'group_961', 'group_962', 'group_964', 'group_969', 'group_974', 'group_976', 'group_979', 'group_984', 'group_987', 'group_992']\n"
     ]
    }
   ],
   "source": [
    "from tqdm.notebook import tqdm\n",
    "\n",
    "list_of_unresolved_groups = []\n",
    "group_dfs = []\n",
    "searched_groups = []\n",
    "columns_to_check = ['product', 'locus_tag', 'gene']\n",
    "\n",
    "# Establish a connection to the database\n",
    "connection = sqlite3.connect(new_db)\n",
    "cursor = connection.cursor()\n",
    "\n",
    "# ok lets start this loop now.\n",
    "for group in tqdm(pangene_groups, desc=\"Searching for those groups!\"):\n",
    "    if \"group\" in group: # pull out the obvious groups that won't be found.\n",
    "        list_of_unresolved_groups.append(group) # put em in a list that we're gonna do differently! using protein_clusters above!\n",
    "    else:\n",
    "        group_df = search_terms_in_db(connection, \"annotations\", group, columns_to_check)\n",
    "        group_dfs.append(group_df)\n",
    "        searched_groups.append(group)\n",
    "\n",
    "# Close the connection\n",
    "connection.close()\n",
    "print(list_of_unresolved_groups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "1dfa9053",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3921\n"
     ]
    }
   ],
   "source": [
    "groups_to_genes = defaultdict(list)\n",
    "with open(clustered_proteins_file, 'r') as infile:\n",
    "    for line in infile.readlines():\n",
    "        splitline = line.split('\\t')\n",
    "        group_id = splitline[0].split(':')[0]\n",
    "        gene1 = [splitline[0].split(':')[1].strip(\" \").strip('\\n')]\n",
    "        rest_of_genes = [gene.strip('\\n') for gene in splitline[1::]] if len(splitline) > 1 else []\n",
    "        groups_to_genes[group_id] = gene1\n",
    "        groups_to_genes[group_id].extend(rest_of_genes)\n",
    "print(len(groups_to_genes.keys()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "30f7eb1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0c433b77957e4f04a17090a55a9916a0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Searching for those groups!:   0%|          | 0/3921 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the current cell or a previous cell. \n",
      "\u001b[1;31mPlease review the code in the cell(s) to identify a possible cause of the failure. \n",
      "\u001b[1;31mClick <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "# Establish a connection to the database\n",
    "connection = sqlite3.connect(new_db)\n",
    "cursor = connection.cursor()\n",
    "\n",
    "all_groups_df = []\n",
    "# ok lets start this loop now.\n",
    "for group in tqdm(groups_to_genes.items(), desc=\"Searching for those groups!\"):\n",
    "    group_df = get_gene_from_baktaID(connection, \"annotations\", item)\n",
    "    group_dfs.append(group_df)\n",
    "    # Concatenate the DataFrames for this group\n",
    "    if group_dfs:\n",
    "        concatenated_group_df = pandas.concat(group_dfs, ignore_index=True)\n",
    "        all_groups_df.append(concatenated_group_df)\n",
    "\n",
    "# Close the connection\n",
    "connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72b514e5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
