{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/mf019/bioinformatics/longread_GWAS/notebooks/metadata/pair_finding\n"
     ]
    }
   ],
   "source": [
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Isolate</th>\n",
       "      <th>MLST</th>\n",
       "      <th>RST_Type</th>\n",
       "      <th>Source</th>\n",
       "      <th>Origin</th>\n",
       "      <th>Location</th>\n",
       "      <th>skin/blood_pairs</th>\n",
       "      <th>Disseminated</th>\n",
       "      <th>method</th>\n",
       "      <th>OspC_Type</th>\n",
       "      <th>AT_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5A18NP1-JBb08-A</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>culture</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nanopore</td>\n",
       "      <td>A</td>\n",
       "      <td>AT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UNY147</td>\n",
       "      <td>34</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>illumina</td>\n",
       "      <td>J</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5A18NP1-JBb08-C</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>culture</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nanopore</td>\n",
       "      <td>A</td>\n",
       "      <td>AT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5A18NP1-JBb08-B</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>culture</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nanopore</td>\n",
       "      <td>A</td>\n",
       "      <td>AT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5A18NP1-JBb08-E</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>culture</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nanopore</td>\n",
       "      <td>A</td>\n",
       "      <td>AT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Isolate MLST  RST_Type   Source Origin Location skin/blood_pairs  \\\n",
       "0  5A18NP1-JBb08-A    1       NaN  culture     US       NY              NaN   \n",
       "1           UNY147   34       3.0      NaN     US       NY              NaN   \n",
       "2  5A18NP1-JBb08-C    1       NaN  culture     US       NY              NaN   \n",
       "3  5A18NP1-JBb08-B    1       NaN  culture     US       NY              NaN   \n",
       "4  5A18NP1-JBb08-E    1       NaN  culture     US       NY              NaN   \n",
       "\n",
       "  Disseminated    method OspC_Type AT_type  \n",
       "0          NaN  nanopore         A     AT1  \n",
       "1            L  illumina         J     NaN  \n",
       "2          NaN  nanopore         A     AT1  \n",
       "3          NaN  nanopore         A     AT1  \n",
       "4          NaN  nanopore         A     AT1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v5_file = '../wrangled/longread_metadata_v5.csv'\n",
    "v6_file = '../wrangled/longread_metadata_v6.csv'\n",
    "metadata_v5 = pandas.read_csv(v5_file)\n",
    "metadata_v5.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Now to fix the hybrid and pacbio assembly naming!\n",
    "\n",
    "Nathan showed me that the hybrids and pacbio assemblies have metadata and a renaming scheme at the very bottom of the jupyter renaming sheet. \n",
    "\n",
    "I will first import that table (jupyter renaming scheme)\n",
    "create a lookup on column A and B,\n",
    "then import the list of assemblies (filenames sans extensions),\n",
    "for each name that is present in col B, add the corresponding value from col A to a new column.\n",
    "download files from bucket\n",
    "rename files as needed\n",
    "then re-upload to bucket\n",
    "then subset assemblies based on method and pairs.\n",
    "then run roary\n",
    "then annotate pangenes\n",
    "then hand off to rest of team.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1215, 54)\n",
      "  Rename_A Original_Name Origin   Location OspC_Type  RST_Type OspC_Allele  \\\n",
      "0     UMA1            N2     US  Nantucket         K       2.0         NaN   \n",
      "1     UMA2            N4     US  Nantucket         F       2.0         NaN   \n",
      "2     UMA3            N7     US  Nantucket         K       2.0         NaN   \n",
      "3     UMA4            N9     US  Nantucket         K       2.0         NaN   \n",
      "4     UMA5           N11     US  Nantucket         N       2.0         NaN   \n",
      "\n",
      "  MLST Disseminated BC + (Y/N)  ... Source.1 Code     Genus    Species.1  \\\n",
      "0    3            L         ND  ...    Jacob  NaN  Borrelia  burgdorferi   \n",
      "1    8            L         ND  ...    Jacob  NaN  Borrelia  burgdorferi   \n",
      "2    3            L         ND  ...    Jacob  NaN  Borrelia  burgdorferi   \n",
      "3    3            L         ND  ...    Jacob  NaN  Borrelia  burgdorferi   \n",
      "4   36            L         ND  ...    Jacob  NaN  Borrelia  burgdorferi   \n",
      "\n",
      "   FastQC_1  BBDuk  FastQC_2 Spades BBMerge sp_identification  \n",
      "0         y      y         y      y       y                 y  \n",
      "1         y      y         y      y       y                 y  \n",
      "2       NaN    NaN       NaN    NaN     NaN               NaN  \n",
      "3         y      y         y      y       y                 y  \n",
      "4         y      y         y      y       y                 y  \n",
      "\n",
      "[5 rows x 54 columns]\n"
     ]
    }
   ],
   "source": [
    "jupyter_renaming_table_file = '/Users/mf019/bioinformatics/longread_GWAS/notebooks/metadata/lymeseq_original_sheets/LymeSeq_SampleTrack - Renaming Scheme Table for Jupyter (1).csv'\n",
    "jupyter_table = pandas.read_csv(jupyter_renaming_table_file)\n",
    "print(jupyter_table.shape)\n",
    "print(jupyter_table.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1215, 2)\n",
      "  Rename_A Original_Name\n",
      "0     UMA1            N2\n",
      "1     UMA2            N4\n",
      "2     UMA3            N7\n",
      "3     UMA4            N9\n",
      "4     UMA5           N11\n"
     ]
    }
   ],
   "source": [
    "\n",
    "renaming_map = jupyter_table[['Rename_A','Original_Name']]\n",
    "print(renaming_map.shape)\n",
    "print(renaming_map.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "renaming_dict = {}\n",
    "for index, row in renaming_map.iterrows():\n",
    "    rename = renaming_map.at[index, 'Rename_A']\n",
    "    original = renaming_map.at[index, 'Original_Name']\n",
    "    renaming_dict[original] = rename\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13\n",
      "19\n",
      "47\n",
      "299\n"
     ]
    }
   ],
   "source": [
    "nanopore_assemblies = metadata_v5[metadata_v5['method'] == 'nanopore']\n",
    "pacbio_assemblies = metadata_v5[metadata_v5['method'] == 'pacbio']\n",
    "hybrid_assemblies = metadata_v5[metadata_v5['method'] == 'hybrid']\n",
    "illumina_assemblies = metadata_v5[metadata_v5['method'] == 'illumina']\n",
    "print(len(nanopore_assemblies))\n",
    "print(len(pacbio_assemblies))\n",
    "print(len(hybrid_assemblies))\n",
    "print(len(illumina_assemblies))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Isolate</th>\n",
       "      <th>MLST</th>\n",
       "      <th>RST_Type</th>\n",
       "      <th>Source</th>\n",
       "      <th>Origin</th>\n",
       "      <th>Location</th>\n",
       "      <th>skin/blood_pairs</th>\n",
       "      <th>Disseminated</th>\n",
       "      <th>method</th>\n",
       "      <th>OspC_Type</th>\n",
       "      <th>AT_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>MR616</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>G</td>\n",
       "      <td>AT7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>MC149</td>\n",
       "      <td>221</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>WI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>O</td>\n",
       "      <td>AT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>B348</td>\n",
       "      <td>19</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>E</td>\n",
       "      <td>AT5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>B331</td>\n",
       "      <td>16</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>I</td>\n",
       "      <td>AT9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>BL522</td>\n",
       "      <td>12</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>M</td>\n",
       "      <td>AT12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>B418</td>\n",
       "      <td>19</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>BL252</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>E</td>\n",
       "      <td>AT5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>B356</td>\n",
       "      <td>12</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>M</td>\n",
       "      <td>AT12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>B483</td>\n",
       "      <td>9</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ND</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>N</td>\n",
       "      <td>AT13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>MC104</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CSF</td>\n",
       "      <td>US</td>\n",
       "      <td>WI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>H</td>\n",
       "      <td>AT8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>B477</td>\n",
       "      <td>38</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>D</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>B500</td>\n",
       "      <td>15</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>I</td>\n",
       "      <td>AT9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>MC123</td>\n",
       "      <td>29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>WI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>L</td>\n",
       "      <td>AT10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>MC105</td>\n",
       "      <td>32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>WI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>H</td>\n",
       "      <td>AT8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>BL546</td>\n",
       "      <td>4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>H</td>\n",
       "      <td>AT8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>B247</td>\n",
       "      <td>3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>BL146</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>K</td>\n",
       "      <td>AT11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>BL224</td>\n",
       "      <td>3</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>K</td>\n",
       "      <td>AT11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>BL515</td>\n",
       "      <td>11</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>334</th>\n",
       "      <td>BL206</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>A</td>\n",
       "      <td>AT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>MR614</td>\n",
       "      <td>11</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Isolate MLST  RST_Type Source Origin Location skin/blood_pairs  \\\n",
       "64    MR616   14       3.0   Skin     US       NY              NaN   \n",
       "93    MC149  221       NaN   Skin     US       WI              NaN   \n",
       "124    B348   19       3.0   Skin     US       NY              NaN   \n",
       "142    B331   16       3.0   Skin     US       NY              NaN   \n",
       "144   BL522   12       3.0    NaN     NY      NaN              NaN   \n",
       "146    B418   19       3.0   Skin     US       NY            BL252   \n",
       "158    B356   12       3.0   Skin     US       NY              NaN   \n",
       "171    B483    9       2.0    NaN     NY      NaN              NaN   \n",
       "188   MC104    4       NaN    CSF     US       WI              NaN   \n",
       "193    B477   38       3.0   Skin     US       NY              NaN   \n",
       "216    B500   15       3.0   Skin     US       NY              NaN   \n",
       "273   MC123   29       NaN   Skin     US       WI              NaN   \n",
       "276   MC105   32       NaN   Skin     US       WI              NaN   \n",
       "285   BL546    4       2.0    NaN     NY      NaN              NaN   \n",
       "290    B247    3       2.0   Skin     US       NY            BL146   \n",
       "302   BL224    3       2.0    NaN     NY      NaN              NaN   \n",
       "309   BL515   11       3.0    NaN     NY      NaN              NaN   \n",
       "334   BL206    1       1.0    NaN     NY      NaN              NaN   \n",
       "345   MR614   11       3.0    NaN     NY      NaN              NaN   \n",
       "\n",
       "    Disseminated  method OspC_Type AT_type  \n",
       "64             L  pacbio         G     AT7  \n",
       "93             L  pacbio         O     AT3  \n",
       "124            D  pacbio         E     AT5  \n",
       "142            L  pacbio         I     AT9  \n",
       "144            D  pacbio         M    AT12  \n",
       "146            D  pacbio         E     AT5  \n",
       "158            L  pacbio         M    AT12  \n",
       "171           ND  pacbio         N    AT13  \n",
       "188            D  pacbio         H     AT8  \n",
       "193            L  pacbio         D     NaN  \n",
       "216            D  pacbio         I     AT9  \n",
       "273            L  pacbio         L    AT10  \n",
       "276            L  pacbio         H     AT8  \n",
       "285            D  pacbio         H     AT8  \n",
       "290            D  pacbio         K    AT11  \n",
       "302            D  pacbio         K    AT11  \n",
       "309            D  pacbio         C     NaN  \n",
       "334            D  pacbio         A     AT1  \n",
       "345            D  pacbio         C     NaN  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pacbio_assemblies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'newname'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m/opt/homebrew/lib/python3.11/site-packages/pandas/core/indexes/base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3802\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.11/site-packages/pandas/_libs/index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.11/site-packages/pandas/_libs/index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'newname'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[16], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m index, row \u001b[38;5;129;01min\u001b[39;00m pacbio_assemblies\u001b[38;5;241m.\u001b[39miterrows():\n\u001b[0;32m----> 2\u001b[0m     newname \u001b[38;5;241m=\u001b[39m \u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mnewname\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mP\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      3\u001b[0m     \u001b[38;5;66;03m#print(str(newname))\u001b[39;00m\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m newname \u001b[38;5;129;01min\u001b[39;00m renaming_dict:\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.11/site-packages/pandas/core/series.py:981\u001b[0m, in \u001b[0;36mSeries.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    978\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[key]\n\u001b[1;32m    980\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m key_is_scalar:\n\u001b[0;32m--> 981\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_value\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    983\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_hashable(key):\n\u001b[1;32m    984\u001b[0m     \u001b[38;5;66;03m# Otherwise index.get_value will raise InvalidIndexError\u001b[39;00m\n\u001b[1;32m    985\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    986\u001b[0m         \u001b[38;5;66;03m# For labels that don't resolve as scalars like tuples and frozensets\u001b[39;00m\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.11/site-packages/pandas/core/series.py:1089\u001b[0m, in \u001b[0;36mSeries._get_value\u001b[0;34m(self, label, takeable)\u001b[0m\n\u001b[1;32m   1086\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[label]\n\u001b[1;32m   1088\u001b[0m \u001b[38;5;66;03m# Similar to Index.get_value, but we do not fall back to positional\u001b[39;00m\n\u001b[0;32m-> 1089\u001b[0m loc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabel\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1090\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindex\u001b[38;5;241m.\u001b[39m_get_values_for_loc(\u001b[38;5;28mself\u001b[39m, loc, label)\n",
      "File \u001b[0;32m/opt/homebrew/lib/python3.11/site-packages/pandas/core/indexes/base.py:3804\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3802\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3804\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3805\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3806\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3808\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3809\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'newname'"
     ]
    }
   ],
   "source": [
    "for index, row in pacbio_assemblies.iterrows():\n",
    "    newname = row['newname'] + 'P'\n",
    "    #print(str(newname))\n",
    "    if newname in renaming_dict:\n",
    "        pacbio_assemblies.at[index, 'paired_sample'] = renaming_dict[newname].str[:-1]\n",
    "    else:\n",
    "        print(f\"NOT FOUND! {newname}\")\n",
    "\n",
    "#print(pacbio_assemblies)\n",
    "\n",
    "for index, row in hybrid_assemblies.iterrows():\n",
    "    hybrid_assemblies.at[index, 'newname'] = row['Isolate'] + 'H'\n",
    "\n",
    "for index, row in hybrid_assemblies.iterrows():\n",
    "    newname = row['newname']\n",
    "    #print(str(newname))\n",
    "    if newname in renaming_dict:\n",
    "        hybrid_assemblies.at[index, 'paired_sample'] = renaming_dict[newname].str[:-1]\n",
    "    else:\n",
    "        print(f\"NOT FOUND! {newname}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Isolate MLST  RST_Type Source Origin   Location skin/blood_pairs  \\\n",
      "64    MR616   14       3.0   Skin     US         NY              NaN   \n",
      "93    MC149  221       NaN   Skin     US         WI              NaN   \n",
      "124    B348   19       3.0   Skin     US         NY              NaN   \n",
      "142    B331   16       3.0   Skin     US         NY              NaN   \n",
      "144   BL522   12       3.0    NaN     NY        NaN              NaN   \n",
      "..      ...  ...       ...    ...    ...        ...              ...   \n",
      "347   URI19   59       1.0    NaN     US         RI              NaN   \n",
      "348   UMA13   59       1.0    NaN     US  Nantucket              NaN   \n",
      "349    UMA2    8       2.0    NaN     US  Nantucket              NaN   \n",
      "356   URI84   59       2.0    NaN     US         RI              NaN   \n",
      "357   URI20   14       3.0    NaN     US         RI              NaN   \n",
      "\n",
      "    Disseminated    method OspC_Type AT_type newname paired_sample  \n",
      "64             L    pacbio         G     AT7  MR616P       UNY203P  \n",
      "93             L    pacbio         O     AT3  MC149P       UWI283P  \n",
      "124            D    pacbio         E     AT5   B348P       UNY169P  \n",
      "142            L    pacbio         I     AT9   B331P           NaN  \n",
      "144            D    pacbio         M    AT12  BL522P      UNY1085P  \n",
      "..           ...       ...       ...     ...     ...           ...  \n",
      "347            L  illumina         B     AT2     NaN           NaN  \n",
      "348            L  illumina         B     NaN     NaN           NaN  \n",
      "349            L  illumina         F     AT6     NaN           NaN  \n",
      "356            L  illumina         B     NaN     NaN           NaN  \n",
      "357            L  illumina         G     AT7     NaN           NaN  \n",
      "\n",
      "[365 rows x 13 columns]\n"
     ]
    }
   ],
   "source": [
    "# Create a list of tables to merge\n",
    "tables_to_merge = [pacbio_assemblies, hybrid_assemblies, illumina_assemblies]\n",
    "\n",
    "# Merge the tables using pd.concat()\n",
    "merged_table = pandas.concat(tables_to_merge)\n",
    "\n",
    "# Print the merged table\n",
    "print(merged_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_table.to_csv('metadata_v6', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Isolate MLST  RST_Type Source Origin Location skin/blood_pairs  \\\n",
      "146    B418   19       3.0   Skin     US       NY            BL252   \n",
      "290    B247    3       2.0   Skin     US       NY            BL146   \n",
      "\n",
      "    Disseminated  method OspC_Type AT_type newname paired_sample  \n",
      "146            D  pacbio         E     AT5   B418P           NaN  \n",
      "290            D  pacbio         K    AT11   B247P       UNY149P  \n"
     ]
    }
   ],
   "source": [
    "print(merged_table.loc[merged_table['skin/blood_pairs'].notnull()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Repeated Original_Name_1\n",
      "0      UMA1              N2\n",
      "1      UMA2              N4\n",
      "2      UMA3              N7\n",
      "3      UMA4              N9\n",
      "4      UMA5             N11\n",
      "5      UMA7             N15\n",
      "6      UMA8             N17\n",
      "7     UMA11             N21\n",
      "8     UMA14             N30\n",
      "9     UMA15             N31\n",
      "10    UMA16             N32\n",
      "11    URI17              R5\n",
      "12    URI18              R6\n",
      "13    URI19              R7\n",
      "14    URI20              R8\n",
      "15    URI21              R9\n",
      "16    URI22             R10\n",
      "17    URI23             R11\n",
      "18    URI24             R12\n",
      "   Skin/Blood Pairs Original_Name_2\n",
      "0      UNY133(BL55)             B48\n",
      "1      UNY134(BL72)             B62\n",
      "2     UNY149(BL146)            B247\n",
      "3     UNY149(BL149)            B250\n",
      "4     UNY153(BL152)            B253\n",
      "5     UNY176(BL220)            B373\n",
      "6     UNY177(BL224)            B379\n",
      "7     UNY178(BL229)            B385\n",
      "8     UNY179(BL231)            B388\n",
      "9     UNY181(BL241)            B400\n",
      "10    UNY182(BL242)            B401\n",
      "11    UNY186(BL252)            B418\n",
      "12    UNY187(BL256)            B427\n",
      "13    UNY188(BL266)            B435\n",
      "14    UNY189(BL280)            B449\n",
      "15    UNY190(BL285)            B454\n",
      "16    UNY191(BL303)            B471\n",
      "17    UNY192(BL307)            B475\n",
      "18    UNY195(BL311)            B479\n",
      "19    UNY202(BL504)           MR604\n",
      "20    UNY204(BL522)           MR620\n",
      "21    UNY205(BL527)           MR625\n",
      "22    UNY207(BL538)           MR635\n",
      "23    UNY208(BL546)           MR641\n",
      "24    UNY209(BL562)           MR656\n",
      "25    UNY215(BL630)           MR715\n",
      "26    UNY216(BL631)           MR716\n",
      "27   ES289a(ES289b)  BSK-Hi 2206/13\n",
      "28   ES289b(ES289a)         2253/11\n"
     ]
    }
   ],
   "source": [
    "repeated_seqs_table = '/Users/mf019/bioinformatics/longread_GWAS/notebooks/metadata/lymeseq_original_sheets/LymeSeq_SampleTrack - Repeated Sequence Sheet.csv'\n",
    "repeated_seqs_df = pandas.read_csv(repeated_seqs_table)\n",
    "#repeated_seqs_df = repeated_seqs_df.drop(axis=1, labels='Unnamed: 2')\n",
    "repeated_seqs_df1 = repeated_seqs_df[['Repeated', 'Original_Name_1']]\n",
    "repeated_seqs_df2 = repeated_seqs_df[['Skin/Blood Pairs', 'Original_Name_2']]\n",
    "repeated_seqs_df1 = repeated_seqs_df1.dropna()\n",
    "print(repeated_seqs_df1)\n",
    "print(repeated_seqs_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/59/fs0fxd1x1wx1j_stw1ccjbcr0000gp/T/ipykernel_60919/2191308876.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  repeated_seqs_df2[['renamed_pair', 'original_name_pair']] = repeated_seqs_df2['Skin/Blood Pairs'].str.split('(', expand=True)\n",
      "/var/folders/59/fs0fxd1x1wx1j_stw1ccjbcr0000gp/T/ipykernel_60919/2191308876.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  repeated_seqs_df2[['renamed_pair', 'original_name_pair']] = repeated_seqs_df2['Skin/Blood Pairs'].str.split('(', expand=True)\n",
      "/var/folders/59/fs0fxd1x1wx1j_stw1ccjbcr0000gp/T/ipykernel_60919/2191308876.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  repeated_seqs_df2['original_name_pair'] = repeated_seqs_df2['original_name_pair'].str.rstrip(')')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Skin/Blood Pairs</th>\n",
       "      <th>Original_Name_2</th>\n",
       "      <th>renamed_pair</th>\n",
       "      <th>original_name_pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UNY133(BL55)</td>\n",
       "      <td>B48</td>\n",
       "      <td>UNY133</td>\n",
       "      <td>BL55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UNY134(BL72)</td>\n",
       "      <td>B62</td>\n",
       "      <td>UNY134</td>\n",
       "      <td>BL72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UNY149(BL146)</td>\n",
       "      <td>B247</td>\n",
       "      <td>UNY149</td>\n",
       "      <td>BL146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UNY149(BL149)</td>\n",
       "      <td>B250</td>\n",
       "      <td>UNY149</td>\n",
       "      <td>BL149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UNY153(BL152)</td>\n",
       "      <td>B253</td>\n",
       "      <td>UNY153</td>\n",
       "      <td>BL152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>UNY176(BL220)</td>\n",
       "      <td>B373</td>\n",
       "      <td>UNY176</td>\n",
       "      <td>BL220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>UNY177(BL224)</td>\n",
       "      <td>B379</td>\n",
       "      <td>UNY177</td>\n",
       "      <td>BL224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>UNY178(BL229)</td>\n",
       "      <td>B385</td>\n",
       "      <td>UNY178</td>\n",
       "      <td>BL229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>UNY179(BL231)</td>\n",
       "      <td>B388</td>\n",
       "      <td>UNY179</td>\n",
       "      <td>BL231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>UNY181(BL241)</td>\n",
       "      <td>B400</td>\n",
       "      <td>UNY181</td>\n",
       "      <td>BL241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>UNY182(BL242)</td>\n",
       "      <td>B401</td>\n",
       "      <td>UNY182</td>\n",
       "      <td>BL242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>UNY186(BL252)</td>\n",
       "      <td>B418</td>\n",
       "      <td>UNY186</td>\n",
       "      <td>BL252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>UNY187(BL256)</td>\n",
       "      <td>B427</td>\n",
       "      <td>UNY187</td>\n",
       "      <td>BL256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>UNY188(BL266)</td>\n",
       "      <td>B435</td>\n",
       "      <td>UNY188</td>\n",
       "      <td>BL266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>UNY189(BL280)</td>\n",
       "      <td>B449</td>\n",
       "      <td>UNY189</td>\n",
       "      <td>BL280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>UNY190(BL285)</td>\n",
       "      <td>B454</td>\n",
       "      <td>UNY190</td>\n",
       "      <td>BL285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>UNY191(BL303)</td>\n",
       "      <td>B471</td>\n",
       "      <td>UNY191</td>\n",
       "      <td>BL303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>UNY192(BL307)</td>\n",
       "      <td>B475</td>\n",
       "      <td>UNY192</td>\n",
       "      <td>BL307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>UNY195(BL311)</td>\n",
       "      <td>B479</td>\n",
       "      <td>UNY195</td>\n",
       "      <td>BL311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>UNY202(BL504)</td>\n",
       "      <td>MR604</td>\n",
       "      <td>UNY202</td>\n",
       "      <td>BL504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>UNY204(BL522)</td>\n",
       "      <td>MR620</td>\n",
       "      <td>UNY204</td>\n",
       "      <td>BL522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>UNY205(BL527)</td>\n",
       "      <td>MR625</td>\n",
       "      <td>UNY205</td>\n",
       "      <td>BL527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>UNY207(BL538)</td>\n",
       "      <td>MR635</td>\n",
       "      <td>UNY207</td>\n",
       "      <td>BL538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>UNY208(BL546)</td>\n",
       "      <td>MR641</td>\n",
       "      <td>UNY208</td>\n",
       "      <td>BL546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>UNY209(BL562)</td>\n",
       "      <td>MR656</td>\n",
       "      <td>UNY209</td>\n",
       "      <td>BL562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>UNY215(BL630)</td>\n",
       "      <td>MR715</td>\n",
       "      <td>UNY215</td>\n",
       "      <td>BL630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>UNY216(BL631)</td>\n",
       "      <td>MR716</td>\n",
       "      <td>UNY216</td>\n",
       "      <td>BL631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ES289a(ES289b)</td>\n",
       "      <td>BSK-Hi 2206/13</td>\n",
       "      <td>ES289a</td>\n",
       "      <td>ES289b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ES289b(ES289a)</td>\n",
       "      <td>2253/11</td>\n",
       "      <td>ES289b</td>\n",
       "      <td>ES289a</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Skin/Blood Pairs Original_Name_2 renamed_pair original_name_pair\n",
       "0      UNY133(BL55)             B48       UNY133               BL55\n",
       "1      UNY134(BL72)             B62       UNY134               BL72\n",
       "2     UNY149(BL146)            B247       UNY149              BL146\n",
       "3     UNY149(BL149)            B250       UNY149              BL149\n",
       "4     UNY153(BL152)            B253       UNY153              BL152\n",
       "5     UNY176(BL220)            B373       UNY176              BL220\n",
       "6     UNY177(BL224)            B379       UNY177              BL224\n",
       "7     UNY178(BL229)            B385       UNY178              BL229\n",
       "8     UNY179(BL231)            B388       UNY179              BL231\n",
       "9     UNY181(BL241)            B400       UNY181              BL241\n",
       "10    UNY182(BL242)            B401       UNY182              BL242\n",
       "11    UNY186(BL252)            B418       UNY186              BL252\n",
       "12    UNY187(BL256)            B427       UNY187              BL256\n",
       "13    UNY188(BL266)            B435       UNY188              BL266\n",
       "14    UNY189(BL280)            B449       UNY189              BL280\n",
       "15    UNY190(BL285)            B454       UNY190              BL285\n",
       "16    UNY191(BL303)            B471       UNY191              BL303\n",
       "17    UNY192(BL307)            B475       UNY192              BL307\n",
       "18    UNY195(BL311)            B479       UNY195              BL311\n",
       "19    UNY202(BL504)           MR604       UNY202              BL504\n",
       "20    UNY204(BL522)           MR620       UNY204              BL522\n",
       "21    UNY205(BL527)           MR625       UNY205              BL527\n",
       "22    UNY207(BL538)           MR635       UNY207              BL538\n",
       "23    UNY208(BL546)           MR641       UNY208              BL546\n",
       "24    UNY209(BL562)           MR656       UNY209              BL562\n",
       "25    UNY215(BL630)           MR715       UNY215              BL630\n",
       "26    UNY216(BL631)           MR716       UNY216              BL631\n",
       "27   ES289a(ES289b)  BSK-Hi 2206/13       ES289a             ES289b\n",
       "28   ES289b(ES289a)         2253/11       ES289b             ES289a"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "repeated_seqs_df2[['renamed_pair', 'original_name_pair']] = repeated_seqs_df2['Skin/Blood Pairs'].str.split('(', expand=True)\n",
    "repeated_seqs_df2['original_name_pair'] = repeated_seqs_df2['original_name_pair'].str.rstrip(')')\n",
    "repeated_seqs_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [],
   "source": [
    "# okay now we need to pull the isolate/sample ids from the notes column\n",
    "wgs_table_file = '/Users/mf019/bioinformatics/longread_GWAS/notebooks/metadata/lymeseq_original_sheets/LymeSeq_SampleTrack - Whole Genome Sequenced Table (1).csv'\n",
    "wgs_table_df = pandas.read_csv(wgs_table_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    New Rename? Rename_A Seq Method? Original_Name Source OspC_Type  MLST  \\\n",
      "47          NaN      NaN      PacBio         BL206  Blood         A   1.0   \n",
      "49          NaN      NaN      PacBio         BL515  Blood         C  11.0   \n",
      "50          NaN      NaN      PacBio         BL522  Blood         M  12.0   \n",
      "51          NaN      NaN      PacBio         BL546  Blood         H   4.0   \n",
      "56          NaN      NaN      PacBio         MR614   Skin         C  11.0   \n",
      "58          NaN   UNY208      PacBio         MR641   Skin         H   4.0   \n",
      "64          NaN      NaN      PacBio         BL224  Blood         K   3.0   \n",
      "\n",
      "   Disseminated                   Notes Skin/Blood Pairs  \n",
      "47            D  Skin isolate is UNY170           UNY170  \n",
      "49            D         Pair with MR614            MR614  \n",
      "50            D  Skin isolate is UNY204           UNY204  \n",
      "51            D         Pair with MR641            MR641  \n",
      "56            D         Pair with BL515            BL515  \n",
      "58            D         Pair with BL546            BL546  \n",
      "64            D  Skin isolate is UNY177           UNY177  \n"
     ]
    }
   ],
   "source": [
    "wgs_table_df['Skin/Blood Pairs'] = wgs_table_df['Notes'].str.split(' ').str[-1]\n",
    "wgs_table_filtered_df = wgs_table_df[['Rename_A','Original_Name', 'Source', 'Seq Method?', 'OspC_Type', 'MLST', 'Disseminated', 'Skin/Blood Pairs']]\n",
    "print(wgs_table_df[wgs_table_df['Skin/Blood Pairs'].notna()])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0         NaN\n",
      "1         NaN\n",
      "2         NaN\n",
      "3       UCT29\n",
      "4       UCT30\n",
      "       ...   \n",
      "61     UNY172\n",
      "62        NaN\n",
      "63     UNY193\n",
      "64    UNY1038\n",
      "65        NaN\n",
      "Name: Original_Name, Length: 66, dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(wgs_table_filtered_df['Original_Name'].map(renaming_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47\n",
      "20\n"
     ]
    }
   ],
   "source": [
    "import glob\n",
    "# Okay maybe that helped some?\n",
    "# Turns out that the repeated sequence sheet is actually two tables in a trench coat.\n",
    "assembly_dir = '/Users/mf019/bioinformatics/longread_GWAS/assemblies'\n",
    "list_of_hybrids = glob.glob(f'{assembly_dir}/hybrid/contigs/*.fasta')\n",
    "list_of_pacbio = glob.glob(f'{assembly_dir}/pacbio/contigs/*.fasta')\n",
    "print(len(list_of_hybrids))\n",
    "print(len(list_of_pacbio))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 438,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file:K7H, rename:URI42H\n",
      "file:K60H, rename:URI87H\n",
      "file:K25H, rename:URI36H\n",
      "file:K71H, rename:URI91H\n",
      "file:K11H, rename:UCT30H\n",
      "file:K22H, rename:URI34H\n",
      "file:K19H, rename:URI48H\n",
      "file:K76H, rename:UCT113H\n",
      "file:K16H, rename:URI33H\n",
      "file:K82H, rename:URI117H\n",
      "file:K10H, rename:UCT31H\n",
      "file:K84H, rename:UCT96H\n",
      "file:K70H, rename:UCT110H\n",
      "file:K24H, rename:UCT35H\n",
      "file:K6H, rename:URI41H\n",
      "file:K17H, rename:URI46H\n",
      "file:K83H, rename:URI118H\n",
      "file:K69H, rename:UCT109H\n",
      "file:EU14_B89613H, rename:ESI403H\n",
      "file:L113H, rename:ESI425H\n",
      "file:K18H, rename:URI47H\n",
      "file:K1H, rename:UCT29H\n",
      "file:K66H, rename:URI107H\n",
      "file:K80H, rename:UCT95H\n",
      "file:K74H, rename:URI111H\n",
      "file:K2H, rename:URI39H\n",
      "file:K13H, rename:URI45H\n",
      "file:K56H, rename:URI86H\n",
      "file:K73H, rename:URI93H\n",
      "file:K59H, rename:URI103H\n",
      "file:EU57_B190413H, rename:ESI26H\n",
      "file:K62H, rename:URI88H\n",
      "file:K3H, rename:URI40H\n",
      "file:K64H, rename:URI89H\n",
      "file:K21H, rename:UCT50H\n",
      "file:K90H, rename:UCT124H\n",
      "file:XD1H, rename:XYZ459H\n",
      "file:EU51_B192411H, rename:ESI361\n",
      "file:K30H, rename:URI56H\n",
      "file:K75H, rename:URI112H\n",
      "file:K15H, rename:UCT32H\n",
      "file:K89H, rename:K89H\n",
      "file:K58H, rename:URI102H\n",
      "file:K72H, rename:UCT92H\n",
      "file:K86H, rename:URI120H\n",
      "file:K12H, rename:URI44H\n",
      "file:K57H, rename:URI101H\n",
      "file:B247P, rename:UNY149P\n",
      "file:B477P, rename:UNY193P\n",
      "file:MC149P, rename:UWI283P\n",
      "file:MR614P, rename:UNY1128P\n",
      "file:BL546P, rename:UNY1090P\n",
      "file:MR616P, rename:UNY203P\n",
      "file:BL522P, rename:UNY1085P\n",
      "file:B348P, rename:UNY169P\n",
      "file:BL515P, rename:UNY1083P\n",
      "file:B500P, rename:B500P\n",
      "file:MC105P, rename:UWI248P\n",
      "file:B483P, rename:UNY990P\n",
      "file:MC123P, rename:UWI263P\n",
      "file:MR641P, rename:UNY208P\n",
      "file:BL206P, rename:UNY1032P\n",
      "file:B331P, rename:B331P\n",
      "file:MC104P, rename:UWI247P\n",
      "file:BL224P, rename:UNY1038P\n",
      "file:B418P, rename:B418P\n",
      "file:B356P, rename:UNY172P\n"
     ]
    }
   ],
   "source": [
    "hybrid_asm_naming_df = pandas.DataFrame(columns=['original', 'renamed'])\n",
    "hybrid_names = []\n",
    "for asm in list_of_hybrids:\n",
    "    filename = asm.split('/')[-1].split('.')[0] + \"H\"\n",
    "    if filename in renaming_dict:\n",
    "        rename = renaming_dict[filename]\n",
    "    else:\n",
    "        rename = filename\n",
    "    hybrid_asm_naming_df.loc[len(hybrid_asm_naming_df)] = [filename, rename]\n",
    "    hybrid_names.append(rename)\n",
    "    print(f'file:{filename}, rename:{rename}')\n",
    "\n",
    "pacbio_asm_naming_df = pandas.DataFrame(columns=['original', 'renamed'])\n",
    "pacbio_names = []\n",
    "for asm in list_of_pacbio:\n",
    "    filename = asm.split('/')[-1].split('.')[0] + \"P\"\n",
    "    if filename in renaming_dict:\n",
    "        rename = renaming_dict[filename]\n",
    "    else:\n",
    "        rename = filename\n",
    "    pacbio_asm_naming_df.loc[len(pacbio_asm_naming_df)] = [filename, rename]\n",
    "    pacbio_names.append(rename)\n",
    "    print(f'file:{filename}, rename:{rename}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 439,
   "metadata": {},
   "outputs": [],
   "source": [
    "hybrid_asm_naming_df\n",
    "hybrid_rename_sed_script = \"\"\n",
    "for index, row in hybrid_asm_naming_df.iterrows():\n",
    "    original_name = f\"{row['original']}.hybrid.assembly.contigs.fasta\"\n",
    "    renamed_name = f\"{row['renamed']}.fasta\"\n",
    "    hybrid_rename_sed_script += f\"s/{original_name}/{renamed_name}/g\\n\"\n",
    "\n",
    "with open('hybrid_rename_script.sed', 'w') as file:\n",
    "    file.write(hybrid_rename_sed_script)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 440,
   "metadata": {},
   "outputs": [],
   "source": [
    "pacbio_asm_naming_df\n",
    "pacbio_rename_sed_script = \"\"\n",
    "for index, row in pacbio_asm_naming_df.iterrows():\n",
    "    original_name = f\"{row['original']}.pacbio.assembly.contigs.fasta\"\n",
    "    renamed_name = f\"{row['renamed']}.fasta\"\n",
    "    pacbio_rename_sed_script += f\"s/{original_name}/{renamed_name}/g\\n\"\n",
    "\n",
    "with open('pacbio_rename_script.sed', 'w') as file:\n",
    "    file.write(pacbio_rename_sed_script)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 441,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ok now I need to parse the jupyter renaming table as well as the rest of everything to identify the paired assemblies\n",
    "# and then separate those out into a single final table....\n",
    "# Then I can deal with the renaming to ensure that the same sample is compared properly.\n",
    "# Example: UNY123 and UNY456 are blood/skin pairs from the same infection.\n",
    "# Lets also say those were redone with pacbio,\n",
    "# thus, UNY123_B_I and UNY123_S_I\n",
    "# and , UNY123_B_P and UNY123_S_P ??\n",
    "# Does this make sense?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#What do I have?\n",
    "- [x] list of assemblies used (filenames)\n",
    "- [x] list of pacbio assemblies and their corresponding names from metadata\n",
    "- [x] list of hybrid assemblies and their corresponding names from metadata\n",
    "- [x] bunch of metadata for all illumina assemblies\n",
    "- [x] list of blood/skin pairs (from metadata)\n",
    "- [x] list of resequenced samples (from metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 442,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pacbio assemblies: 20\n",
      "['UNY149P', 'UNY193P', 'UWI283P', 'UNY1128P', 'UNY1090P', 'UNY203P', 'UNY1085P', 'UNY169P', 'UNY1083P', 'B500P', 'UWI248P', 'UNY990P', 'UWI263P', 'UNY208P', 'UNY1032P', 'B331P', 'UWI247P', 'UNY1038P', 'B418P', 'UNY172P']\n",
      "hybrid assemblies: 47\n",
      "['URI42H', 'URI87H', 'URI36H', 'URI91H', 'UCT30H', 'URI34H', 'URI48H', 'UCT113H', 'URI33H', 'URI117H', 'UCT31H', 'UCT96H', 'UCT110H', 'UCT35H', 'URI41H', 'URI46H', 'URI118H', 'UCT109H', 'ESI403H', 'ESI425H', 'URI47H', 'UCT29H', 'URI107H', 'UCT95H', 'URI111H', 'URI39H', 'URI45H', 'URI86H', 'URI93H', 'URI103H', 'ESI26H', 'URI88H', 'URI40H', 'URI89H', 'UCT50H', 'UCT124H', 'XYZ459H', 'ESI361', 'URI56H', 'URI112H', 'UCT32H', 'K89H', 'URI102H', 'UCT92H', 'URI120H', 'URI44H', 'URI101H']\n"
     ]
    }
   ],
   "source": [
    "print(f'pacbio assemblies: {len(pacbio_names)}')\n",
    "print(pacbio_names)\n",
    "print(f'hybrid assemblies: {len(hybrid_names)}')\n",
    "print(hybrid_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 443,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ok so now lets pull in that jupyter table and pull all of the samples in...\n",
    "renaming_scheme_jupyter_file = '/Users/mf019/bioinformatics/longread_GWAS/notebooks/metadata/lymeseq_original_sheets/LymeSeq_SampleTrack - Renaming Scheme Table for Jupyter (1).csv'\n",
    "jupyter_df = pandas.read_csv(renaming_scheme_jupyter_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 444,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/59/fs0fxd1x1wx1j_stw1ccjbcr0000gp/T/ipykernel_60919/2462978438.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filtered_jupyter_df['original_pair_name'] = filtered_jupyter_df['Original_Name'].str[:-1:]\n",
      "/var/folders/59/fs0fxd1x1wx1j_stw1ccjbcr0000gp/T/ipykernel_60919/2462978438.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filtered_jupyter_df['renamed_pair_name'] = filtered_jupyter_df['Rename_A'].str[:-1:]\n"
     ]
    }
   ],
   "source": [
    "filtered_jupyter_df = jupyter_df[jupyter_df['Rename_A'].isin(pacbio_names) | jupyter_df['Rename_A'].isin(hybrid_names)]\n",
    "filtered_jupyter_df['original_pair_name'] = filtered_jupyter_df['Original_Name'].str[:-1:]\n",
    "filtered_jupyter_df['renamed_pair_name'] = filtered_jupyter_df['Rename_A'].str[:-1:]\n",
    "renamed_names_pbh = list(filtered_jupyter_df['renamed_pair_name'])\n",
    "original_names_pbh = list(filtered_jupyter_df['original_pair_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 445,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Isolate</th>\n",
       "      <th>MLST</th>\n",
       "      <th>RST_Type</th>\n",
       "      <th>Source</th>\n",
       "      <th>Origin</th>\n",
       "      <th>Location</th>\n",
       "      <th>skin/blood_pairs</th>\n",
       "      <th>Disseminated</th>\n",
       "      <th>method</th>\n",
       "      <th>OspC_Type</th>\n",
       "      <th>AT_type</th>\n",
       "      <th>newname</th>\n",
       "      <th>paired_sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>MR616</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>G</td>\n",
       "      <td>AT7</td>\n",
       "      <td>MR616P</td>\n",
       "      <td>UNY203P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>MC149</td>\n",
       "      <td>221</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>WI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>O</td>\n",
       "      <td>AT3</td>\n",
       "      <td>MC149P</td>\n",
       "      <td>UWI283P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>B348</td>\n",
       "      <td>19</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>E</td>\n",
       "      <td>AT5</td>\n",
       "      <td>B348P</td>\n",
       "      <td>UNY169P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>BL522</td>\n",
       "      <td>12</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>M</td>\n",
       "      <td>AT12</td>\n",
       "      <td>BL522P</td>\n",
       "      <td>UNY1085P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>B356</td>\n",
       "      <td>12</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>M</td>\n",
       "      <td>AT12</td>\n",
       "      <td>B356P</td>\n",
       "      <td>UNY172P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>URI111</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>US</td>\n",
       "      <td>RI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>illumina</td>\n",
       "      <td>A</td>\n",
       "      <td>AT1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>URI91</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>US</td>\n",
       "      <td>RI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>illumina</td>\n",
       "      <td>A</td>\n",
       "      <td>AT1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>URI88</td>\n",
       "      <td>4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>US</td>\n",
       "      <td>RI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>illumina</td>\n",
       "      <td>H</td>\n",
       "      <td>AT8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>URI101</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>US</td>\n",
       "      <td>RI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>illumina</td>\n",
       "      <td>A</td>\n",
       "      <td>AT1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>URI89</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>US</td>\n",
       "      <td>RI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>illumina</td>\n",
       "      <td>A</td>\n",
       "      <td>AT1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows  13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Isolate MLST  RST_Type Source Origin Location skin/blood_pairs  \\\n",
       "64    MR616   14       3.0   Skin     US       NY              NaN   \n",
       "93    MC149  221       NaN   Skin     US       WI              NaN   \n",
       "124    B348   19       3.0   Skin     US       NY              NaN   \n",
       "144   BL522   12       3.0    NaN     NY      NaN              NaN   \n",
       "158    B356   12       3.0   Skin     US       NY              NaN   \n",
       "..      ...  ...       ...    ...    ...      ...              ...   \n",
       "314  URI111    1       1.0    NaN     US       RI              NaN   \n",
       "325   URI91    1       1.0    NaN     US       RI              NaN   \n",
       "341   URI88    4       2.0    NaN     US       RI              NaN   \n",
       "342  URI101    1       1.0    NaN     US       RI              NaN   \n",
       "344   URI89    1       1.0    NaN     US       RI              NaN   \n",
       "\n",
       "    Disseminated    method OspC_Type AT_type newname paired_sample  \n",
       "64             L    pacbio         G     AT7  MR616P       UNY203P  \n",
       "93             L    pacbio         O     AT3  MC149P       UWI283P  \n",
       "124            D    pacbio         E     AT5   B348P       UNY169P  \n",
       "144            D    pacbio         M    AT12  BL522P      UNY1085P  \n",
       "158            L    pacbio         M    AT12   B356P       UNY172P  \n",
       "..           ...       ...       ...     ...     ...           ...  \n",
       "314            D  illumina         A     AT1     NaN           NaN  \n",
       "325            L  illumina         A     AT1     NaN           NaN  \n",
       "341            L  illumina         H     AT8     NaN           NaN  \n",
       "342            D  illumina         A     AT1     NaN           NaN  \n",
       "344            D  illumina         A     AT1     NaN           NaN  \n",
       "\n",
       "[111 rows x 13 columns]"
      ]
     },
     "execution_count": 445,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_table1 = merged_table[merged_table['Isolate'].isin(renamed_names_pbh) | merged_table['Isolate'].isin(original_names_pbh)]\n",
    "filtered_table1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 446,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47\n",
      "20\n"
     ]
    }
   ],
   "source": [
    "print(len(list_of_hybrids))\n",
    "print(len(list_of_pacbio))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Isolate</th>\n",
       "      <th>MLST</th>\n",
       "      <th>RST_Type</th>\n",
       "      <th>Source</th>\n",
       "      <th>Origin</th>\n",
       "      <th>Location</th>\n",
       "      <th>skin/blood_pairs</th>\n",
       "      <th>Disseminated</th>\n",
       "      <th>method</th>\n",
       "      <th>OspC_Type</th>\n",
       "      <th>AT_type</th>\n",
       "      <th>newname</th>\n",
       "      <th>paired_sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MR616</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>G</td>\n",
       "      <td>AT7</td>\n",
       "      <td>MR616P</td>\n",
       "      <td>UNY203P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MC149</td>\n",
       "      <td>221</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>WI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>O</td>\n",
       "      <td>AT3</td>\n",
       "      <td>MC149P</td>\n",
       "      <td>UWI283P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B348</td>\n",
       "      <td>19</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>E</td>\n",
       "      <td>AT5</td>\n",
       "      <td>B348P</td>\n",
       "      <td>UNY169P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B331</td>\n",
       "      <td>16</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Skin</td>\n",
       "      <td>US</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>I</td>\n",
       "      <td>AT9</td>\n",
       "      <td>B331P</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BL522</td>\n",
       "      <td>12</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D</td>\n",
       "      <td>pacbio</td>\n",
       "      <td>M</td>\n",
       "      <td>AT12</td>\n",
       "      <td>BL522P</td>\n",
       "      <td>UNY1085P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360</th>\n",
       "      <td>URI19</td>\n",
       "      <td>59</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>US</td>\n",
       "      <td>RI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>illumina</td>\n",
       "      <td>B</td>\n",
       "      <td>AT2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>UMA13</td>\n",
       "      <td>59</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>US</td>\n",
       "      <td>Nantucket</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>illumina</td>\n",
       "      <td>B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>UMA2</td>\n",
       "      <td>8</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>US</td>\n",
       "      <td>Nantucket</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>illumina</td>\n",
       "      <td>F</td>\n",
       "      <td>AT6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>URI84</td>\n",
       "      <td>59</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>US</td>\n",
       "      <td>RI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>illumina</td>\n",
       "      <td>B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>URI20</td>\n",
       "      <td>14</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>US</td>\n",
       "      <td>RI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>illumina</td>\n",
       "      <td>G</td>\n",
       "      <td>AT7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>365 rows  13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Isolate MLST  RST_Type Source Origin   Location skin/blood_pairs  \\\n",
       "0     MR616   14       3.0   Skin     US         NY              NaN   \n",
       "1     MC149  221       NaN   Skin     US         WI              NaN   \n",
       "2      B348   19       3.0   Skin     US         NY              NaN   \n",
       "3      B331   16       3.0   Skin     US         NY              NaN   \n",
       "4     BL522   12       3.0    NaN     NY        NaN              NaN   \n",
       "..      ...  ...       ...    ...    ...        ...              ...   \n",
       "360   URI19   59       1.0    NaN     US         RI              NaN   \n",
       "361   UMA13   59       1.0    NaN     US  Nantucket              NaN   \n",
       "362    UMA2    8       2.0    NaN     US  Nantucket              NaN   \n",
       "363   URI84   59       2.0    NaN     US         RI              NaN   \n",
       "364   URI20   14       3.0    NaN     US         RI              NaN   \n",
       "\n",
       "    Disseminated    method OspC_Type AT_type newname paired_sample  \n",
       "0              L    pacbio         G     AT7  MR616P       UNY203P  \n",
       "1              L    pacbio         O     AT3  MC149P       UWI283P  \n",
       "2              D    pacbio         E     AT5   B348P       UNY169P  \n",
       "3              L    pacbio         I     AT9   B331P           NaN  \n",
       "4              D    pacbio         M    AT12  BL522P      UNY1085P  \n",
       "..           ...       ...       ...     ...     ...           ...  \n",
       "360            L  illumina         B     AT2     NaN           NaN  \n",
       "361            L  illumina         B     NaN     NaN           NaN  \n",
       "362            L  illumina         F     AT6     NaN           NaN  \n",
       "363            L  illumina         B     NaN     NaN           NaN  \n",
       "364            L  illumina         G     AT7     NaN           NaN  \n",
       "\n",
       "[365 rows x 13 columns]"
      ]
     },
     "execution_count": 448,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_v6_df = pandas.read_csv('/Users/mf019/bioinformatics/longread_GWAS/notebooks/metadata/pair_finding/metadata_v6')\n",
    "metadata_v6_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 453,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset1 = metadata_v6_df[metadata_v6_df['method'].isin(['pacbio', 'hybrid'])]\n",
    "subset1a = jupyter_df[jupyter_df['Rename_A'].isin(subset1['paired_sample'])]\n",
    "subset1b = jupyter_df[jupyter_df['Rename_A'].isin(subset1['paired_sample'].str[:-1])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
