{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9f1ce2a3-b8e3-497b-96dc-2df1d2835ea9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import shutil\n",
    "import pandas\n",
    "import pprint\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "from Bio            import Entrez\n",
    "from Bio            import SeqIO\n",
    "from Bio.Blast      import NCBIXML\n",
    "from collections import defaultdict\n",
    "\n",
    "############################################\n",
    "Entrez.email = \"mfoster11@mgh.harvard.edu\" #\n",
    "############################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "59bd523e-b226-4794-b662-3cef4ae88f81",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Appl Environ Microbiol. 2008 Jun 6;74(16):5008–5014. doi: 10.1128/AEM.00479-08\n",
    "# \"... For ospC, alignments were made on the translated amino acid sequences and then back-translated to nucleotide sequences to ensure in-frame nucleotide alignment...\"\n",
    "# I was told just use nucl identity since most are incomplete and frame cannot be guaranteed.\n",
    "\n",
    "iras_types = { \n",
    "    'A' : 'X69596',   'Ba' : 'EF537413', 'Bb' : 'NC_011724', 'C' : 'DQ437462',\n",
    "    'Da' : 'AF029863', 'Db' : 'GQ478283',  'E' : 'AY275221', 'Fa' : 'AY275225',\n",
    "    'Fb' : 'EF537433', 'Fc' : 'GQ478285',  'G' : 'AY275223', 'Ha' : 'EU377781',\n",
    "    'Hb' : 'GQ478286', 'Ia' : 'AY275219', 'Ib' : 'EU377752',  'J' : 'CP001535',\n",
    "     'K' : 'AY275214',  'L' : 'EU375832',  'M' : 'CP001550',  'N' : 'EU377775',\n",
    "     'O' : 'FJ997281',  'T' : 'AY275222', 'Ua' : 'EU377769', 'Ub' : 'GQ478287',\n",
    "    'A3' : 'EF592541', 'B3' : 'EF592542', 'C3' : 'EF592543', 'D3' : 'EF592544',\n",
    "    'E3' : 'EF592545', 'F3' : 'EF592547', 'H3' : 'FJ932733', 'I3' : 'FJ932734',\n",
    "    'B_bissettii25015': 'U04282'\n",
    "}\n",
    "\n",
    "#Appl Environ Microbiol. 2013 Mar;79(5):1444–1453. doi: 10.1128/AEM.02749-12\n",
    "ospc_ref_file = \"reference_types.tsv\"\n",
    "hanincovas_types = pandas.read_csv(ospc_ref_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8c083def-3b91-4db8-ad56-a93a84b540d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs = \"ref_ospc_seqs_v2\"\n",
    "if not os.path.exists(seqs):\n",
    "    os.mkdir(seqs)\n",
    "\n",
    "type_to_acc = defaultdict(list)\n",
    "\n",
    "for _, data in hanincovas_types.iterrows():\n",
    "    ospc_type = data['ospC_type']\n",
    "    gb_acc = data[\"gb_acc\"]\n",
    "    type_to_acc[ospc_type].append(gb_acc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a1742a6a-393c-484f-b727-7ad114fc5224",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'AF029863': {'dict1': ['D'], 'dict2': ['Da']},\n",
      " 'DQ437462': {'dict1': ['C'], 'dict2': ['C']},\n",
      " 'EF537413': {'dict1': ['B1'], 'dict2': ['Ba']},\n",
      " 'FJ932733': {'dict1': ['H3'], 'dict2': ['H3']},\n",
      " 'FJ932734': {'dict1': ['I3'], 'dict2': ['I3']}}\n"
     ]
    }
   ],
   "source": [
    "value_to_keys = {}\n",
    "\n",
    "for key, values in type_to_acc.items():\n",
    "    for acc in values:\n",
    "        if acc not in value_to_keys:\n",
    "            value_to_keys[acc] = {'dict1': [key], 'dict2': []}\n",
    "        else:\n",
    "            value_to_keys[acc]['dict1'].append(key)\n",
    "\n",
    "for key, value in iras_types.items():\n",
    "    if value not in value_to_keys:\n",
    "        value_to_keys[value] = {'dict1': [], 'dict2': [key]}\n",
    "    else:\n",
    "        value_to_keys[value]['dict2'].append(key)\n",
    "\n",
    "unique_values = set(list(item for value in type_to_acc.values() for item in value)) | set(iras_types.values())\n",
    "\n",
    "duplicates = {\n",
    "    value: keys for value, keys in value_to_keys.items() if keys['dict1'] and keys['dict2']\n",
    "}\n",
    "\n",
    "pprint.pprint(duplicates)\n",
    "# {'AF029863': {'dict1': ['D'], 'dict2': ['Da']},\n",
    "# 'DQ437462': {'dict1': ['C'], 'dict2': ['C']},\n",
    "# 'EF537413': {'dict1': ['B1'], 'dict2': ['Ba']},\n",
    "# 'FJ932733': {'dict1': ['H3'], 'dict2': ['H3']},\n",
    "# 'FJ932734': {'dict1': ['I3'], 'dict2': ['I3']}}\n",
    "\n",
    "#print(unique_values)\n",
    "\n",
    "acc_to_types = {\n",
    "    value: key for value, key in value_to_keys.items() #if value not in duplicates.keys()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "980bfaf3-a305-4e3d-a083-61f7016a4104",
   "metadata": {},
   "outputs": [],
   "source": [
    "weird_types = []\n",
    "fixed_type_to_ref = defaultdict(list)\n",
    "for acc, dicts in acc_to_types.items():\n",
    "    types = [val for val in dicts.values() if val != []]\n",
    "    if len(types) > 1:\n",
    "        if types[0] == types[1]:\n",
    "            #pass This is why my counts were off!! \n",
    "            key = types[0][0]\n",
    "            fixed_type_to_ref[key].append(acc)\n",
    "        else:\n",
    "            #print(acc, types)\n",
    "            weird_types.append(f\"{acc}: {types}\")\n",
    "            \n",
    "    else:\n",
    "        #print(acc, *types[0])\n",
    "        key = types[0][0]\n",
    "        fixed_type_to_ref[key].append(acc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ae168fde-6d51-4b00-a6ce-ce31a0032c80",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\"EF537413: [['B1'], ['Ba']]\", \"AF029863: [['D'], ['Da']]\"]\n"
     ]
    }
   ],
   "source": [
    "print(weird_types) # [\"EF537413: [['B1'], ['Ba']]\", \"AF029863: [['D'], ['Da']]\"]\n",
    "#pprint.pprint(fixed_type_to_ref)\n",
    "# I could do a coin flip but might as well preserve both rather than one or the other? this is irritating.\n",
    "fixed_type_to_ref['B1'] = ['EF537413'] # weird one that's just gonna have to live like this. I hope I don't split types by dashes downstream...\n",
    "fixed_type_to_ref['Da'] = ['AF029863'] # This can just be Da."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4ecf9ac6-efca-452d-8537-266367eda7f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A', 'A3', 'B', 'B(nt59)', 'B1', 'B2', 'B3', 'Bb', 'C', 'C3', 'D', 'D3', 'Da', 'Db', 'E', 'E3', 'F', 'F3', 'Fa', 'Fb', 'Fc', 'G', 'H', 'H3', 'Ha', 'Hb', 'I', 'I3', 'Ia', 'Ib', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'Ua', 'Ub', 'V', 'W', 'X']\n",
      "47\n"
     ]
    }
   ],
   "source": [
    "ospc_types = list(fixed_type_to_ref.keys())\n",
    "print(sorted(ospc_types))\n",
    "print(len(ospc_types)) # 47\n",
    "# Input: \n",
    "# ['A', 'A3', 'B', 'B(nt59)', 'B1', 'B2', 'B3', 'Bb', 'C', 'C3', 'D', 'D3', 'Da', 'Db', 'E', 'E3', \n",
    "# 'F', 'F3', 'Fa', 'Fb', 'Fc', 'G', 'H', 'H3', 'Ha', 'Hb', 'I', 'I3', 'Ia', 'Ib', 'J', 'K', 'L', 'M', \n",
    "# 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'Ua', 'Ub', 'V', 'W', 'X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "28e51e21-f513-44c4-82d6-bba454facadd",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NOT fetching record for A !! ALREADY EXISTS!!\n",
      "NOT fetching record for B(nt59) !! ALREADY EXISTS!!\n",
      "NOT fetching record for B2 !! ALREADY EXISTS!!\n",
      "NOT fetching record for B !! ALREADY EXISTS!!\n",
      "fetching records for C : ['AF029862', 'DQ437462', 'EU482043']\n",
      "NOT fetching record for D !! ALREADY EXISTS!!\n",
      "NOT fetching record for E3 !! ALREADY EXISTS!!\n",
      "NOT fetching record for E !! ALREADY EXISTS!!\n",
      "NOT fetching record for F !! ALREADY EXISTS!!\n",
      "NOT fetching record for G !! ALREADY EXISTS!!\n",
      "NOT fetching record for H !! ALREADY EXISTS!!\n",
      "fetching records for H3 : ['FJ932733']\n",
      "NOT fetching record for I !! ALREADY EXISTS!!\n",
      "fetching records for I3 : ['FJ932734']\n",
      "NOT fetching record for J !! ALREADY EXISTS!!\n",
      "NOT fetching record for K !! ALREADY EXISTS!!\n",
      "NOT fetching record for L !! ALREADY EXISTS!!\n",
      "NOT fetching record for M !! ALREADY EXISTS!!\n",
      "NOT fetching record for N !! ALREADY EXISTS!!\n",
      "NOT fetching record for O !! ALREADY EXISTS!!\n",
      "NOT fetching record for P !! ALREADY EXISTS!!\n",
      "NOT fetching record for Q !! ALREADY EXISTS!!\n",
      "NOT fetching record for R !! ALREADY EXISTS!!\n",
      "NOT fetching record for S !! ALREADY EXISTS!!\n",
      "NOT fetching record for T !! ALREADY EXISTS!!\n",
      "NOT fetching record for U !! ALREADY EXISTS!!\n",
      "NOT fetching record for V !! ALREADY EXISTS!!\n",
      "NOT fetching record for W !! ALREADY EXISTS!!\n",
      "NOT fetching record for X !! ALREADY EXISTS!!\n",
      "NOT fetching record for Bb !! ALREADY EXISTS!!\n",
      "NOT fetching record for Db !! ALREADY EXISTS!!\n",
      "NOT fetching record for Fa !! ALREADY EXISTS!!\n",
      "NOT fetching record for Fb !! ALREADY EXISTS!!\n",
      "NOT fetching record for Fc !! ALREADY EXISTS!!\n",
      "NOT fetching record for Ha !! ALREADY EXISTS!!\n",
      "NOT fetching record for Hb !! ALREADY EXISTS!!\n",
      "NOT fetching record for Ia !! ALREADY EXISTS!!\n",
      "NOT fetching record for Ib !! ALREADY EXISTS!!\n",
      "NOT fetching record for Ua !! ALREADY EXISTS!!\n",
      "NOT fetching record for Ub !! ALREADY EXISTS!!\n",
      "NOT fetching record for A3 !! ALREADY EXISTS!!\n",
      "NOT fetching record for B3 !! ALREADY EXISTS!!\n",
      "NOT fetching record for C3 !! ALREADY EXISTS!!\n",
      "NOT fetching record for D3 !! ALREADY EXISTS!!\n",
      "NOT fetching record for F3 !! ALREADY EXISTS!!\n",
      "NOT fetching record for B1a !! ALREADY EXISTS!!\n",
      "NOT fetching record for Da !! ALREADY EXISTS!!\n"
     ]
    }
   ],
   "source": [
    "for ospc_type, accs in fixed_type_to_ref.items():\n",
    "    out_file = f\"{seqs}/{ospc_type}.fasta\"\n",
    "    if not os.path.exists(out_file):\n",
    "        with open(out_file, \"w\") as f:\n",
    "            print(f\"fetching records for {ospc_type} : {accs}\")\n",
    "            for acc in accs:\n",
    "                handle = Entrez.efetch(db=\"nucleotide\", id=acc, rettype=\"fasta\", retmode=\"text\")\n",
    "                record = handle.read()\n",
    "                #print(f\"writing record: {acc} to file!\")\n",
    "                f.write(record[0::])\n",
    "                handle.close()\n",
    "    else:\n",
    "        print(f\"NOT fetching record for {ospc_type} !! ALREADY EXISTS!!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "95ef46ac-f6f2-45ab-86e5-229ab79ee1c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "writing record: U04282 to file!\n"
     ]
    }
   ],
   "source": [
    "#strain = 'B_bissettii25015'\n",
    "#acc = 'U04282'\n",
    "#\n",
    "#with open(f'{seqs}/{strain}.fasta', 'w') as f:\n",
    "#    handle = Entrez.efetch(db=\"nucleotide\", id=acc, rettype=\"fasta\", retmode=\"text\")\n",
    "#    record = handle.read()\n",
    "#    print(f\"writing record: {acc} to file!\")\n",
    "#    f.write(record[0::])\n",
    "#    handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "55f6ad17-a9fa-479e-b040-60d128c4f9bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['NC_011724']\n",
      "['AF029870', 'DQ437444', 'EU482050', 'U91802', 'CP001535']\n",
      "['EU482052', 'U01892', 'CP001550']\n"
     ]
    }
   ],
   "source": [
    "# ok so some of those entries include the entire plasmid sequence not just the ospc gene, so we will need to extract the ospc gene from the plasmid sequence\n",
    "# the ones that are plasmid sequences are as follows:\n",
    "# Bb, J, M\n",
    "# Bb is only one seq, \n",
    "# J has multiple ospC seqs and then cp26, header of plasmid is CP001535.1\n",
    "# manually changing file name from J.fasta to J_good.fasta\n",
    "# Extracting the full cp26 sequence to J.fasta,\n",
    "# running the below code to append to J_good.fasta\n",
    "# renaming J_good.fasta back to J.fasta\n",
    "\n",
    "# M has multiple ospC seqs and then cp26, header of plasmid is >CP001550.1\n",
    "# manually changing file name from M.fasta to M_good.fasta\n",
    "# Extracting the full cp26 sequence to M.fasta,\n",
    "# running the below code to append to M_good.fasta,\n",
    "# renaming M_good.fasta back to M.fasta\n",
    "\n",
    "# while I'm manually editing sequences, I also added the following to the B1a header: \n",
    "# \"ospC-B1/ospC-Ba allele\" to reflect that it is actually both. \n",
    "\n",
    "for ospc_type in ['Bb', 'J', 'M']:\n",
    "    print(fixed_type_to_ref[ospc_type])\n",
    "coords = {\n",
    "    # THESE ARE ZERO BASED!!!\n",
    "    'Bb': (16904,17540), # from 16905..17540\n",
    "    'J': (16909,17545),  # from 16910..17545\n",
    "    'M': (16916,17555),  # from 16917..17555\n",
    "}\n",
    "\n",
    "# because Bb is a single sequence, we gotta do some shenanigans.\n",
    "with open(f\"{seqs}/Bb_good.fasta\", 'a'):\n",
    "    pass\n",
    "\n",
    "#\n",
    "## ok we already have the full plasmid seqs for each of these three types, we also now have the coordinates\n",
    "for ospc_type in ['Bb', 'J', 'M']:\n",
    "    #print(f\"extracting ospc gene from {ospc_type} sequence\")\n",
    "    with open(f\"{seqs}/{ospc_type}.fasta\", \"r\") as f:\n",
    "        record = SeqIO.read(f, \"fasta\")\n",
    "        #print(record)\n",
    "        ospc_gene = record[coords[ospc_type][0]:coords[ospc_type][1]]\n",
    "        ospc_gene.description = ospc_gene.description.split('plasmid')[0]+f'outer surface protein C (ospC) gene, ospC-{ospc_type} allele'\n",
    "        with open(f\"{seqs}/{ospc_type}_good.fasta\", \"a\") as f:\n",
    "            SeqIO.write(ospc_gene, f, \"fasta\")\n",
    "        shutil.move(f\"{seqs}/{ospc_type}_good.fasta\", f\"{seqs}/{ospc_type}.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1fc2952b-37d1-4321-9700-500a55ab7078",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_best_frame_both_strands(sequence):\n",
    "    \"\"\"Check both strands and all frames, preserve stop codons for inspection\"\"\"\n",
    "    forward_frames = [(i, str(sequence[i:].translate(table=11)), 'forward') for i in range(3)]\n",
    "    reverse_frames = [(i, str(sequence.reverse_complement()[i:].translate(table=11)), 'reverse') for i in range(3)]\n",
    "    \n",
    "    all_frames = forward_frames + reverse_frames\n",
    "    \n",
    "    # Sort by number of stops\n",
    "    sorted_frames = sorted(all_frames, key=lambda x: x[1].count('*'))\n",
    "    \n",
    "    #for frame, translation, direction in sorted_frames:\n",
    "    #    stops = translation.count('*')\n",
    "    #    if stops > 1:\n",
    "    #        print(f\"{direction} frame {frame}: {stops} stops\")\n",
    "    #        print(translation)\n",
    "    #        print()\n",
    "    \n",
    "    return sorted_frames[0]  # (frame, translation, direction) of best frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "ad5fad66-1d10-42ad-a15e-5b2e8c5bc0d9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "128\n",
      "OspC nucleotide file written!\n"
     ]
    }
   ],
   "source": [
    "# Now lets make our multifasta:\n",
    "rec_count = 0\n",
    "written_ids = []\n",
    "ospc_seqs = glob.glob(f\"{seqs}/*.fasta\")\n",
    "ospc_seq_file_nucl = []\n",
    "ospc_seq_file_aa = []\n",
    "for seq in ospc_seqs:\n",
    "    records = SeqIO.parse(seq, \"fasta\")\n",
    "    for record in records:\n",
    "        written_ids.append(record.id)\n",
    "        rec_count += 1\n",
    "        record.id = f\"{record.id}_OspC-\" + seq.split(\"/\")[-1].split(\".\")[0]\n",
    "        record.description = \"\"\n",
    "        ospc_seq_file_nucl.append(record)\n",
    "        \n",
    "        #while len(record.seq) % 3 != 0:\n",
    "        #   record.seq += 'N'\n",
    "        ####record.seq = record.seq.translate(table=11) # use for outputting amino acids.\n",
    "        #protein = find_best_frame_both_strands(record.seq)\n",
    "        #record.seq = protein[1].replace('*', '')\n",
    "        #ospc_seq_file_aa.append(record)\n",
    "        \n",
    "print(rec_count)\n",
    "\n",
    "with open(\"all_ospc_raw_v5.fna\", \"w\") as f:\n",
    "    SeqIO.write(ospc_seq_file_nucl, f, \"fasta\")\n",
    "print(\"OspC nucleotide file written!\")\n",
    "\n",
    "#with open(\"all_ospc_raw_v4.faa\", \"w\") as f:\n",
    "#    SeqIO.write(ospc_seq_file_aa, f, \"fasta\")\n",
    "#print(\"OspC amino acid file written!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "115ccc1f-f3ab-44ff-9566-f265d72fee4b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "127\n",
      "127\n",
      "128\n",
      "127\n"
     ]
    }
   ],
   "source": [
    "print(len(value_to_keys.keys())) # 127\n",
    "print(len([item[0][0] for value in fixed_type_to_ref.values() for item in value])) # 124\n",
    "print(len([rec.id for rec in SeqIO.parse(\"all_ospc_raw_v5.fna\", \"fasta\")]))\n",
    "print(len([rec.id for rec in SeqIO.parse(\"all_ospc_raw_v4.faa\", \"fasta\")]))\n",
    "all_ids = sorted(value_to_keys.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a1d005c6-f094-4c70-98bb-c0da3dfb8cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#missing = set(all_ids) - set([x.split('.')[0] for x in written_ids]) # {'DQ437462', 'FJ932733', 'FJ932734'}\n",
    "#for acc in missing:\n",
    "#   print(value_to_keys[acc])\n",
    "# {'dict1': ['C'], 'dict2': ['C']}\n",
    "# {'dict1': ['H3'], 'dict2': ['H3']}\n",
    "# {'dict1': ['I3'], 'dict2': ['I3']}\n",
    "\n",
    "# THIS IS FIXED BY COMMENTING OUT THE PASS IN DICT CREATION AND ACTUALLY INCLUDING IT."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ebcdb61a-0f3c-4157-988f-c11dff742efb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit-est -i /data/all_ospc_raw_v5.fna -o\n",
      "         /data/ospC_ref_clustering_v5/nucl/90/ospC_v5_90 -c 0.9\n",
      "         -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:16 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 128\n",
      "longest and shortest : 1150 and 476\n",
      "Total letters: 72904\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 12M = 12M\n",
      "Table           : 1 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 12M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 1184090\n",
      "Max number of word counting entries: 98449733\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        128\n",
      "\n",
      "      128  finished         31  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 12M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.23\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit -i /data/all_ospc_raw_v4.faa -o\n",
      "         /data/ospC_ref_clustering_v4/aa/90/ospC_v4_90 -c 0.9\n",
      "         -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:17 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 127\n",
      "longest and shortest : 369 and 158\n",
      "Total letters: 23913\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 75M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 2510812\n",
      "Max number of word counting entries: 90505071\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        127\n",
      "\n",
      "      127  finished         35  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 76M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.03\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit-est -i /data/all_ospc_raw_v5.fna -o\n",
      "         /data/ospC_ref_clustering_v5/nucl/91/ospC_v5_91 -c\n",
      "         0.91 -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:17 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "Your word length is 5, using 8 may be faster!\n",
      "total seq: 128\n",
      "longest and shortest : 1150 and 476\n",
      "Total letters: 72904\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 12M = 12M\n",
      "Table           : 1 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 12M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 1184090\n",
      "Max number of word counting entries: 98449733\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        128\n",
      "\n",
      "      128  finished         31  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 12M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.22\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit -i /data/all_ospc_raw_v4.faa -o\n",
      "         /data/ospC_ref_clustering_v4/aa/91/ospC_v4_91 -c 0.91\n",
      "         -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:18 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 127\n",
      "longest and shortest : 369 and 158\n",
      "Total letters: 23913\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 75M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 2510812\n",
      "Max number of word counting entries: 90505071\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        127\n",
      "\n",
      "      127  finished         36  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 76M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.03\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit-est -i /data/all_ospc_raw_v5.fna -o\n",
      "         /data/ospC_ref_clustering_v5/nucl/92/ospC_v5_92 -c\n",
      "         0.92 -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:19 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "Your word length is 5, using 8 may be faster!\n",
      "total seq: 128\n",
      "longest and shortest : 1150 and 476\n",
      "Total letters: 72904\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 12M = 12M\n",
      "Table           : 1 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 12M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 1184090\n",
      "Max number of word counting entries: 98449733\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        128\n",
      "\n",
      "      128  finished         32  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 12M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.18\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit -i /data/all_ospc_raw_v4.faa -o\n",
      "         /data/ospC_ref_clustering_v4/aa/92/ospC_v4_92 -c 0.92\n",
      "         -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:20 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 127\n",
      "longest and shortest : 369 and 158\n",
      "Total letters: 23913\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 75M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 2510812\n",
      "Max number of word counting entries: 90505071\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        127\n",
      "\n",
      "      127  finished         36  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 76M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.03\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit-est -i /data/all_ospc_raw_v5.fna -o\n",
      "         /data/ospC_ref_clustering_v5/nucl/93/ospC_v5_93 -c\n",
      "         0.93 -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:20 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "Your word length is 5, using 8 may be faster!\n",
      "total seq: 128\n",
      "longest and shortest : 1150 and 476\n",
      "Total letters: 72904\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 12M = 12M\n",
      "Table           : 1 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 12M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 1184090\n",
      "Max number of word counting entries: 98449733\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        128\n",
      "\n",
      "      128  finished         32  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 12M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.12\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit -i /data/all_ospc_raw_v4.faa -o\n",
      "         /data/ospC_ref_clustering_v4/aa/93/ospC_v4_93 -c 0.93\n",
      "         -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:21 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 127\n",
      "longest and shortest : 369 and 158\n",
      "Total letters: 23913\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 75M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 2510812\n",
      "Max number of word counting entries: 90505071\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        127\n",
      "\n",
      "      127  finished         36  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 76M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.03\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit-est -i /data/all_ospc_raw_v5.fna -o\n",
      "         /data/ospC_ref_clustering_v5/nucl/94/ospC_v5_94 -c\n",
      "         0.94 -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:22 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "Your word length is 5, using 8 may be faster!\n",
      "total seq: 128\n",
      "longest and shortest : 1150 and 476\n",
      "Total letters: 72904\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 12M = 12M\n",
      "Table           : 1 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 12M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 1184090\n",
      "Max number of word counting entries: 98449733\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        128\n",
      "\n",
      "      128  finished         33  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 12M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.06\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit -i /data/all_ospc_raw_v4.faa -o\n",
      "         /data/ospC_ref_clustering_v4/aa/94/ospC_v4_94 -c 0.94\n",
      "         -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:22 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 127\n",
      "longest and shortest : 369 and 158\n",
      "Total letters: 23913\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 75M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 2510812\n",
      "Max number of word counting entries: 90505071\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        127\n",
      "\n",
      "      127  finished         36  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 76M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.03\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit-est -i /data/all_ospc_raw_v5.fna -o\n",
      "         /data/ospC_ref_clustering_v5/nucl/95/ospC_v5_95 -c\n",
      "         0.95 -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:23 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "Your word length is 5, using 8 may be faster!\n",
      "total seq: 128\n",
      "longest and shortest : 1150 and 476\n",
      "Total letters: 72904\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 12M = 12M\n",
      "Table           : 1 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 12M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 1184090\n",
      "Max number of word counting entries: 98449733\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        128\n",
      "\n",
      "      128  finished         34  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 12M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.04\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit -i /data/all_ospc_raw_v4.faa -o\n",
      "         /data/ospC_ref_clustering_v4/aa/95/ospC_v4_95 -c 0.95\n",
      "         -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:23 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 127\n",
      "longest and shortest : 369 and 158\n",
      "Total letters: 23913\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 75M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 2510812\n",
      "Max number of word counting entries: 90505071\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        127\n",
      "\n",
      "      127  finished         36  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 76M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.03\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit-est -i /data/all_ospc_raw_v5.fna -o\n",
      "         /data/ospC_ref_clustering_v5/nucl/96/ospC_v5_96 -c\n",
      "         0.96 -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:24 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "Your word length is 5, using 8 may be faster!\n",
      "total seq: 128\n",
      "longest and shortest : 1150 and 476\n",
      "Total letters: 72904\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 12M = 12M\n",
      "Table           : 1 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 12M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 1184090\n",
      "Max number of word counting entries: 98449733\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        128\n",
      "\n",
      "      128  finished         38  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 12M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.03\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit -i /data/all_ospc_raw_v4.faa -o\n",
      "         /data/ospC_ref_clustering_v4/aa/96/ospC_v4_96 -c 0.96\n",
      "         -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:25 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 127\n",
      "longest and shortest : 369 and 158\n",
      "Total letters: 23913\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 75M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 2510812\n",
      "Max number of word counting entries: 90505071\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        127\n",
      "\n",
      "      127  finished         43  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 76M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.03\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit-est -i /data/all_ospc_raw_v5.fna -o\n",
      "         /data/ospC_ref_clustering_v5/nucl/97/ospC_v5_97 -c\n",
      "         0.97 -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:25 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "Your word length is 5, using 8 may be faster!\n",
      "total seq: 128\n",
      "longest and shortest : 1150 and 476\n",
      "Total letters: 72904\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 12M = 12M\n",
      "Table           : 1 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 12M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 1184090\n",
      "Max number of word counting entries: 98449733\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        128\n",
      "\n",
      "      128  finished         41  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 12M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.02\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit -i /data/all_ospc_raw_v4.faa -o\n",
      "         /data/ospC_ref_clustering_v4/aa/97/ospC_v4_97 -c 0.97\n",
      "         -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:26 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 127\n",
      "longest and shortest : 369 and 158\n",
      "Total letters: 23913\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 75M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 2510812\n",
      "Max number of word counting entries: 90505071\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        127\n",
      "\n",
      "      127  finished         48  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 76M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.03\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit-est -i /data/all_ospc_raw_v5.fna -o\n",
      "         /data/ospC_ref_clustering_v5/nucl/98/ospC_v5_98 -c\n",
      "         0.98 -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:27 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "Your word length is 5, using 8 may be faster!\n",
      "total seq: 128\n",
      "longest and shortest : 1150 and 476\n",
      "Total letters: 72904\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 12M = 12M\n",
      "Table           : 1 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 12M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 1184090\n",
      "Max number of word counting entries: 98449733\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        128\n",
      "\n",
      "      128  finished         42  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 12M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.02\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit -i /data/all_ospc_raw_v4.faa -o\n",
      "         /data/ospC_ref_clustering_v4/aa/98/ospC_v4_98 -c 0.98\n",
      "         -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:27 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 127\n",
      "longest and shortest : 369 and 158\n",
      "Total letters: 23913\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 75M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 2510812\n",
      "Max number of word counting entries: 90505071\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        127\n",
      "\n",
      "      127  finished         53  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 76M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.03\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit-est -i /data/all_ospc_raw_v5.fna -o\n",
      "         /data/ospC_ref_clustering_v5/nucl/99/ospC_v5_99 -c\n",
      "         0.99 -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:28 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "Your word length is 5, using 8 may be faster!\n",
      "total seq: 128\n",
      "longest and shortest : 1150 and 476\n",
      "Total letters: 72904\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 12M = 12M\n",
      "Table           : 1 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 12M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 1184090\n",
      "Max number of word counting entries: 98449733\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        128\n",
      "\n",
      "      128  finished         49  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 12M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.02\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit -i /data/all_ospc_raw_v4.faa -o\n",
      "         /data/ospC_ref_clustering_v4/aa/99/ospC_v4_99 -c 0.99\n",
      "         -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:28 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 127\n",
      "longest and shortest : 369 and 158\n",
      "Total letters: 23913\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 75M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 2510812\n",
      "Max number of word counting entries: 90505071\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        127\n",
      "\n",
      "      127  finished         63  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 76M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.03\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit-est -i /data/all_ospc_raw_v5.fna -o\n",
      "         /data/ospC_ref_clustering_v5/nucl/100/ospC_v5_100 -c\n",
      "         1.0 -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:29 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "Your word length is 5, using 8 may be faster!\n",
      "total seq: 128\n",
      "longest and shortest : 1150 and 476\n",
      "Total letters: 72904\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 12M = 12M\n",
      "Table           : 1 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 12M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 1184090\n",
      "Max number of word counting entries: 98449733\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        128\n",
      "\n",
      "      128  finished         76  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 12M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.02\n",
      "\n",
      "STDERR:\n",
      " \n",
      "STDOUT:\n",
      " ================================================================\n",
      "Program: CD-HIT, V4.8.1 (+OpenMP), Nov 22 2024, 20:06:10\n",
      "Command: cd-hit -i /data/all_ospc_raw_v4.faa -o\n",
      "         /data/ospC_ref_clustering_v4/aa/100/ospC_v4_100 -c 1.0\n",
      "         -n 5 -g 1 -d 40\n",
      "\n",
      "Started: Mon Dec  9 22:31:29 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total seq: 127\n",
      "longest and shortest : 369 and 158\n",
      "Total letters: 23913\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 1 X 10M = 10M\n",
      "Table           : 1 X 65M = 65M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 75M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 2510812\n",
      "Max number of word counting entries: 90505071\n",
      "\n",
      "\n",
      "comparing sequences from          0  to        127\n",
      "\n",
      "      127  finished         76  clusters\n",
      "\n",
      "Approximated maximum memory consumption: 76M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 0.03\n",
      "\n",
      "STDERR:\n",
      " \n"
     ]
    }
   ],
   "source": [
    "# OKAY NOW TIME FOR CD-HIT\n",
    "# Run inside of docker container ;)\n",
    "current_dir = os.getcwd()\n",
    "for ident in range(90,101):\n",
    "    output_path_nucl = f'ospC_ref_clustering_v5/nucl/{ident}'\n",
    "    output_path_aa = f'ospC_ref_clustering_v4/aa/{ident}'\n",
    "    \n",
    "    os.makedirs(output_path_nucl, exist_ok=True)\n",
    "    os.makedirs(output_path_aa, exist_ok=True)\n",
    "        \n",
    "    docker_cmd_nucl = [\n",
    "        'docker', 'run', '-v', f'{current_dir}:/data', '-w', '/opt', 'mjfos2r/cd-hit',\n",
    "        '/bin/bash', '-c',\n",
    "        f'cd-hit-est -i /data/all_ospc_raw_v5.fna -o /data/{output_path_nucl}/ospC_v5_{ident} -c {ident/100} -n 5 -g 1 -d 40'\n",
    "    ]\n",
    "    docker_cmd_aa = [\n",
    "        'docker', 'run', '-v', f'{current_dir}:/data', '-w', '/opt', 'mjfos2r/cd-hit',\n",
    "        '/bin/bash', '-c',\n",
    "        f'cd-hit -i /data/all_ospc_raw_v4.faa -o /data/{output_path_aa}/ospC_v4_{ident} -c {ident/100} -n 5 -g 1 -d 40 ',\n",
    "    ]\n",
    "        \n",
    "    result = subprocess.run(docker_cmd_nucl, capture_output=True, text=True)\n",
    "    print('STDOUT:\\n', result.stdout)\n",
    "    print('STDERR:\\n', result.stderr)\n",
    "    result = subprocess.run(docker_cmd_aa, capture_output=True, text=True)\n",
    "    print('STDOUT:\\n', result.stdout)\n",
    "    print('STDERR:\\n', result.stderr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "b995c1f6-5ec0-4fa9-afb5-2b4b4ff50485",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48\n",
      "41\n",
      "['B1', 'Db', 'Fa', 'Ha', 'Ia', 'Ua']\n"
     ]
    }
   ],
   "source": [
    "# cat ospC_v4_100.fna | grep \">\" | sed 's/.*-\\(.*\\)/\\1/' | sort | uniq\n",
    "# Lets check our clusters!\n",
    "# V4_all:\n",
    "# ['A', 'A3', 'B', 'B(nt59)', 'B.bissettii_25015', 'B2', 'B3', 'Bb', 'C', 'C3', 'D', 'D3', 'Db', 'E', 'E3', \n",
    "# 'F', 'F3', 'Fa', 'Fb', 'Fc', 'G', 'H', 'Ha', 'Hb', 'I', 'Ia', 'Ib', 'J', 'K', 'L', 'M', 'N', 'O', 'P', \n",
    "# 'Q', 'R', 'S', 'T', 'U', 'Ua', 'Ub', 'V', 'W', 'X']\n",
    "\n",
    "# V5_all:\n",
    "#[ 'A', 'A3', 'B', 'B(nt59)', 'B1', 'B2', 'B3', 'Bb', 'C', 'C3', 'D', 'D3', 'Da', 'Db', 'E', 'E3', \n",
    "# 'F', 'F3', 'Fa', 'Fb', 'Fc', 'G', 'H', 'H3', 'Ha', 'Hb', 'I', 'I3', 'Ia', 'Ib', 'J', 'K', 'L', 'M', 'N', 'O', 'P', \n",
    "# 'Q', 'R', 'S', 'T', 'U', 'Ua', 'Ub', 'V', 'W', 'X' ] \n",
    "\n",
    "# V5_nucl Consensus: (at 100% ident)\n",
    "# ['A', 'A3', 'B', 'B(nt59)', 'B2', 'B3', 'Bb', 'C', 'C3', 'D', 'D3', 'Da', 'E', 'E3', \n",
    "#  'F', 'F3', 'Fb', 'Fc', 'G', 'H', 'H3', 'Hb', 'I', 'I3', 'Ib', 'J', 'K', 'L', 'M', 'N', 'O', 'P', \n",
    "#  'Q', 'R', 'S', 'T', 'U', 'Ub', 'V', 'W', 'X']\n",
    "\n",
    "v5_input = [ 'A', 'A3', 'B', 'B(nt59)', 'B1', 'B2', 'B3', 'Bb', 'C', 'C3', 'D', 'D3', 'Da', 'Db', 'E', 'E3', \n",
    " 'F', 'F3', 'Fa', 'Fb', 'Fc', 'G', 'H', 'H3', 'Ha', 'Hb', 'I', 'I3', 'Ia', 'Ib', 'J', 'K', 'L', 'M', 'N', 'O', 'P', \n",
    " 'Q', 'R', 'S', 'T', 'U', 'Ua', 'Ub', 'V', 'W', 'X', 'B_bissettii25015'] \n",
    "\n",
    "v5_types = ['A', 'A3', 'B', 'B(nt59)', 'B2', 'B3', 'Bb', 'C', 'C3', 'D', 'D3', 'Da', 'E', 'E3', \n",
    "  'F', 'F3', 'Fb', 'Fc', 'G', 'H', 'H3', 'Hb', 'I', 'I3', 'Ib', 'J', 'K', 'L', 'M', 'N', 'O', 'P', \n",
    "  'Q', 'R', 'S', 'T', 'U', 'Ub', 'V', 'W', 'X', 'B_bissettii25015']\n",
    "\n",
    "print(len(v5_input))\n",
    "print(len(v5_types_nucl))\n",
    "\n",
    "print(sorted(set(v5_input) - set(v5_types))) # ['B1', 'Db', 'Fa', 'Ha', 'Ia', 'Ua']\n",
    "# FIXING MANUALLY SEE BELOW."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "id": "9f853cac-8dc5-4bcf-94cc-97fd96684141",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['docker', 'run', '-v', '/home/mf019/longread_pangenome/OspC_typing:/data', '-w', '/data', 'dnalinux/clustalo', '/bin/bash', '-c', 'clustalo -i /data/all_ospc_raw_v4.faa -o /data/all_ospc_raw_v4.faa.aln --force'], returncode=0)"
      ]
     },
     "execution_count": 342,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ok time to run clustal and see what's going on. \n",
    "clustal_cmd = [\n",
    "    'docker', 'run',\n",
    "    '-v', f'{current_dir}:/data',\n",
    "    '-w', '/data',\n",
    "    'dnalinux/clustalo',\n",
    "    '/bin/bash', '-c',\n",
    "    f'clustalo -i /data/all_ospc_raw_v4.faa -o /data/all_ospc_raw_v4.faa.aln --force'\n",
    "]\n",
    "subprocess.run(clustal_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "24386d1f-8008-4f25-bf36-0adf48831c53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# B1 is clustered with other B types so we shall name it B1 since we have B2 and B3. There is no Ba. \n",
    "#>Cluster 16\n",
    "#0\t583nt, >AF029861.1_OspC-B... *\n",
    "#1\t504nt, >EU482042.1_OspC-B... at +/100.00%\n",
    "#2\t537nt, >EF537413.1_OspC-B1... at +/100.00%\n",
    "\n",
    "# B2 and Bb are also the same cluster. \n",
    "# Bb is B2 and will be numbered as such.\n",
    "#>Cluster 6\n",
    "#0\t636nt, >NC_011724.1_OspC-Bb... *\n",
    "#1\t522nt, >U91795.1_OspC-B... at +/100.00%\n",
    "#2\t579nt, >L42868.1_OspC-B2... at +/100.00%\n",
    "\n",
    "# D is also causing trouble.. \n",
    "# D renamed to Db,\n",
    "#>Cluster 0\n",
    "#0\t1150nt, >L25413.1_OspC-D... *\n",
    "#1\t531nt, >GQ478283.1_OspC-Db... at +/100.00%\n",
    "\n",
    "## D renamed to Da (cluster named correctly)\n",
    "#>Cluster 41\n",
    "#0\t504nt, >EU482044.1_OspC-D... at +/100.00%\n",
    "#1\t555nt, >AF029863.1_OspC-Da... *\n",
    "\n",
    "## D3 is unchanged but should be updated to Dc for consistency.\n",
    "#>Cluster 59\n",
    "#0\t531nt, >EF592544.1_OspC-D3... *\n",
    "\n",
    "# Actually none of this makes any sense because these classifications are all over the damn place.\n",
    "# lets return to Iras' set and look at the lengths he provided.\n",
    "# update: they're mostly incomplete CDS as well. \n",
    "# ok. Amine recommended that we just roll with 98% and rename the clusters. I think I'm going to just drop off the subtype\n",
    "# and stick to whatever the major type is, regardless of cluster identity. \n",
    "# # If it's cluster 16- B or if its Cluster 6- B, it's type B.\n",
    "# # This is not for me to resolve at this time."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
