{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import pandas\n",
    "import pprint\n",
    "from Bio            import Entrez\n",
    "from Bio            import SeqIO\n",
    "from Bio.Blast      import NCBIXML\n",
    "\n",
    "############################################\n",
    "Entrez.email = \"mfoster11@mgh.harvard.edu\" # Always tell NCBI who you are\n",
    "############################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GET_REFERENCE_SEQS = False\n",
    "FIX_REFERENCE_SEQS = True\n",
    "BUILD_BLAST_DB = False\n",
    "RUN_BLAST = False\n",
    "PARSE_BLAST_RESULTS = False\n",
    "MERGE_RESULTS = False\n",
    "FINALIZE_AND_WRITE_METADATA = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_input_metadata(metadata : str) -> pandas.DataFrame:\n",
    "    input_metadata = pandas.read_csv(metadata)\n",
    "    return input_metadata\n",
    "\n",
    "def parse_blast_results(blast_results : str) -> dict:\n",
    "    file_size = os.path.getsize(blast_results)\n",
    "    if file_size == 0:\n",
    "        return None\n",
    "    results = {}\n",
    "    with open(blast_results, \"r\") as f:\n",
    "        blast_records = NCBIXML.parse(f)\n",
    "        for record in blast_records:\n",
    "            results[record.query] = []\n",
    "            for alignment in record.alignments:\n",
    "                for hsp in alignment.hsps:\n",
    "                    results[record.query].append((alignment.title, hsp.expect, hsp.score,))\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblies = glob.glob(\"../../assemblies/*/annotation/*/*.gbff\")\n",
    "#print(assemblies)\n",
    "\n",
    "metadata = \"../metadata/longread_metadata_v2.csv\"\n",
    "input_metadata = load_input_metadata(metadata)\n",
    "known_types = input_metadata[['Isolate', 'OspC_Type']]\n",
    "#print(known_types)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Next Steps\n",
    "Ok, so now what we'll do is take the table above, pull all isolates that have a known ospc_type, then pull the ospc annotations from each isolate, then make a tree, then confirm tree layout with known types, then run tree of all known ospc annotations, then use previous labels as classifications for each cluster.\n",
    "\n",
    "> What are the characteristics that differentiate each OspC type.\n",
    ">> Can I use motif scanning? Definitely MSA."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Okay, Ira has sent me a large file of OspC types and AA seqs\n",
    "I will be taking this, making a blast DB and then I need to isolate the OspC seqs from each assembly, then blast against the OspC database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ospc_genotype_to_ref = {\n",
    "     'A' : 'X69596',\n",
    "    'Ba' : 'EF537413',\n",
    "    'Bb' : 'NC_011724',\n",
    "     'C' : 'DQ437462',\n",
    "    'Da' : 'AF029863',\n",
    "    'Db' : 'GQ478283',\n",
    "     'E' : 'AY275221',\n",
    "    'Fa' : 'AY275225',\n",
    "    'Fb' : 'EF537433',\n",
    "    'Fc' : 'GQ478285',\n",
    "     'G' : 'AY275223',\n",
    "    'Ha' : 'EU377781',\n",
    "    'Hb' : 'GQ478286',\n",
    "    'Ia' : 'AY275219',\n",
    "    'Ib' : 'EU377752',\n",
    "     'J' : 'CP001535',\n",
    "     'K' : 'AY275214',\n",
    "     'L' : 'EU375832',\n",
    "     'M' : 'CP001550',\n",
    "     'N' : 'EU377775',\n",
    "     'O' : 'FJ997281',\n",
    "     'T' : 'AY275222',\n",
    "    'Ua' : 'EU377769',\n",
    "    'Ub' : 'GQ478287',\n",
    "    'A3' : 'EF592541',\n",
    "    'B3' : 'EF592542',\n",
    "    'C3' : 'EF592543',\n",
    "    'D3' : 'EF592544',\n",
    "    'E3' : 'EF592545',\n",
    "    'F3' : 'EF592547',\n",
    "    'H3' : 'FJ932733',\n",
    "    'I3' : 'FJ932734',\n",
    "}\n",
    "# okay so per Ira, these ATs correspond to OspC types. tbh I should blast these ATs against the OspC types and see how they match up... (2024-03-05 - MJF)\n",
    "ospc_AT_to_ref = {\n",
    "   'AT1' : 'EU482041',\n",
    "   'AT2' : 'EU482042',\n",
    "   'AT3' : 'EU482043',\n",
    "   'AT4' : 'EU482044',\n",
    "   'AT5' : 'EU482045',\n",
    "   'AT6' : 'EU482046',\n",
    "   'AT7' : 'EU482047',\n",
    "   'AT8' : 'EU482048',\n",
    "   'AT9' : 'EU482049',\n",
    "  'AT10' : 'EU482050',\n",
    "  'AT11' : 'EU482051',\n",
    "  'AT12' : 'EU482052',\n",
    "  'AT13' : 'EU482053',\n",
    "  'AT14' : 'EU482054',\n",
    "  'AT15' : 'EU482055',\n",
    "  'AT16' : 'EU482056',\n",
    "  'B.bissettii_25015' : 'U04282'\n",
    "\n",
    "}\n",
    "len(ospc_genotype_to_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if GET_REFERENCE_SEQS is True:\n",
    "    # ok now to pull all of those seqs via entrez direct and then make a multifasta file upon which to build a blastdb from\n",
    "    # we will use the following command to pull the sequences from the NCBI nucleotide database using our defined dictionary\n",
    "    # and we will also call the command using the biopython wrapper for entrez direct\n",
    "    for ospc_type, ref in ospc_genotype_to_ref.items():\n",
    "        print(f\"fetching record for {ospc_type} : {ref}\")\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=ref, rettype=\"fasta\", retmode=\"text\")\n",
    "        record = handle.read()\n",
    "        with open(f\"ospc_seqs/{ospc_type}.fasta\", \"w\") as f:\n",
    "            print(\"writing record to file!\")\n",
    "            f.write(record[0::])\n",
    "        handle.close()\n",
    "    print(\"done!\")\n",
    "    for ospc_AT, ref in ospc_AT_to_ref.items():\n",
    "        print(f\"fetching record for {ospc_AT} : {ref}\")\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=ref, rettype=\"fasta\", retmode=\"text\")\n",
    "        record = handle.read()\n",
    "        with open(f\"AT_seqs/{ospc_AT}.fasta\", \"w\") as f:\n",
    "            print(\"writing record to file!\")\n",
    "            f.write(record[0::])\n",
    "        handle.close()\n",
    "    print(\"done!\")\n",
    "else:\n",
    "    print(\"skipping reference sequence retrieval\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if FIX_REFERENCE_SEQS is True:\n",
    "    # ok so some of those entries include the entire plasmid sequence not just the ospc gene, so we will need to extract the ospc gene from the plasmid sequence\n",
    "    # the ones that are plasmid sequences are as follows:\n",
    "    # Bb, J, M,\n",
    "    # we will need to extract the ospc gene from these sequences\n",
    "    for ospc_type in ['Bb', 'J', 'M']:\n",
    "        print(ospc_genotype_to_ref[ospc_type])\n",
    "\n",
    "    coords = {\n",
    "        'Bb' : (16904,17540), # DON'T FORGET TO CHANGE TO 0 BASED COORDINATES!\n",
    "        'J' : (16909,17545), # DON'T FORGET TO CHANGE TO 0 BASED COORDINATES!\n",
    "        'M' : (16916,17555), # DON'T FORGET TO CHANGE TO 0 BASED COORDINATES!\n",
    "    }\n",
    "    #\n",
    "    ## ok we already have the full plasmid seqs for each of these three types, we also now have the coordinates\n",
    "    for ospc_type in ['Bb', 'J', 'M']:\n",
    "        print(f\"extracting ospc gene from {ospc_type} sequence\")\n",
    "        with open(f\"ospc_seqs/{ospc_type}.fasta\", \"r\") as f:\n",
    "            record = SeqIO.read(f, \"fasta\")\n",
    "            print(record)\n",
    "            ospc_gene = record[coords[ospc_type][0]:coords[ospc_type][1]]\n",
    "            with open(f\"ospc_seqs/{ospc_type}_ospc.fasta\", \"w\") as f:\n",
    "                SeqIO.write(ospc_gene, f, \"fasta\")\n",
    "else:\n",
    "    print(\"skipping ospc gene extraction and record repair\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if WRITE_REF_SEQS is True:\n",
    "    # Now lets write out all of our OspC sequences to a single file for blastdb creation\n",
    "    ospc_seqs = glob.glob(\"ospc_seqs/*.fasta\")\n",
    "    #print(ospc_seqs)\n",
    "    ospc_seq_file = []\n",
    "    for seq in ospc_seqs:\n",
    "        record = SeqIO.read(seq, \"fasta\")\n",
    "        record.id = \"OspC_Type-\" + seq.split(\"/\")[-1].split(\".\")[0]\n",
    "        record.description = \"\"\n",
    "        #print(record.id)\n",
    "        ospc_seq_file.append(record)\n",
    "    with open(\"all_ospc.fasta\", \"w\") as f:\n",
    "        SeqIO.write(ospc_seq_file, f, \"fasta\")\n",
    "    print(\"OspC file written!\")\n",
    "else:\n",
    "    print(\"skipping writing out reference OspC sequences\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if WRITE_REF_SEQS is True:\n",
    "    # Okay now lets write the AT sequences to a single file for blastdb creation\n",
    "    AT_seqs = glob.glob(\"AT_seqs/*.fasta\")\n",
    "    #print(AT_seqs)\n",
    "    AT_seq_file = []\n",
    "    for seq in AT_seqs:\n",
    "        record = SeqIO.read(seq, \"fasta\")\n",
    "        record.id = \"OspC_Type-\" + seq.split(\"/\")[-1].split(\".\")[0]\n",
    "        record.description = \"\"\n",
    "        #print(record.id)\n",
    "        AT_seq_file.append(record)\n",
    "    with open(\"all_ospc_ATs.fasta\", \"w\") as f:\n",
    "        SeqIO.write(AT_seq_file, f, \"fasta\")\n",
    "    print(\"AT file written!\")\n",
    "else:\n",
    "    print(\"skipping writing out reference AT sequences\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if MAKE_BLASTDB is True:\n",
    "    # ok now lets make a local blast DB for our OspC types\n",
    "    from Bio.Blast.Applications import NcbimakeblastdbCommandline\n",
    "    cmd = NcbimakeblastdbCommandline(input_file=\"all_ospc.fasta\", dbtype=\"nucl\", out=\"blastdb/ospc/ospc\")\n",
    "    print(cmd)\n",
    "    stdout, stderr = cmd()\n",
    "    print(stdout)\n",
    "    print(stderr)\n",
    "    # now lets make a local blast DB for our AT types\n",
    "    from Bio.Blast.Applications import NcbimakeblastdbCommandline\n",
    "    cmd = NcbimakeblastdbCommandline(input_file=\"all_ospc_ATs.fasta\", dbtype=\"nucl\", out=\"blastdb/ospcAT/ospcAT\")\n",
    "    print(cmd)\n",
    "    stdout, stderr = cmd()\n",
    "    print(stdout)\n",
    "    print(stderr)\n",
    "else:\n",
    "    print(\"skipping making blastdb for AT\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ok now lets pull all of the ospCs from all assemblies and then blast them against our local blastdb\n",
    "ospc_identifiers = [\n",
    "    #\"S2/P23\", # This may not be right. I need to figure out why some assemblies lack OspC.\n",
    "    \"P12\", # this is what it's called in garinii per Matteo.\n",
    "    \"ospC\",\n",
    "    \"OspC\",\n",
    "    \"Ospc\",\n",
    "    \"ospc\",\n",
    "    \"outer surface protein C\",\n",
    "    \"Outer surface protein C\",\n",
    "    \"outer surface lipoprotein C\",\n",
    "    \"Outer surface lipoprotein C\",\n",
    "    \"outer surface protein c\",\n",
    "]\n",
    "\n",
    "ospc_genes = {}\n",
    "for file in assemblies:\n",
    "    #print(file)\n",
    "    sample_id = file.split(\"/\")[-1].split(\".\")[0].replace(\"_200\",\"\" )\n",
    "    ospc_genes[sample_id] = []\n",
    "    with open(file, \"r\") as f:\n",
    "        record = SeqIO.parse(f, \"genbank\")\n",
    "        for rec in record:\n",
    "            #print(rec.id)\n",
    "            for feature in rec.features:\n",
    "                if feature.type == \"CDS\":\n",
    "                    product = feature.qualifiers['product']\n",
    "                    if any(x in product[0] for x in ospc_identifiers):\n",
    "                        #feature.qualifiers['product']\n",
    "                        # get sequence\n",
    "                        #print(feature.location)\n",
    "                        # Now lets make a new SeqRecord object for the gene and append it to our list of OspC Genes!\n",
    "                        seq = feature.extract(rec.seq)\n",
    "                        gene = SeqIO.SeqRecord(seq, id=product[0], name=product[0], description=f\"gene: {product[0]} from isolate {sample_id}\", dbxrefs=None)\n",
    "                        ospc_genes[sample_id].append(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Okay now lets pull the ospC genes from all of the assemblies and write them to a file for blasting!\n",
    "for sample in ospc_genes:\n",
    "    #print(sample)\n",
    "    #print(ospc_genes[sample])\n",
    "    with open(f\"temp_in/{sample}.osps.fasta\", \"w\") as f:\n",
    "        SeqIO.write(ospc_genes[sample], f\"temp_in/{sample}.osps.fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if RUN_BLAST is True:\n",
    "    # ok now we can blast each file against our local blastdb\n",
    "    from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "    for sample in ospc_genes:\n",
    "        print(f\"blasting {sample}\")\n",
    "        cmd = NcbiblastnCommandline(query=f\"temp_in/{sample}.osps.fasta\", db=\"blastdb/ospc/ospc\", out=f\"temp_out/{sample}.osps_blast.xml\", outfmt=5)\n",
    "        print(cmd)\n",
    "        stdout, stderr = cmd()\n",
    "        print(stdout)\n",
    "        print(stderr)\n",
    "else:\n",
    "    print(\"skipping blasting\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if RUN_BLAST is True:\n",
    "    # ok now we can blast each file against our local blastdb\n",
    "    from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "    for sample in ospc_genes:\n",
    "        print(f\"blasting {sample}\")\n",
    "        cmd = NcbiblastnCommandline(query=f\"temp_in/{sample}.osps.fasta\", db=\"blastdb/ospcAT/ospcAT\", out=f\"temp_out/{sample}.ospsAT_blast.xml\", outfmt=5)\n",
    "        print(cmd)\n",
    "        stdout, stderr = cmd()\n",
    "        print(stdout)\n",
    "        print(stderr)\n",
    "else:\n",
    "    print(\"skipping blasting\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ok now we can parse the blast results and then add the ospc type to the metadata\n",
    "# we will use the function defined at the beginning to parse the blast results!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if PARSE_BLAST_RESULTS is True:\n",
    "    checked_types_ospc = pandas.DataFrame(columns=[\"Isolate\", \"ospc_newtype\"])\n",
    "    blast_results_ospc = glob.glob(\"temp_out/*-osps_blast.xml\")\n",
    "\n",
    "    for result in blast_results_ospc:\n",
    "        sample_id = result.split(\"/\")[-1].split(\".\")[0].replace(\"-osps_blast\", \"\")\n",
    "        isolate = sample_id\n",
    "        ospc_type = \"unknown\"\n",
    "        parsed_results = parse_blast_results(result)\n",
    "        if parsed_results is not None:\n",
    "            for result in parsed_results:\n",
    "                if len(parsed_results[result]) > 0:\n",
    "                    best_hit = parsed_results[result][0][0]\n",
    "                    coords = parsed_results[result][0][0].find(\"OspC_Type-\")\n",
    "                    ospc_type = parsed_results[result][0][0][coords+10:coords+14]\n",
    "                    #print(isolate, ospc_type)\n",
    "                    checked_types_ospc.loc[len(checked_types_ospc)] = [isolate, ospc_type]\n",
    "                    break\n",
    "        else:\n",
    "            #print(isolate, ospc_type)\n",
    "            checked_types_ospc.loc[len(checked_types_ospc)] = [isolate, \"unknown\"]\n",
    "else:\n",
    "    print(\"skipping parsing blast results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if WRITE_RESULTS is True:\n",
    "    checked_types_ospc.to_csv(\"new_ospC_types.csv\", index=False)\n",
    "else:\n",
    "    print(\"skipping writing initial results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if PARSE_BLAST_RESULTS is True:\n",
    "    checked_types_AT = pandas.DataFrame(columns=[\"Isolate\", \"AT_type\"])\n",
    "    blast_results_AT = glob.glob(\"temp_out/*-ospsAT_blast.xml\")\n",
    "\n",
    "    for result in blast_results_AT:\n",
    "        #print(result)\n",
    "        sample_id = result.split(\"/\")[-1].replace(\"-ospsAT_blast.xml\", \"\")\n",
    "        isolate = sample_id\n",
    "        parsed_results = parse_blast_results(result)\n",
    "        if parsed_results is not None:\n",
    "            for result in parsed_results:\n",
    "                if len(parsed_results[result]) > 0:\n",
    "                    best_hit = parsed_results[result][0][0]\n",
    "                    coords = parsed_results[result][0][0].find(\"OspC_Type-\")\n",
    "                    ospc_type = parsed_results[result][0][0][coords+10:coords+14]\n",
    "                    #print(isolate, ospc_type)\n",
    "                    checked_types_AT.loc[len(checked_types_AT)] = [isolate, ospc_type]\n",
    "                    break\n",
    "        else:\n",
    "            #print(isolate, ospc_type)\n",
    "            checked_types_AT.loc[len(checked_types_AT)] = [isolate, \"unknown\"]\n",
    "else:\n",
    "    print(\"skipping parsing blast results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is to check how many have been duplicated to see what is messing up above...\n",
    "print(len(checked_types_ospc))\n",
    "print(len(checked_types_AT))\n",
    "duplicatesOC = checked_types_ospc[checked_types_ospc['Isolate'].duplicated()]\n",
    "print(duplicatesOC)\n",
    "duplicatesAT = checked_types_AT[checked_types_AT['Isolate'].duplicated()]\n",
    "print(duplicatesAT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pprint.pprint(known_types)\n",
    "pprint.pprint(checked_types_ospc)\n",
    "pprint.pprint(checked_types_AT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if MERGE_RESULTS is True:\n",
    "    merged_table = []\n",
    "    # Merge the two tables on the 'Isolate' column\n",
    "    merged_table = pandas.merge(known_types, checked_types_ospc, on='Isolate', how='left')\n",
    "    #print(merged_table)\n",
    "    merged_table = pandas.merge(merged_table, checked_types_AT, on='Isolate', how='left')\n",
    "    #print(merged_table)\n",
    "    for i in range(0, len(merged_table)):\n",
    "        if pandas.isnull(merged_table.at[i, 'OspC_Type']):\n",
    "            merged_table.at[i, 'OspC_Type'] = merged_table.at[i, 'ospc_newtype']\n",
    "    merged_table.drop(columns=['ospc_newtype'], inplace=True)\n",
    "    #print(merged_table)\n",
    "    print(\"Known types and new types tables have been merged!\")\n",
    "else:\n",
    "    print(\"skipping merging\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if FINALIZE_AND_WRITE_METADATA is True:\n",
    "    #input_metadata.drop(columns=['OspC_Type'], inplace=True)\n",
    "    output_metadata = pandas.merge(input_metadata, merged_table, on='Isolate', how='left')\n",
    "    #print(output_metadata)\n",
    "    output_metadata.to_csv(\"longread_metadata_v3.csv\", index=False)\n",
    "    print(\"metadata written!\")\n",
    "else:\n",
    "    print(\"skipping writing final metadata\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
