# Manuscript Meeting 12 June 2024
***
    ##Topics of discussion:
***
    ### Authorship and order of authors
Michael, Rachel, Ben, Amine <- Co-first authors.
    - Klemen, Ira, and rest as second/collaborating authors.
    - Lemieux as senior/supervisory author. (Last author)
>>Each project to be determined individually, not an IOU system of placement.
***
    ### Main sections and their assigned authors.
*(DUE END OF DAY TOMORROW 13JUNE2024)*
    - After this, Amine will finish the introduction and polish the manuscript and finish/write discussion.
    - Ben and Michael will reformat figures and rerun analyses that require it.
    - By Friday at 5 we will distribute to coauthors, we can push back to Monday at 5 if needed.
        a. 1st part of paper - Ben
            - (1000 words across 3 paragraphs summarizing the assembly metrics and the figures.)
            - comparison between methods and laying out long-read superiority.
        b. 2nd part of paper - Michael
            - (1000 words on genome annotation + challenges of plasmid annotations + solutions + our findings on presence to markers and the phylogenetic tree)
            - annotation and plasmid id process among other parts of the dataset processing and generation.
        c. 3rd part of paper - Amine
            - Two main points: Complex population structure + Recombination Structure.
            - (1000 words on genetic diversity in the genome + recombination within genome and description of pangenome + ways that lipoproteins differ from others in these metrics.)
            - population genetics + biology + recombination + etc
        d. Polishing/Partitioning/Organization/Flow - Amine
***
    ### Section Overview
***
    #### Sections 1 + 2: Sequencing Method + Plasmid Analysis
Figure 1. Contiguity + Gene prediction plots by method
    - Could also be shuffled to do:
        a. waterfall,
        b. gene number
        c. plasmid stuff.
Figure 2. Plasmid Presence/Absence tree.
    - (to be figure 3)
Figure 3. Plasmid Presence/Absence tree comparing against sequencing method.
    - (to be figure two to bridge between method comparison and the analysis we performed.)
Figure 4. Plasmid Correlation Graph
    - +/- correlations of plasmids.
    - how interconnected are these plasmids. At what point do these networks begin breaking away.
    - Now lets look at RST specific networks.
Figure 5. Gene Graph
    - Core genes, RST1 specific genes
    - what point does this graphg begin breaking down.
    - larger genetic structure influences network connectedness.

#### Sections 2-2.9? Network Stuff:
This is gonna need to be condensed and moved around to flow with the emergent narrative.
*Priorities*
>>Ben to accept the baton for working with Rachel's figures, Michael and Ben have access to Rachel's codebase.
    1. Translate +/- correlations into biology.
        - Wheel & tire vs wheel A | wheel B
        - Plasmids that are positively correlated are part of the same functional processes, negatively correlated plasmids may be able to substitute.
        - Ex. Necessary to have together, + corr (wheel + tire)
        - Ex. Others are repellant to each other. (tire brands)
Network Figures.
Figure 1.
Figure 2.
Figures 3a & 3b.
Figures 4a & 4b. <- maybe >
    - Network stuff on gene + gene to plasmid analysis.
BAPS + Recombination Figures:
Figure 1.
    - recombination figure (amine)
Figure 2.
    - BAPS figure (amine)
Figure 3.
    - Pangenome openness discussion? <- still evaluating openness, initial findings suggest openness not previously resolved. ->
