{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import csv\n",
    "import glob\n",
    "from Bio import SeqIO\n",
    "from collections import defaultdict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "b31_reference_genbank = '/Users/mf019/bioinformatics/longread_pangenome/ref/renamed_GCF_000008685.2.gbff'\n",
    "# def some vars for paths\n",
    "annotations_dir = '/Users/mf019/bioinformatics/longread_pangenome/assemblies/paired_assemblies/paired_only/longread/annotation'\n",
    "alignments_dir = '/Users/mf019/bioinformatics/longread_pangenome/synteny-checking/pgv_mummer_output_2'\n",
    "output_dir = '/Users/mf019/bioinformatics/longread_pangenome/synteny-checking/output'\n",
    "\n",
    "# Okay let's def some functs to parse the aln tsv from mummer.\n",
    "# we also need to parse the B31 genome and store that in a dictionary for each plasmid.\n",
    "\n",
    "# Basically, {Plasmid: {gene: {coords: int:int, strand: int, name: str, etc...} ... } ... }\n",
    "# And the same thing for contigs.\n",
    "# then parse the aln file and go : okay we have this assembly, this contig, and these coordinates that align with a reference plasmid.\n",
    "# What genes are within these coordinates, and what genes are on the reference between those coordinates.\n",
    "\n",
    "# :)\n",
    "\n",
    "# We will also do this one at a time. Probably also move blast results into each asm directory so each assembly has everything in one place.\n",
    "\n",
    "# That is the goal at least.\n",
    "\n",
    "# I will go for a walk to think about what exactly I'm gonna do. Or rather, how exactly I'm gonna do this.\n",
    "\n",
    "# ultimately I just need to *actually do it* but manually checking is probably the best way forward in the meantime.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# VSCODE JUST DELETED MY FUNCTION DEFINTION CELL OMFG\n",
    "# okay so i need to rewrite:\n",
    "\n",
    "def parse_genbank(path):\n",
    "    records = defaultdict()\n",
    "    for record in SeqIO.parse(path, 'genbank'):\n",
    "        records[record.id] = record\n",
    "    return records\n",
    "\n",
    "def parse_alignment(path):\n",
    "    alignments = []\n",
    "    with open(path, 'r') as infile:\n",
    "        lines = infile.readlines()\n",
    "    keys = lines[0].strip().split('\\t')\n",
    "    for line in lines[1::]:\n",
    "        values = line.strip().split('\\t')\n",
    "        alignments.append({key:value for key,value in zip(keys, values)})\n",
    "    return alignments\n",
    "\n",
    "def is_within_range(feature, start, end):\n",
    "    #print(feature)\n",
    "    start_check = int(feature.location.start) >= start\n",
    "    end_check = int(feature.location.end) <= end\n",
    "    checked = (start_check and end_check)\n",
    "    #if start_check is False and end_check is True:\n",
    "    #    fragment = \"3'-truncation\"\n",
    "    #if  end_check is True and start_check is False:\n",
    "    #    fragment = \"5'-truncation\"\n",
    "    #else:\n",
    "    #    fragment = 'NA'\n",
    "    return checked\n",
    "\n",
    "def get_coverage(length, coords):\n",
    "    covered_positions = set()\n",
    "    for position in coords:\n",
    "        start = position[0]\n",
    "        end = position[1]\n",
    "\n",
    "        if start > end:  # If the start position is higher than the end position - as in cases of reverse strand alignment :) - swap em\n",
    "            start, end = end, start\n",
    "\n",
    "        covered_positions.update(range(start, end+1))\n",
    "    percent_coverage = (len(covered_positions)/length)*100\n",
    "    return percent_coverage\n",
    "\n",
    "def get_features_from_range(record, start, end):\n",
    "    # gonna pull the whole feature\n",
    "    if start > end:  # If the start position is higher than the end position - as in cases of reverse strand alignment :) - swap em\n",
    "        start, end = end, start\n",
    "    features = [feature for feature in record.features if is_within_range(feature, start, end) and feature.type == 'CDS']\n",
    "    return features\n",
    "\n",
    "def get_genes_from_alignments(alignments, ref_dict, asm_dict):\n",
    "    genes_dict = defaultdict()\n",
    "    for alignment in alignments:\n",
    "        # for whatever reason, mummer indicates the reference as the query and the assembly as the ref. Whatever. just be mindful.\n",
    "        ref_id = alignment['QUERY_ID']\n",
    "        ref_name = alignment['QUERY_NAME']\n",
    "        ref_start = int(alignment['QUERY_START'])\n",
    "        ref_end = int(alignment['QUERY_END'])\n",
    "        ref_aln_length = int(alignment['QUERY_LENGTH']) # ALIGNED LENGTH)\n",
    "        ref_length = len(ref_dict[ref_name].seq)\n",
    "        assembly_id = alignment['REF_ID']\n",
    "        contig_name = alignment['REF_NAME'].replace(\"0000\", \"_\").replace(\"_0\", \"_\")\n",
    "        contig_start = int(alignment['REF_START'])\n",
    "        contig_end = int(alignment['REF_END'])\n",
    "        contig_aln_length = int(alignment['REF_LENGTH']) # ALIGNED LENGTH)\n",
    "        aln_identity = alignment['IDENTITY']\n",
    "        contig_length = len(asm_dict[contig_name].seq)\n",
    "        #print(ref_name, ref_length)\n",
    "        # ref_features = [feature.qualifiers['locus_tag'] for feature in ref_dict[query_name].features if is_within_range(feature, aln_start_ref, aln_end_ref) and feature.type == 'CDS']\n",
    "        ref_features = get_features_from_range(ref_dict[ref_name], ref_start, ref_end)\n",
    "        asm_features = get_features_from_range(asm_dict[contig_name], contig_start, contig_end)\n",
    "        ref_genes = simplify_genes_for_contig(ref_features)\n",
    "        asm_genes = simplify_genes_for_contig(asm_features)\n",
    "        #asm_genes = \"placeholder :)\"\n",
    "        # Get percent coverage for each alignment\n",
    "        # UPDATE DON'T DO THAT HERE THE RANGES GET ALL WEIRD.\n",
    "        #ref_coverage = get_coverage(ref_length, ref_start, ref_end)\n",
    "        #contig_coverage = get_coverage(contig_length, contig_start, contig_end)\n",
    "\n",
    "        if contig_name not in genes_dict:\n",
    "            genes_dict[contig_name] = defaultdict(dict)\n",
    "            genes_dict[contig_name]['contig_length'] = int(contig_length)\n",
    "\n",
    "        if ref_name not in genes_dict[contig_name]:\n",
    "            genes_dict[contig_name][ref_name] = []\n",
    "\n",
    "        alignment_dict = {\n",
    "                        'ref_aln': {\n",
    "                                'ref_length': int(ref_length),\n",
    "                                'start': int(ref_start),\n",
    "                                'end': int(ref_end),\n",
    "                                'aln_length': ref_aln_length,\n",
    "                                'percent_cov': int(),\n",
    "                                'features': ref_genes,\n",
    "                            },\n",
    "                        'asm_aln': {\n",
    "                                'start': int(contig_start),\n",
    "                                'end': int(contig_end),\n",
    "                                'aln_length': int(contig_aln_length),\n",
    "                                'aln_identity': aln_identity, # to the reference, this will probably get confusing downstream :)\n",
    "                                'percent_cov': int(),\n",
    "                                'features': asm_genes,\n",
    "                            }\n",
    "                    }\n",
    "\n",
    "        genes_dict[contig_name][ref_name].append(alignment_dict)\n",
    "    #\n",
    "    return genes_dict\n",
    "\n",
    "def simplify_genes_for_contig(features):\n",
    "    genes = []\n",
    "    for feature in features:\n",
    "        locus_tag = feature.qualifiers['locus_tag'][0].strip(\"'\").strip('[').strip(']')\n",
    "        product = ' '.join(feature.qualifiers['product'])\n",
    "        gene = (locus_tag, product)\n",
    "        genes.append(gene)\n",
    "    return genes\n",
    "\n",
    "def make_table_for_asm(genes_dict, output_path):\n",
    "    lines = []\n",
    "    header_row = 'contig_id\\tcontig_len\\tref\\tref_len\\tcontig_cov\\treference_cov\\tgenes_on_ref\\tgenes_on_contig\\n'\n",
    "    with open(output_path, 'w') as outfile:\n",
    "        lines.append(header_row)\n",
    "        for contig in genes_dict:\n",
    "            ref_genes = []\n",
    "            contig_genes = []\n",
    "            for aln in genes_dict[contig].keys():\n",
    "                if aln != 'contig_length':\n",
    "                    contig_len = genes_dict[contig]['contig_length']\n",
    "                    aln_coords = [(cov['asm_aln']['start'],cov['asm_aln']['end']) for cov in genes_dict[contig][aln]]\n",
    "                    ref_aln_coords = [(cov['ref_aln']['start'],cov['ref_aln']['end']) for cov in genes_dict[contig][aln]]\n",
    "                    ref_len = genes_dict[contig][aln][0]['ref_aln']['ref_length']\n",
    "                    for index, item in enumerate(genes_dict[contig][aln]):\n",
    "                        ref_genes.extend([gene[0] for gene in item['ref_aln']['features']])\n",
    "                        contig_genes.extend([gene[0] for gene in item['asm_aln']['features']])\n",
    "                    contig_cov = get_coverage(contig_len, aln_coords)\n",
    "                    #print(contig_cov, aln_coords)\n",
    "                    reference_cov = get_coverage(ref_len, ref_aln_coords)\n",
    "                    output_row = f'{contig}\\t{contig_len}\\t{aln}\\t{ref_len}\\t{contig_cov:.2f}\\t{reference_cov:.2f}\\t{len(ref_genes)}\\t{len(contig_genes)}\\n'\n",
    "                    lines.append(output_row)\n",
    "        outfile.writelines(lines)\n",
    "\n",
    "def run_all(alignments_dir, annotations_dir, reference_genbank, output_dir):\n",
    "    ref_dict = parse_genbank(reference_genbank)\n",
    "    alignment_files = glob.glob(f'{alignments_dir}/*/align_coords.tsv')\n",
    "    for alignment in alignment_files:\n",
    "        sample_id = alignment.split('/')[-2]\n",
    "        print(f'parsing alignments for {sample_id}!')\n",
    "        assembly = f'{annotations_dir}/{sample_id}.gbff'\n",
    "        print(f'gbff for {sample_id} exists: {os.path.exists(assembly)}')\n",
    "        print(assembly)\n",
    "        asm_dict = parse_genbank(assembly)\n",
    "        alignments = parse_alignment(alignment)\n",
    "        genes = get_genes_from_alignments(alignments, ref_dict, asm_dict)\n",
    "        output_path = f'{output_dir}/{sample_id}_B31_Synteny.tsv'\n",
    "        make_table_for_asm(genes, output_path)\n",
    "        print(f'Finished! moving on')\n",
    "\n",
    "run_all(alignments_dir, annotations_dir, b31_reference_genbank, output_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "error_gb = parse_genbank('/Users/mf019/bioinformatics/longread_pangenome/assemblies/paired_assemblies/paired_only/longread/annotation/UWI263P.gbff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "error_gb['contig_1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
