{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import pandas\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "from collections import defaultdict\n",
    "from pygenomeviz import GenomeViz\n",
    "from pygenomeviz.parser import Genbank\n",
    "from pygenomeviz.align import AlignCoord\n",
    "from pygenomeviz.utils import ColorCycler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lets start with some functions to process the contig, this will come from iterating through the align_coords.tsv dataframe\n",
    "# and grouping by contig_id.\n",
    "\n",
    "def process_assembly(homologies_file, syntenies_file, reference_gb, assembly_gb, output_dir, **kwargs):\n",
    "    # process the alignments and spit out a plot for each contig :)\n",
    "    gbk_assembly = list(SeqIO.parse(open(assembly_gb, 'r'), 'genbank'))\n",
    "    gbk_reference = list(SeqIO.parse(open(reference_gb, 'r'), 'genbank'))\n",
    "    assembly_id = Path(assembly_gb).stem\n",
    "    reference_id = '.'.join(os.path.basename(reference_gb).split('.')[:-1])\n",
    "\n",
    "    output_path = os.path.join(output_dir, assembly_id)\n",
    "    png_path = os.path.join(output_path, 'png')\n",
    "    html_path = os.path.join(output_path, 'html')\n",
    "\n",
    "    check_n_make(output_path)\n",
    "    check_n_make(png_path)\n",
    "    check_n_make(html_path)\n",
    "\n",
    "    homologies = pandas.read_csv(homologies_file, sep='\\t')\n",
    "    syntenies = pandas.read_csv(syntenies_file, sep = '\\t')\n",
    "\n",
    "    grouped_homologies = homologies.groupby('REF_NAME', ) # THIS WILL REQUIRE CHANGING IN THE FUTURE {{TO-DO: FIX FOR NON B31 CASE}}\n",
    "    grouped_syntenies  = syntenies.groupby('REF_NAME', ) # THIS WILL REQUIRE CHANGING IN THE FUTURE {{TO-DO: FIX FOR NON B31 CASE}}\n",
    "\n",
    "    for contig_id, contig_df in grouped_homologies:\n",
    "        num_homologies = len(contig_df)\n",
    "        if check_group_exists(grouped_syntenies, contig_id):\n",
    "            num_syntenies = len(grouped_syntenies.get_group(contig_id))\n",
    "        else:\n",
    "            num_syntenies = 0\n",
    "        print(f'Processing {contig_id}: homologies:{num_homologies}, syntenies:{num_syntenies}')\n",
    "        if num_syntenies > 0:\n",
    "            process_and_plot(contig_id, contig_df, grouped_syntenies, gbk_assembly, gbk_reference, assembly_id, reference_id, output_path)\n",
    "        else:\n",
    "            print(f\"\\tError plotting {contig_id}! No syntenies! Attempting to only plot homology!\\n\")\n",
    "            process_and_plot(contig_id, contig_df, grouped_syntenies, gbk_assembly, gbk_reference, assembly_id, reference_id, output_path, homology_only=True)\n",
    "    print(f\"Finished plotting contigs for {assembly_id}! Have a wonderful day :)\")\n",
    "\n",
    "def process_and_plot(contig_id, contig_df, grouped_syntenies, gbk_assembly, gbk_reference, assembly_id, reference_id, output_path, **kwargs):\n",
    "    # Process contig alignments for homology and synteny\n",
    "    homology_only = kwargs.get('homology_only', False)\n",
    "    # ok now let's parse this out,\n",
    "    ref_homology_ids, homology_coords = process_contig_alignments(contig_df)\n",
    "    if homology_only:\n",
    "        reference_ids = ref_homology_ids\n",
    "    else:\n",
    "        ref_synteny_ids, synteny_coords = process_contig_alignments(grouped_syntenies.get_group(contig_id))\n",
    "        reference_ids = ref_homology_ids.union(ref_synteny_ids)\n",
    "\n",
    "    # Parse GenBank files for assembly and reference\n",
    "    gbks = []\n",
    "    gbk_asm = parse_split_genbank(gbk_assembly, contig_id, assembly_id)\n",
    "    gbks.append(gbk_asm)\n",
    "    gbk_ref = parse_split_genbank(gbk_reference, reference_ids, reference_id)\n",
    "    gbks.append(gbk_ref)\n",
    "\n",
    "    # Initialize GenomeViz\n",
    "    gv = GenomeViz(track_align_type=\"center\")\n",
    "    gv.set_scale_bar()\n",
    "    # Add tracks for each GenBank file\n",
    "    track_number = 0\n",
    "    for gbk in gbks:\n",
    "        if track_number == 0:\n",
    "            sublabel_pos = \"top-center\"\n",
    "        else:\n",
    "            sublabel_pos = \"bottom-center\"\n",
    "        color = ColorCycler()\n",
    "        #print(gbk.name)\n",
    "        track = gv.add_feature_track(gbk.name, gbk.get_seqid2size(), space=0.05, label_kws=dict(color=color), align_label=False)\n",
    "        for seqid, features in gbk.get_seqid2features(feature_type=\"CDS\").items():\n",
    "            segment = track.get_segment(seqid)\n",
    "\n",
    "            segment.add_sublabel(f\"{segment.name}: {segment.start:,}-{segment.end:,}bp\", size = 8, pos = sublabel_pos, ymargin=0.3)\n",
    "            for feature in features:\n",
    "                segment.add_features(feature, fc=\"blue\", lw=0.077, ignore_outside_range = True, plotstyle=\"arrow\")\n",
    "\n",
    "        track_number += 1\n",
    "    # Add homology and synteny links\n",
    "    if len(homology_coords) > 0:\n",
    "        min_ident = int(min([ac.identity for ac in homology_coords if ac.identity]))\n",
    "        color, inverted_color = \"grey\", \"red\"\n",
    "        for ac in homology_coords:\n",
    "            gv.add_link(ac.query_link, ac.ref_link, color=color, inverted_color=inverted_color, filter_length=250, curve=True, v=ac.identity, vmin=min_ident)\n",
    "        gv.set_colorbar([color, inverted_color], vmin=min_ident, bar_label=\"Identity\")\n",
    "    if not homology_only:\n",
    "        color, inverted_color = \"blue\", \"purple\"\n",
    "        for sc in synteny_coords:\n",
    "            gv.add_link(sc.query_link, sc.ref_link, color=color, inverted_color=inverted_color, curve=True)\n",
    "\n",
    "    # Save the plot as PNG and HTML\n",
    "    output_png = os.path.join(output_path, 'png', f'{contig_id}_synteny_plot.png')\n",
    "    output_html = os.path.join(output_path, 'html', f'{contig_id}_synteny_plot.html')\n",
    "    gv.savefig(output_png)\n",
    "    gv.savefig_html(output_html)\n",
    "    print(f'Finished plotting {contig_id}\\n')\n",
    "\n",
    "def process_contig_alignments(contig_df):\n",
    "    #keys = ['QUERY_ID', 'QUERY_NAME', 'QUERY_START', 'QUERY_END', 'QUERY_LENGTH',\n",
    "    #   'REF_ID', 'REF_NAME', 'REF_START', 'REF_END', 'REF_LENGTH', 'IDENTITY',\n",
    "    #   'EVALUE']\n",
    "    reference_ids = set()\n",
    "    acs = []\n",
    "    for index, row in contig_df.iterrows():\n",
    "        row_dict = {k:v for k,v in row.items()}\n",
    "        reference_ids.add(row_dict['QUERY_NAME']) # change this to query_id for all_v_all ? doesn't matter for B31\n",
    "        acs.append(\n",
    "            AlignCoord(\n",
    "                row_dict['QUERY_ID'],\n",
    "                row_dict['QUERY_NAME'],\n",
    "                row_dict['QUERY_START'],\n",
    "                row_dict['QUERY_END'],\n",
    "                row_dict['REF_ID'],\n",
    "                row_dict['REF_NAME'],\n",
    "                row_dict['REF_START'],\n",
    "                row_dict['REF_END'],\n",
    "                row_dict['IDENTITY'],\n",
    "            )\n",
    "        )\n",
    "    return reference_ids, acs\n",
    "\n",
    "def parse_genbanks_for_contig(reference_ids):\n",
    "    files = []\n",
    "    for genbank_id in reference_ids:\n",
    "        genbank_file = get_genbank_file(genbank_id)\n",
    "        files.append(genbank_file)\n",
    "    combined_temp_file = concatenate_genbank_files(files)\n",
    "    gbk_ref = Genbank(combined_temp_file)\n",
    "    return gbk_ref\n",
    "\n",
    "def parse_split_genbank(records, record_ids, temp_file_name):\n",
    "    temp_dir = tempfile.gettempdir()\n",
    "    temp_file_path = os.path.join(temp_dir, temp_file_name)\n",
    "    filtered_records = [record for record in records if record.id in record_ids]\n",
    "    with open(temp_file_path, 'w') as outfile:\n",
    "        SeqIO.write(filtered_records, outfile, 'genbank')\n",
    "    gbk_ref = Genbank(temp_file_path, name=temp_file_name)\n",
    "    return gbk_ref\n",
    "\n",
    "def check_group_exists(grouped_df, group_name):\n",
    "    return group_name in grouped_df.groups\n",
    "\n",
    "def check_n_make(path):\n",
    "    if not os.path.exists(path):\n",
    "        os.makedirs(path)\n",
    "\n",
    "def concatenate_genbank_files(file_list):\n",
    "    # Create a temporary file\n",
    "    temp_file = tempfile.NamedTemporaryFile(delete=False, mode='w', suffix='.gb')\n",
    "\n",
    "    with open(temp_file.name, 'w') as outfile:\n",
    "        for file_name in file_list:\n",
    "            with open(file_name, 'r') as infile:\n",
    "                records = SeqIO.parse(infile, 'genbank')\n",
    "                SeqIO.write(records, outfile, 'genbank')\n",
    "    # Return the name of the temporary file\n",
    "    return temp_file.name\n",
    "\n",
    "def get_genbank_file(genbank_id, where_to_look):\n",
    "    file_path = os.path.join(where_to_look, f'{genbank_id}.gbff')\n",
    "    if os.path.exists(file_path):\n",
    "        return genbank_file\n",
    "    else:\n",
    "        raise FileNotFoundError(f\"The file at path '{file_path}' does not exist.\")\n",
    "\n",
    "def make_track_from_gb(gv, track_id, lens, features):\n",
    "    track = gv.add_feature_track(track_id, lens)\n",
    "    track.add_sublabel()\n",
    "    track.add_features(features)\n",
    "    return track\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "cwd = os.getcwd()\n",
    "test_output_dir = os.path.join(cwd,'script_testing_output')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "assembly_genbanks_dir = '/home/mf019/borrelia_plasmid_classifier_v3/assemblies/longread/annotation'\n",
    "assembly_genbanks = glob.glob(f'{assembly_genbanks_dir}/**/*.gbff', recursive=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_homology_file = 'asms_vs_B31_v2/ESI26H/align_coords.tsv'\n",
    "test_homology = pandas.read_csv(test_homology_file, sep = '\\t', )\n",
    "test_homology = test_homology.drop(['EVALUE'], axis=1)\n",
    "\n",
    "test_synteny_file = 'asms_vs_B31_prot/ESI26H/align_coords.tsv'\n",
    "test_synteny = pandas.read_csv(test_synteny_file, sep = '\\t', )\n",
    "test_synteny = test_synteny.drop(['EVALUE'], axis=1)\n",
    "\n",
    "test_assembly_genbank = '/home/mf019/borrelia_plasmid_classifier_v3/assemblies/longread/annotation/ESI26H/ESI26H.gbff'\n",
    "b31_genbank = 'renamed_GCF_000008685.2.gbff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing contig000001: homologies:3, syntenies:746\n",
      "Finished plotting contig000001\n",
      "\n",
      "Processing contig000002: homologies:12, syntenies:54\n",
      "Finished plotting contig000002\n",
      "\n",
      "Processing contig000003: homologies:15, syntenies:31\n",
      "Finished plotting contig000003\n",
      "\n",
      "Processing contig000004: homologies:2, syntenies:26\n",
      "Finished plotting contig000004\n",
      "\n",
      "Processing contig000005: homologies:3, syntenies:29\n",
      "Finished plotting contig000005\n",
      "\n",
      "Processing contig000006: homologies:1, syntenies:1\n",
      "Finished plotting contig000006\n",
      "\n",
      "Processing contig000007: homologies:6, syntenies:9\n",
      "Finished plotting contig000007\n",
      "\n",
      "Processing contig000008: homologies:7, syntenies:14\n",
      "Finished plotting contig000008\n",
      "\n",
      "Processing contig000009: homologies:3, syntenies:23\n",
      "Finished plotting contig000009\n",
      "\n",
      "Processing contig000010: homologies:7, syntenies:13\n",
      "Finished plotting contig000010\n",
      "\n",
      "Processing contig000011: homologies:2, syntenies:18\n",
      "Finished plotting contig000011\n",
      "\n",
      "Processing contig000012: homologies:4, syntenies:8\n",
      "Finished plotting contig000012\n",
      "\n",
      "Processing contig000013: homologies:1, syntenies:1\n",
      "Finished plotting contig000013\n",
      "\n",
      "Processing contig000016: homologies:1, syntenies:0\n",
      "\tError plotting contig000016! No syntenies! Attempting to only plot homology!\n",
      "\n",
      "Finished plotting contig000016\n",
      "\n",
      "Processing contig000019: homologies:1, syntenies:1\n",
      "Finished plotting contig000019\n",
      "\n",
      "Processing contig000020: homologies:1, syntenies:0\n",
      "\tError plotting contig000020! No syntenies! Attempting to only plot homology!\n",
      "\n",
      "Finished plotting contig000020\n",
      "\n",
      "Processing contig000029: homologies:1, syntenies:0\n",
      "\tError plotting contig000029! No syntenies! Attempting to only plot homology!\n",
      "\n",
      "Finished plotting contig000029\n",
      "\n",
      "Processing contig000033: homologies:1, syntenies:0\n",
      "\tError plotting contig000033! No syntenies! Attempting to only plot homology!\n",
      "\n",
      "Finished plotting contig000033\n",
      "\n",
      "Processing contig000040: homologies:1, syntenies:0\n",
      "\tError plotting contig000040! No syntenies! Attempting to only plot homology!\n",
      "\n",
      "Finished plotting contig000040\n",
      "\n",
      "Processing contig000041: homologies:1, syntenies:0\n",
      "\tError plotting contig000041! No syntenies! Attempting to only plot homology!\n",
      "\n",
      "Finished plotting contig000041\n",
      "\n",
      "Processing contig000044: homologies:1, syntenies:0\n",
      "\tError plotting contig000044! No syntenies! Attempting to only plot homology!\n",
      "\n",
      "Finished plotting contig000044\n",
      "\n",
      "Processing contig000051: homologies:1, syntenies:0\n",
      "\tError plotting contig000051! No syntenies! Attempting to only plot homology!\n",
      "\n",
      "Finished plotting contig000051\n",
      "\n",
      "Processing contig000055: homologies:1, syntenies:0\n",
      "\tError plotting contig000055! No syntenies! Attempting to only plot homology!\n",
      "\n",
      "Finished plotting contig000055\n",
      "\n",
      "Processing contig000062: homologies:1, syntenies:0\n",
      "\tError plotting contig000062! No syntenies! Attempting to only plot homology!\n",
      "\n",
      "Finished plotting contig000062\n",
      "\n",
      "Finished plotting contigs for ESI26H! Have a wonderful day :)\n"
     ]
    }
   ],
   "source": [
    "fig = process_assembly(test_homology_file, test_synteny_file,\n",
    "                       b31_genbank_file, test_assembly_genbank, test_output_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
