{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import igraph as ig\n",
    "import chart_studio.plotly as ply\n",
    "import plotly.graph_objects as go\n",
    "import plotly.express as px\n",
    "import plotly.io as pio\n",
    "from scipy.spatial import ConvexHull\n",
    "import random\n",
    "import os\n",
    "import pathlib\n",
    "import glob\n",
    "import kaleido"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: kaleido in /opt/homebrew/lib/python3.11/site-packages (0.2.1)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_matrix = 'prot_v1/matrices/shared_gene_matrix_90percent_v1.tsv'\n",
    "edge_matrix = 'prot_v1/matrices/shared_gene_matrix_90percent_v1_edges.tsv'\n",
    "id_mapping = '/Users/mf019/bioinformatics/longread_pangenome/plasmid_id/output/best_matches_v9_1kb.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def edges_from_matrix(matrix, output):\n",
    "    df = pd.read_csv(matrix, sep='\\t', index_col=0, header=0)\n",
    "\n",
    "    source = []\n",
    "    target = []\n",
    "    weight = []\n",
    "\n",
    "    for i in range(len(df.index)):\n",
    "        for j in range(i, len(df.columns)):\n",
    "            if df.iloc[i,j] > 0:\n",
    "                source.append(df.index[i])\n",
    "                target.append(df.columns[j])\n",
    "                score = df.iloc[i,j]\n",
    "                weight.append(score)\n",
    "                #if df.iloc[i,i] >= 1000 and df.iloc[j,j] > 1000:\n",
    "                #    source.append(df.index[i])\n",
    "                #    target.append(df.columns[j])\n",
    "                #    score = 2*df.iloc[i,j]/sum([df.iloc[i,i], df.iloc[j,j]])\n",
    "                #    weight.append(score)\n",
    "\n",
    "\n",
    "    edge_df = pd.DataFrame({\n",
    "        'Source': source,\n",
    "        'Target': target,\n",
    "        'weight': weight,\n",
    "        'interaction': ['interacts'] * len(source)  # Add this line apparently.\n",
    "    })\n",
    "    edge_df.to_csv(output, index=False)\n",
    "    print(f'saved edges to {output}')\n",
    "    return edge_df\n",
    "\n",
    "def create_igraph(edge_df):\n",
    "    nodes = list(set(edge_df['Source'].to_list() + edge_df['Target'].to_list()))\n",
    "    nodes = [str(node) for node in nodes]  # Convert all nodes to strings\n",
    "    G = ig.Graph()\n",
    "    G.add_vertices(nodes)\n",
    "    edges = [(str(row['Source']), str(row['Target'])) for _, row in edge_df[['Source', 'Target']].iterrows()]\n",
    "    G.add_edges(edges)\n",
    "    G.es['weight'] = edge_df['weight'].tolist()\n",
    "    return G, nodes  # Return both the graph and the order of nodes\n",
    "\n",
    "def align_labels_and_groups(idmap, vertex_order):\n",
    "    longread_idmap = idmap[idmap['method'] == 'longread']\n",
    "    # Create a dictionary mapping contig names to their labels and groups\n",
    "    node_dict = longread_idmap.apply(lambda x: f\"{x['name']}_{x['contig']}\", axis=1)\n",
    "    node_dict = dict(zip(node_dict, longread_idmap['best_hit']))\n",
    "\n",
    "    aligned_labels = []\n",
    "    aligned_groups = []\n",
    "\n",
    "    print(\"Sample of idmap:\")\n",
    "    print(longread_idmap.head())\n",
    "\n",
    "    print(\"\\nSample of vertex_order:\")\n",
    "    print(vertex_order[:10])\n",
    "\n",
    "    for vertex in vertex_order:\n",
    "        if vertex in node_dict:\n",
    "            aligned_labels.append(vertex)\n",
    "            aligned_groups.append(node_dict[vertex])\n",
    "        else:\n",
    "            print(f'cannot find {vertex} in node_dict!')\n",
    "            aligned_labels.append(vertex)\n",
    "            aligned_groups.append('unknown')\n",
    "\n",
    "    print(f\"\\nTotal vertices: {len(vertex_order)}\")\n",
    "    print(f\"Aligned labels: {len(aligned_labels)}\")\n",
    "    print(f\"Aligned groups: {len(aligned_groups)}\")\n",
    "    print(f\"Sample of aligned labels: {aligned_labels[:10]}\")\n",
    "    print(f\"Sample of aligned groups: {aligned_groups[:10]}\")\n",
    "    print(f\"Unique groups: {set(aligned_groups)}\")\n",
    "\n",
    "    return aligned_labels, aligned_groups\n",
    "\n",
    "# random colors\n",
    "# def create_color_mapping(groups):\n",
    "#    unique_groups = list(set(groups))\n",
    "#    n_colors = len(unique_groups)\n",
    "#    colors = px.colors.qualitative.Plotly * (n_colors // len(px.colors.qualitative.Plotly) + 1)\n",
    "#    #colors = px.colors.qualitative.Plotly[:n_colors]\n",
    "#    return {group: colors[i] for i, group in enumerate(unique_groups)}\n",
    "\n",
    "def create_color_mapping(groups):\n",
    "    color_map = {\n",
    "        \"cp26\": \"#d60000\", \"lp54\": \"#018700\", \"lp17\": \"#b500ff\", \"lp28-3\": \"#05acc6\",\n",
    "        \"lp28-4\": \"#97ff00\", \"lp38\": \"#ffa52f\", \"cp32-7\": \"#ff8ec8\", \"cp32-4\": \"#79525e\",\n",
    "        \"lp36\": \"#00fdcf\", \"cp32-6\": \"#afa5ff\", \"lp25\": \"#93ac83\", \"lp28-1\": \"#9a6900\",\n",
    "        \"cp32-3\": \"#366962\", \"cp32-9\": \"#d3008c\", \"cp32-5\": \"#fdf490\", \"cp32-11\": \"#c86e66\",\n",
    "        \"cp32-12\": \"#9ee2ff\", \"cp32-10\": \"#00c846\", \"lp28-2\": \"#a877ac\", \"lp28-6\": \"#b8ba01\",\n",
    "        \"lp21\": \"#f4bfb1\", \"lp28-5\": \"#ff28fd\", \"cp32-8\": \"#f2cdff\", \"cp32-1\": \"#009e7c\",\n",
    "        \"cp9\": \"#ff6200\", \"lp56\": \"#56642a\", \"lp28-7\": \"#953f1f\", \"cp32-13\": \"#90318e\",\n",
    "        \"lp5\": \"#ff3464\", \"lp21-cp9\": \"#a0e491\", \"cp9-3\": \"#8c9ab1\", \"lp28-8\": \"#829026\",\n",
    "        \"lp28-9\": \"#ae083f\", \"chromosome\": \"#9eecff\", \"Absent\": \"white\"\n",
    "    }\n",
    "\n",
    "    # For any groups not in the predefined color_map, assign a random color\n",
    "    unique_groups = set(groups)\n",
    "    for group in unique_groups:\n",
    "        if group not in color_map:\n",
    "            color_map[group] = \"#{:06x}\".format(random.randint(0, 0xFFFFFF))\n",
    "\n",
    "    return color_map\n",
    "\n",
    "\n",
    "def make_3d_plot(G, labels, groups, output_file):\n",
    "    layout = G.layout('kk', dim=3)\n",
    "    Xn = [layout[k][0] for k in range(len(G.vs))]\n",
    "    Yn = [layout[k][1] for k in range(len(G.vs))]\n",
    "    Zn = [layout[k][2] for k in range(len(G.vs))]\n",
    "\n",
    "    Xe = []\n",
    "    Ye = []\n",
    "    Ze = []\n",
    "\n",
    "    for e in G.es:\n",
    "        Xe += [layout[e.source][0],layout[e.target][0], None]\n",
    "        Ye += [layout[e.source][1],layout[e.target][1], None]\n",
    "        Ze += [layout[e.source][2],layout[e.target][2], None]\n",
    "\n",
    "    color_map = create_color_mapping(groups)\n",
    "    color_values = [color_map[group] for group in groups]\n",
    "\n",
    "    trace1 = go.Scatter3d(\n",
    "        x = Xe,\n",
    "        y = Ye,\n",
    "        z = Ze,\n",
    "        mode = 'lines',\n",
    "        line = dict(\n",
    "            color = 'rgb(125,125,125)',\n",
    "            width=0.5,\n",
    "            ),\n",
    "        opacity=0.7,\n",
    "        hoverinfo = 'none'\n",
    "    )\n",
    "\n",
    "    trace2 = go.Scatter3d(\n",
    "        x = Xn,\n",
    "        y = Yn,\n",
    "        z = Zn,\n",
    "        mode = 'markers',\n",
    "        name = 'contigs',\n",
    "        marker = dict(\n",
    "            symbol='circle',\n",
    "            size=8,\n",
    "            color=color_values,\n",
    "            colorscale='Viridis',\n",
    "            line=dict(color='rgb(50,50,50)', width=0.5),\n",
    "        ),\n",
    "        text=labels,\n",
    "        hoverinfo = 'text'\n",
    "    )\n",
    "    # create opacity slider\n",
    "    steps = []\n",
    "    for step in np.arange(0, 1.1, 0.1):\n",
    "        step = round(step, 2)\n",
    "        steps.append(\n",
    "            dict(\n",
    "                method=\"update\",\n",
    "                args=[{\"opacity\": [step, 1]}],  # Update only trace1 (edges)\n",
    "                label=str(step)\n",
    "            )\n",
    "        )\n",
    "\n",
    "    # edge opacity slider\n",
    "    sliders = [dict(\n",
    "        active=7,\n",
    "        currentvalue={\"prefix\": \"Edge Opacity: \"},\n",
    "        pad={\"t\": 40},\n",
    "        steps=steps\n",
    "    )]\n",
    "\n",
    "    layout3d = go.Layout(\n",
    "        title=\"Sequence Homology between contigs across 49 long-read assemblies (3D)\",\n",
    "        width=1900,\n",
    "        height=750,\n",
    "        showlegend=False,\n",
    "        scene=dict(\n",
    "            xaxis=dict(title=''),\n",
    "            yaxis=dict(title=''),\n",
    "            zaxis=dict(title=''),\n",
    "        ),\n",
    "        margin=dict(r=0, l=0, b=0, t=100),\n",
    "        hovermode='closest',\n",
    "        annotations=[\n",
    "            dict(\n",
    "                showarrow=False,\n",
    "                text=\"MJF-2024\",\n",
    "                xref='paper',\n",
    "                yref='paper',\n",
    "                x=0,\n",
    "                y=0.1,\n",
    "                xanchor='left',\n",
    "                yanchor='bottom',\n",
    "                font=dict(\n",
    "                    size=8\n",
    "                )\n",
    "            )\n",
    "        ],\n",
    "        sliders=sliders\n",
    "        )\n",
    "    data=[trace1, trace2]\n",
    "    fig = go.Figure(data, layout=layout3d)\n",
    "    fig.update_layout(\n",
    "        scene_aspectmode='data',\n",
    "        autosize=True\n",
    "        )\n",
    "    config = {\n",
    "        'responsive': True,\n",
    "        'scrollZoom': True,\n",
    "    }\n",
    "    pio.write_html(fig, file=output_file, full_html=False, include_plotlyjs='cdn', config=config)\n",
    "    print(\"html written!\")\n",
    "# useful lambda\n",
    "#labels = idmap.apply(lambda x: f\"{x['name']}_{x['contig']}\" if x['method'] == 'longread' else '', axis=1)\n",
    "#labels = labels[labels != '']\n",
    "#groups = list(set(idmap['best_hit']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#def create_color_mapping(groups):\n",
    "#    custom_palette = [\n",
    "#    '#FF5733', '#33FF57', '#3357FF', '#F3FF33', '#33FFF6', '#F633FF', '#FF33A1', '#33FFA1',\n",
    "#    '#A1FF33', '#FFA133', '#33A1FF', '#A133FF', '#FF33F6', '#bdc258', '#FF33D4', '#33FFD4',\n",
    "#    '#D4FF33', '#FFD433', '#33D4FF', '#D433FF', '#FF5733', '#33FF57', '#3357FF', '#F3FF33',\n",
    "#    '#33FFF6', '#F633FF', '#FF33A1', '#33FFA1', '#A1FF33', '#FFA133', '#33A1FF', '#A133FF',\n",
    "#    '#FF33F6', '#F6FF33', '#FF33D4', '#33FFD4', '#D4FF33', '#FFD433', '#33D4FF', '#D433FF',\n",
    "#    '#FF5733', '#33FF57', '#3357FF', '#F3FF33', '#33FFF6', '#F633FF', '#FF33A1', '#33FFA1',\n",
    "#    '#A1FF33', '#FFA133'\n",
    "#    ]\n",
    "#    unique_groups = list(set(groups))\n",
    "#    n_colors = len(unique_groups)\n",
    "#\n",
    "#    # Repeat the custom palette as needed to match the number of unique groups\n",
    "#    colors = custom_palette * (n_colors // len(custom_palette) + 1)\n",
    "#\n",
    "#    # Slice to ensure only n_colors are used\n",
    "#    return {group: colors[i] for i, group in enumerate(unique_groups)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_3d_plot_v2(G, labels, groups, output_file, layout):\n",
    "    layout = G.layout(layout, dim=3)\n",
    "    Xn = [layout[k][0] for k in range(len(G.vs))]\n",
    "    Yn = [layout[k][1] for k in range(len(G.vs))]\n",
    "    Zn = [layout[k][2] for k in range(len(G.vs))]\n",
    "\n",
    "    color_map = create_color_mapping(groups)\n",
    "    color_values = [color_map[group] for group in groups]\n",
    "\n",
    "    # Create a list of unique groups\n",
    "    unique_groups = list(set(groups))\n",
    "    print(unique_groups)\n",
    "\n",
    "    # Create a single edge trace for all connections\n",
    "    Xe, Ye, Ze = [], [], []\n",
    "    for e in G.es:\n",
    "        Xe += [layout[e.source][0], layout[e.target][0], None]\n",
    "        Ye += [layout[e.source][1], layout[e.target][1], None]\n",
    "        Ze += [layout[e.source][2], layout[e.target][2], None]\n",
    "\n",
    "    edge_trace = go.Scatter3d(\n",
    "        x=Xe,\n",
    "        y=Ye,\n",
    "        z=Ze,\n",
    "        mode='lines',\n",
    "        name='Edges',\n",
    "        line=dict(color='rgb(125,125,125)', width=0.35),\n",
    "        hoverinfo='none',\n",
    "        showlegend=False\n",
    "    )\n",
    "\n",
    "    # Create node traces for each group\n",
    "    node_traces = []\n",
    "    for group in unique_groups:\n",
    "        group_indices = [i for i, g in enumerate(groups) if g == group]\n",
    "        node_traces.append(\n",
    "            go.Scatter3d(\n",
    "                x=[Xn[i] for i in group_indices],\n",
    "                y=[Yn[i] for i in group_indices],\n",
    "                z=[Zn[i] for i in group_indices],\n",
    "                mode='markers',\n",
    "                name=group,\n",
    "                marker=dict(\n",
    "                    symbol='circle',\n",
    "                    size=6,\n",
    "                    color=color_map[group],\n",
    "                    line=dict(color='rgb(38, 38, 38)', width=0.35),\n",
    "                ),\n",
    "                #text=[labels[i] for i in group_indices],\n",
    "                text=[f'Label: {labels[i]}<br>Group: {group}' for i in group_indices],\n",
    "                hoverinfo='text'\n",
    "            )\n",
    "        )\n",
    "\n",
    "    # Combine all traces\n",
    "    data = [edge_trace] + node_traces\n",
    "\n",
    "    # Create slider for edge opacity\n",
    "    steps = []\n",
    "    for step in np.arange(0, 1.1, 0.1):\n",
    "        step = round(step, 2)\n",
    "        steps.append(\n",
    "            dict(\n",
    "                method=\"update\",\n",
    "                args=[{\"opacity\": [step, *[1]*len(node_traces)]}],  # First trace is edges, rest are nodes\n",
    "                label=str(step)\n",
    "            )\n",
    "        )\n",
    "\n",
    "    sliders = [dict(\n",
    "        active=5,\n",
    "        currentvalue={\"prefix\": \"Edge Opacity: \"},\n",
    "        pad={\"t\": 5, \"b\": 10},\n",
    "        steps=steps\n",
    "    )]\n",
    "\n",
    "    layout3d = go.Layout(\n",
    "        title=\"Sequence Homology between contigs across 49 long-read assemblies (3D)\",\n",
    "        scene=dict(\n",
    "            xaxis=dict(title=''),\n",
    "            yaxis=dict(title=''),\n",
    "            zaxis=dict(title=''),\n",
    "        ),\n",
    "        margin=dict(r=0, l=0, b=0, t=100),\n",
    "        hovermode='closest',\n",
    "        legend=dict(\n",
    "            itemsizing='constant',\n",
    "            title_text='Plasmids',\n",
    "            bgcolor='rgba(255,255,255,0.5)',\n",
    "            bordercolor='rgba(0,0,0,0)',\n",
    "            borderwidth=2\n",
    "        ),\n",
    "        sliders=sliders,\n",
    "        annotations=[\n",
    "            dict(\n",
    "                showarrow=False,\n",
    "                text=\"MJF-2024\",\n",
    "                xref='paper',\n",
    "                yref='paper',\n",
    "                x=0,\n",
    "                y=0.1,\n",
    "                xanchor='left',\n",
    "                yanchor='bottom',\n",
    "                font=dict(\n",
    "                    size=8\n",
    "                )\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    fig = go.Figure(data=data, layout=layout3d)\n",
    "\n",
    "    fig.update_layout(\n",
    "        scene_aspectmode='data',\n",
    "        autosize=True,\n",
    "        uirevision=True\n",
    "    )\n",
    "\n",
    "    config = {\n",
    "        'responsive': True,\n",
    "        'scrollZoom': True,\n",
    "    }\n",
    "    html_outf = f'{output_file}.html'\n",
    "    png_outf = f'{output_file}.png'\n",
    "    pio.to_image(fig, file=png_outf,\n",
    "    pio.write_html(fig, file=output_file, full_html=False, include_plotlyjs='cdn', config=config)\n",
    "    print(\"html written!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_3d_plot_v3(G, labels, groups, output_file):\n",
    "    layout = G.layout('kk', dim=3)\n",
    "    Xn = [layout[k][0] for k in range(len(G.vs))]\n",
    "    Yn = [layout[k][1] for k in range(len(G.vs))]\n",
    "    Zn = [layout[k][2] for k in range(len(G.vs))]\n",
    "    color_map = create_color_mapping(groups)\n",
    "\n",
    "    # Create a list of unique groups\n",
    "    unique_groups = list(set(groups))\n",
    "    print(len(unique_groups))\n",
    "    # Create edge traces for connections between each pair of groups\n",
    "    edge_traces = []\n",
    "    for i, group1 in enumerate(unique_groups):\n",
    "        for j, group2 in enumerate(unique_groups[i:], i):  # Start from i to avoid duplicates\n",
    "            Xe, Ye, Ze = [], [], []\n",
    "            for e in G.es:\n",
    "                if (groups[e.source] == group1 and groups[e.target] == group2) or (groups[e.source] == group2 and groups[e.target] == group1):\n",
    "                    Xe += [layout[e.source][0], layout[e.target][0], None]\n",
    "                    Ye += [layout[e.source][1], layout[e.target][1], None]\n",
    "                    Ze += [layout[e.source][2], layout[e.target][2], None]\n",
    "            if Xe:  # Only add the trace if there are edges\n",
    "                edge_traces.append(\n",
    "                    go.Scatter3d(\n",
    "                        x=Xe,\n",
    "                        y=Ye,\n",
    "                        z=Ze,\n",
    "                        mode='lines',\n",
    "                        name=group1,\n",
    "                        line=dict(color='rgb(125,125,125)', width=0.5),\n",
    "                        hoverinfo='skip',\n",
    "                        showlegend=False,\n",
    "                        visible=True\n",
    "                    )\n",
    "                )\n",
    "    print(len(edge_traces))\n",
    "    # Create node traces for each group\n",
    "    node_traces = []\n",
    "    for group in unique_groups:\n",
    "        group_indices = [i for i, g in enumerate(groups) if g == group]\n",
    "        node_traces.append(\n",
    "            go.Scatter3d(\n",
    "                x=[Xn[i] for i in group_indices],\n",
    "                y=[Yn[i] for i in group_indices],\n",
    "                z=[Zn[i] for i in group_indices],\n",
    "                mode='markers',\n",
    "                name=group,\n",
    "                marker=dict(\n",
    "                    symbol='circle',\n",
    "                    size=8,\n",
    "                    color=color_map[group],\n",
    "                    line=dict(color='rgb(50,50,50)', width=0.5),\n",
    "                ),\n",
    "                text=[labels[i] for i in group_indices],\n",
    "                hoverinfo='text'\n",
    "            )\n",
    "        )\n",
    "    print(len(node_traces))\n",
    "\n",
    "    # Combine all traces\n",
    "    data = edge_traces + node_traces\n",
    "\n",
    "    updatemenus = [\n",
    "        dict(\n",
    "            buttons=[\n",
    "                dict(\n",
    "                    args=[{'visible': [True] * len(data)}],\n",
    "                    label='Show All',\n",
    "                    method='update'\n",
    "                ),\n",
    "                dict(\n",
    "                    args=[{'visible': [False] * len(edge_traces) + [True] * len(node_traces)}],\n",
    "                    label='Hide All Edges',\n",
    "                    method='update'\n",
    "                )\n",
    "            ] + [\n",
    "                dict(\n",
    "                    # Set False for edge traces with None as name, or check group in valid names\n",
    "                    args=[{'visible': [(trace.name is not None and group in trace.name) for trace in edge_traces] + [True] * len(node_traces)}],\n",
    "                    label=f'Show only {group} edges',\n",
    "                    method='update'\n",
    "                ) for group in unique_groups\n",
    "            ],\n",
    "            direction='down',\n",
    "            pad={'r': 10, 't': 10},\n",
    "            showactive=True,\n",
    "            x=0.1,\n",
    "            xanchor='right',\n",
    "            y=1.1,\n",
    "            yanchor='top'\n",
    "        )\n",
    "    ]\n",
    "\n",
    "    # Create slider for edge opacity\n",
    "    steps = []\n",
    "    for step in np.arange(0, 1.1, 0.1):\n",
    "        step = round(step, 2)\n",
    "        steps.append(\n",
    "            dict(\n",
    "                method=\"update\",\n",
    "                args=[{\"opacity\": [step] * len(edge_traces) + [1] * len(node_traces)}],\n",
    "                label=str(step)\n",
    "            )\n",
    "        )\n",
    "\n",
    "    sliders = [dict(\n",
    "        active=5,\n",
    "        currentvalue={\"prefix\": \"Edge Opacity: \"},\n",
    "        pad={\"t\": 50},\n",
    "        steps=steps\n",
    "    )]\n",
    "\n",
    "    layout3d = go.Layout(\n",
    "        title=\"Sequence Homology between contigs across 49 long-read assemblies (3D)\",\n",
    "        scene=dict(\n",
    "            xaxis=dict(title=''),\n",
    "            yaxis=dict(title=''),\n",
    "            zaxis=dict(title=''),\n",
    "        ),\n",
    "        updatemenus=updatemenus,\n",
    "        margin=dict(r=0, l=0, b=0, t=100),\n",
    "        hovermode='closest',\n",
    "        legend=dict(\n",
    "            itemsizing='constant',\n",
    "            title_text='Plasmids',\n",
    "            bgcolor='rgba(255,255,255,0.5)',\n",
    "            bordercolor='rgba(0,0,0,0)',\n",
    "            borderwidth=2\n",
    "        ),\n",
    "        sliders=sliders,\n",
    "        annotations=[\n",
    "            dict(\n",
    "                showarrow=False,\n",
    "                text=\"MJF-2024\",\n",
    "                xref='paper',\n",
    "                yref='paper',\n",
    "                x=0,\n",
    "                y=0.1,\n",
    "                xanchor='left',\n",
    "                yanchor='bottom',\n",
    "                font=dict(\n",
    "                    size=8\n",
    "                )\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    fig = go.Figure(data=data, layout=layout3d)\n",
    "    fig.update_layout(\n",
    "        scene_aspectmode='data',\n",
    "        autosize=True,\n",
    "        uirevision=True\n",
    "    )\n",
    "\n",
    "    config = {\n",
    "        'responsive': True,\n",
    "        'scrollZoom': True,\n",
    "    }\n",
    "\n",
    "    pio.write_html(fig, file=output_file, full_html=False, include_plotlyjs='cdn', config=config)\n",
    "    print(\"html written!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_coordinates(G):\n",
    "    layout = G.layout('kk', dim=3)\n",
    "    Xn = [layout[k][0] for k in range(len(G.vs))]\n",
    "    Yn = [layout[k][1] for k in range(len(G.vs))]\n",
    "    Zn = [layout[k][2] for k in range(len(G.vs))]\n",
    "\n",
    "    Xe, Ye, Ze = [], [], []\n",
    "    for e in G.es:\n",
    "        Xe += [layout[e.source][0], layout[e.target][0], None]\n",
    "        Ye += [layout[e.source][1], layout[e.target][1], None]\n",
    "        Ze += [layout[e.source][2], layout[e.target][2], None]\n",
    "\n",
    "    return Xn, Yn, Zn, Xe, Ye, Ze\n",
    "\n",
    "def generate_meshes(Xn, Yn, Zn, groups, color_map):\n",
    "    mesh_traces = []\n",
    "    unique_groups = set(groups)\n",
    "\n",
    "    for group in unique_groups:\n",
    "        group_indices = [i for i, g in enumerate(groups) if g == group]\n",
    "        if len(group_indices) < 4:\n",
    "            continue\n",
    "        group_points = np.array(\n",
    "            [[Xn[i], Yn[i], Zn[i]] for i in group_indices]\n",
    "        )\n",
    "        vertices = hull.simplices\n",
    "\n",
    "        mesh_trace = go.Mesh3d(\n",
    "            x=group_points[:, 0],\n",
    "            y=group_points[:, 1],\n",
    "            z=group_points[:, 2],\n",
    "            i=vertices[:, 0],\n",
    "            j=vertices[:, 1],\n",
    "            k=vertices[:, 2],\n",
    "            color=color_map[group],\n",
    "            opacity=0.3,\n",
    "            name=f'Group {group}'\n",
    "        )\n",
    "        mesh_traces.append(mesh_trace)\n",
    "    return mesh_traces\n",
    "\n",
    "def make_3d_mesh_plot(G, labels, groups, output_file):\n",
    "    # Generate coordinates\n",
    "    Xn, Yn, Zn, Xe, Ye, Ze = generate_coordinates(G)\n",
    "\n",
    "    # Create color mapping\n",
    "    color_map = create_color_mapping(groups)\n",
    "    color_values = [color_map[group] for group in groups]\n",
    "\n",
    "    # Create edge trace\n",
    "    trace1 = go.Scatter3d(\n",
    "        x=Xe, y=Ye, z=Ze,\n",
    "        mode='lines',\n",
    "        line=dict(color='rgb(125,125,125)', width=0.5),\n",
    "        opacity=0.7,\n",
    "        hoverinfo='none'\n",
    "    )\n",
    "\n",
    "    # Create node trace\n",
    "    trace2 = go.Scatter3d(\n",
    "        x=Xn, y=Yn, z=Zn,\n",
    "        mode='markers',\n",
    "        name='contigs',\n",
    "        marker=dict(\n",
    "            symbol='circle',\n",
    "            size=5,\n",
    "            color=color_values,\n",
    "            colorscale='Viridis',\n",
    "            line=dict(color='rgb(50,50,50)', width=0.5),\n",
    "        ),\n",
    "        text=labels,\n",
    "        hoverinfo='text'\n",
    "    )\n",
    "\n",
    "    # Generate meshes\n",
    "    mesh_traces = generate_meshes(Xn, Yn, Zn, groups, color_map)\n",
    "\n",
    "    # Combine all traces\n",
    "    data = [trace1, trace2] + mesh_traces\n",
    "\n",
    "    # Create layout\n",
    "    layout3d = go.Layout(\n",
    "        title=\"Sequence Homology between contigs across 49 long-read assemblies (3D)\",\n",
    "        width=1900,\n",
    "        height=750,\n",
    "        showlegend=True,\n",
    "        scene=dict(\n",
    "            xaxis=dict(title=''),\n",
    "            yaxis=dict(title=''),\n",
    "            zaxis=dict(title=''),\n",
    "        ),\n",
    "        margin=dict(r=0, l=0, b=0, t=100),\n",
    "        hovermode='closest',\n",
    "        annotations=[\n",
    "            dict(\n",
    "                showarrow=False,\n",
    "                text=\"MJF-2024\",\n",
    "                xref='paper',\n",
    "                yref='paper',\n",
    "                x=0,\n",
    "                y=0.1,\n",
    "                xanchor='left',\n",
    "                yanchor='bottom',\n",
    "                font=dict(size=8)\n",
    "            )\n",
    "        ],\n",
    "    )\n",
    "\n",
    "    # Create figure and update layout\n",
    "    fig = go.Figure(data, layout=layout3d)\n",
    "    fig.update_layout(\n",
    "        scene_aspectmode='data',\n",
    "        autosize=True\n",
    "    )\n",
    "\n",
    "    # Set configuration\n",
    "    config = {\n",
    "        'responsive': True,\n",
    "        'scrollZoom': True,\n",
    "    }\n",
    "\n",
    "    # Write to HTML file\n",
    "    pio.write_html(fig, file=output_file, full_html=False, include_plotlyjs='cdn', config=config)\n",
    "    print(\"HTML written!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "saved edges to prot_v1/matrices/shared_gene_matrix_90percent_v1_edges.tsv\n"
     ]
    }
   ],
   "source": [
    "edge_df = edges_from_matrix(raw_matrix, edge_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Edge Matrix read from: prot_v1/matrices/shared_gene_matrix_90percent_v1_edges.tsv\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(edge_matrix):\n",
    "    edge_df = edges_from_matrix(raw_matrix, edge_matrix)\n",
    "else:\n",
    "    edge_df = pd.read_csv(edge_matrix)\n",
    "    print(f'Edge Matrix read from: {edge_matrix}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "G, vertex_order = create_igraph(edge_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "idmap = pd.read_csv(id_mapping, sep='\\t', header=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_bad_name(x):\n",
    "    if isinstance(x, str):\n",
    "        if '|' in x:\n",
    "            parts = x.split('|')\n",
    "            fix = parts[-1] if len(parts) == 3 else x\n",
    "            return fix\n",
    "        return x\n",
    "\n",
    "idmap['best_hit'] = idmap['best_hit'].apply(fix_bad_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample of idmap:\n",
      "       name    method        contig  contig_len wp_alignment_id  \\\n",
      "256  URI47H  longread  contig000001      921203      CP124080.1   \n",
      "257  URI47H  longread  contig000002       83849      CP094592.1   \n",
      "258  URI47H  longread  contig000003       53886      CP124079.1   \n",
      "259  URI47H  longread  contig000004       52987      CP002308.1   \n",
      "260  URI47H  longread  contig000005       30859      CP094604.1   \n",
      "\n",
      "            pf_alignment_id wp_plasmid_name pf_plasmid_name    best_hit  \n",
      "256  B379_chromosome_ParA_2      chromosome      chromosome  chromosome  \n",
      "257       JD1_lp28-3_ParA_1          lp28-5          lp28-3      lp28-3  \n",
      "258                B31_lp54            lp54            lp54        lp54  \n",
      "259              B31_lp28-4            lp25          lp28-4      lp28-4  \n",
      "260              JD1_cp32-8          cp32-8          cp32-8      cp32-8  \n",
      "\n",
      "Sample of vertex_order:\n",
      "['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "\n",
      "Total vertices: 1103\n",
      "Aligned labels: 1103\n",
      "Aligned groups: 1103\n",
      "Sample of aligned labels: ['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "Sample of aligned groups: ['cp26', 'lp28-2', 'lp38', 'lp36', 'cp26', 'lp28-7', 'cp26', 'lp28-1', 'cp32-6', 'lp17']\n",
      "Unique groups: {'lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-9', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17'}\n"
     ]
    }
   ],
   "source": [
    "aligned_labels, aligned_groups = align_labels_and_groups(idmap, vertex_order)\n",
    "label2group = dict(zip(aligned_labels, aligned_groups))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "def igraph_to_coords_file(G, aligned_labels, groups):\n",
    "    color_map = create_color_mapping(groups)\n",
    "    color_values = [color_map[group] for group in groups]\n",
    "    layout = G.layout('kk', dim=3)\n",
    "    Xn = [layout[k][0] for k in range(len(G.vs))]\n",
    "    Yn = [layout[k][1] for k in range(len(G.vs))]\n",
    "    Zn = [layout[k][2] for k in range(len(G.vs))]\n",
    "    nodes = {\n",
    "            i: {\n",
    "                \"location\": [\n",
    "                    Xn[i], Yn[i], Zn[i]\n",
    "                ],\n",
    "                \"name\": aligned_labels[i],\n",
    "                \"color\": color_values[i],\n",
    "                \"plasmid\": groups[i],\n",
    "            } for i in range(G.vcount())\n",
    "     } #\n",
    "    edges = [{\"source\": e.source, \"target\": e.target} for e in G.es]\n",
    "\n",
    "    return {\"nodes\": nodes, \"edges\": edges}\n",
    "\n",
    "def generate_cutoff_gene_graphs(dir, id2map, output):\n",
    "    if not os.path.exists(output):\n",
    "        os.mkdir(output)\n",
    "    files = glob.glob(f'{dir}/matrix/*.tsv')\n",
    "    for file in files:\n",
    "        curr_matrix = pathlib.Path(file).stem\n",
    "        print(curr_matrix)\n",
    "        edge_matrix = f'{dir}/edges/{curr_matrix}_edges.tsv'\n",
    "        if not os.path.exists(edge_matrix):\n",
    "            edge_df = edges_from_matrix(file, edge_matrix)\n",
    "        else:\n",
    "            edge_df = pd.read_csv(edge_matrix)\n",
    "            print(f'Edge Matrix read from: {edge_matrix}')\n",
    "        G, vertex_order = create_igraph(edge_df)\n",
    "        aligned_labels, aligned_groups = align_labels_and_groups(idmap, vertex_order)\n",
    "        label2group = dict(zip(aligned_labels, aligned_groups))\n",
    "        layout = 'kk3d'\n",
    "        output_file = f'{output}/{curr_matrix}_{layout}.html'\n",
    "        make_3d_plot_v2(G, aligned_labels, aligned_groups, output_file, layout)\n",
    "        # Convert the graph and save to JSON\n",
    "        print(\"plot generated\")\n",
    "        graph_data = igraph_to_coords_file(G, aligned_labels, aligned_groups)\n",
    "        with open(f'{dir}/json/{curr_matrix}_{layout}.json', 'w') as f:\n",
    "            json.dump(graph_data, f, indent=4)\n",
    "            print(\"JSON dumped!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shared_gene_matrix_50percent_v1\n",
      "Edge Matrix read from: prot_v1/matrices/edges/shared_gene_matrix_50percent_v1_edges.tsv\n",
      "Sample of idmap:\n",
      "       name    method        contig  contig_len wp_alignment_id  \\\n",
      "256  URI47H  longread  contig000001      921203      CP124080.1   \n",
      "257  URI47H  longread  contig000002       83849      CP094592.1   \n",
      "258  URI47H  longread  contig000003       53886      CP124079.1   \n",
      "259  URI47H  longread  contig000004       52987      CP002308.1   \n",
      "260  URI47H  longread  contig000005       30859      CP094604.1   \n",
      "\n",
      "            pf_alignment_id wp_plasmid_name pf_plasmid_name    best_hit  \n",
      "256  B379_chromosome_ParA_2      chromosome      chromosome  chromosome  \n",
      "257       JD1_lp28-3_ParA_1          lp28-5          lp28-3      lp28-3  \n",
      "258                B31_lp54            lp54            lp54        lp54  \n",
      "259              B31_lp28-4            lp25          lp28-4      lp28-4  \n",
      "260              JD1_cp32-8          cp32-8          cp32-8      cp32-8  \n",
      "\n",
      "Sample of vertex_order:\n",
      "['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "\n",
      "Total vertices: 1119\n",
      "Aligned labels: 1119\n",
      "Aligned groups: 1119\n",
      "Sample of aligned labels: ['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "Sample of aligned groups: ['cp26', 'lp28-2', 'lp38', 'lp36', 'cp26', 'lp28-7', 'cp26', 'lp28-1', 'cp32-6', 'lp17']\n",
      "Unique groups: {'lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-9', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-3+10', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17'}\n",
      "['lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-9', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-3+10', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17']\n",
      "html written!\n",
      "plot generated\n",
      "JSON dumped!\n",
      "shared_gene_matrix_100percent_v1\n",
      "Edge Matrix read from: prot_v1/matrices/edges/shared_gene_matrix_100percent_v1_edges.tsv\n",
      "Sample of idmap:\n",
      "       name    method        contig  contig_len wp_alignment_id  \\\n",
      "256  URI47H  longread  contig000001      921203      CP124080.1   \n",
      "257  URI47H  longread  contig000002       83849      CP094592.1   \n",
      "258  URI47H  longread  contig000003       53886      CP124079.1   \n",
      "259  URI47H  longread  contig000004       52987      CP002308.1   \n",
      "260  URI47H  longread  contig000005       30859      CP094604.1   \n",
      "\n",
      "            pf_alignment_id wp_plasmid_name pf_plasmid_name    best_hit  \n",
      "256  B379_chromosome_ParA_2      chromosome      chromosome  chromosome  \n",
      "257       JD1_lp28-3_ParA_1          lp28-5          lp28-3      lp28-3  \n",
      "258                B31_lp54            lp54            lp54        lp54  \n",
      "259              B31_lp28-4            lp25          lp28-4      lp28-4  \n",
      "260              JD1_cp32-8          cp32-8          cp32-8      cp32-8  \n",
      "\n",
      "Sample of vertex_order:\n",
      "['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "\n",
      "Total vertices: 1049\n",
      "Aligned labels: 1049\n",
      "Aligned groups: 1049\n",
      "Sample of aligned labels: ['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "Sample of aligned groups: ['cp26', 'lp28-2', 'lp38', 'lp36', 'cp26', 'lp28-7', 'cp26', 'lp28-1', 'cp32-6', 'lp17']\n",
      "Unique groups: {'lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17'}\n",
      "['lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17']\n",
      "html written!\n",
      "plot generated\n",
      "JSON dumped!\n",
      "shared_gene_matrix_90percent_v1\n",
      "Edge Matrix read from: prot_v1/matrices/edges/shared_gene_matrix_90percent_v1_edges.tsv\n",
      "Sample of idmap:\n",
      "       name    method        contig  contig_len wp_alignment_id  \\\n",
      "256  URI47H  longread  contig000001      921203      CP124080.1   \n",
      "257  URI47H  longread  contig000002       83849      CP094592.1   \n",
      "258  URI47H  longread  contig000003       53886      CP124079.1   \n",
      "259  URI47H  longread  contig000004       52987      CP002308.1   \n",
      "260  URI47H  longread  contig000005       30859      CP094604.1   \n",
      "\n",
      "            pf_alignment_id wp_plasmid_name pf_plasmid_name    best_hit  \n",
      "256  B379_chromosome_ParA_2      chromosome      chromosome  chromosome  \n",
      "257       JD1_lp28-3_ParA_1          lp28-5          lp28-3      lp28-3  \n",
      "258                B31_lp54            lp54            lp54        lp54  \n",
      "259              B31_lp28-4            lp25          lp28-4      lp28-4  \n",
      "260              JD1_cp32-8          cp32-8          cp32-8      cp32-8  \n",
      "\n",
      "Sample of vertex_order:\n",
      "['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "\n",
      "Total vertices: 1103\n",
      "Aligned labels: 1103\n",
      "Aligned groups: 1103\n",
      "Sample of aligned labels: ['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "Sample of aligned groups: ['cp26', 'lp28-2', 'lp38', 'lp36', 'cp26', 'lp28-7', 'cp26', 'lp28-1', 'cp32-6', 'lp17']\n",
      "Unique groups: {'lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-9', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17'}\n",
      "['lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-9', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17']\n",
      "html written!\n",
      "plot generated\n",
      "JSON dumped!\n",
      "shared_gene_matrix_60percent_v1\n",
      "Edge Matrix read from: prot_v1/matrices/edges/shared_gene_matrix_60percent_v1_edges.tsv\n",
      "Sample of idmap:\n",
      "       name    method        contig  contig_len wp_alignment_id  \\\n",
      "256  URI47H  longread  contig000001      921203      CP124080.1   \n",
      "257  URI47H  longread  contig000002       83849      CP094592.1   \n",
      "258  URI47H  longread  contig000003       53886      CP124079.1   \n",
      "259  URI47H  longread  contig000004       52987      CP002308.1   \n",
      "260  URI47H  longread  contig000005       30859      CP094604.1   \n",
      "\n",
      "            pf_alignment_id wp_plasmid_name pf_plasmid_name    best_hit  \n",
      "256  B379_chromosome_ParA_2      chromosome      chromosome  chromosome  \n",
      "257       JD1_lp28-3_ParA_1          lp28-5          lp28-3      lp28-3  \n",
      "258                B31_lp54            lp54            lp54        lp54  \n",
      "259              B31_lp28-4            lp25          lp28-4      lp28-4  \n",
      "260              JD1_cp32-8          cp32-8          cp32-8      cp32-8  \n",
      "\n",
      "Sample of vertex_order:\n",
      "['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "\n",
      "Total vertices: 1117\n",
      "Aligned labels: 1117\n",
      "Aligned groups: 1117\n",
      "Sample of aligned labels: ['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "Sample of aligned groups: ['cp26', 'lp28-2', 'lp38', 'lp36', 'cp26', 'lp28-7', 'cp26', 'lp28-1', 'cp32-6', 'lp17']\n",
      "Unique groups: {'lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-9', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-3+10', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17'}\n",
      "['lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-9', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-3+10', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17']\n",
      "html written!\n",
      "plot generated\n",
      "JSON dumped!\n",
      "shared_gene_matrix_70percent_v1\n",
      "saved edges to prot_v1/matrices/edges/shared_gene_matrix_70percent_v1_edges.tsv\n",
      "Sample of idmap:\n",
      "       name    method        contig  contig_len wp_alignment_id  \\\n",
      "256  URI47H  longread  contig000001      921203      CP124080.1   \n",
      "257  URI47H  longread  contig000002       83849      CP094592.1   \n",
      "258  URI47H  longread  contig000003       53886      CP124079.1   \n",
      "259  URI47H  longread  contig000004       52987      CP002308.1   \n",
      "260  URI47H  longread  contig000005       30859      CP094604.1   \n",
      "\n",
      "            pf_alignment_id wp_plasmid_name pf_plasmid_name    best_hit  \n",
      "256  B379_chromosome_ParA_2      chromosome      chromosome  chromosome  \n",
      "257       JD1_lp28-3_ParA_1          lp28-5          lp28-3      lp28-3  \n",
      "258                B31_lp54            lp54            lp54        lp54  \n",
      "259              B31_lp28-4            lp25          lp28-4      lp28-4  \n",
      "260              JD1_cp32-8          cp32-8          cp32-8      cp32-8  \n",
      "\n",
      "Sample of vertex_order:\n",
      "['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "\n",
      "Total vertices: 1117\n",
      "Aligned labels: 1117\n",
      "Aligned groups: 1117\n",
      "Sample of aligned labels: ['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "Sample of aligned groups: ['cp26', 'lp28-2', 'lp38', 'lp36', 'cp26', 'lp28-7', 'cp26', 'lp28-1', 'cp32-6', 'lp17']\n",
      "Unique groups: {'lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-9', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-3+10', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17'}\n",
      "['lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-9', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-3+10', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17']\n",
      "html written!\n",
      "plot generated\n",
      "JSON dumped!\n",
      "shared_gene_matrix_80percent_v1\n",
      "saved edges to prot_v1/matrices/edges/shared_gene_matrix_80percent_v1_edges.tsv\n",
      "Sample of idmap:\n",
      "       name    method        contig  contig_len wp_alignment_id  \\\n",
      "256  URI47H  longread  contig000001      921203      CP124080.1   \n",
      "257  URI47H  longread  contig000002       83849      CP094592.1   \n",
      "258  URI47H  longread  contig000003       53886      CP124079.1   \n",
      "259  URI47H  longread  contig000004       52987      CP002308.1   \n",
      "260  URI47H  longread  contig000005       30859      CP094604.1   \n",
      "\n",
      "            pf_alignment_id wp_plasmid_name pf_plasmid_name    best_hit  \n",
      "256  B379_chromosome_ParA_2      chromosome      chromosome  chromosome  \n",
      "257       JD1_lp28-3_ParA_1          lp28-5          lp28-3      lp28-3  \n",
      "258                B31_lp54            lp54            lp54        lp54  \n",
      "259              B31_lp28-4            lp25          lp28-4      lp28-4  \n",
      "260              JD1_cp32-8          cp32-8          cp32-8      cp32-8  \n",
      "\n",
      "Sample of vertex_order:\n",
      "['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "\n",
      "Total vertices: 1114\n",
      "Aligned labels: 1114\n",
      "Aligned groups: 1114\n",
      "Sample of aligned labels: ['UNY149P_B247_cp26', 'UCT109H_contig000002', 'UCT50H_contig000004', 'URI87H_contig000004', 'URI93H_contig000016', 'URI48H_contig000012', 'ESI26H_contig000009', 'UNY208P_MR641_lp28-1', 'URI101H_contig000014', 'URI111H_contig000019']\n",
      "Sample of aligned groups: ['cp26', 'lp28-2', 'lp38', 'lp36', 'cp26', 'lp28-7', 'cp26', 'lp28-1', 'cp32-6', 'lp17']\n",
      "Unique groups: {'lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-9', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-3+10', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17'}\n",
      "['lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-9', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-3+10', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17']\n",
      "html written!\n",
      "plot generated\n",
      "JSON dumped!\n"
     ]
    }
   ],
   "source": [
    "\n",
    "generate_cutoff_gene_graphs('prot_v1/matrices', idmap, 'prot_v1/3d_graphs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    layout = 'kk3d'\n",
    "    output_file = f'{curr_matrix}_{layout}.html'\n",
    "    make_3d_plot_v2(G, aligned_labels, aligned_groups, output_file, layout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['lp36', 'cp32-1', 'cp32-2', 'lp28-1', 'cp32-11', 'chromosome', 'cp32-9-4', 'cp9-3', 'lp21', 'cp32-8', 'cp32-7', 'cp32-13', 'lp28-7', 'lp25', 'lp28-11', 'lp56', 'lp28-2', 'lp28-9', 'lp28-3', 'cp32-10', 'lp5', 'cp32-12', 'lp28-8', 'cp32-4', 'lp28-6', 'cp32-3', 'lp21-cp9', None, 'lp28-5', 'cp32-5', 'cp26', 'lp38', 'lp54', 'cp32-6', 'cp32-9', 'lp28-4', 'cp9', 'lp17']\n",
      "html written!\n"
     ]
    }
   ],
   "source": [
    "#layouts = [\n",
    "#    'circle', 'dh', 'drl', 'fr', 'fr3d', 'graphopt',\n",
    "#    'grid', 'kk', 'kk3d', 'large', 'mds', 'random',\n",
    "#    'random_3d', 'rt', 'rt_circular', 'circular_3d',\n",
    "#]\n",
    "#layout = 'fr3d'\n",
    "layout = 'kk3d'\n",
    "output_file = f'igraph_asm_ava_gene_homology_90_v1_{layout}.html'\n",
    "make_3d_plot_v2(G, aligned_labels, aligned_groups, output_file, layout)\n",
    "#make_3d_plot_v3(G, aligned_labels, aligned_groups, output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[9.403121174629154, -9.130347882842578, 0.2505539477838669, 1.812445970962161, -6.633880406381925, 0.23333900348159398, 4.996536318601759, 0.28644564888471963, 10.509958796401806, 6.903121041251271, -2.2922755887779283, 5.973553919530981, 1.9954178087434382, 7.81889986357552, 11.432936709793843, 5.230509491858685, 9.81345468218701, -8.019293817015631, -7.618072590656357, 8.861674567662176, -7.339671730038228, -13.580560364492944, 5.2381594950757515, 8.602203097414922, -13.533463781615358, -10.807775090861982, 4.8655151375410135, 4.912752500412865, 4.614462944568486, -3.422202729398015, -6.992479009266176, -4.670364155761025, 5.091006383388406, 9.05071467386509, -14.152021207952197, 5.154697847839958, 5.218093829697477, 4.836744279895769, 5.450601755358445, -3.8567865361722418, -9.362103091575895, -9.910467348562152, -5.722229063736414, 7.740778935344506, -1.0102386155507757, 6.814961824733492, -7.931570100324443, 5.712375513571463, 2.875673370653343, -8.811921320400417, -13.175109177236024, -4.705835892963397, -7.573381217884844, -8.948919865303859, 4.966614245051989, 6.540565243515696, 6.025949115437518, -5.6610448407229015, 8.84808640219667, -10.15527145801714, 7.9066043779060795, 9.092297609252446, 1.9413333253643499, -4.167707979466067, 3.5907566016358374, 9.27074525351589, -13.58553634946806, -5.414534228372785, 5.973170943050884, -5.053525848151566, -11.388344949197391, 1.8123963522433955, -7.914146301262957, -6.8431785277618165, -6.4667417216496865, -6.137755506712285, 5.864860140368514, -18.541175346009535, 5.3500199079853825, 7.806192532298056, 5.163083081405128, 4.424895769635986, -16.871309553232145, 3.7905237469181974, 6.913251943128996, 8.643460325820241, -1.1274758310029358, 5.3567569413844245, 9.141382875477047, -3.585328008690718, -4.197670560935655, -7.797079036157223, 8.075405009717539, -6.389210400968478, -5.795346763036371, 8.103580931140211, 2.4003113613199836, 0.2762104348574422, -10.135136648983076, -2.2186225719813883, -3.953510724245775, 6.497943540055366, -13.76012253951721, -13.435192098106464, 4.750020379323772, -9.916366992147335, 5.81116786084966, 2.370753008075068, -16.6577117918658, -9.993990394486254, 6.700969308816771, -4.022463948517561, -6.968699045436069, 1.803166890622465, 7.886950684679845, -7.604377550183822, 13.031125098644583, -12.48286742911507, -14.178420530095662, -6.907860067013194, -7.423516515291297, -4.596258891311374, -8.954041252751606, -6.8505589762837245, 9.736699973133378, -4.143211736304761, -7.261498054725648, 9.042848332466997, 4.053348732743224, -6.118687677433504, -10.060409467820232, -6.21470178855318, -7.013482107220371, 5.3960731470809735, 8.213107922064244, 7.270351962427854, -10.298440594894174, 9.139432284390315, -0.21250426157651317, -0.7130727891884454, -8.948049589690715, 4.781024170549654, 3.0633058954362973, -2.7525473387162744, -10.23082359920368, 4.1452881671998325, 1.861279033110157, -13.28142596890379, -4.019541797629222, 11.496942607366183, -8.242186458248112, 5.461882866030441, -8.11061683423261, -9.421022564924922, -9.902576408680869, -0.029052112481259086, -0.5963016531775608, 6.711639281591296, -5.988849828144025, 7.9090744374097435, 5.286423522101583, -7.4264980786224, 4.12324344407611, 9.907318842156789, 8.815835157458986, -4.614266411387692, -2.5261778735828906, -12.270454861585717, -0.8660927584318708, 3.302477247547746, -7.012051228241194, 5.401908687244887, 0.7905796802063058, 4.944341989947421, 2.759562506611725, -4.745707657586787, 2.313803395979845, -4.7439163049428235, -0.9244818214909485, -8.459114813085568, -8.883950554942931, 5.234545471835972, -7.623758883062372, -9.075231269467231, 3.2230824325799623, 2.6476935593914708, 2.599454877108239, 4.218124858202014, 8.447575871384963, 7.865806389831381, -7.351186437932386, -7.6735714736535625, -4.7777563437461765, -12.490347329104114, -5.618307182426931, -4.291459234988076, -6.353596348380592, 5.754668242294236, 10.053910009151389, 0.9483094780920207, -0.4386561309244013, -10.47716794108519, 2.2451082106840583, -13.04065203146724, 8.824380658095233, -4.085945587934781, -13.916500856040349, 2.31571500406815, 6.306989210368611, 5.462635920070409, 2.7598949143237146, 3.8909291447226257, 1.6425037876202773, -12.887270418947674, 6.814240844895231, -6.85632931933789, -3.5526256624519204, -4.4934662416409505, 6.37749442710203, -16.391081955517798, 2.799013681787993, -4.136410723024528, 1.6236622997092776, -5.0003205164301106, 5.50823741202029, -3.0029611768424433, 5.136071705083549, 8.567536675619815, 5.497146026721464, -12.039575898891762, -10.290707644452185, 6.739789071479805, -0.7085249001058002, -8.139484073111682, -9.547659606531576, -0.5142855322084281, 6.249458291753491, -14.00814687978471, 6.616819428056508, 2.2230697096796366, -10.645550118683166, 4.752843252650337, 0.07170426705522859, 5.701921833963117, -10.536916838371642, -6.740855546879985, 2.43053324331386, -0.1513968397921242, 0.7185267195790345, 5.700789331440418, 6.937811104101717, 4.54488134025838, 2.9803734423006847, -8.011357746074347, -9.782689147005259, 4.785827209335437, -8.66162868840771, -13.008968154606658, -10.21963144439075, -7.431134055841181, -2.015906948326309, 10.776744062504939, 7.989783895196206, -8.558491114956128, -10.233022463689931, 9.078282720104704, -8.114623640558108, -4.977932558521064, -3.8975060045651015, -6.362784633269688, 2.4742530587976717, -4.847160890342283, -8.227452125549691, 0.82170972977429, -4.212695315706195, -18.513511203950316, -5.209210611878971, 9.730070170416, 3.5485553612585212, -9.81686478640285, 8.65447583348745, 8.163642203500396, -1.1767695346953206, 1.7201383429175616, -5.178026597443036, -4.496137251745675, 3.0111028589412507, -4.172660246390564, 5.828461984652952, -9.846474247745038, 4.3531593866891525, -9.307727573553706, 2.8895840284739083, -6.184932605386703, 7.532946570412457, -1.0712750134704474, -5.114942675304981, 4.961128226338203, -5.978857051732581, -9.497345759339378, 4.842959317455359, 8.12313983186403, 5.378099853637946, 5.430711381966029, 2.833790639148149, 0.14862858192391898, 8.555266963501683, 7.075700398783017, -4.344983930669751, 5.902835829464433, 8.421187888579073, 3.596931670822314, 0.10397327163816382, 3.32402640463172, 5.799973668979735, -3.347177978544831, 5.337781911039476, 6.239248687844741, -8.466220427042067, 6.313055762471194, -10.38032452345719, -8.002954360609408, 2.95852777599993, -13.682895503397154, 9.276515495533285, 6.596283645747461, -8.530648391869967, -13.107630886268778, 0.41346608132413254, 2.149942462718882, -8.035261676296825, 6.364149187862007, 11.951922349371944, 5.0895448650460935, -6.211069526475026, -1.4438688026438329, -8.302468018592618, -3.3496811463914367, 8.02533495347074, 3.91003809346072, -6.320880360423986, 6.047917662759215, 1.7862691537010411, 8.002814690601335, -10.08688255156214, -5.952128113300439, 5.819943307745815, -11.87162747764147, -13.259958682416256, -8.738544112771068, 3.15749208185675, 5.602569464303677, 4.947052544193601, -4.928453677895648, 9.262785800403906, 5.905216544535319, -12.215818109862406, -3.7374791114847254, -4.1456685180409085, -13.303638125136121, -5.163840340768765, 2.3290410590896853, -5.244023168676702, 2.6889121811343815, 3.6240723468728246, -8.53589601177165, -2.0305675551144855, -6.149040774654012, -6.9137382494160935, -3.1570148735628094, 1.9444744100935103, 8.875417682547658, 7.867551900880713, 6.033341925424162, -5.981642459519501, 8.415600305276847, 6.934488748322425, -10.250097684386214, 5.324200861898766, -7.920137880992263, 6.870715939893567, -10.488551121804537, -9.961534235299107, 7.615810060860509, 4.077542792235041, 2.4102395035058026, -10.026066384625738, -10.803048656320374, 7.082484882763196, 8.694890788684631, -7.967278450205117, -10.318101929569776, -8.230021650761154, -10.585163499040066, 4.096777174496778, 8.301006202135627, 3.70186253537682, -8.179893563025649, 5.356613354522007, -6.851910692899226, -3.925332748015217, -9.140205371334757, 4.7167536584946985, 0.6087844051414599, -11.025597100491066, 7.8063850942052655, -10.44535129444251, -18.083075106848668, -6.770870443415714, -11.425832107519987, 1.9001281600802242, 3.0270838022143227, 5.198755658292658, -6.199706416034605, -0.012548252799608271, 6.274087888811637, -4.279154355807304, 2.200992978932108, -5.103663222170083, 4.09347127835361, 5.38828835508704, -0.622576501768253, 3.274971329402589, 6.564260639477753, 6.146429067488785, -4.69592297264736, 1.4813881765705337, -11.76258145129706, 1.6770817242186888, -4.6083207791992375, 5.878759325742889, 2.4906488068585486, -6.018994203597462, -5.933826348935825, -7.271905296542503, 11.939279927736656, -7.52969533613081, -12.821933436762885, -8.731453545501338, -4.412987532655369, -0.9721561213489267, -4.187800028959598, 6.138588654453139, 1.8335404337027257, 3.07408879747317, -0.08534205783822923, -4.739250624157661, -20.033179195255595, 4.672645801762361, -4.704361508467907, 5.381028458976075, 8.008876306691683, 4.530189866727814, -0.9182689660919177, 3.4597450535907956, 10.2376287914565, 6.009345143611351, -7.14008865342244, -5.392533864116176, 8.540231708491826, 9.006602710847263, 2.851802782473748, -10.36707021002246, 4.6003222428871915, 4.9892738612228005, 6.604351276739788, -1.1229503760405009, 6.771980068814475, 9.308750285975883, 3.6251719195710725, 4.014336852576619, 5.045345292644659, -10.027993865822586, 3.1789821223465933, 8.048866286178923, 12.020464619510685, 4.930925916682298, 3.763504030244591, -12.50079669335506, -7.606346906560821, 8.723099485741981, 9.905331026170252, -0.8774029158912899, -6.092888771739461, -5.0881322925777726, 2.8288274687731993, -18.69203488958245, -5.351886941657914, 9.083886727124986, -3.0021812802932617, 8.636960024334554, 5.0233814060555675, -5.072539242913051, -3.6794005503190417, 2.912079820150947, 6.689196325580304, -8.413143533653312, -4.586907584683707, -9.188867744958795, -6.180194383165109, 0.816355728059156, 2.871396931458435, -9.631079609630984, 8.311849134699152, 10.274548574448817, 8.367548536835539, -9.196939231457547, 6.558464533188441, -4.841565246800582, 3.767718872979109, 2.666086659475161, -6.033099593487646, -9.895385040477132, -5.314841516188385, -7.695679571776854, 5.000816785690068, -5.168497706908116, -5.016554216252751, 4.109069649277662, -5.659102966817399, -2.2332045028118865, -0.7139368141754061, -0.9153734495613042, -0.7674561550081527, 6.7786164462756995, -6.4069490698415095, -7.20412291274606, -6.043993263576845, 2.7006926053035567, -3.5316727256140044, 4.110226852563708, -12.440751107292225, 2.5626551090316947, -6.0132809431687315, -4.241201203723025, 5.281587685917203, -5.603346136304934, -5.988242699808023, -7.72124789646892, -4.45056210526838, 2.2472620212409575, -7.8579259262755965, -7.826806717380602, -5.0759708510018084, 7.041478281467915, 1.7949938334012703, 9.05528692838183, 7.882660932854643, 2.4603196225137425, -10.30145153894888, -6.933056124347621, 6.31650768952519, 9.52826500631863, 4.349315527488894, 5.556608551122717, -5.00251835868601, 6.045441349408944, -2.858973556309468, 5.833462187856911, 0.9790078608739236, -8.285747889023316, 6.551538320019157, -8.639521356696326, 1.9740898426712583, -4.6517495239746225, -9.440680700129766, -6.068813718715749, -8.773211738696942, -0.803876349524024, -13.742329213612274, 1.0436542856889939, 7.38516976790378, -8.837264415312966, -5.844938002170032, 1.8342206833278898, -9.88985576689892, 3.2188550731675334, -4.9326665877749925, 7.032930289731938, 5.3095648140097795, -5.489954695192298, 5.084979855971196, 5.272264072348766, -9.052594860085906, 7.727727927363788, 3.4627775158854446, 7.03845157299264, -13.452553322447283, 1.116819589521081, 7.53516647554033, -6.196857040298365, 4.555450444537047, -9.631495866403942, 5.605783608306989, -6.892213443264632, 0.2172964044707717, -11.557736251288244, -9.182697924821266, -10.49278199291569, -4.3662408386102, -10.226047713711754, 5.247735238631159, 5.770079414428423, -11.31093721309317, 4.153626218697044, -9.218175323980704, 2.6855819413013022, -7.66872424292062, 8.577138932925047, -0.41201940807137355, 8.650053626805464, -6.359756416732983, -18.986699581833303, 5.388892050425771, 2.472042851913595, -4.474187672126792, 2.1214061872894714, 8.375127422902176, -8.123823679038182, 4.126149524859957, 1.1816314126526477, 0.14723407460206223, -12.156169111500484, -6.623228603439564, -8.921480131798454, -5.6689549465197135, -4.030777341286072, -12.870702432108393, 3.100658442720261, 5.710872953471497, -12.302118273688848, -12.14195508463908, 4.869610936271507, -4.013554998422081, 8.557762886498272, 6.6606255622534745, -12.803342538777425, -8.183599937037187, 3.648530904105029, -7.4558382970797865, 4.709733708959725, 4.93592339010906, 1.7896820117291303, 2.0724421042594936, 3.5528810990438044, 2.9858003018307233, -4.792558931896694, 2.69311982510871, -4.804295222538167, -5.9841913289027024, 7.286347719363768, 1.9589245497525896, -12.814711264074461, 3.0246322770713054, -6.52216981912749, 1.6052691121500302, -11.286569616157582, -7.612209265439125, 8.146495352424047, 7.42217915828718, -6.411153509455011, 7.885540572323007, 2.8211269237240257, -3.0757426835044486, -3.7984680934995105, -18.10025627664542, -9.584973024046533, -7.432502472601409, -0.2392291261387656, 4.39276907778471, -3.3222004247738894, -7.124156802191909, -14.768433034237402, 7.848968603575083, 8.215115617703551, 7.00432449394303, -6.3798420603164745, -8.43348085543174, 6.3848722293677085, 2.6751269159202655, -8.384806050750115, 9.38842454578495, 9.326702354103919, -9.445595524733568, -3.9485335639418757, 4.996720268723983, -4.102973232311517, 9.403359067118693, -9.885572878402076, -4.3427090205967165, -3.5998865803358613, 4.968649792563365, -7.455294973744944, 1.2115246012487813, 2.94832493435339, -6.240445537666829, 10.039542941302207, 0.926588527929339, -10.309213724127881, -3.2589838904005157, -7.408838192311318, 2.999583046114546, 7.863211721379181, -0.5078287821190834, -7.489344714580361, -8.366446139433037, -9.810050714435643, 6.2535194705922015, -7.723547238061654, 2.8879434859222006, -14.533378009603993, 5.123181331326333, 5.869254773221662, -12.1024862835668, -6.163164856823132, -8.4914949845721, 4.417408743048177, -3.327703885840532, -10.31989278745534, -2.5729109574881317, 3.0565237002519527, 9.23355520361247, 1.8490126875776967, -10.013193292578137, -5.68240543819612, 7.620702291898667, -7.834495605872492, 10.893027643278954, 6.642969845088353, -3.9448439518533642, 5.942250907661794, 6.797091640020763, -11.834835536195715, 9.082214697272759, -4.090231240937953, -8.257146133166039, 8.99470448146637, 9.05006568924392, -10.14927802195782, 3.019075287998075, 0.5244568930645626, 1.5272861862748373, -4.309894546892006, -10.309239647515128, 8.894629013473228, -14.03974599772417, 5.886197492749165, 6.169228414572947, 1.0104221040665178, -8.157369131336067, 0.7035382330723052, -18.331376837020887, -12.921736820759163, 5.868893038414036, -5.064978516617217, -7.765980458864354, 5.610080399427072, -9.241841947135631, -1.1597913758865102, 5.851541126912133, 0.7409310732564529, -5.578382504339301, 2.78590895460421, -14.426524518003397, 7.679679139361843, -5.640656401704985, -2.3680982258653196, 3.4223928344260646, 2.574152169983988, -11.546189251158562, -9.759255383095429, -10.481421690791963, -10.544180571050974, -7.242406301246436, 5.822971272929417, -10.531074535979664, -3.786011509687567, 8.735471635634116, -8.191507348194756, -8.816834217579446, 7.84938701027794, 2.71413012603004, 4.644202456844496, 4.80172605991596, 8.732790159532167, 10.431116744805168, -10.337533432569364, 9.259716333141533, 6.470961919092244, 1.6368828811776583, -6.789877044238385, -6.420765836303733, -10.842050895825277, 2.4797150333662596, -5.900188691315221, -11.600459815149344, -3.38089790284285, -9.59969545594708, 7.132069986435519, 7.924532108150099, 1.7976121386300663, -11.2384181104938, 3.3908878594753205, 5.809569759846686, 5.061502802634806, 2.046922638335356, 7.200093170782798, -11.050518333280246, 3.710270748523238, -10.877845378601709, -6.743749714750075, -8.058554120156055, -13.102596065044242, -3.757536225261645, 1.0580300959213955, 9.257728839298938, 2.678046024131689, -9.361570962442924, 3.8211929413839747, -9.327740525351274, -6.693085612629587, -5.383205970883062, -6.400709009814358, 0.06410867402444775, -9.538738686979054, 9.254116706032168, -8.034589129883292, 0.09744679489238028, -10.041053598132512, -3.879470808576216, 0.44336418169358854, -4.1281165481407385, -9.319188500532938, 2.7601467969126925, 1.6595493388546887, -7.35955687151967, 7.644662780408685, -8.905613770534659, -10.340923201299102, -19.07885244359701, 7.500065091682751, -8.529777038245275, -8.098463370862602, 6.306344835320269, -5.660094282929045, 3.1614714572101836, -6.986845668233866, -9.151981394702654, 1.5734176075018076, -10.033492703827088, 10.507298527445954, 5.112548921069584, -10.40659975549925, -11.729797716315185, 5.425253576727625, 9.222246145936818, 6.343642435435062, 7.923985292420126, -5.489451055017915, -4.267296081997489, -5.665281937286335, 3.0674341127588516, 0.5657955699196348, -11.427694468173971, 2.104039918683479, -7.447918681453641, -19.042039781977653, 12.67756535868795, 7.960006439525772, 7.672194318405161, 5.654703131085235, -9.08826833877149, -3.977051677922233, -3.081525909239379, -10.512277123436483, 4.990179735680413, -11.352824567290982, 1.5483979908529986, 3.4466528340943716, -7.734444092121994, -0.9447601745489772, 7.861699941912174, -14.225006854108257, 7.270859544620738, -9.147323308188406, -5.148770237424784, -5.134148294834367, 3.2878997936774965, 4.0313244559278045, -7.123926907766098, -9.568846577384686, -5.71338014711878, 9.290976102494238, 7.414960121124791, 2.8438615667180023, -13.63096807222618, 1.8388626818939935, 6.240311289919429, 4.035545328593296, 5.296703159243147, 6.292765892467494, 6.142851623255508, 6.547418528736562, 9.38527881266383, -10.249229559397435, -8.969009106311857, -7.516104264734494, 8.572212792004647, -10.571713529213119, -7.199525382091741, 4.990896598970267, -8.998727994622117, 1.5703062173972724, -3.7060386198197914, -8.39827744038772, 2.4161825729497814, -6.964597775253351, 6.414314314204524, 6.766682974247969, -0.6508454660025353, -12.600302826178918, 5.820194772221227, -13.699590640396975, 6.2971523070027144, -1.35647514864797, -3.1235353997069137, 8.074619306520729, 4.9099750745554935, -2.2698200872802947, -11.782921297510821, -4.326943193807088, -3.33459081704749, 2.9491021126017367, 7.840555373434492, 5.02193755701041, 9.028934967386425, 6.681334582845788, 5.582934149137138, -5.920110615661303, -3.9128487741522324, 6.814254207045295, 8.225584982631199, 4.715164568993835, 9.67984827867603, -0.7922505958862975, 12.405688617779525, -14.672538367132288, 3.601888945908506, -9.42570201871053, -5.129431758719764, -1.9175477839646518, -7.636237256055812, -3.8182091158466105, 6.202633985717747, -3.870943955828492, -5.316711999198217, 10.49400257803457, -5.777701581530724, 6.805363112596311, 6.271187497421512, -2.452550958589289, -4.677564722466083, -0.15920021179935342, 1.3410205713072525, -9.157658622147013, -9.625595941154069, 9.859036696307067, 9.021119912689857, 9.623043276950108, 4.569356971130892, 2.16824892475096, 4.076495775878547, -10.375494712975538, 3.83859219805967, 4.886954289423275, 2.460729244558435, 3.096185576550386, -3.6153354026439612, -4.894571456464403, -12.509311316407402, -11.39725196566152, -8.906189351461382, -10.627881889306181, 0.450193314434643, -6.166883563249055, -8.544166368707023, -10.539971671149527, -10.49691070096361, -12.439266591916068, -4.762003509005812, -3.3989351094033626, 6.151064048100813, -3.875513725765364, 4.889977138455987, -9.93937320269432, 1.1229532231072574, -1.6619295823586586, -9.55600468248547, -11.684939414999208, -5.240186003150207, -11.064012673951565, 0.6851959212407135, -9.698095594074928, 5.669847765773439, 5.933965338025662, 5.997517233804027, -7.8952974626913655, -12.90081101228924, -5.6661785646579705, 5.681417746961404, 4.691404537113327, 6.211638065865591, 7.87354699520942, -10.02678147856641, -12.61340869705402, -2.4623974893611402, 6.833710278341965, 6.650049233309715, 5.081686188974539, 2.4877697655855293, 8.915008601656746, 2.066388886825687, 5.454621703641722, -10.45515405447934, 2.800005580623349, -7.623324039727259, -1.9635522155947147, 6.875409529487117, -3.65744623846986, 1.3048991501081024, -9.22012471466019, -4.064071970633755, 8.979003293936263, 6.323971791887019, 1.826073036819639, 9.046428517606644, 2.92099013548797, 8.899931685052362, 2.017723370771403, 4.1342584554566, -17.537557878060355, 6.964277297761882, -7.362680974733322, 5.55952185162006, -5.057162005987118, 3.733253327050629, -9.58832489141505, -5.511700219994239, -10.16056820552306, -7.22469921577004, 5.271499779862919, -14.510380421804724, 6.914263117971594, -9.839795275104741, 0.6559833505425271, 2.6442232583929277, -0.17848001032422953, 8.051245404932416, 0.6587329608531987, -8.036746450134697, -7.907520619852667, -8.078801598092873, -19.54529332317338, 8.953773705151551, 4.887085093169498, 6.343269936554012, -5.459164541233027, 6.4952328940170325, -4.9002238778846126, 5.611680356914627, -7.054325279425319, -9.749968699616446, 7.534222046183344, -3.7370244913387483, -9.554416362433704, -4.548835575493337, -4.008125487061455, -10.697055565784233, -7.22396048212922, -0.0917228488599717, -5.092393911192391, -5.036802100370543, 8.813447484417974, 5.002071552718432, -10.10801015383268, -0.8689918366775413, 8.073737538960476, 5.412379343056266, 8.26693332172411, -9.80804626164608, 3.079615345038937, 9.60148913754303, 5.179126792668553, -8.670870625245595, -2.4676152229369253, 2.0075646254361574, -8.5776794373621, -9.485942694807862, 7.4771268551225925, -5.1605111506226296, 1.4223212173889879]\n"
     ]
    }
   ],
   "source": [
    "layout = G.layout('kk', dim=3)\n",
    "Xn = [layout[k][0] for k in range(len(G.vs))]\n",
    "Yn = [layout[k][1] for k in range(len(G.vs))]\n",
    "Zn = [layout[k][2] for k in range(len(G.vs))]\n",
    "print(Xn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prune_and_subsample_graph(graph, weight_cutoff, sample_size):\n",
    "    # Create a new graph to hold the pruned version\n",
    "    pruned_graph = graph.copy()\n",
    "\n",
    "    # Get the edge weights\n",
    "    weights = pruned_graph.es['weight']\n",
    "\n",
    "    # Create a list of edges to delete (those below the cutoff)\n",
    "    edges_to_delete = [edge.index for edge, weight in zip(pruned_graph.es, weights) if weight < weight_cutoff]\n",
    "\n",
    "    # Delete the edges\n",
    "    pruned_graph.delete_edges(edges_to_delete)\n",
    "\n",
    "    # Remove any isolated vertices (nodes with no edges)\n",
    "    pruned_graph.vs.select(_degree=0).delete()\n",
    "\n",
    "    # Randomly subsample nodes\n",
    "    total_nodes = pruned_graph.vcount()\n",
    "    if sample_size < total_nodes:\n",
    "        nodes_to_keep = random.sample(range(total_nodes), sample_size)\n",
    "        pruned_graph = pruned_graph.subgraph(nodes_to_keep)\n",
    "\n",
    "    return pruned_graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "weight_cutoff = 0.8\n",
    "sample_size = 200\n",
    "pG = prune_and_subsample_graph(G, weight_cutoff, sample_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "200\n",
      "1572\n"
     ]
    }
   ],
   "source": [
    "print(pG.vcount())\n",
    "print(pG.ecount())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "def igraph_to_coords_file(G, aligned_labels, groups):\n",
    "    color_map = create_color_mapping(groups)\n",
    "    color_values = [color_map[group] for group in groups]\n",
    "    layout = G.layout('kk', dim=3)\n",
    "    Xn = [layout[k][0] for k in range(len(G.vs))]\n",
    "    Yn = [layout[k][1] for k in range(len(G.vs))]\n",
    "    Zn = [layout[k][2] for k in range(len(G.vs))]\n",
    "    nodes = {\n",
    "            i: {\n",
    "                \"location\": [\n",
    "                    Xn[i], Yn[i], Zn[i]\n",
    "                ],\n",
    "                \"name\": aligned_labels[i],\n",
    "                \"color\": color_values[i],\n",
    "                \"plasmid\": groups[i],\n",
    "            } for i in range(G.vcount())\n",
    "     } #\n",
    "    edges = [{\"source\": e.source, \"target\": e.target} for e in G.es]\n",
    "\n",
    "    return {\"nodes\": nodes, \"edges\": edges}\n",
    "\n",
    "# Convert the graph and save to JSON\n",
    "graph_data = igraph_to_coords_file(G, aligned_labels, aligned_groups)\n",
    "with open('network.json', 'w') as f:\n",
    "    json.dump(graph_data, f, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
