{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9caf0b1a-c3b5-4352-8928-4858a306ef4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "from Bio import SeqIO, SeqFeature, SeqRecord\n",
    "from collections import defaultdict\n",
    "import pickle\n",
    "import glob\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f7417072-48bc-420e-838e-194651d49877",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_assemblies = glob.glob('/home/mf019/longread_pangenome/longread_analysis/paired_assemblies/paired_only/*/annotation/*/*.gbff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "f8e4fa90-546a-497a-aa3f-db3a768059fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lets set up our parsing function. This is just the barebones function. I'll expand it with additional helpers\n",
    "# once I reach that point :)\n",
    "# primarily to include the plasmid name for the replicon.\n",
    "# and to\n",
    "def parse_genbank(file_path):\n",
    "    assembly_id = os.path.basename(file_path).split('.')[0]\n",
    "    assemblies = SeqIO.parse(file_path, \"genbank\")\n",
    "    genes = []\n",
    "    for record in assemblies:\n",
    "        for feature in record.features:\n",
    "            if feature.type in [\"gene\", \"CDS\", \"mRNA\", \"tRNA\", \"rRNA\", \"signal_peptide\"]:\n",
    "                gene_info = {\n",
    "                    \"feature_type\": feature.type,\n",
    "                    \"gene_id\": feature.id,\n",
    "                    \"gene\": feature.qualifiers.get(\"gene\", [\"unknown\"])[0],\n",
    "                    \"gene_name\": feature.qualifiers.get(\"name\", [\"unknown\"])[0],\n",
    "                    \"gene_family\": feature.qualifiers.get(\"note\", [\"unknown\"])[0],\n",
    "                    \"protein_id\": feature.qualifiers.get(\"protein_id\", [\"unknown\"])[0],\n",
    "                    \"product\": feature.qualifiers.get(\"product\", [\"unknown\"])[0],\n",
    "                    \"sequence\": str(feature.extract(record.seq)),\n",
    "                    \"contig\": record.id,\n",
    "                    \"replicon_name\": \"Placeholder for merging with calls :)\",\n",
    "                    \"start\": feature.location.start,\n",
    "                    \"end\": feature.location.end,\n",
    "                    \"strand\": feature.location.strand,\n",
    "                    \"dbxrefs\": feature.qualifiers.get(\"dbxrefs\",[\"No crossrefs?\"])[0],\n",
    "                    \"assembly\": assembly_id,\n",
    "                    \"refseq_id\": feature.qualifiers.get(\"RefSeq\", [\"No RefSeq?\"])[0],\n",
    "                    \"gene_seq\": str(feature.extract(record.seq)),\n",
    "                }\n",
    "                genes.append(gene_info)\n",
    "    return genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "1627a4a0-4316-4e1a-90b0-8006b6551c48",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "URI87H\n",
      "3009\n",
      "URI34H\n",
      "2667\n",
      "URI88H\n",
      "3081\n",
      "URI33H\n",
      "2621\n",
      "UCT110H\n",
      "2763\n",
      "URI39H\n",
      "3161\n",
      "URI91H\n",
      "3225\n",
      "UCT35H\n",
      "2633\n",
      "UWI247P\n",
      "2677\n",
      "URI120H\n",
      "3167\n",
      "URI107H\n",
      "3035\n",
      "UWI263P\n",
      "2821\n",
      "URI89H\n",
      "3129\n",
      "URI42H\n",
      "2921\n",
      "URI44H\n",
      "2759\n",
      "UCT109H\n",
      "2953\n",
      "URI40H\n",
      "3017\n",
      "URI117H\n",
      "2671\n",
      "URI47H\n",
      "3031\n",
      "URI86H\n",
      "2735\n",
      "URI36H\n",
      "2923\n",
      "UNY208P\n",
      "2829\n",
      "ESI26H\n",
      "2503\n",
      "UCT31H\n",
      "3105\n",
      "URI56H\n",
      "2451\n",
      "UCT30H\n",
      "2847\n",
      "URI103H\n",
      "2925\n",
      "UCT29H\n",
      "3173\n",
      "URI112H\n",
      "2809\n",
      "UWI248P\n",
      "2751\n",
      "UNY203P\n",
      "2655\n",
      "UCT96H\n",
      "2979\n",
      "UCT32H\n",
      "2995\n",
      "UNY193P\n",
      "2793\n",
      "UCT113H\n",
      "2455\n",
      "URI93H\n",
      "2929\n",
      "UNY169P\n",
      "2685\n",
      "UWI283P\n",
      "2631\n",
      "URI102H\n",
      "2863\n",
      "URI41H\n",
      "2951\n",
      "UNY172P\n",
      "2673\n",
      "UNY149P\n",
      "2855\n",
      "UCT92H\n",
      "3101\n",
      "URI118H\n",
      "2711\n",
      "UCT50H\n",
      "2647\n",
      "URI101H\n",
      "3105\n",
      "URI46H\n",
      "2803\n",
      "URI48H\n",
      "2723\n",
      "URI111H\n",
      "3103\n",
      "URI47\n",
      "2665\n",
      "URI103\n",
      "2571\n",
      "UCT31\n",
      "2625\n",
      "UCT92\n",
      "2649\n",
      "URI46\n",
      "2633\n",
      "UWI247\n",
      "2481\n",
      "UCT30\n",
      "2597\n",
      "URI44\n",
      "2591\n",
      "URI87\n",
      "2673\n",
      "UCT50\n",
      "2611\n",
      "UWI283\n",
      "2511\n",
      "URI107\n",
      "2679\n",
      "UCT110\n",
      "2659\n",
      "URI48\n",
      "2665\n",
      "UNY203\n",
      "2513\n",
      "UWI263\n",
      "2465\n",
      "URI117\n",
      "2443\n",
      "UCT35\n",
      "2571\n",
      "URI111\n",
      "2723\n",
      "URI33\n",
      "2601\n",
      "URI86\n",
      "2567\n",
      "UNY149\n",
      "2617\n",
      "UCT96\n",
      "2499\n",
      "URI56\n",
      "2579\n",
      "URI118\n",
      "2383\n",
      "UNY172\n",
      "2499\n",
      "URI112\n",
      "2567\n",
      "UNY208\n",
      "2485\n",
      "URI89\n",
      "2751\n",
      "URI102\n",
      "2659\n",
      "URI36\n",
      "2623\n",
      "UWI248\n",
      "2487\n",
      "ESI26\n",
      "2581\n",
      "URI40\n",
      "2715\n",
      "URI39\n",
      "2679\n",
      "UNY169\n",
      "2593\n",
      "UCT113\n",
      "2637\n",
      "URI101\n",
      "2775\n",
      "URI42\n",
      "2617\n",
      "UCT109\n",
      "2671\n",
      "UCT32\n",
      "2623\n",
      "URI41\n",
      "2647\n",
      "URI88\n",
      "2747\n",
      "URI91\n",
      "2715\n",
      "UCT29\n",
      "2717\n",
      "URI34\n",
      "3248\n",
      "URI120\n",
      "2497\n",
      "UNY193\n",
      "2977\n",
      "URI93\n",
      "2727\n"
     ]
    }
   ],
   "source": [
    "all_genes = []\n",
    "total_feats = 0\n",
    "for assembly in list_of_assemblies:\n",
    "    current_genes = []\n",
    "    assembly_id = os.path.basename(assembly).split('.')[0]\n",
    "    print(assembly_id)\n",
    "    current_genes = parse_genbank(assembly)\n",
    "    num_feats = len(current_genes)\n",
    "    print(num_feats)\n",
    "    total_feats += num_feats\n",
    "    all_genes.append(current_genes)\n",
    "all_genes_flat = [feature for features in all_genes for feature in features]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "35597b7d-7156-48a0-a887-d3d6c9eb86bf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "list: 268857\n",
      "actual: 268857\n"
     ]
    }
   ],
   "source": [
    "all_feats = len(all_genes_flat)\n",
    "print(f'list: {all_feats}')\n",
    "print(f'actual: {total_feats}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "7dd0f159-90d9-4bae-bbd9-9819e958eae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_database(db_name):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    cursor = conn.cursor()\n",
    "    cursor.execute('''CREATE TABLE IF NOT EXISTS genes (\n",
    "                      id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "                      feature_type TEXT,\n",
    "                      gene_id TEXT,\n",
    "                      gene TEXT,\n",
    "                      gene_name TEXT,\n",
    "                      gene_family TEXT,\n",
    "                      protein_id TEXT,\n",
    "                      product TEXT,\n",
    "                      sequence TEXT,\n",
    "                      contig TEXT,\n",
    "                      replicon_name TEXT,\n",
    "                      start INTEGER,\n",
    "                      end INTEGER,\n",
    "                      strand INTEGER,\n",
    "                      dbxrefs TEXT,\n",
    "                      assembly TEXT,\n",
    "                      refseq_id TEXT,\n",
    "                      gene_seq TEXT)''')\n",
    "    conn.commit()\n",
    "    conn.close()\n",
    "    \n",
    "def populate_database(db_name, genes):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    cursor = conn.cursor()\n",
    "    for gene in genes:\n",
    "        cursor.execute('''INSERT INTO genes\n",
    "                          (feature_type, gene_id, gene, gene_name, gene_family, protein_id, product, sequence, contig, replicon_name, start, end, strand, dbxrefs, assembly, refseq_id, gene_seq)\n",
    "                          VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)''',\n",
    "                       (gene[\"feature_type\"], gene[\"gene_id\"], gene[\"gene\"], gene[\"gene_name\"], gene[\"gene_family\"],\n",
    "                        gene[\"protein_id\"], gene[\"product\"], gene[\"sequence\"], gene[\"contig\"],\n",
    "                        gene[\"replicon_name\"], gene[\"start\"], gene[\"end\"], gene[\"strand\"], gene[\"dbxrefs\"],\n",
    "                        gene[\"assembly\"], gene[\"refseq_id\"], gene[\"gene_seq\"]))\n",
    "    conn.commit()\n",
    "    conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "10fa434d-5aeb-4e2b-83da-cdbe902190ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example usage\n",
    "def query_database(db_name, **kwargs):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    cursor = conn.cursor()\n",
    "    \n",
    "    query = \"SELECT * FROM genes WHERE \"\n",
    "    conditions = []\n",
    "    values = []\n",
    "    \n",
    "    for key, value in kwargs.items():\n",
    "        conditions.append(f\"{key} = ?\")\n",
    "        values.append(value)\n",
    "    \n",
    "    query += \" AND \".join(conditions)\n",
    "    cursor.execute(query, values)\n",
    "    results = cursor.fetchall()\n",
    "    conn.close()\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "715fb19b-5e02-426e-bb3b-12232ed72cc2",
   "metadata": {},
   "outputs": [
    {
     "ename": "ProgrammingError",
     "evalue": "Error binding parameter 17: type 'Seq' is not supported",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mProgrammingError\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[61], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m create_database(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgenes.db\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m----> 2\u001b[0m \u001b[43mpopulate_database\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgenes.db\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mall_genes_flat\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[59], line 30\u001b[0m, in \u001b[0;36mpopulate_database\u001b[0;34m(db_name, genes)\u001b[0m\n\u001b[1;32m     28\u001b[0m cursor \u001b[38;5;241m=\u001b[39m conn\u001b[38;5;241m.\u001b[39mcursor()\n\u001b[1;32m     29\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m gene \u001b[38;5;129;01min\u001b[39;00m genes:\n\u001b[0;32m---> 30\u001b[0m     \u001b[43mcursor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'''\u001b[39;49m\u001b[38;5;124;43mINSERT INTO genes\u001b[39;49m\n\u001b[1;32m     31\u001b[0m \u001b[38;5;124;43m                      (feature_type, gene_id, gene, gene_name, gene_family, protein_id, product, sequence, contig, replicon_name, start, end, strand, dbxrefs, assembly, refseq_id, gene_seq)\u001b[39;49m\n\u001b[1;32m     32\u001b[0m \u001b[38;5;124;43m                      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)\u001b[39;49m\u001b[38;5;124;43m'''\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     33\u001b[0m \u001b[43m                   \u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfeature_type\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene_id\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene_name\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene_family\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     34\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mprotein_id\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mproduct\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43msequence\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcontig\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     35\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mreplicon_name\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstart\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mend\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrand\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdbxrefs\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     36\u001b[0m \u001b[43m                    \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43massembly\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mrefseq_id\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgene_seq\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     37\u001b[0m conn\u001b[38;5;241m.\u001b[39mcommit()\n\u001b[1;32m     38\u001b[0m conn\u001b[38;5;241m.\u001b[39mclose()\n",
      "\u001b[0;31mProgrammingError\u001b[0m: Error binding parameter 17: type 'Seq' is not supported"
     ]
    }
   ],
   "source": [
    "create_database(\"genes.db\")\n",
    "populate_database(\"genes.db\",all_genes_flat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3e47e8f-d95c-443d-a12a-ee2b83914deb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
