{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a109cccb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pprint\n",
    "#from BCBio import GFF\n",
    "import pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "fc0246c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## none of these are currently being used\n",
    "#baf_gff = 'genomes/bafPKo.gff'\n",
    "#bb31_gff = 'genomes/bbB31.gff'\n",
    "#bga_gff = 'genomes/bgaPBi.gff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "b81c8e91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/mf019/longread_pangenome/expanded_dataset_analysis/group2BB/notebook\n"
     ]
    }
   ],
   "source": [
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "fac942f2-6879-4e8c-949c-3d51b57bc331",
   "metadata": {},
   "outputs": [],
   "source": [
    "rs2bb = '/home/mf019/longread_pangenome/expanded_dataset_analysis/group2BB/B31_refseq_id_to_old_name.tsv'\n",
    "rs2bb_df = pandas.read_csv(rs2bb, sep='\\t')\n",
    "rs2bb_df = rs2bb_df.rename(columns={\n",
    "    \"refseq_id\": \"gene\",\n",
    "    \"old_name\": \"old_locus_tag\"\n",
    "})\n",
    "rs2bb_df['old_locus_tag'] = rs2bb_df['old_locus_tag'].apply(lambda x: x.split('%2C')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "7d3696f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "sp_pangenome_ref = '/home/mf019/longread_pangenome/expanded_dataset_analysis/output/results/v8/v8_filtered_sp/roary_v8_filtered_sp/pan_genome_reference.fa'\n",
    "nsp_pangenome_ref = '/home/mf019/longread_pangenome/expanded_dataset_analysis/output/results/v8/v8_filtered_nsp/roary_v8_filtered_nsp/pan_genome_reference.fa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "0dfbb51a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set blast results\n",
    "v8_sp_blast_results = '/home/mf019/longread_pangenome/expanded_dataset_analysis/output/results/v8/v8_filtered_sp/alignments/pangenome_vs_B31.topHit.tsv'\n",
    "v8_nsp_blast_results = '/home/mf019/longread_pangenome/expanded_dataset_analysis/output/results/v8/v8_filtered_nsp/alignments/pangenome_vs_B31.topHit.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "75fc253f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results:                  0                1        2     3    4   5    6     7    8   \\\n",
      "0      CFHKPG_00001  gene-BB_RS00005  100.000   564    0   0    1   564   10   \n",
      "1      CFHKPG_00002  gene-BB_RS00010  100.000  1029    0   0    1  1029    1   \n",
      "2      CFHKPG_00003  gene-BB_RS00015   99.932  1467    0   1    1  1467    1   \n",
      "3      CFHKPG_00004  gene-BB_RS00020  100.000  1794    0   0    1  1794    1   \n",
      "4      CFHKPG_00005  gene-BB_RS00025  100.000  1056    0   0    1  1056    1   \n",
      "...             ...              ...      ...   ...  ...  ..  ...   ...  ...   \n",
      "27507  PHNMIL_01551  gene-BB_RS06635   70.404   446  120   5    1   440  127   \n",
      "27508  PHNMIL_01551  gene-BB_RS05785   69.802   404  114   4   41   440  161   \n",
      "27509  PHNMIL_01551  gene-BB_RS04455   69.974   383  111   4   61   441  184   \n",
      "27510  PHNMIL_01551  gene-BB_RS05530   67.025   279   82   5  119   392  197   \n",
      "27511  PHNMIL_01551  gene-BB_RS07780   77.215    79   18   0  230   308   98   \n",
      "\n",
      "         9             10      11  \n",
      "0       573  0.000000e+00  1018.0  \n",
      "1      1029  0.000000e+00  1856.0  \n",
      "2      1466  0.000000e+00  2638.0  \n",
      "3      1794  0.000000e+00  3236.0  \n",
      "4      1056  0.000000e+00  1905.0  \n",
      "...     ...           ...     ...  \n",
      "27507   566  1.480000e-51   198.0  \n",
      "27508   560  7.170000e-43   168.0  \n",
      "27509   564  1.300000e-39   159.0  \n",
      "27510   470  1.210000e-14    76.1  \n",
      "27511   176  7.660000e-11    62.6  \n",
      "\n",
      "[27512 rows x 12 columns]\n",
      "id2group:                 ID                                              group\n",
      "0     CFHKPG_00001                    Uncharacterized protein BB_0001\n",
      "1     CFHKPG_00002  Glycoside hydrolase family 3 N-terminal domain...\n",
      "2     CFHKPG_00003                                         group_2101\n",
      "3     CFHKPG_00004                                               manB\n",
      "4     CFHKPG_00005                                               trpS\n",
      "...            ...                                                ...\n",
      "4500  PHNMIL_01144                                         group_2880\n",
      "4501  PHNMIL_01214                                         group_3513\n",
      "4502  PHNMIL_01425                                          group_551\n",
      "4503  PHNMIL_01427                                         group_3787\n",
      "4504  PHNMIL_01551                                         group_1397\n",
      "\n",
      "[4505 rows x 2 columns]\n",
      "top_hits:                 ID        gene  percent_ident  alignment_length  \\\n",
      "0     AEPIPJ_00937  BB_RS05470         81.005               637   \n",
      "1     AEPIPJ_00957  BB_RS05925         99.579               951   \n",
      "2     AEPIPJ_00959  BB_RS05915         91.975              1296   \n",
      "3     AEPIPJ_00962  BB_RS05900         97.669               429   \n",
      "4     AEPIPJ_01042  BB_RS05260         71.951               246   \n",
      "...            ...         ...            ...               ...   \n",
      "4337  PPOBBG_01091  BB_RS05885         99.744               390   \n",
      "4338  PPOBBG_01252  BB_RS06040         66.294               537   \n",
      "4339  PPOBBG_01294  BB_RS06035        100.000              1137   \n",
      "4340  PPOBBG_01479  BB_RS05425         99.225               129   \n",
      "4341  PPOBBG_01480  BB_RS05425        100.000               144   \n",
      "\n",
      "            E-score       group old_locus_tag  \n",
      "0     3.430000e-170  group_4190        BB_C11  \n",
      "1      0.000000e+00  group_4156        BB_P17  \n",
      "2      0.000000e+00     group_1        BB_P15  \n",
      "3      0.000000e+00  group_4119        BB_P12  \n",
      "4      7.330000e-30   group_216        BB_A38  \n",
      "...             ...         ...           ...  \n",
      "4337   0.000000e+00  group_3329        BB_P09  \n",
      "4338   4.480000e-40   group_989        BB_P40  \n",
      "4339   0.000000e+00  group_1658        BB_P39  \n",
      "4340   2.270000e-61  group_1615        BB_A74  \n",
      "4341   4.340000e-71  group_1614        BB_A74  \n",
      "\n",
      "[4342 rows x 7 columns]\n",
      "translatedhits:                 ID       group        gene old_locus_tag  percent_ident  \\\n",
      "0     AEPIPJ_00937  group_4190  BB_RS05470        BB_C11         81.005   \n",
      "1     AEPIPJ_00957  group_4156  BB_RS05925        BB_P17         99.579   \n",
      "2     AEPIPJ_00959     group_1  BB_RS05915        BB_P15         91.975   \n",
      "3     AEPIPJ_00962  group_4119  BB_RS05900        BB_P12         97.669   \n",
      "4     AEPIPJ_01042   group_216  BB_RS05260        BB_A38         71.951   \n",
      "...            ...         ...         ...           ...            ...   \n",
      "4337  PPOBBG_01091  group_3329  BB_RS05885        BB_P09         99.744   \n",
      "4338  PPOBBG_01252   group_989  BB_RS06040        BB_P40         66.294   \n",
      "4339  PPOBBG_01294  group_1658  BB_RS06035        BB_P39        100.000   \n",
      "4340  PPOBBG_01479  group_1615  BB_RS05425        BB_A74         99.225   \n",
      "4341  PPOBBG_01480  group_1614  BB_RS05425        BB_A74        100.000   \n",
      "\n",
      "      alignment_length        E-score  \n",
      "0                  637  3.430000e-170  \n",
      "1                  951   0.000000e+00  \n",
      "2                 1296   0.000000e+00  \n",
      "3                  429   0.000000e+00  \n",
      "4                  246   7.330000e-30  \n",
      "...                ...            ...  \n",
      "4337               390   0.000000e+00  \n",
      "4338               537   4.480000e-40  \n",
      "4339              1137   0.000000e+00  \n",
      "4340               129   2.270000e-61  \n",
      "4341               144   4.340000e-71  \n",
      "\n",
      "[4342 rows x 7 columns]\n",
      "Missing Groups:\n",
      "                ID                          group\n",
      "346   CFHKPG_00354                     group_2259\n",
      "408   CFHKPG_00423                     group_2644\n",
      "452   CFHKPG_00474  NADH dehydrogenase subunit 4L\n",
      "853   CFHKPG_00894                     group_2385\n",
      "986   CFHKPG_01028                     group_1598\n",
      "...            ...                            ...\n",
      "4401  MGPBEI_01445                     group_2961\n",
      "4467  FEGGFG_00873                     group_2808\n",
      "4469  FEGGFG_00875                      aac(3)-Ia\n",
      "4470  FEGGFG_01126                     aph(3')-Ia\n",
      "4497  JEKLJL_01268                     group_2719\n",
      "\n",
      "[163 rows x 2 columns]\n",
      "results:                 0                1        2     3   4   5   6     7   8   \\\n",
      "0     CFHKPG_00001  gene-BB_RS00005  100.000   564   0   0   1   564  10   \n",
      "1     CFHKPG_00002  gene-BB_RS00010  100.000  1029   0   0   1  1029   1   \n",
      "2     CFHKPG_00003  gene-BB_RS00015   99.932  1467   0   1   1  1467   1   \n",
      "3     CFHKPG_00004  gene-BB_RS00020  100.000  1794   0   0   1  1794   1   \n",
      "4     CFHKPG_00005  gene-BB_RS00025  100.000  1056   0   0   1  1056   1   \n",
      "...            ...              ...      ...   ...  ..  ..  ..   ...  ..   \n",
      "8248  PHNMIL_01144  gene-BB_RS06360   98.361   183   2   1   1   183   1   \n",
      "8249  PHNMIL_01144  gene-BB_RS07205   98.361   183   2   1   1   183   1   \n",
      "8250  PHNMIL_01144  gene-BB_RS06575   97.814   183   3   1   1   183   1   \n",
      "8251  PHNMIL_01144  gene-BB_RS07530   97.268   183   4   1   1   183   1   \n",
      "8252  PHNMIL_01144  gene-BB_RS06780   97.268   183   4   1   1   183   1   \n",
      "\n",
      "        9             10      11  \n",
      "0      573  0.000000e+00  1018.0  \n",
      "1     1029  0.000000e+00  1856.0  \n",
      "2     1466  0.000000e+00  2638.0  \n",
      "3     1794  0.000000e+00  3236.0  \n",
      "4     1056  0.000000e+00  1905.0  \n",
      "...    ...           ...     ...  \n",
      "8248   182  1.030000e-86   314.0  \n",
      "8249   182  1.030000e-86   314.0  \n",
      "8250   182  1.260000e-85   309.0  \n",
      "8251   182  5.340000e-84   305.0  \n",
      "8252   182  5.340000e-84   305.0  \n",
      "\n",
      "[8253 rows x 12 columns]\n",
      "id2group:                 ID                                              group\n",
      "0     CFHKPG_00001                    Uncharacterized protein BB_0001\n",
      "1     CFHKPG_00002  Glycoside hydrolase family 3 N-terminal domain...\n",
      "2     CFHKPG_00003                                          group_490\n",
      "3     CFHKPG_00004                                         group_2077\n",
      "4     CFHKPG_00005                                               trpS\n",
      "...            ...                                                ...\n",
      "2085  FEGGFG_01126                                         aph(3')-Ia\n",
      "2086  FEGGFG_01127                                         group_1090\n",
      "2087  HKCNAF_01204                                         group_1135\n",
      "2088  JEKLJL_01268                                         group_1006\n",
      "2089  PHNMIL_01144                                         group_1148\n",
      "\n",
      "[2090 rows x 2 columns]\n",
      "top_hits:                 ID        gene  percent_ident  alignment_length  \\\n",
      "0     AEPIPJ_01198  BB_RS05980         85.679               405   \n",
      "1     AEPIPJ_01339  BB_RS05985         93.458               107   \n",
      "2     AEPIPJ_01570  BB_RS06030         86.957                69   \n",
      "3     AEPIPJ_01576  BB_RS04520         79.505               444   \n",
      "4     AGEDGC_01059  BB_RS06030         87.500                64   \n",
      "...            ...         ...            ...               ...   \n",
      "1980  PHNMIL_01144  BB_RS05935         98.907               183   \n",
      "1981  PKEGNH_00854  BB_RS03720         78.947                76   \n",
      "1982  PKEGNH_00912  BB_RS05980         88.889               153   \n",
      "1983  PKEGNH_00997  BB_RS05980         77.701               435   \n",
      "1984  PKEGNH_01154  BB_RS04605         89.599               798   \n",
      "\n",
      "            E-score                   group old_locus_tag  \n",
      "0     1.570000e-133              group_1332        BB_P28  \n",
      "1      2.090000e-41              group_1333        BB_P29  \n",
      "2      5.620000e-17              group_1334        BB_P38  \n",
      "3     4.670000e-109              group_1335        BB_G06  \n",
      "4      4.540000e-16              group_1039        BB_P38  \n",
      "...             ...                     ...           ...  \n",
      "1980   2.430000e-88              group_1148        BB_P19  \n",
      "1981   2.340000e-12              group_1040       BB_0734  \n",
      "1982   2.540000e-50              group_1041        BB_P28  \n",
      "1983   5.300000e-95  Mlp lipofamily protein        BB_P28  \n",
      "1984   0.000000e+00              group_1044        BB_G22  \n",
      "\n",
      "[1985 rows x 7 columns]\n",
      "translatedhits:                 ID                   group        gene old_locus_tag  \\\n",
      "0     AEPIPJ_01198              group_1332  BB_RS05980        BB_P28   \n",
      "1     AEPIPJ_01339              group_1333  BB_RS05985        BB_P29   \n",
      "2     AEPIPJ_01570              group_1334  BB_RS06030        BB_P38   \n",
      "3     AEPIPJ_01576              group_1335  BB_RS04520        BB_G06   \n",
      "4     AGEDGC_01059              group_1039  BB_RS06030        BB_P38   \n",
      "...            ...                     ...         ...           ...   \n",
      "1980  PHNMIL_01144              group_1148  BB_RS05935        BB_P19   \n",
      "1981  PKEGNH_00854              group_1040  BB_RS03720       BB_0734   \n",
      "1982  PKEGNH_00912              group_1041  BB_RS05980        BB_P28   \n",
      "1983  PKEGNH_00997  Mlp lipofamily protein  BB_RS05980        BB_P28   \n",
      "1984  PKEGNH_01154              group_1044  BB_RS04605        BB_G22   \n",
      "\n",
      "      percent_ident  alignment_length        E-score  \n",
      "0            85.679               405  1.570000e-133  \n",
      "1            93.458               107   2.090000e-41  \n",
      "2            86.957                69   5.620000e-17  \n",
      "3            79.505               444  4.670000e-109  \n",
      "4            87.500                64   4.540000e-16  \n",
      "...             ...               ...            ...  \n",
      "1980         98.907               183   2.430000e-88  \n",
      "1981         78.947                76   2.340000e-12  \n",
      "1982         88.889               153   2.540000e-50  \n",
      "1983         77.701               435   5.300000e-95  \n",
      "1984         89.599               798   0.000000e+00  \n",
      "\n",
      "[1985 rows x 7 columns]\n",
      "Missing Groups:\n",
      "                ID                          group\n",
      "345   CFHKPG_00354                      group_607\n",
      "407   CFHKPG_00423                      group_941\n",
      "451   CFHKPG_00474  NADH dehydrogenase subunit 4L\n",
      "852   CFHKPG_00894                      group_705\n",
      "974   CFHKPG_01028                      group_169\n",
      "...            ...                            ...\n",
      "2069  MGPBEI_01445                     group_1227\n",
      "2082  FEGGFG_00873                     group_1086\n",
      "2084  FEGGFG_00875                      aac(3)-Ia\n",
      "2085  FEGGFG_01126                     aph(3')-Ia\n",
      "2088  JEKLJL_01268                     group_1006\n",
      "\n",
      "[105 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "ver = 'v8'\n",
    "typeof = ['filtered_sp', 'filtered_nsp']\n",
    "\n",
    "for t in typeof:\n",
    "    g2idmap = f'/home/mf019/longread_pangenome/expanded_dataset_analysis/output/results/{ver}/{ver}_{t}/alignments/{ver}_{t}_group_to_ID_mapping.txt'\n",
    "    g2BBmap = f'/home/mf019/longread_pangenome/expanded_dataset_analysis/output/results/{ver}/{ver}_{t}/alignments/{ver}_{t}_groupID_to_BBgene.tsv'\n",
    "    mg_out = f'/home/mf019/longread_pangenome/expanded_dataset_analysis/output/results/{ver}/{ver}_{t}/alignments/{ver}_{t}_missing_groups.tsv'\n",
    "    pg_ref = f'/home/mf019/longread_pangenome/expanded_dataset_analysis/output/results/{ver}/{ver}_{t}/roary_{ver}_{t}/pan_genome_reference.fa'\n",
    "    blast_results = f'/home/mf019/longread_pangenome/expanded_dataset_analysis/output/results/{ver}/{ver}_{t}/alignments/pangenome_vs_B31.topHit.tsv'\n",
    "    \n",
    "    if not os.path.exists(g2idmap):\n",
    "        # Create the group-to-ID mapping\n",
    "        with open(pg_ref) as f, open(g2idmap, 'w') as out:\n",
    "            for line in f:\n",
    "                if line.startswith('>'):\n",
    "                    line = line.replace('>', '').split(' ')\n",
    "                    seqID = line[0]\n",
    "                    groupID = ' '.join(line[1:])\n",
    "                    if 'BB' in groupID:\n",
    "                        groupID = 'BB'.join(groupID.split('BB'))\n",
    "                    out.write(f'{seqID}\\t{groupID}\\n')\n",
    "\n",
    "    # Read BLAST results\n",
    "    results = pandas.read_csv(blast_results, sep='\\t', header=None)\n",
    "    print(\"results:\", results)\n",
    "\n",
    "    # Read ID-to-group mapping\n",
    "    id2group = pandas.read_csv(g2idmap, sep='\\t', header=None, names=[\"ID\", \"group\"])\n",
    "    print(\"id2group:\", id2group)\n",
    "\n",
    "    # Blast TSV Header: \n",
    "    #(0, 'query_acc.ver') (1, 'subject_acc.ver') (2, '%_identity') (3, 'alignment') \n",
    "    #(4, 'length') (5, 'mismatches') (6, 'gap_opens') (7, 'q._start') \n",
    "    #(8, 'q._end') (9, 's._start') (10, 's._end') (11, 'evalue') (12, 'bit_score')\n",
    "    # Create the hits DataFrame\n",
    "    hits = pandas.DataFrame({\n",
    "        \"ID\": results[0],\n",
    "        \"gene\": results[1],\n",
    "        \"percent_ident\": results[2],\n",
    "        \"alignment_length\": results[3],\n",
    "        \"E-score\": results[10]\n",
    "    })\n",
    "    hits['gene'] = hits['gene'].apply(lambda x: x.strip('gene-'))\n",
    "    # Merge hits with id2group to map groups\n",
    "    hits = hits.merge(id2group, on=\"ID\", how=\"left\")\n",
    "\n",
    "    # Drop duplicates\n",
    "    hits = hits.drop_duplicates()\n",
    "\n",
    "    hits = hits.merge(rs2bb_df, on=\"gene\", how=\"left\")\n",
    "    \n",
    "    # Filter for genes that start with \"gene-BB\"\n",
    "    #hits = hits.loc[hits[\"gene\"].astype(str).str.startswith(\"gene-BB\")]\n",
    "\n",
    "    # Sort and group by query ID\n",
    "    top_hits = hits.sort_values(\n",
    "        by=[\"ID\", \"group\", \"gene\", \"old_locus_tag\", \"E-score\", \"percent_ident\", \"alignment_length\"],\n",
    "        ascending=[True, True, True, False, False, False, False]\n",
    "    ).groupby(\"ID\", as_index=False).first()\n",
    "        # Handle duplicate columns (group_x and group_y)\n",
    "    if \"group_y\" in top_hits.columns:\n",
    "        top_hits = top_hits.drop(columns=[\"group_y\"]).rename(columns={\"group_x\": \"group\"})\n",
    "    print(\"top_hits:\", top_hits)\n",
    "    # Reorder columns\n",
    "    translatedhits = top_hits[[\"ID\", \"group\", \"gene\", \"old_locus_tag\", \"percent_ident\", \"alignment_length\", \"E-score\"]]\n",
    "    print(\"translatedhits:\", translatedhits)\n",
    "    # Find groups in id2group that are not in translatedhits\n",
    "    missing_groups = id2group[~id2group[\"group\"].isin(translatedhits[\"group\"])]\n",
    "\n",
    "    # Output missing groups\n",
    "    print(\"Missing Groups:\")\n",
    "    print(missing_groups)\n",
    "    # Write the final results to a TSV file\n",
    "    translatedhits.to_csv(g2BBmap, sep='\\t', header=True, index=False)\n",
    "    missing_groups.to_csv(mg_out, sep='\\t', header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3478d4e-0bd4-4757-b090-d0bad8b67f1b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
