{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3bf9432d-e7ae-4b05-a91a-0c84d8581bfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pathlib\n",
    "import sqlite3\n",
    "import pickle\n",
    "import pandas\n",
    "from tqdm.notebook import tqdm\n",
    "from Bio import SeqIO\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "046be1d5-f834-46d4-a667-e41e263bf85e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pickle_dict(input_object, output_file):\n",
    "    with open(output_file, 'wb') as jar:\n",
    "        pickle.dump(input_object, jar)\n",
    "    print(f\"Successfully Pickled to {output_file}\")\n",
    "\n",
    "######################################################################################################\n",
    "#::::::::::::::::::::::::::::::::::GENBANK PARSER AND DICT CREATOR:::::::::::::::::::::::::::::::::::#\n",
    "######################################################################################################\n",
    "# Lets set up our parsing function. This is just the barebones function.\n",
    "# I'll expand it with additional helpers once I reach that point :)\n",
    "# primarily to include the plasmid name for the replicon.and more?\n",
    "def parse_genbank(file_path, assembly_id, acc2name, contig2plasmid):\n",
    "    set_of_keys = set()\n",
    "    ref_flag = True if assembly_id == \"REF\" else False\n",
    "    assemblies = SeqIO.parse(file_path, \"genbank\")\n",
    "    genes = []\n",
    "    for record in assemblies:\n",
    "        if ref_flag == False and assembly_id not in contig2plasmid.keys():\n",
    "            break\n",
    "        for feature in record.features:\n",
    "            if feature.type in [\"gene\", \"CDS\", \"mRNA\", \"tRNA\", \"rRNA\", \"signal_peptide\"]:\n",
    "                if ref_flag:\n",
    "                    assembly_id = acc_id2name[record.id]['strain']\n",
    "                    replicon_name = acc_id2name[record.id]['name']\n",
    "                else:\n",
    "                    if record.id not in contig2plasmid[assembly_id]:\n",
    "                        replicon_name = 'NA'\n",
    "                    else:\n",
    "                        replicon_name = contig2plasmid[assembly_id][record.id]['plasmid_name']\n",
    "\n",
    "                locus_tag = feature.qualifiers.get(\"locus_tag\", ['unknown'])[0]\n",
    "                db_xrefs = feature.qualifiers.get(\"db_xref\", [])\n",
    "                db_xref_dict = {xref.split(\":\")[0]: xref.split(\":\")[1] for xref in db_xrefs}\n",
    "\n",
    "                gene_info = {\n",
    "                    \"feature_type\": feature.type,\n",
    "                    \"gene\": feature.qualifiers.get(\"gene\", [\"unknown\"])[0],\n",
    "                    \"locus_tag\": locus_tag,\n",
    "                    \"note\": feature.qualifiers.get(\"note\", [\"unknown\"])[0],\n",
    "                    \"protein_id\": feature.qualifiers.get(\"protein_id\", [\"unknown\"])[0],\n",
    "                    \"product\": feature.qualifiers.get(\"product\", [\"unknown\"])[0],\n",
    "                    \"sequence\": str(feature.extract(record.seq)),\n",
    "                    \"replicon\": record.id,\n",
    "                    \"replicon_name\": replicon_name,\n",
    "                    \"start\": feature.location.start,\n",
    "                    \"end\": feature.location.end,\n",
    "                    \"strand\": feature.location.strand,\n",
    "                    \"assembly\": assembly_id,\n",
    "                    \"translation\": str(feature.qualifiers.get('translation',[''])),\n",
    "                    \"inference\": feature.qualifiers.get(\"inference\",[\"\"])[0],\n",
    "                    \"transl_table\": feature.qualifiers.get(\"transl_table\",[\"\"])[0],\n",
    "                    'db_xrefs': feature.qualifiers.get(\"db_xrefs\",[\"\"])[0],\n",
    "                    **db_xref_dict  # Unpack db_xref_dict to include db_xrefs as columns\n",
    "                }\n",
    "                genes.append(gene_info)\n",
    "                set_of_keys.update(gene_info.keys())\n",
    "\n",
    "    # fill out empty dicts with None!\n",
    "    for gene in genes:\n",
    "        for key in set_of_keys:\n",
    "            if key not in gene:\n",
    "                gene[key] = None\n",
    "\n",
    "    return genes\n",
    "\n",
    "######################################################################################################\n",
    "#::::::::::::::::::::::::::::::HARD SET SCHEMA METHODS (First attempt):::::::::::::::::::::::::::::::#\n",
    "######################################################################################################\n",
    "def create_table(db_name, table_name):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    cursor = conn.cursor()\n",
    "    cursor.execute(f'''CREATE TABLE IF NOT EXISTS {table_name} (\n",
    "                      id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "                      feature_type TEXT,\n",
    "                      gene TEXT,\n",
    "                      locus_tag TEXT,\n",
    "                      note TEXT,\n",
    "                      protein_id TEXT,\n",
    "                      product TEXT,\n",
    "                      sequence TEXT,\n",
    "                      replicon TEXT,\n",
    "                      replicon_name TEXT,\n",
    "                      start INTEGER,\n",
    "                      end INTEGER,\n",
    "                      strand INTEGER,\n",
    "                      db_xrefs TEXT,\n",
    "                      assembly TEXT,\n",
    "                      translation TEXT,\n",
    "                      inference TEXT,\n",
    "                      transl_table INT)''')\n",
    "    conn.commit()\n",
    "    conn.close()\n",
    "\n",
    "def populate_table(db_name, table_name, genes):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    cursor = conn.cursor()\n",
    "    for gene in genes:\n",
    "        cursor.execute(f'''INSERT INTO {table_name}\n",
    "                          (feature_type, gene, locus_tag, note, protein_id, product, sequence, replicon, replicon_name, start, end, strand, db_xrefs, assembly, translation, inference, transl_table)\n",
    "                          VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)''',\n",
    "                       (gene[\"feature_type\"], gene[\"gene\"], gene[\"locus_tag\"], gene[\"note\"],\n",
    "                        gene[\"protein_id\"], gene[\"product\"], gene[\"sequence\"], gene[\"replicon\"],\n",
    "                        gene[\"replicon_name\"], gene[\"start\"], gene[\"end\"], gene[\"strand\"], gene[\"db_xrefs\"],\n",
    "                        gene[\"assembly\"], gene[\"translation\"], gene[\"inference\"], gene[\"transl_table\"]))\n",
    "    conn.commit()\n",
    "    conn.close()\n",
    "    \n",
    "######################################################################################################\n",
    "#::::::::::::::::::::::::::::::AUTOMATED CREATION METHOD (2nd attempt):::::::::::::::::::::::::::::::#\n",
    "######################################################################################################\n",
    "# Make table and populate dynamically since we don't know all of the db_xrefs\n",
    "# (or rather haven't used them in SCHEMA creation)\n",
    "def create_and_populate_database(genes, db_name):\n",
    "    df = pandas.DataFrame(genes)\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    df.to_sql('annotations', conn, if_exists='replace', index=False)\n",
    "    conn.close()\n",
    "\n",
    "\n",
    "#####################################################################################################\n",
    "#:::::::::::::::::::::::::::::::THESE HELPERS WORK WITH BOTH METHODS::::::::::::::::::::::::::::::::#\n",
    "#####################################################################################################\n",
    "def get_tables(db_name):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    tables = pandas.read_sql_query(\"SELECT name FROM sqlite_master WHERE type='table';\", conn)\n",
    "    conn.close()\n",
    "    return tables\n",
    "\n",
    "def get_columns(db_name, table):\n",
    "    # Describe the structure of the 'genes' table\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    columns = pandas.read_sql_query(f\"PRAGMA table_info({table});\", conn)\n",
    "    conn.close()\n",
    "    return columns\n",
    "\n",
    "def table_head(db_name, table):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    table_head = pandas.read_sql_query(f\"SELECT * FROM {table} LIMIT 10;\", conn)\n",
    "    conn.close()\n",
    "    return table_head\n",
    "\n",
    "def get_all_of_gene(db_name, table, gene):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    all_genes = pandas.read_sql_query(f\"SELECT * FROM {table} WHERE feature_type = 'CDS' AND gene='{gene}'\", conn)\n",
    "    conn.close()\n",
    "    return all_genes\n",
    "\n",
    "def get_all_of_product(db_name, table, product):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    all_products = pandas.read_sql_query(f\"SELECT * FROM {table} WHERE feature_type = 'CDS' AND product='{product}'\", conn)\n",
    "    conn.close()\n",
    "    return all_products\n",
    "\n",
    "def dump_table_to_df(db_name, table):\n",
    "    conn = sqlite3.connect(db_name)\n",
    "    dataframe = pandas.read_sql_query(f\"SELECT * FROM {table};\", conn)\n",
    "    conn.close()\n",
    "    return dataframe\n",
    "\n",
    "def connect_to_db(db_or_conn):\n",
    "    # is our input a db path or a connection to a db?\n",
    "    if isinstance(db_or_conn, str):\n",
    "        # it is string, it must be a db path. Create a new connection\n",
    "        conn = sqlite3.connect(db_or_conn)\n",
    "        close_conn = True\n",
    "    else:\n",
    "        # Ain't a string so must be connection. Use the existing connection\n",
    "        conn = db_or_conn\n",
    "        close_conn = False\n",
    "    return conn, close_conn\n",
    "\n",
    "def get_gene_from_baktaID(db_or_conn, table, bakta_id):\n",
    "    \"\"\"\n",
    "    example:\n",
    "    print(get_gene_from_baktaID(db, 'assemblies', 'CKDGNL_04300'))\n",
    "    \"\"\"\n",
    "    # check input db_or_conn and handle it.\n",
    "    conn, close_conn = connect_to_db(db_or_conn)\n",
    "    all_genes = pandas.read_sql_query(f\"SELECT * FROM {table} WHERE locus_tag='{bakta_id}'\", conn)\n",
    "    # Pack it in pack it out!\n",
    "    if close_conn:\n",
    "        conn.close()\n",
    "    return all_genes\n",
    "\n",
    "# SQL should never be written more than once and here is my function.\n",
    "def query_term_vs_multi_column(connection, table, term, columns):\n",
    "    # Construct the SQL query\n",
    "    query = f\"\"\"\n",
    "    SELECT *\n",
    "    FROM {table}\n",
    "    WHERE {' OR '.join([f\"{col} = ?\" for col in columns])}\n",
    "    \"\"\"\n",
    "    # try to find the term in each column.\n",
    "    try:\n",
    "        result_df = pandas.read_sql_query(query, connection, params=[term] * len(columns))\n",
    "        #print(f\"Query for term '{term}' succeeded.\")\n",
    "        return result_df # return our dataframe!\n",
    "    except Exception as e:\n",
    "        print(f\"Query for term '{term}' failed with error: {e}\")\n",
    "        return pandas.DataFrame() # Empty on error.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1a81324c-52bb-44b3-a306-af8fd73e2df0",
   "metadata": {},
   "outputs": [],
   "source": [
    "db_v = 'v3.1'                  # database_version\n",
    "ds_v = 'v5'                  # dataset_version\n",
    "pg_v = 'v8'                  # pangenome_version\n",
    "pc_v = 'v10.2'                  # plasmid calls version\n",
    "pg_t = f'{pg_v}_filtered_sp' # pangenome type [('split paralogs', '*_sp'), ('no split paralogs', '*_nsp')]\n",
    "db_n = f'Bbss_db_{db_v}.db' # database filename\n",
    "\n",
    "# Set core paths\n",
    "analysis_dir = pathlib.Path('/home/mf019/longread_pangenome/expanded_dataset_analysis')\n",
    "assemblies_dir = analysis_dir.joinpath('assemblies', f'dataset_{ds_v}')\n",
    "ref_dir = analysis_dir.joinpath('ref')\n",
    "db = ref_dir.joinpath('asm_db', db_n) # database path definition!\n",
    "roary_results = analysis_dir.joinpath('results', pg_v, pg_t , f'roary_{pg_t}/')\n",
    "\n",
    "# get input assemblies\n",
    "list_of_assemblies = list(assemblies_dir.glob('*/*.gbff')) # Bakta genbanks for our assemblies!\n",
    "\n",
    "# Plasmid Calls\n",
    "plasmid_calls_csv = analysis_dir.joinpath('genotyping', 'replicons', 'calls_v10', 'best_hits_1000bp_v10.2.csv') # replicon ids parsed from output of plasmid_caller.py :) # MUST USE ALL HITS TO FULLY PARSE EACH ASSEMBLY\n",
    "blast_parsing_pkl = ref_dir.joinpath('replicons', 'wp', 'wp_v2_info.pkl')\n",
    "# define roary output paths\n",
    "clustered_proteins_file = roary_results.joinpath('clustered_proteins')\n",
    "roary_gene_presence_absence_v4 = roary_results.joinpath('gene_presence_absence.Rtab')\n",
    "\n",
    "# Define file names for pickles\n",
    "contig2plasmid_pickle = ref_dir.joinpath('asm_db', f'dataset_{ds_v}_contig2plasmid_dict_1kb_{pc_v}.pkl') # dict to map from contig_id to plasmid_id\n",
    "assemblies_genbank_pickle = ref_dir.joinpath('asm_db', f'dataset_{ds_v}_assembly_genbank_dict_{pc_v}.pkl') # dict with all of our parsed genbanks within."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a11f44d9-42be-4522-8ee2-c631231ad525",
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's parse our plasmid parsing dictionary using ole (somewhat)reliable.\n",
    "# REMINDER: Structure is: {NCBI_ID : {'ID', 'name', 'length', 'strain'}}\n",
    "# was renamed FROM: 'blast_parsing_dict.pickle' TO: 'wp_v2_info.pkl'\n",
    "# {{TODO: RENAME FILE IN PLASMID CALLER CONTAINER/UPDATE SCRIPTS!}} <- done for v6\n",
    "\n",
    "with open(blast_parsing_pkl, 'rb') as infile:\n",
    "    acc2name = pickle.load(infile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "186d5b6f-d23e-41a8-999f-513c388891d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########################################\n",
    "# Parse best_hits for contig2plasmid dict #\n",
    "###########################################\n",
    "# Let's pull the plasmid calls in and  make us a dataframe\n",
    "contig2plasmid_df = pandas.read_csv(plasmid_calls_csv, delimiter=',')\n",
    "# subset that to just what we care about (for now)\n",
    "contig2plasmid_df = contig2plasmid_df[[ 'assembly_id', 'contig_id', 'plasmid_name' ]] # dropped 'completeness'\n",
    "contig2plasmid_df['contig_header'] = contig2plasmid_df['contig_id']\n",
    "contig2plasmid_df['contig_id'] = contig2plasmid_df['contig_id'].apply(lambda x: x.split(' ')[0])\n",
    "# and now let's flip it to a dict so we can use key lookups to simply rename them.\n",
    "contig2plasmid = defaultdict(lambda: defaultdict(dict))\n",
    "for _, row in contig2plasmid_df.iterrows():\n",
    "    contig2plasmid[row['assembly_id']][row['contig_id']] = {\n",
    "        'contig_header' : row ['contig_header'],\n",
    "        'plasmid_name': row['plasmid_name'],\n",
    "    }\n",
    "contig2plasmid = {name: dict(contigs) for name, contigs in contig2plasmid.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "33bd8f3b-9e38-4ed8-a263-6e970a8be639",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "28823c400d2a424ca3d5ec12e0ca222b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Parsing assembly genbanks!:   0%|          | 0/82 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_040790765.1_ASM4079076v1_genomic\n",
      "B418P\n",
      "GCF_002151465.1_ASM215146v1_genomic\n",
      "URI87H\n",
      "URI34H\n",
      "URI88H\n",
      "URI33H\n",
      "UCT110H\n",
      "URI39H\n",
      "URI91H\n",
      "UCT35H\n",
      "ESI361H\n",
      "UWI247P\n",
      "URI120H\n",
      "URI107H\n",
      "GCF_002151505.1_ASM215150v1_genomic\n",
      "UWI263P\n",
      "GCF_040790795.1_ASM4079079v1_genomic\n",
      "GCF_040790785.1_ASM4079078v1_genomic\n",
      "URI89H\n",
      "URI42H\n",
      "URI44H\n",
      "UCT109H\n",
      "URI40H\n",
      "B500P\n",
      "URI117H\n",
      "URI47H\n",
      "URI86H\n",
      "GCF_040790715.1_ASM4079071v1_genomic\n",
      "GCF_040790755.1_ASM4079075v1_genomic\n",
      "URI36H\n",
      "UNY208P\n",
      "GCF_040790745.1_ASM4079074v1_genomic\n",
      "ESI26H\n",
      "UCT31H\n",
      "GCF_040790805.1_ASM4079080v1_genomic\n",
      "ESI403H\n",
      "URI56H\n",
      "XYZ459H\n",
      "UCT30H\n",
      "GCF_040790735.1_ASM4079073v1_genomic\n",
      "GCF_040819585.1_PFhe_I_PB_Ill_cons_genomic\n",
      "URI103H\n",
      "UCT29H\n",
      "UNY1128P\n",
      "URI112H\n",
      "UNY1032P\n",
      "UWI248P\n",
      "UNY203P\n",
      "UCT96H\n",
      "GCF_019134655.1_ASM1913465v1_genomic\n",
      "ESI425H\n",
      "UCT32H\n",
      "UNY990P\n",
      "UNY193P\n",
      "UCT113H\n",
      "URI93H\n",
      "GCF_024662155.1_ASM2466215v1_genomic\n",
      "UNY169P\n",
      "UWI283P\n",
      "UNY1090P\n",
      "UNY1083P\n",
      "B331P\n",
      "URI102H\n",
      "URI41H\n",
      "UNY172P\n",
      "GCF_040790775.1_ASM4079077v1_genomic\n",
      "UNY149P\n",
      "GCF_024662195.1_ASM2466219v1_genomic\n",
      "UCT92H\n",
      "URI118H\n",
      "UCT50H\n",
      "URI101H\n",
      "UNY1085P\n",
      "URI46H\n",
      "URI48H\n",
      "GCF_002151485.1_ASM215148v1_genomic\n",
      "GCF_003367295.1_ASM336729v1_genomic\n",
      "GCF_024662175.1_ASM2466217v1_genomic\n",
      "URI111H\n",
      "UCT124H\n",
      "UNY1038P\n",
      "Making sure features match between input and final object\n",
      "list: 229962\n",
      "actual: 229962\n",
      "Successfully Pickled to /home/mf019/longread_pangenome/expanded_dataset_analysis/ref/asm_db/dataset_v5_assembly_genbank_dict_v10.2.pkl\n"
     ]
    }
   ],
   "source": [
    "####################\n",
    "# Parse assemblies #\n",
    "####################\n",
    "all_assembly_genes = []\n",
    "total_feats = 0\n",
    "for assembly in tqdm(list_of_assemblies, desc=\"Parsing assembly genbanks!\"):\n",
    "    current_genes = []\n",
    "    assembly_id = '.'.join(os.path.basename(assembly).split('.')[0:-1])\n",
    "    print(assembly_id)\n",
    "    current_genes = parse_genbank(assembly, assembly_id, acc2name, contig2plasmid)\n",
    "    num_feats = len(current_genes)\n",
    "    total_feats += num_feats\n",
    "    all_assembly_genes.append(current_genes)\n",
    "all_assembly_genes_flat = [feature for features in all_assembly_genes for feature in features]\n",
    "all_feats = len(all_assembly_genes_flat)\n",
    "print(\"Making sure features match between input and final object\")\n",
    "print(f'list: {all_feats}')\n",
    "print(f'actual: {total_feats}')\n",
    "pickle_dict(all_assembly_genes_flat, assemblies_genbank_pickle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "766e8937-aa48-47de-ad64-d46da7d81e00",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create_table(db, \"assemblies\")\n",
    "#populate_table(db,\"assemblies\", all_assembly_genes_flat)\n",
    "create_and_populate_database(all_assembly_genes_flat, db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7afd95a7-abb0-4b39-8af7-672d12c868c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_type</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>note</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>product</th>\n",
       "      <th>sequence</th>\n",
       "      <th>replicon</th>\n",
       "      <th>replicon_name</th>\n",
       "      <th>start</th>\n",
       "      <th>...</th>\n",
       "      <th>UniRef</th>\n",
       "      <th>UniParc</th>\n",
       "      <th>RFAM</th>\n",
       "      <th>GO</th>\n",
       "      <th>BlastRules</th>\n",
       "      <th>KEGG</th>\n",
       "      <th>PFAM</th>\n",
       "      <th>EC</th>\n",
       "      <th>COG</th>\n",
       "      <th>NCBIProtein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BFHFJM_00001</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>ATATATAATTTAATAGTATATATATATAATTTAATAGATAAAAAAT...</td>\n",
       "      <td>NZ_CP161107.1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDS</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BFHFJM_00001</td>\n",
       "      <td>(5' truncated)</td>\n",
       "      <td>gnl|Bakta|BFHFJM_00001</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>ATATATAATTTAATAGTATATATATATAATTTAATAGATAAAAAAT...</td>\n",
       "      <td>NZ_CP161107.1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BFHFJM_00002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>ATGAAATATAGTGCTATTTTATTAATATGTAGCGTTAATTTATTTT...</td>\n",
       "      <td>NZ_CP161107.1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>102</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDS</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BFHFJM_00002</td>\n",
       "      <td>unknown</td>\n",
       "      <td>gnl|Bakta|BFHFJM_00002</td>\n",
       "      <td>Uncharacterized protein BB_0001</td>\n",
       "      <td>ATGAAATATAGTGCTATTTTATTAATATGTAGCGTTAATTTATTTT...</td>\n",
       "      <td>NZ_CP161107.1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>102</td>\n",
       "      <td>...</td>\n",
       "      <td>UniRef90_O51035</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BFHFJM_00003</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>ATGGACTTTTTAAAAACCTTTTCTTTTTTGTTTTTTAGCTTTTTTT...</td>\n",
       "      <td>NZ_CP161107.1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>782</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CDS</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BFHFJM_00003</td>\n",
       "      <td>unknown</td>\n",
       "      <td>gnl|Bakta|BFHFJM_00003</td>\n",
       "      <td>Glycoside hydrolase family 3 N-terminal domain...</td>\n",
       "      <td>ATGGACTTTTTAAAAACCTTTTCTTTTTTGTTTTTTAGCTTTTTTT...</td>\n",
       "      <td>NZ_CP161107.1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>782</td>\n",
       "      <td>...</td>\n",
       "      <td>UniRef90_O54536</td>\n",
       "      <td>UPI000D02875C</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>gene</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BFHFJM_00004</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>ATGAATGTTAAAGTTGATAAAATTTTTTCTGAAATGATACTTGAAA...</td>\n",
       "      <td>NZ_CP161107.1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1798</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CDS</td>\n",
       "      <td>unknown</td>\n",
       "      <td>BFHFJM_00004</td>\n",
       "      <td>unknown</td>\n",
       "      <td>gnl|Bakta|BFHFJM_00004</td>\n",
       "      <td>UTP--glucose-1-phosphate uridylyltransferase</td>\n",
       "      <td>ATGAATGTTAAAGTTGATAAAATTTTTTCTGAAATGATACTTGAAA...</td>\n",
       "      <td>NZ_CP161107.1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1798</td>\n",
       "      <td>...</td>\n",
       "      <td>UniRef90_A0A0H3C280</td>\n",
       "      <td>UPI00016C372B</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>gene</td>\n",
       "      <td>manB</td>\n",
       "      <td>BFHFJM_00005</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>ATGCTTAAACAATATTCACTTAACATGAAAAATTTTAAAAAAGCTT...</td>\n",
       "      <td>NZ_CP161107.1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>3410</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDS</td>\n",
       "      <td>manB</td>\n",
       "      <td>BFHFJM_00005</td>\n",
       "      <td>unknown</td>\n",
       "      <td>gnl|Bakta|BFHFJM_00005</td>\n",
       "      <td>Phosphomannomutase</td>\n",
       "      <td>ATGCTTAAACAATATTCACTTAACATGAAAAATTTTAAAAAAGCTT...</td>\n",
       "      <td>NZ_CP161107.1</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>3410</td>\n",
       "      <td>...</td>\n",
       "      <td>UniRef90_O51892</td>\n",
       "      <td>UPI00016C4D34</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>G</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  feature_type     gene     locus_tag            note              protein_id  \\\n",
       "0         gene  unknown  BFHFJM_00001         unknown                 unknown   \n",
       "1          CDS  unknown  BFHFJM_00001  (5' truncated)  gnl|Bakta|BFHFJM_00001   \n",
       "2         gene  unknown  BFHFJM_00002         unknown                 unknown   \n",
       "3          CDS  unknown  BFHFJM_00002         unknown  gnl|Bakta|BFHFJM_00002   \n",
       "4         gene  unknown  BFHFJM_00003         unknown                 unknown   \n",
       "5          CDS  unknown  BFHFJM_00003         unknown  gnl|Bakta|BFHFJM_00003   \n",
       "6         gene  unknown  BFHFJM_00004         unknown                 unknown   \n",
       "7          CDS  unknown  BFHFJM_00004         unknown  gnl|Bakta|BFHFJM_00004   \n",
       "8         gene     manB  BFHFJM_00005         unknown                 unknown   \n",
       "9          CDS     manB  BFHFJM_00005         unknown  gnl|Bakta|BFHFJM_00005   \n",
       "\n",
       "                                             product  \\\n",
       "0                                            unknown   \n",
       "1                               hypothetical protein   \n",
       "2                                            unknown   \n",
       "3                    Uncharacterized protein BB_0001   \n",
       "4                                            unknown   \n",
       "5  Glycoside hydrolase family 3 N-terminal domain...   \n",
       "6                                            unknown   \n",
       "7       UTP--glucose-1-phosphate uridylyltransferase   \n",
       "8                                            unknown   \n",
       "9                                 Phosphomannomutase   \n",
       "\n",
       "                                            sequence       replicon  \\\n",
       "0  ATATATAATTTAATAGTATATATATATAATTTAATAGATAAAAAAT...  NZ_CP161107.1   \n",
       "1  ATATATAATTTAATAGTATATATATATAATTTAATAGATAAAAAAT...  NZ_CP161107.1   \n",
       "2  ATGAAATATAGTGCTATTTTATTAATATGTAGCGTTAATTTATTTT...  NZ_CP161107.1   \n",
       "3  ATGAAATATAGTGCTATTTTATTAATATGTAGCGTTAATTTATTTT...  NZ_CP161107.1   \n",
       "4  ATGGACTTTTTAAAAACCTTTTCTTTTTTGTTTTTTAGCTTTTTTT...  NZ_CP161107.1   \n",
       "5  ATGGACTTTTTAAAAACCTTTTCTTTTTTGTTTTTTAGCTTTTTTT...  NZ_CP161107.1   \n",
       "6  ATGAATGTTAAAGTTGATAAAATTTTTTCTGAAATGATACTTGAAA...  NZ_CP161107.1   \n",
       "7  ATGAATGTTAAAGTTGATAAAATTTTTTCTGAAATGATACTTGAAA...  NZ_CP161107.1   \n",
       "8  ATGCTTAAACAATATTCACTTAACATGAAAAATTTTAAAAAAGCTT...  NZ_CP161107.1   \n",
       "9  ATGCTTAAACAATATTCACTTAACATGAAAAATTTTAAAAAAGCTT...  NZ_CP161107.1   \n",
       "\n",
       "  replicon_name  start  ...               UniRef        UniParc  RFAM    GO  \\\n",
       "0    chromosome      2  ...                 None           None  None  None   \n",
       "1    chromosome      2  ...                 None           None  None  None   \n",
       "2    chromosome    102  ...                 None           None  None  None   \n",
       "3    chromosome    102  ...      UniRef90_O51035           None  None  None   \n",
       "4    chromosome    782  ...                 None           None  None  None   \n",
       "5    chromosome    782  ...      UniRef90_O54536  UPI000D02875C  None  None   \n",
       "6    chromosome   1798  ...                 None           None  None  None   \n",
       "7    chromosome   1798  ...  UniRef90_A0A0H3C280  UPI00016C372B  None  None   \n",
       "8    chromosome   3410  ...                 None           None  None  None   \n",
       "9    chromosome   3410  ...      UniRef90_O51892  UPI00016C4D34  None  None   \n",
       "\n",
       "  BlastRules  KEGG  PFAM    EC   COG NCBIProtein  \n",
       "0       None  None  None  None  None        None  \n",
       "1       None  None  None  None  None        None  \n",
       "2       None  None  None  None  None        None  \n",
       "3       None  None  None  None  None        None  \n",
       "4       None  None  None  None  None        None  \n",
       "5       None  None  None  None  None        None  \n",
       "6       None  None  None  None  None        None  \n",
       "7       None  None  None  None  None        None  \n",
       "8       None  None  None  None  None        None  \n",
       "9       None  None  None  None     G        None  \n",
       "\n",
       "[10 rows x 29 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_tables(db)\n",
    "get_columns(db, 'annotations')\n",
    "table_head(db, 'annotations')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d4004df-262a-4624-947f-95164a6db565",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
